{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c169068b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import warnings\n",
    "from warnings import simplefilter\n",
    "warnings.filterwarnings(\"ignore\", category=UserWarning)\n",
    "warnings.filterwarnings('ignore', category=FutureWarning)\n",
    "warnings.filterwarnings(\"ignore\", category=PendingDeprecationWarning)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "62042f97-a76e-4b34-805f-3f1ac81e7288",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import rpy2\n",
    "import logging\n",
    "import anndata2ri\n",
    "import scanpy as sc\n",
    "import decoupler as dc\n",
    "import matplotlib.pyplot as plt\n",
    "import rpy2.robjects as robjects\n",
    "from itertools import chain\n",
    "from rpy2.robjects import pandas2ri\n",
    "from functions import helper_functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "447b6d1c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-----\n",
      "anndata     0.11.1\n",
      "scanpy      1.9.3\n",
      "-----\n",
      "PIL                         9.5.0\n",
      "anndata2ri                  1.1\n",
      "asttokens                   NA\n",
      "backcall                    0.2.0\n",
      "cffi                        1.15.1\n",
      "comm                        0.1.3\n",
      "cycler                      0.10.0\n",
      "cython_runtime              NA\n",
      "dateutil                    2.8.2\n",
      "debugpy                     1.6.7\n",
      "decorator                   5.1.1\n",
      "decoupler                   1.4.0\n",
      "dot_parser                  NA\n",
      "exceptiongroup              1.1.1\n",
      "executing                   1.2.0\n",
      "functions                   NA\n",
      "google                      NA\n",
      "gsva_prep                   NA\n",
      "h5py                        3.9.0\n",
      "igraph                      0.10.4\n",
      "ipykernel                   6.23.2\n",
      "ipywidgets                  8.0.6\n",
      "jedi                        0.18.2\n",
      "jinja2                      3.1.2\n",
      "joblib                      1.2.0\n",
      "kiwisolver                  1.4.4\n",
      "leidenalg                   0.9.1\n",
      "llvmlite                    0.39.1\n",
      "louvain                     0.8.0\n",
      "markupsafe                  2.1.3\n",
      "matplotlib                  3.7.1\n",
      "mpl_toolkits                NA\n",
      "mpmath                      1.3.0\n",
      "natsort                     8.4.0\n",
      "numba                       0.56.4\n",
      "numexpr                     2.8.4\n",
      "numpy                       1.23.5\n",
      "nvfuser                     NA\n",
      "opt_einsum                  v3.3.0\n",
      "packaging                   23.1\n",
      "pandas                      2.0.2\n",
      "parso                       0.8.3\n",
      "patsy                       0.5.3\n",
      "pexpect                     4.8.0\n",
      "pickleshare                 0.7.5\n",
      "pkg_resources               NA\n",
      "platformdirs                3.7.0\n",
      "plotly                      5.24.1\n",
      "prompt_toolkit              3.0.38\n",
      "psutil                      5.9.5\n",
      "ptyprocess                  0.7.0\n",
      "pure_eval                   0.2.2\n",
      "pyarrow                     12.0.1\n",
      "pycparser                   2.21\n",
      "pydev_ipython               NA\n",
      "pydevconsole                NA\n",
      "pydevd                      2.9.5\n",
      "pydevd_file_utils           NA\n",
      "pydevd_plugins              NA\n",
      "pydevd_tracing              NA\n",
      "pydot                       1.4.2\n",
      "pygments                    2.15.1\n",
      "pyparsing                   3.1.0\n",
      "pytz                        2023.3\n",
      "pytz_deprecation_shim       NA\n",
      "rpy2                        3.5.12\n",
      "scipy                       1.10.1\n",
      "seaborn                     0.12.2\n",
      "session_info                1.0.0\n",
      "setuptools                  68.0.0\n",
      "six                         1.16.0\n",
      "sklearn                     1.2.2\n",
      "sparse                      0.14.0\n",
      "stack_data                  0.6.2\n",
      "statsmodels                 0.14.0\n",
      "sympy                       1.12\n",
      "texttable                   1.6.7\n",
      "threadpoolctl               3.1.0\n",
      "torch                       2.0.1+cu117\n",
      "tornado                     6.3.2\n",
      "tqdm                        4.65.0\n",
      "traitlets                   5.9.0\n",
      "typing_extensions           NA\n",
      "tzlocal                     NA\n",
      "vscode                      NA\n",
      "wcwidth                     0.2.6\n",
      "yaml                        6.0\n",
      "zmq                         25.1.0\n",
      "zoneinfo                    NA\n",
      "-----\n",
      "IPython             8.14.0\n",
      "jupyter_client      8.2.0\n",
      "jupyter_core        5.3.1\n",
      "-----\n",
      "Python 3.10.6 (main, Oct 24 2022, 16:07:47) [GCC 11.2.0]\n",
      "Linux-6.8.0-49-generic-x86_64-with-glibc2.35\n",
      "-----\n",
      "Session information updated at 2025-01-17 14:34\n"
     ]
    }
   ],
   "source": [
    "# # Ignore R warning messages\n",
    "#Note: this can be commented out to get more verbose R output\n",
    "rpy2.rinterface_lib.callbacks.logger.setLevel(logging.ERROR)\n",
    "\n",
    "# Automatically convert rpy2 outputs to pandas dataframes\n",
    "pandas2ri.activate()\n",
    "anndata2ri.activate()\n",
    "%load_ext rpy2.ipython\n",
    "\n",
    "sc.settings.verbosity = 3\n",
    "#sc.set_figure_params(dpi=200, dpi_save=300)\n",
    "sc.logging.print_versions()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "048c2e09",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "    WARNING: The R package \"reticulate\" only fixed recently\n",
      "    an issue that caused a segfault when used with rpy2:\n",
      "    https://github.com/rstudio/reticulate/pull/1188\n",
      "    Make sure that you use a version of that package that includes\n",
      "    the fix.\n",
      "    Allowing parallel execution with up to 40 working processes.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "In addition: Warning messages:\n",
       "1: replacing previous import ‘GenomicRanges::intersect’ by ‘SeuratObject::intersect’ when loading ‘hdWGCNA’ \n",
       "2: replacing previous import ‘GenomicRanges::union’ by ‘dplyr::union’ when loading ‘hdWGCNA’ \n",
       "3: replacing previous import ‘GenomicRanges::setdiff’ by ‘dplyr::setdiff’ when loading ‘hdWGCNA’ \n",
       "4: replacing previous import ‘dplyr::as_data_frame’ by ‘igraph::as_data_frame’ when loading ‘hdWGCNA’ \n",
       "5: replacing previous import ‘Seurat::components’ by ‘igraph::components’ when loading ‘hdWGCNA’ \n",
       "6: replacing previous import ‘dplyr::groups’ by ‘igraph::groups’ when loading ‘hdWGCNA’ \n",
       "7: replacing previous import ‘dplyr::union’ by ‘igraph::union’ when loading ‘hdWGCNA’ \n",
       "8: replacing previous import ‘GenomicRanges::subtract’ by ‘magrittr::subtract’ when loading ‘hdWGCNA’ \n",
       "9: replacing previous import ‘Matrix::as.matrix’ by ‘proxy::as.matrix’ when loading ‘hdWGCNA’ \n",
       "10: replacing previous import ‘igraph::groups’ by ‘tidygraph::groups’ when loading ‘hdWGCNA’ \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "suppressPackageStartupMessages({\n",
    "    library(Matrix)\n",
    "    library(viridis)\n",
    "    library(harmony)\n",
    "    library(ggpubr)\n",
    "    library(tictoc)\n",
    "    library(RColorBrewer)\n",
    "    library(Hmisc)\n",
    "    library(corrplot)\n",
    "    library(grid)\n",
    "    library(gridExtra)\n",
    "    library(igraph)\n",
    "    library(ggrepel)\n",
    "    library(readxl)\n",
    "    library(conflicted)\n",
    "    library(plyr)\n",
    "    library(dplyr)\n",
    "    library(parallel)\n",
    "    library(stringr)\n",
    "    library(BiocParallel)\n",
    "\n",
    "    # single-cell analysis package\n",
    "    library(Seurat)\n",
    "    library(zellkonverter)   \n",
    "    library(SingleCellExperiment)\n",
    "    library(tidyr)\n",
    "    library(readxl)\n",
    "    library(GSA)\n",
    "    library(limma)\n",
    "\n",
    "    # plotting and data science packages\n",
    "    library(tidyverse)\n",
    "    library(cowplot)\n",
    "    library(patchwork)\n",
    "    library(ggplot2)\n",
    "\n",
    "    # co-expression network analysis packages:\n",
    "    library(WGCNA)\n",
    "    library(hdWGCNA)\n",
    "\n",
    "    # gene enrichment packages\n",
    "    library(enrichR)\n",
    "    library(GeneOverlap)\n",
    "    library(GSEABase)\n",
    "    library(GSVA) \n",
    "\n",
    "    # cell-cell communication\n",
    "    library(nichenetr)\n",
    "\n",
    "# needs to be run every time you start R and want to use %>%\n",
    "})\n",
    "\n",
    "# using the cowplot theme for ggplot\n",
    "theme_set(theme_cowplot())\n",
    "\n",
    "# set random seed for reproducibility\n",
    "set.seed(12345)\n",
    "\n",
    "# optionally enable multithreading\n",
    "enableWGCNAThreads(nThreads = 40)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f21b7795",
   "metadata": {},
   "source": [
    "# **Systematic differential analysis of pathway activity**\n",
    "\n",
    "Pathway activity scores were computed in accordance with the overall protocol outlined in [**Joel W. Blanchard et. al.**](https://doi.org/10.1038/s41586-022-05439-w).\n",
    "\n",
    "- Aggregate gene expression profiles into `pseudo-cells` or `pseudo-bulks` depending on the desired strategy in **[Gazestani. et. al. 2023](https://www.sciencedirect.com/science/article/pii/S0092867423008590?via%3Dihub)** (`pseudo-cells`), [**Joel W. Blanchard et. al.**](https://doi.org/10.1038/s41586-022-05439-w) (`pseudo-bulks`), or [**Morabito et al.**](https://www.sciencedirect.com/science/article/pii/S2667237523001273?via%3Dihub) (`meta-cells`)\n",
    " \n",
    "\n",
    "- Use `PAGODA2`, `GSVA`, or `AUCell`, to obtain pathway activity scores for pathways (`GO Ontology Biological Processes`) obtained from [**Maayan Lab**](https://maayanlab.cloud/Enrichr/#libraries), [**MSigDB**](https://www.gsea-msigdb.org/gsea/msigdb/human/collections.jsp#C5) or custom gene sets.\n",
    "\n",
    "- Perform differential pathway activity analysis using limma"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a070e8e",
   "metadata": {},
   "source": [
    "\n",
    "## **Data Prep Parameters**\n",
    "\n",
    "- `test_names`: List of the different test names of interest.\n",
    "\n",
    "- `save_prefix`: Preferred prefix for saving critical files. Ideally chosen to be in the format `{source name}_{brain region}`. e.g `mathys_pfc`\n",
    "\n",
    "- `subject_id`: Column name for Subject/Patient ID in both metadata and `.obs`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "53693e9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "save_prefix = 'seaad_mtg'                           # this takes the format '{StudyName}_{ThreeLetterAccronymForBrainRegion}'\n",
    "cell_group = 'excitatory'\n",
    "cell_group_number = 7\n",
    "subject_id = 'Donor ID' \n",
    "cell_type_column = 'Subclass'                       # 'Supertype (non-expanded)', 'Subclass'\n",
    "pseudobulk_level = 'Supertype'                      # Level at which cells are aggregated into pseudobulks,\n",
    "                                                    # 'Class', 'Subclass', or 'Supertype' \n",
    "factor = 'pathology.group'                          # pathology.group # Continuous Pseudoprogression Score\n",
    "factor = factor.replace(\" \", \"\").replace(\"-\", \"\")\n",
    "test_names = ['late_vs_early']                      # test categories\n",
    "region_name = save_prefix.split('_')[-1].upper()\n",
    "\n",
    "data_dir = f'/media/tadeoye/Volume1/data/seq/SEA-AD/{region_name}/RNAseq'\n",
    "save_dir = f'../results/'\n",
    "\n",
    "if not os.path.exists(save_dir):\n",
    "    os.makedirs(save_dir)\n",
    "\n",
    "subclass = {\n",
    "    'excitatory': ['L5 IT', 'L2/3 IT', 'L4 IT', 'L6 IT', 'L6 IT Car3', 'L5/6 NP', 'L6b', 'L6 CT', 'L5 ET'],\n",
    "    'inhibitory': ['Pvalb', 'Sst', 'Lamp5 Lhx6', 'Vip', 'Lamp5', 'Sncg', 'Chandelier', 'Sst Chodl', 'Pax6'],\n",
    "    'astrocyte': ['Astrocyte'],\n",
    "    'microglia': ['Microglia-PVM'],\n",
    "    'opc': ['OPC'],\n",
    "    'oligodendrocyte': ['Oligodendrocyte'],\n",
    "    'endothelial': ['Endothelial'],\n",
    "    'vlmc': ['VLMC'],\n",
    "    }\n",
    "\n",
    "if cell_group==\"all\":\n",
    "    cell_supertypes = list(chain(*list(subclass.values()))) \n",
    "else:\n",
    "    cell_supertypes = [subclass[cell_group][cell_group_number]]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fae59c52",
   "metadata": {},
   "source": [
    "## **Load Standardized Data**\n",
    "\n",
    "We have previously standardized and filtered the `adata` obj for each cell sypertype (`00_Standardize_and_Split_RNAseq.ipynb`). We implement the comprehensive covariate standardization and categorization employed **[here](https://github.com/AllenInstitute/SEA-AD_2024/blob/main/Single%20nucleus%20omics/04_Differential%20expression%20analysis/00_Split%20AnnData%20for%20nebula.ipynb)** to ensure robust statistical comparisons. \n",
    "\n",
    "Continuous variables were normalized to a [0,1] interval using min-max scaling, including `post-mortem interval (PMI)` and `the number of genes detected per cell`. \n",
    "\n",
    "`Age at death` was discretized into five equal-width bins and subsequently normalized to account for non-linear age effects.\n",
    "\n",
    "Categorical variables underwent systematic preprocessing to ensure consistent encoding. `Sex` was binary-encoded as F/M, while `APOE4` carrier status was derived from genotype information and encoded as Y/N. Race information, specifically White identification, was encoded as Checked/Unchecked. Technical covariates, including sequencing method and donor ID, were retained with cleaned category levels. \n",
    "\n",
    "For quality control metrics, we normalized the `number of unique molecular identifiers (UMIs)` and `genes detected per cell`,to enabling direct comparisons across samples.\n",
    "\n",
    "Here, we load the standardized `h5ad` file. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "3888255a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "loading data for Pvalb...\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 76958 × 36601\n",
       "    obs: 'method', 'Subclass', 'Supertype (non-expanded)', 'Supertype', 'continuous_pseudo_progression_score', 'age_at_death_binned_codes', 'sex', 'race_choice_white', 'genes_detected', 'donor_id', 'number_of_umis', 'pmi', 'apoe4_status'\n",
       "    var: 'gene_ids'\n",
       "    uns: 'APOE4 Status_colors', 'Braak_colors', 'CERAD score_colors', 'Cognitive Status_colors', 'Great Apes Metadata', 'Highest Lewy Body Disease_colors', 'LATE_colors', 'Overall AD neuropathological Change_colors', 'Sex_colors', 'Subclass_colors', 'Supertype_colors', 'Thal_colors', 'UW Clinical Metadata', 'X_normalization', 'batch_condition', 'default_embedding', 'neighbors', 'title', 'umap'\n",
       "    obsm: 'X_scVI', 'X_umap'\n",
       "    layers: 'UMIs'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "if cell_group == \"all\":\n",
    "    print(f'loading data for all {cell_type_column} cell types...')\n",
    "    adata_annot = sc.read_h5ad(data_dir+f'counts/anndata/all_subclass_standardized_anndata.h5ad')\n",
    "    adata_annot.X = adata_annot.X.toarray()\n",
    "    adata_annot.layers['counts'] = adata_annot.layers['UMIs'].toarray()\n",
    "    del adata_annot.layers['UMIs']\n",
    "else:\n",
    "    print(f'loading data for {cell_supertypes[0]}...')\n",
    "    adata_annot = sc.read_h5ad(os.path.join(\n",
    "        data_dir,\n",
    "        'counts',\n",
    "        cell_group,\n",
    "        helper_functions.clean_strings(cell_supertypes[0], preserve_case=True),\n",
    "        f'{helper_functions.clean_strings(cell_supertypes[0], preserve_case=True)}_standardized_anndata.h5ad'\n",
    "        )\n",
    "    )\n",
    "\n",
    "adata_annot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "63debebe",
   "metadata": {},
   "source": [
    "# **Parameters for Testing Differential Pathway Activity**\n",
    "\n",
    "The primary differential pathway activity analaysis:\n",
    "\n",
    "- `covariates`: Specifies variables to account for in the analysis:\n",
    "    - Must include `Continuous Pseudo-progression Score` which will later on be matched to`pathology.group` as the primary variable of interest.\n",
    "\n",
    "    - Can include additional confounding variables like:\n",
    "        - Demographic factors (e.g., `Sex`, `Age`)\n",
    "        - Technical factors (e.g., `Sample Batch`)\n",
    "\n",
    "    - If there are no additional confounders, set `covariates = ['Continuous Pseudo-progression Score']`.\n",
    "\n",
    "    - For the best results, ensure:\n",
    "        - continuous covariates are similarly scaled.\n",
    "        - uniformity in categorical data. E.g., Avoid having 'M', 'Male', and 'male' in the same dataset.\n",
    "        - regularly assess the impact of covariates by running the analyses with and without each covariate. Watch out for covariates unduly influencing results. Watch out for covariates unduly influencing results.\n",
    "\n",
    "    - **`The available covariates in the standardized AnnData object are strictly limited to those established during the initial preprocessing step in 00_Standardize_and_Split_RNAseq.ipynb`**. To incorporate additional covariates, return to `00_Standardize_and_Split_RNAseq.ipynb` and add desired covariates during standardization then re-run the complete preprocessing pipeline.\n",
    "\n",
    "- `random_effect`: Identifies the technical variable to be treated as random effects (in this case `subject_id` or `Sample_batch`). These effects are regressed out using `duplicatecorrection`. This would help account or technical variation not relevant to the biological question.\n",
    "\n",
    "- `jack_knifing`: Boolean parameter controlling resampling analysis. When set to `True`, the script performs repeated analysis on data subsets to assess robustness of identified pathway activity patterns. This would help identify patterns that might be driven by outliers or specific samples."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "687feb23",
   "metadata": {},
   "outputs": [],
   "source": [
    "######################### Differential Pathway Activity arguments #########################\n",
    "jack_knifing = True                                                 # Boolean parameter controlling resampling analysis\n",
    "random_effect = helper_functions.clean_strings([subject_id])[0]     # Technical covariate to be included as random effect not of interest (regressed out by duplicatecorrection)\n",
    "subject_id = random_effect\n",
    "\n",
    "covariates = [\n",
    "    'Continuous Pseudo-progression Score',\n",
    "    'Age at Death binned codes',\n",
    "    'Sex',\n",
    "    'PMI',\n",
    "    'Genes detected',\n",
    "    'Number of UMIs',\n",
    "    'method',\n",
    "    'Race (choice=White)',\n",
    "    'APOE4_Status'\n",
    "]\n",
    "\n",
    "covariates = helper_functions.clean_strings(covariates, replace_hyphen=True)\n",
    "\n",
    "numeric_covariates = helper_functions.clean_strings(['Age at Death binned codes',\n",
    "                                            'Continuous Pseudo-progression Score', \n",
    "                                            'PMI',\n",
    "                                            'Genes detected',\n",
    "                                            'Number of UMIs'],\n",
    "                                            replace_hyphen=True)\n",
    "\n",
    "# confirm that all covariates and numeric covariates are in adata.obs\n",
    "\n",
    "for covariate in covariates + numeric_covariates:\n",
    "    assert covariate in adata_annot.obs.columns, f\"Missing covariate: {covariate}\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53e506d2",
   "metadata": {},
   "source": [
    "# **Count Aggregation**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "218a48f8",
   "metadata": {},
   "source": [
    "### **Count Aggregation Parameters**\n",
    "\n",
    "Since we plan on using a pseudobulking strategy to aggregate single-cell data into more robust representative samples, we specify the following parameters:\n",
    "\n",
    "- `count_agg_strategy`: Defines the pseudobulking strategy and includes options such as \n",
    "    - `standardbulk`: Uses `decoupler` to aggregate cells by summing raw counts for each gene across all cells that share the same sample ID and cell type.\n",
    "\n",
    "    - `blanchardbulk`: Implements the pseudobulking strategy From [Blanchard et al. Nature. 2022](https://doi.org/10.1038/s41586-022-05439-w)\n",
    "\n",
    "    - `network`, `smaller_network`, or `random`: A set of pseudocell strategies adopted from [Gazestani. et. al. Cell. 2023](https://www.sciencedirect.com/science/article/pii/S0092867423008590?via%3Dihub) to either randomly assign cells to pseudocells in each cell type from each individual (`random`); or create a similarity network between these cells and use that similarity network to group cells with highly similar expression patterns to each other (`network`) at varying sizes (`smaller network`).\n",
    "    \n",
    "    - `metacell`: Implements the hdWGCNA [Morabito et al. Cell Reports Methods. 2023](https://www.sciencedirect.com/science/article/pii/S2667237523001273?via%3Dihub) bootstrapped aggregation (bagging) strategy to reduce matrix sparsity. LIke [Gazestani. et. al. Cell. 2023](https://www.sciencedirect.com/science/article/pii/S0092867423008590?via%3Dihub) it uses k-nearest neighbors to identify transcriptionally similar cells in a reduced dimensional space.\n",
    "\n",
    "- `reconstruct_embedding`: Controls whether to reconstruct the embedding based on the specified `count_agg_strategy`. Cellular embedding reconstruction is performed distinctly for `network`, `smaller network`, and `metacell` aggregation strategies. For `network` and `smaller network` strategies, final embedding is extracted from top principal components, derived from the normalized and scaled data. The `metacell` strategy incorporates additional `Harmony` correction to account for batch effects across subjects.\n",
    "\n",
    "- `emedding_name`: This allows for selection of pre-computed dimensional reductions stored within the AnnData object. If `reconstruct_embedding = False`, it specifies the key corresponding to the desired embedding matrix in the `.obsm` attribute of the AnnData object. The specified key must exist within the `.obsm` dictionary and is analogous to the `reducedDimNames` nomenclature used for SIngleCellEXperiment objects.\n",
    "\n",
    "- `gene_celltype_threshold`: Determines the number of cells a gene must be expressed in.\n",
    "\n",
    "- `pathway_gene_threshold`: Determines the number of genes that must be present in a pathway.\n",
    "\n",
    "- `min_size_limit`: Specifies the minimum acceptable size of the pseudobulk data, with pseudobulks below this cell threshold excluded from the analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "34687ad6",
   "metadata": {},
   "outputs": [],
   "source": [
    "filter_genes = \"TRUE\"\n",
    "count_agg_strategy = 'metacell'   # options 'network', 'random', 'pseudobulk', 'smaller_network', 'standardbulk', 'metacell', `blanchardbulk``\n",
    "reconstruct_embedding = 'FALSE'\n",
    "embedding_name = 'X_scVI'\n",
    "gene_celltype_threshold = 0.10    # determines number of cells the gene must be expressed in\n",
    "pathway_gene_threshold = 0.33     # determines number of genes that must be present in that pathway\n",
    "min_size_limit = 30               # min_size_limit: the minimum acceptable size of the pseudobulk data.\n",
    "                                  # min_size_limit: pseudobulks with cells less than this threshold are excluded from the analysis "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a90ecc10",
   "metadata": {},
   "source": [
    "### **Send parameters to R interface with rpy2**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a85685c6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"loaded data into memory for recursive use\"\n",
      "class: SingleCellExperiment \n",
      "dim: 36601 76958 \n",
      "metadata(19): APOE4 Status_colors Braak_colors ... title umap\n",
      "assays(2): X counts\n",
      "rownames(36601): MIR1302-2HG FAM138A ... AC007325.4 AC007325.2\n",
      "rowData names(1): gene_ids\n",
      "colnames(76958): CCCTTAGCATTAAAGG-L8TX_210826_01_B06-1153814368\n",
      "  TAAGCCATCGCATTAG-L8TX_210708_01_D09-1153814279 ...\n",
      "  GAGTTACAGCGTGTCC-L8TX_210430_01_B04-1153814208\n",
      "  TACACCCAGGTTCCGC-L8TX_210429_01_D03-1142430416\n",
      "colData names(13): method Subclass ... pmi apoe4_status\n",
      "reducedDimNames(2): X_scVI UMAP\n",
      "mainExpName: NULL\n",
      "altExpNames(0):\n"
     ]
    }
   ],
   "source": [
    "%%R -i adata_annot -i subject_id -i cell_type_column -i cell_group -i gene_celltype_threshold -i cell_supertypes -i pseudobulk_level -i test_names -i save_prefix -i save_dir -i data_dir -i subject_id -i covariates -i random_effect -i min_size_limit -i count_agg_strategy -i numeric_covariates -i reconstruct_embedding -i embedding_name\n",
    "\n",
    "# adata_annot <- readH5AD(paste0(data_dir, \"anndata/\", save_prefix, \"_standardized_anndata.h5ad\"), use_hdf5=TRUE, reader='R')\n",
    "\n",
    "names(assays(adata_annot)) = c(\"X\", \"counts\")\n",
    "\n",
    "print('loaded data into memory for recursive use')\n",
    "print(adata_annot)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "64ff536c",
   "metadata": {},
   "source": [
    "## **Standard Pseudobulk Approach**\n",
    "\n",
    "To create pseudobulk data per donor and cell_type\n",
    "\n",
    "### **Pseudo-bulk With Decoupler**\n",
    "\n",
    "#### **Requied Parameters when using decoupler**\n",
    "\n",
    "\n",
    "- `min_counts`: Filter to remove samples by a minimum number of summed counts in a sample-group pair when using standard pseudobulk method\n",
    "\n",
    "- `min_count`: Minimum count requiered per gene for at least some samples.\n",
    "\n",
    "- `min_total_count`: Minimum total count required per gene across all samples.\n",
    "\n",
    "- `large_n`: Number of samples per group that is considered to be \"large\".\n",
    "    \n",
    "- `min_prop`: Minimum proportion of samples in the smallest group that express the gene."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "23402d2c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get pseudo-bulk profile\n",
    "\n",
    "min_counts = 1000       #    Filter to remove samples by a minimum number of summed counts in a sample-group pair when using standard pseudobulk method\n",
    "min_count = 10          #    Minimum count requiered per gene for at least some samples.\n",
    "min_total_count = 15    #    Minimum total count required per gene across all samples.\n",
    "large_n = 1             #    Number of samples per group that is considered to be \"large\".\n",
    "min_prop = 1            #    Minimum proportion of samples in the smallest group that express the gene.\n",
    "\n",
    "if count_agg_strategy == 'standardbulk':\n",
    "    \n",
    "    pdata = dc.get_pseudobulk(\n",
    "        adata_annot,\n",
    "        sample_col=subject_id,\n",
    "        groups_col=pseudobulk_level,\n",
    "        # layer='counts',\n",
    "        mode='sum',\n",
    "        min_cells=min_size_limit, \n",
    "        min_counts=min_counts,\n",
    "    )\n",
    "\n",
    "    # Sample-cell-type QC visualization\n",
    "    dc.plot_psbulk_samples(pdata, \n",
    "                           groupby=[subject_id, pseudobulk_level], \n",
    "                           figsize=(12, 6))\n",
    "    \n",
    "    summed_counts_per_celltype = {}\n",
    "    expressed_genes_per_celltype = {}\n",
    "\n",
    "    fig, axs = plt.subplots(ncols=5, nrows=len(pdata.obs[pseudobulk_level].unique())//5+1, figsize=(14, 8))\n",
    "    axs = axs.flatten()\n",
    "    for i, cell_type in enumerate(pdata.obs[pseudobulk_level].unique()):\n",
    "        summed_counts_per_celltype[cell_type] = pdata[pdata.obs[pseudobulk_level]==cell_type].copy()\n",
    "        \n",
    "        # gene QC visualization\n",
    "\n",
    "        dc.plot_filter_by_expr(summed_counts_per_celltype[cell_type], \n",
    "                               group='pathology.group', \n",
    "                               min_count=min_count,\n",
    "                               min_total_count=min_total_count, \n",
    "                               large_n=large_n, \n",
    "                               min_prop=min_prop, \n",
    "                               ax=axs[i],\n",
    "                               )\n",
    "        \n",
    "        axs[i].set_title(cell_type)\n",
    "\n",
    "        \n",
    "        expressed_genes_per_celltype[cell_type] = dc.filter_by_expr(summed_counts_per_celltype[cell_type],\n",
    "                                                                    group='pathology.group', \n",
    "                                                                    min_count=min_count,\n",
    "                                                                    min_total_count=min_total_count, \n",
    "                                                                    large_n=large_n, \n",
    "                                                                    min_prop=min_prop)\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "8e6539da",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "\n",
    "if count_agg_strategy == 'standardbulk':\n",
    "    summed_counts_per_celltype =  robjects.ListVector( {\n",
    "                                                        cell_type: summed_counts_per_celltype[cell_type]\n",
    "                                                    \n",
    "                                                        for cell_type in summed_counts_per_celltype.keys()\n",
    "                                                        }\n",
    "                                                    )\n",
    "\n",
    "    expressed_genes_per_celltype =  robjects.ListVector( {\n",
    "                                                        cell_type: expressed_genes_per_celltype[cell_type]\n",
    "                                                    \n",
    "                                                        for cell_type in expressed_genes_per_celltype.keys()\n",
    "                                                        }\n",
    "                                                    )\n",
    "else:\n",
    "    summed_counts_per_celltype = 'NULL'\n",
    "    expressed_genes_per_celltype = 'NULL'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "5aa5a888",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Standardbulk method not implemented.\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i summed_counts_per_celltype -i expressed_genes_per_celltype\n",
    "\n",
    "if (summed_counts_per_celltype == 'NULL'){\n",
    "    print('Standardbulk method not implemented.')\n",
    "} else {\n",
    "    print('done')\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b31d3038",
   "metadata": {},
   "source": [
    "## **Pseuobulking With Strategy From [Joel W. Blanchard et al. Nature. 2022](https://doi.org/10.1038/s41586-022-05439-w)**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "03b62009",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Blanchardbulk method not implemented.\"\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "if (count_agg_strategy == 'blanchardbulk'){\n",
    "    \n",
    "    conflicts_prefer(base::unname)\n",
    "    \n",
    "    source('../scripts/functions/qc_and_annotation_aux_functions.r')\n",
    "\n",
    "    sce <- adata_annot\n",
    "\n",
    "    print('normalizing counts')\n",
    "    sce = normalize.default(sce)\n",
    "\n",
    "    # standardize subject IDs\n",
    "    meta = colData(sce)\n",
    "    meta[, subject_id] <- gsub(\"\\\\.\", \"_\", meta[, subject_id])\n",
    "    cell_labels = rownames(meta)\n",
    "\n",
    "    # use matrix multiplication to summarize (sum) across counts per cell (including all individuals)\n",
    "    print('sum counts by individual')\n",
    "    labels = as.data.frame(as.character(interaction(meta[, pseudobulk_level], meta[, subject_id])))\n",
    "    summed_logcounts_cellxind = sum_counts(logcounts(sce), labels, cell_labels)\n",
    "    \n",
    "    # get averages corresponding to both count matrices\n",
    "    avs_logcounts_cellxind = t(apply(summed_logcounts_cellxind$summed_counts, 1, function(x){x/summed_logcounts_cellxind$ncells}))\n",
    "\n",
    "    # split the averages by celltype\n",
    "    x = (strsplit(colnames(avs_logcounts_cellxind), '[.]'))\n",
    "    celltype = unlist(lapply(1:length(x), function(i) x[[i]][[1]]))\n",
    "    individual = unlist(lapply(1:length(x), function(i) x[[i]][[2]]))\n",
    "    celltype_unique = unique(celltype)\n",
    "\n",
    "    avs_by_ind_out = list()\n",
    "    summed_counts_per_celltype = list()\n",
    "    expressed_genes_per_celltype = list()\n",
    "\n",
    "    for(i in celltype_unique){\n",
    "\n",
    "        index = celltype==i\n",
    "        df = avs_logcounts_cellxind[, index]\n",
    "        colnames(df) = individual[index]\n",
    "        avs_by_ind_out[[i]] = df\n",
    "\n",
    "        # create SingleCellExperiment object for summed counts\n",
    "        summed_counts_per_celltype[[i]] = SingleCellExperiment(assays = list(logcounts = as.matrix(df)))\n",
    "        rownames(colData(summed_counts_per_celltype[[i]])) = colnames(df)\n",
    "        \n",
    "        metadata <- meta[!duplicated(meta[, subject_id]), ]\n",
    "        colData(summed_counts_per_celltype[[i]]) <- merge(colData(summed_counts_per_celltype[[i]]), \n",
    "                                                    metadata, \n",
    "                                                    by = NULL,\n",
    "                                                    by.x = \"row.names\",\n",
    "                                                    by.y = subject_id, \n",
    "                                                    all.x = TRUE)\n",
    "\n",
    "        rownames(rowData(summed_counts_per_celltype[[i]])) <- rownames(df)  \n",
    "\n",
    "        colData(summed_counts_per_celltype[[i]])[, subject_id] = colData(summed_counts_per_celltype[[i]])$Row.names\n",
    "        \n",
    "        # in how many cells per celltype is each gene detected?\n",
    "        print('getting nonzero counts')\n",
    "\n",
    "        mask <- colData(sce)[, pseudobulk_level] == i\n",
    "        sce_temp <- sce[, mask]\n",
    "        meta_temp <- as.data.frame(colData(sce_temp)[, pseudobulk_level])\n",
    "\n",
    "        counts_nonzero = counts(sce_temp)>0\n",
    "        \n",
    "        # determine expressed genes based on threshold\n",
    "        detected_genes_cell = sum_counts(counts_nonzero, label = meta_temp, cell_labels)\n",
    "        fraction_detected_genes_cell = t(apply(detected_genes_cell$summed_counts, 1, function(x){x/detected_genes_cell$ncells}))\n",
    "\n",
    "        # get expression list\n",
    "        fraction_detected_genes_cell_binary = fraction_detected_genes_cell>gene_celltype_threshold\n",
    "        genes = rownames(fraction_detected_genes_cell)\n",
    "        expressed = lapply(colnames(fraction_detected_genes_cell_binary), function(x) genes[unname(fraction_detected_genes_cell_binary[,x])])\n",
    "        names(expressed) = colnames(fraction_detected_genes_cell_binary)\n",
    "\n",
    "        expressed_genes_per_celltype[[i]] <- names(expressed)\n",
    "            \n",
    "    }\n",
    "\n",
    "} else {\n",
    "  summed_counts_per_celltype <- NULL\n",
    "  names(summed_counts_per_celltype) <- NULL\n",
    "\n",
    "  print('Blanchardbulk method not implemented.')\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c8b11465",
   "metadata": {},
   "source": [
    "## **Pseudocell approach [Gazestani. et. al. 2023](https://www.sciencedirect.com/science/article/pii/S0092867423008590?via%3Dihub)**\n",
    "\n",
    "### **Pseudocell construction** \n",
    "\n",
    "On one hand, cell-based methods for differential expression are time-consuming and sensitive to single-cell data dropouts. On the other hand, pseudobulk methods, which aggregate cell expressions into a single sample (as with the `standardbulk` approach above), address these issues but may lead to a loss of statistical power and neglect variations in confidence levels between individuals with different cell-type sampling sizes.\n",
    "\n",
    "To address these challenges pseudocell/metacell methods have been introduced this method lies between the cell-based and pseudobulk-based approaches, so it can take advantage of each method while remedying the issues attributed to them.\n",
    "\n",
    "According to **[Gazestani. et. al. 2023](https://www.sciencedirect.com/science/article/pii/S0092867423008590?via%3Dihub)**, `we first aggregate the raw UMI count` of, on average, every 30 cells per subject and cell type. \n",
    "\n",
    "We construct one pseudocell for cell types that have between `15 to 45 cells in a donor` and exclude cell types that had less than `15 cells`. This pseudocell-based analysis reduces the impact of dropout and technical variability, while ameliorating low statistical power and high variation in sample size issues attributed to the pseudobulk approaches. \n",
    "\n",
    "**[Gazestani. et. al. 2023](https://www.sciencedirect.com/science/article/pii/S0092867423008590?via%3Dihub)**, provide functions:\n",
    "\n",
    "- `To generate pseudocells that are composed of, on average, 20 cells or higher. This function can also generate pseudobulk samples.`\n",
    "\n",
    "- `To generate pseudocells that are on average composed of 10 cells`"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5d3c5a72",
   "metadata": {},
   "source": [
    "### **Available Aggregration Methods**\n",
    "\n",
    "\n",
    "1. #### **Pseudocells from low-dim network embedding**\n",
    "\n",
    "The main arguments for running the function:\n",
    "\n",
    "- `parsing.col.names`: a vector of column names to be used to parse the data. Usually the columns related to the donor id and the cell type annotation (see above figure).\n",
    "\n",
    "- `pseudocell.size`: the average size of pseudcoells. If NULL, generates pseudobulk data.\n",
    "\n",
    "- `inputExpData`: Input single cell data object in format.\n",
    "\n",
    "- `min_size_limit`: minimum acceptable size of the pseudocells. usually 10 or 15.\n",
    "\n",
    "- `inputPhenoData`: the meta data on the cells that matches in order with the input PC space to use to create the similarity network. It's highly recommended to be used if. defualt: NULL.\n",
    "\n",
    "- `nPCs`: Number of pcs/dimensions to use to create the similarity network. Used only if human or mouse\n",
    "\n",
    "- `rand_pseudobulk_mod`: To randomly assign cells to pseudocells in each cell type from each individual; or create a similarity network between these cells and use that similarity net to group cells with highly similar expression patterns to each other. `default: TRUE, but Here we set to FALSE``\n",
    "\n",
    "The authors reiterate that it's highly recommended to provide if the function will generate the embedding for each cell type in each individual. it's possible that such embedding is driven by the quality of the cells in the group as opposed to biological variation, and hence quality of the pseudocells and inference on them would be limited.\n",
    "\n",
    "\n",
    "2. #### **Pseudocells with random cell assignment**\n",
    "\n",
    "Alternatively, we can create pseudocells using a method that randomly assign cells to pseudocells in each cell type from each individual by setting `rand_pseudobulk_mod=TRUE`. Rather than the previous method which creates a similarity network between these cells and uses that similarity network to group cells with highly similar expression patterns to each other.\n",
    "\n",
    "\n",
    "3. #### **Pseudocells with smaller size**\n",
    "\n",
    "We can also generate pseudocells using that allows us to have pseudocells of smaller size. The main arguments for this function:\n",
    "\n",
    "- `inputExpData`: Input single cell data object in SingleCellExperiment format\n",
    "\n",
    "- `embeddings`: Input PC space to use to create the similarity network. It can't be set to NULL\n",
    "\n",
    "- `pseudobulk_split_col`: The column name to parse the single cell data.\n",
    "\n",
    "- `min_dataset_size`: Minimum number of acceptable pseudocell size. default: 4.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "b144d494",
   "metadata": {},
   "outputs": [],
   "source": [
    "############## PARAMETERS FOR NETWORK AND RANDOM #####################\n",
    "\n",
    "pseudocell_size = 30    # the average size of pseudcoells. If NULL, generates pseudobulk data.\n",
    "nDims = 20              # Number of dimensions to use to create the similarity network. \n",
    "organism = 'Human'      # Used only if human or mouse\n",
    "nCores = 40 \n",
    "\n",
    "############### PARAMETERS FOR 'SMALL NETWORK' #####################\n",
    "\n",
    "pseudobulk_split_col = \"lib_anno\"\n",
    "min_dataset_size = 4 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "442d4b83",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[conflicted] Will prefer GSEABase::setdiff over any other package.\n",
      "[conflicted] Will prefer SingleCellExperiment::counts over any other package.\n",
      "Main Functions:\n",
      ".myRead10X()\n",
      ".myRead10X_h5()\n",
      ".myLigerToExpSet()\n",
      ".mycBindFn()\n",
      ".myExpSetCreatorFn()\n",
      ".myIntegrative_oneline()\n",
      ".myFindAllMarkers()\n",
      ".myAnnotateFn()\n",
      ".my2dPlot()\n",
      ".myPseudoCellfn_v2()\n",
      ".myLabelTransfer_harmony()\n",
      ".myLabelTransfer_liger()\n",
      ".myMapToHuman()\n",
      ".myRiverPlotFn()\n",
      ".myClusteringOptimizerFn()\n",
      ".myMarkerBasedAnalysisFn()\n",
      ".mycellAssignHeatmap()\n",
      ".myMetaMarkerFn()\n",
      ".myFindNeighbors()\n",
      ".myVlnPlot()\n",
      ".myFeaturePlot()\n",
      ".myheatmap.3()\n",
      ".myEvalMarkers()\n",
      ".myReadGMT()\n",
      ".mySplitObject()\n",
      ".myRTNgsea(two-sided GSEA implementation)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Loading required package: GenomicFeatures\n",
       "Loading required package: AnnotationFilter\n",
       "qs 0.27.2. Announcement: https://github.com/qsbase/qs/issues/103\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R -i organism -i nDims -i pseudocell_size -i pseudobulk_split_col -i min_dataset_size -i nCores\n",
    "\n",
    "library(scuttle)\n",
    "library(Matrix)\n",
    "library(ensembldb)\n",
    "library(EnsDb.Hsapiens.v86)\n",
    "\n",
    "conflicts_prefer(GSEABase::setdiff)\n",
    "conflicts_prefer(SingleCellExperiment::counts)\n",
    "\n",
    "source(\"../scripts/functions/sconline_code.R\")\n",
    "\n",
    "# generating the embedding space\n",
    "if (count_agg_strategy %in% c('network', 'smaller_network', 'random')){\n",
    "\n",
    "  if (as.logical(reconstruct_embedding)){\n",
    "    exp_seurat = .extraExport2SeuratFn(adata_annot) %>%\n",
    "                  Seurat::NormalizeData() %>%\n",
    "                      FindVariableFeatures() %>% \n",
    "                      ScaleData() %>% \n",
    "                      RunPCA(verbose=F)\n",
    "\n",
    "    embedding_data = exp_seurat@reductions$pca@cell.embeddings[,1:nDims]\n",
    "  } else{\n",
    "    embedding_data = reducedDim(adata_annot, embedding_name)\n",
    "  }\n",
    "\n",
    "}\n",
    "\n",
    "\n",
    "if (count_agg_strategy == 'network'){\n",
    "\n",
    "  # generating the pseudocells\n",
    "    \n",
    "  # Function adds/modifies three annotation columns: pseuodcell_size, QC_Gene_total_count, QC_Gene_unique_count\n",
    "  # QC_Gene_total_count: equivalant to nUMI for the pseudobulk samples\n",
    "  # QC_Gene_unique_count: equivalant to nGene for the pseudobulk samples\n",
    "  # use scale(tst$QC_Gene_total_count) and scale(tst$pseudocell_size) as additional covariates for the DE analysis\n",
    "\n",
    "  res = suppressWarnings(.sconline.PseudobulkGeneration(argList = NULL, \n",
    "                                  # The columns in the pheonData that will be used to parse the expression data \n",
    "                                  # and generate the pseudocell/pseudobulk data\n",
    "                                    parsing.col.names = c(subject_id, pseudobulk_level), \n",
    "                                  # average pseudocell size.\n",
    "                                    pseudocell.size = pseudocell_size,\n",
    "                                    inputExpData = adata_annot,\n",
    "                                  # minimum acceptable size (ie, #cells) for each pseudobulk\n",
    "                                    min_size_limit = min_size_limit,\n",
    "                                  # in case we want to run the function outside sconline space\n",
    "                                    inputPhenoData = as.data.frame(colData(adata_annot)),\n",
    "                                  # the embedding space to be used for the generation of the pseudobulk.\n",
    "                                  # only needed when pseudocell.size is not null\n",
    "                                    inputEmbedding = embedding_data, \n",
    "                                  # the dimension of the embedding space for the construction of pseudobulk data\n",
    "                                    nPCs = nDims, \n",
    "                                    ncores = nCores,\n",
    "                                    rand_pseudobulk_mod = F,\n",
    "                                  # used to identify and estimate Mitochondrial Genes\n",
    "                                    organism = organism))\n",
    "                                    \n",
    "} else  if (count_agg_strategy == 'random'){\n",
    "\n",
    "  res = suppressWarnings(.sconline.PseudobulkGeneration(argList=NULL, \n",
    "                                  # The columns in the pheonData that will be used to parse the expression data \n",
    "                                  # and generate the pseudocell/pseudobulk data\n",
    "                                    parsing.col.names = c(subject_id, pseudobulk_level), \n",
    "                                  # average pseudocell size.\n",
    "                                    pseudocell.size = pseudocell_size,\n",
    "                                    inputExpData = adata_annot,\n",
    "                                  # minimum acceptable size (ie, #cells) for each pseudobul\n",
    "                                    min_size_limit = min_size_limit,\n",
    "                                  # in case we want to run the function outside sconline space\n",
    "                                    inputPhenoData = as.data.frame(colData(adata_annot)),\n",
    "                                    ncores = nCores,\n",
    "                                    rand_pseudobulk_mod = T,\n",
    "                                  # used to identify and estimate Mitochondrial Genes\n",
    "                                    organism = organism))   \n",
    "                                \n",
    "}  else if (count_agg_strategy == 'pseudobulk'){\n",
    "\n",
    "  res = suppressWarnings(.sconline.PseudobulkGeneration(argList = NULL, \n",
    "                                  # The columns in the pheonData that will be used to parse the expression data \n",
    "                                  # and generate the pseudocell/pseudobulk data\n",
    "                                    parsing.col.names = c(subject_id, pseudobulk_level),  \n",
    "                                  # average pseudocell size.\n",
    "                                    pseudocell.size = NULL,\n",
    "                                    inputExpData = adata_annot,\n",
    "                                  # minimum acceptable size (ie, #cells) for each pseudobulk\n",
    "                                    min_size_limit = min_size_limit,\n",
    "                                  # in case we want to run the function outside sconline space\n",
    "                                    inputPhenoData = as.data.frame(colData(adata_annot)),\n",
    "                                  # the embedding space to be used for the generation of the pseudobulk.\n",
    "                                  # only needed when pseudocell.size is not null\n",
    "                                    inputEmbedding = NULL, \n",
    "                                  # the dimension of the embedding space for the construction of pseudobulk data\n",
    "                                    nPCs = NULL, \n",
    "                                    ncores = nCores,\n",
    "                                    rand_pseudobulk_mod = F,\n",
    "                                    organism = organism))\n",
    "\n",
    "} else  if (count_agg_strategy == 'smaller_network') {   \n",
    "\n",
    "  adata_annot$lib_anno = paste0(adata_annot[[subject_id]], \"_\", adata_annot[[pseudobulk_level]])\n",
    "\n",
    "  # creates pseudobulk samples of median size 10\n",
    "  res = .sconline.Pseudobulk10(inputExpData=adata_annot,\n",
    "                              embeddings = embedding_data,\n",
    "                              pseudobulk_split_col = \"lib_anno\",\n",
    "                              min_dataset_size = min_dataset_siz,\n",
    "                              organism = organism)   \n",
    "\n",
    "}\n",
    "\n",
    "if (count_agg_strategy %in% c('network', 'smaller_network', 'random', 'pseudobulk')){\n",
    "  res$QC_Gene_total_log=log2(res$QC_Gene_total_count)\n",
    "  res$QC_Gene_unique_log=log2(res$QC_Gene_unique_count)\n",
    "  res$QC_MT_log=log2(res$QC_MT.pct)\n",
    "  \n",
    "  summed_counts_per_celltype <- lapply(unique(colData(res)[, pseudobulk_level]), function(cell_type) res[, colData(res)[, pseudobulk_level] == cell_type,])\n",
    "  names(summed_counts_per_celltype) <- unique(colData(res)[, pseudobulk_level]) \n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe5a5bbf",
   "metadata": {},
   "source": [
    "## **Metacell approach from [Morabito et al. Cell Reports Methods. 2023](https://www.sciencedirect.com/science/article/pii/S2667237523001273?via%3Dihub)**\n",
    "\n",
    "We implemented the metacell aggregation approach from the hdWGCNA framework, originally developed to address the inherent sparsity of single-cell expression data for co-expression network construction. The method employs a bootstrapped aggregation (bagging) strategy that preserves biological context by constraining cell groupings within matched sample and cell type boundaries. The method first projects cells into a reduced dimensional space where k-nearest neighbors (KNN) analysis identifies transcriptionally similar cells. Then, using a bootstrapped sampling procedure, cells are iteratively selected and aggregated with their `k nearest neighbors`, where `k ≥ 2`. To prevent redundancy in the final metacell expression matrix, the method uses an overlap threshold (set as `max_shared`) that restricts the selection of cells whose neighbor sets substantially overlap with previously selected cells. The procedure continues until either reaching a target number of metacells or exhausting the pool of valid cells.\n",
    "\n",
    "This approach is governed by three key parameters: \n",
    "\n",
    "- `n_cells_per_metacell`: determines the number of cells per metacell\n",
    "\n",
    "- `max_shared`: controls the permitted overlap between metacells, and \n",
    "\n",
    "- `min_cells` establishes a minimum cell count threshold for group inclusion.\n",
    "\n",
    "- `gene_selection`: Required only if `pseudobulking_strategy=='metacell'`, indicating the gene selection method for setting up a Seurat object for WGCNA.\n",
    "\n",
    "- `geneSet`: Required only if `pseudobulking_strategy=='metacell'` and `gene_selction='custom'`, specifying a custom gene set.\n",
    "\n",
    "The resulting metacell expression profiles demonstrate significantly reduced sparsity while maintaining the underlying biological structure and ameliorating technical artifacts such as dropout events. Gene selection for metacell construction can be specified through either preset criteria or custom gene sets, which enables a flexible adaptation to diverse analytical contexts. The robustness of the derived metacell profiles makes them particularly suitable for downstream applications including correlation analyses, network construction, and pathway activity quantification."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "2c7a1a94",
   "metadata": {},
   "outputs": [],
   "source": [
    "n_cells_per_metacell = 30           # determines the number of cells per metacell\n",
    "max_shared = 10                     # controls the permitted overlap between metacells, and \n",
    "min_cells = 100                     # establishes a minimum cell count threshold for group inclusion. (should be greater than n_cells_per_metacells)\n",
    "gene_selection = 'fraction'         # Only required if `pseudobulking_strategy=='metacell'``. specifies the gene selection method when setting up seurat object for WGCNA. \n",
    "gene_set = ''                       # Only required if `pseudobulking_strategy=='metacell'`` and `gene_selction='custom'`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "ceaa02b7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Fetching embeddings from X_scVI\"\n",
      "[1] \"Setting up data for metacell construction\"\n",
      "[1] \"11928 WGCNA Genes\"\n",
      "[1] \"Constructing MetaCells\"\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "In addition: Warning messages:\n",
       "1: Keys should be one or more alphanumeric characters followed by an underscore, setting key from X_scVI_ to XscVI_ \n",
       "2: Layer 'data' is empty \n",
       "3: Layer 'scale.data' is empty \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R -i gene_set -i gene_selection -i n_cells_per_metacell -i max_shared -i min_cells\n",
    "\n",
    "if (count_agg_strategy == 'metacell'){\n",
    "\n",
    "  seurat_obj <- as.Seurat(adata_annot, counts = \"counts\", data = \"counts\")\n",
    "  \n",
    "  # Perform dimensionality reduction and plot\n",
    "  if (as.logical(reconstruct_embedding)){\n",
    "    \n",
    "    print('Reconstructing cell embeddings')\n",
    "    seurat_obj <- FindVariableFeatures(seurat_obj)\n",
    "    seurat_obj <- ScaleData(seurat_obj)\n",
    "    seurat_obj <- RunPCA(seurat_obj)\n",
    "    seurat_obj <- RunHarmony(seurat_obj, group.by.vars = subject_id)\n",
    "    seurat_obj <- RunUMAP(seurat_obj, reduction='harmony', n.neighbors=15, dims=1:30, min.dist=0.1)\n",
    "    reduction = 'harmony'\n",
    "    p <- DimPlot(seurat_obj, group.by = pseudobulk_level, label = TRUE) +\n",
    "        umap_theme() + ggtitle(subject_id) + NoLegend()\n",
    "\n",
    "    fig_dir = paste0(save_dir, \"/plots/UMAP/\")\n",
    "\n",
    "    if (!dir.exists(fig_dir)) {\n",
    "      dir.create(fig_dir, recursive=TRUE)\n",
    "    }\n",
    "\n",
    "    # Save plot to PDF\n",
    "    pdf(file = paste0(fig_dir, save_prefix, \"_integrated_umap.pdf\"), width = 4, height = 4, useDingbats = FALSE)\n",
    "    print(p)\n",
    "    dev.off()\n",
    "  } else {\n",
    "    print(paste0('Fetching embeddings from ', embedding_name))\n",
    "    reduction = embedding_name\n",
    "  }\n",
    "\n",
    "  print('Setting up data for metacell construction')\n",
    "\n",
    "  if (gene_selection == 'custom') {\n",
    "    seurat_obj <- SetupForWGCNA(\n",
    "      seurat_obj,\n",
    "      gene_select = \"custom\",            # the gene selection approach\n",
    "      gene_list = as.vector(gene_set),   # list of genes to be included\n",
    "      group.by = pseudobulk_level,       # grouping parameter\n",
    "      wgcna_name = 'pseudo_cell'         # the name of the hdWGCNA experiment\n",
    "    )\n",
    "  } else {\n",
    "    seurat_obj <- SetupForWGCNA(\n",
    "      seurat_obj,\n",
    "      gene_select = \"fraction\",           # the gene selection approach\n",
    "      fraction = gene_celltype_threshold, # fraction of cells for gene inclusion\n",
    "      group.by = pseudobulk_level,        # grouping parameter\n",
    "      wgcna_name = 'pseudo_cell'          # the name of the hdWGCNA experiment\n",
    "    )\n",
    "  }\n",
    "\n",
    "  print(paste0(length(GetWGCNAGenes(seurat_obj)), \" WGCNA Genes\"))\n",
    "\n",
    "  print('Constructing MetaCells')\n",
    "  \n",
    "  seurat_obj <- MetacellsByGroups(\n",
    "    seurat_obj = seurat_obj,\n",
    "    group.by = c(pseudobulk_level, subject_id),   # specify the columns in seurat_obj@meta.data to group by\n",
    "    reduction = reduction,                        # select the dimensionality reduction to perform KNN on\n",
    "    k = n_cells_per_metacell ,                    # nearest-neighbors parameter\n",
    "    max_shared = max_shared,                      # maximum number of shared cells between two metacells\n",
    "    min_cells = min_cells,                        # the minimum number of cells in a particular grouping to construct metacells\n",
    "    ident.group = pseudobulk_level,               # set the Idents of the metacell seurat object\n",
    "    wgcna_name = 'pseudo_cell',                   # the name of the hdWGCNA experiment\n",
    "    mode = \"sum\"                                  # determines how to make gene expression profiles for metacells from their constituent single cells. \n",
    "                                                  # Options are \"average\" or \"sum\".\n",
    "    )\n",
    "\n",
    "  # perform logCPM normalization of metacell expression matrix.\n",
    "  # print('Normalizing (log Counts Per Million Normalization) of MetaCells')\n",
    "  # seurat_obj <- NormalizeMetacells(\n",
    "  #   seurat_obj, \n",
    "  #   wgcna_name = 'pseudo_cell',\n",
    "  #   normalization.method = \"LogNormalize\",\n",
    "  #   scale.factor = 1e6)\n",
    "\n",
    "  ############ Merge Metadata from raw adata to metacells #################\n",
    "  sce <- as.SingleCellExperiment(GetMetacellObject(seurat_obj))\n",
    "\n",
    "  metadata_cols <- setdiff(colnames(colData(adata_annot)), pseudobulk_level)\n",
    "  metadata_raw <- colData(adata_annot)[, metadata_cols]\n",
    "\n",
    "  metadata_raw <- as.data.frame(metadata_raw)\n",
    "\n",
    "  metadata_raw <- metadata_raw[!duplicated(metadata_raw[, subject_id]), ]\n",
    "\n",
    "  metacell_coldata <- as.data.frame(colData(sce))\n",
    "  merged_coldata <- merge(metacell_coldata, \n",
    "                        metadata_raw, \n",
    "                        by.x = subject_id, \n",
    "                        by.y = subject_id, \n",
    "                        all.x = TRUE)\n",
    "\n",
    "  rownames(merged_coldata) <- rownames(colData(sce))\n",
    "  colData(sce) <- DataFrame(merged_coldata)\n",
    "\n",
    "  ####### Split SCE object into named list by cell type ###############\n",
    "  cell_types <- unique(colData(sce)[[pseudobulk_level]])\n",
    "  summed_counts_per_celltype <- list()\n",
    "  \n",
    "  for (cell_type in cell_types) {\n",
    "    cells_idx <- colData(sce)[[pseudobulk_level]] == cell_type\n",
    "    sce_subset <- sce[, cells_idx]\n",
    "    summed_counts_per_celltype[[cell_type]] <- sce_subset\n",
    "  }\n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "82548f5f",
   "metadata": {},
   "source": [
    "## **Gene Filtering and Log Normalization**\n",
    "\n",
    "Gene set variation analysis (`GSVA`) employs a kernel estimator for the cumulative density function, which necessitates input gene expression data in logarithmic scale measurements (`log-CPMs`, `log-RPKMs`, or `log-TPMs`). To prepare the `pseudo-bulked` data for this, we first `filterByExpr` function from `edgeR`. Genes were retained based on two criteria: a minimum count threshold of 0.9 and a minimum total count threshold determined as the maximum of either 2% of the sample size or 10 counts. The filtered expression data was then normalized to account for library size differences and subsequently transformed to log-CPM values using a prior count of 0.25 to manage zero counts. \n",
    "\n",
    "The resulting log-transformed values were stored in the `logcounts` assay."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "14cf99bb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[conflicted] Will prefer base::intersect over any other package.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "In addition: Warning message:\n",
       "In asMethod(object) :\n",
       "  sparse->dense coercion: allocating vector of size 2.5 GiB\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R -o expressed_genes_per_celltype -o summed_counts_per_celltype\n",
    "\n",
    "library(edgeR)\n",
    "\n",
    "if (count_agg_strategy != 'blanchardbulk'){\n",
    "    conflicts_prefer(base::intersect)\n",
    "    \n",
    "    expressed_genes_per_celltype <- list()\n",
    "\n",
    "    for (cell_type in names(summed_counts_per_celltype)) {\n",
    "\n",
    "        inputExpData <- summed_counts_per_celltype[[cell_type]]\n",
    "        # keep <- filterByExpr(counts(inputExpData),min.count = 0.9, min.total.count = max(0.02*ncol(inputExpData ), 10))\n",
    "\n",
    "        dge <- DGEList(counts=counts(inputExpData))\n",
    "        # dge <- dge[keep, ,keep.lib.sizes=FALSE]\n",
    "        dge <- calcNormFactors(dge)\n",
    "        \n",
    "        # Create logCPM object for log-transformed expression values\n",
    "        logCPM <- new(\"EList\")\n",
    "        logCPM$E <- edgeR::cpm(dge, normalized.lib.sizes = TRUE, log = TRUE, prior.count = 0.25)\n",
    "        \n",
    "        assays(summed_counts_per_celltype[[cell_type]])[[\"logcounts\"]]  <- logCPM$E\n",
    "        expressed_genes_per_celltype[[cell_type]] <- as.vector(rownames(logCPM$E))\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf632702",
   "metadata": {},
   "source": [
    "### **Save Summed Counts and Expressed Genes**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "0993854e",
   "metadata": {},
   "outputs": [],
   "source": [
    "if cell_group==\"all\":\n",
    "    cell_supertypes = list(chain(*list(subclass.values()))) \n",
    "else:\n",
    "    cell_supertypes = [subclass[cell_group][cell_group_number]]\n",
    "\n",
    "subclass = robjects.ListVector(subclass)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "9df03a70",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i data_dir -i subclass -i cell_supertypes\n",
    "\n",
    "source('../scripts/functions/helper_functions.r')\n",
    "\n",
    "for (cell_type in names(summed_counts_per_celltype)){\n",
    "  # Save summed counts\n",
    "  saveRDS(\n",
    "    summed_counts_per_celltype[[cell_type]],\n",
    "    file.path(\n",
    "      data_dir,\n",
    "      \"counts\",\n",
    "      cell_group,\n",
    "      clean_strings(cell_supertypes[1], preserve_case = TRUE),\n",
    "      paste0(clean_strings(cell_type, preserve_case = TRUE), \"_\", count_agg_strategy,\"_count_data.rds\")\n",
    "    )\n",
    "  )\n",
    "  \n",
    "  # Save expressed genes\n",
    "  saveRDS(\n",
    "    expressed_genes_per_celltype[[cell_type]],\n",
    "    file.path(\n",
    "      data_dir,\n",
    "      \"counts\",\n",
    "      cell_group,\n",
    "      clean_strings(cell_supertypes[1], preserve_case = TRUE),\n",
    "      paste0(clean_strings(cell_type, preserve_case = TRUE), \"_\", count_agg_strategy,\"_expressed_genes.rds\")\n",
    "    )\n",
    "  )\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "e7c9cafc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "$Pvalb\n",
      "class: SingleCellExperiment \n",
      "dim: 36601 9175 \n",
      "metadata(0):\n",
      "assays(2): counts logcounts\n",
      "rownames(36601): MIR1302-2HG FAM138A ... AC007325.4 AC007325.2\n",
      "rowData names(0):\n",
      "colnames(9175): Pvalb#H19.33.004_1 Pvalb#H19.33.004_2 ...\n",
      "  Pvalb#H21.33.047_130 Pvalb#H21.33.047_131\n",
      "colData names(18): donor_id orig.ident ... pmi apoe4_status\n",
      "reducedDimNames(0):\n",
      "mainExpName: originalexp\n",
      "altExpNames(0):\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "The assay named 'counts' will be used to determine the proportion of zeroes for each batch\n",
       "In addition: Warning message:\n",
       "In asMethod(object) :\n",
       "  sparse->dense coercion: allocating vector of size 2.5 GiB\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R \n",
    "\n",
    "library(scMerge)\n",
    "source('../scripts/functions/helper_functions.r')\n",
    "\n",
    "# Convert assays to matrix format\n",
    "for(i in seq_along(summed_counts_per_celltype)) {\n",
    "  # Convert counts\n",
    "  if(\"counts\" %in% assayNames(summed_counts_per_celltype[[i]])) {\n",
    "    assay(summed_counts_per_celltype[[i]], \"counts\") <- \n",
    "      as.matrix(assay(summed_counts_per_celltype[[i]], \"counts\"))\n",
    "  }\n",
    "  \n",
    "  # Convert logcounts\n",
    "  if(\"logcounts\" %in% assayNames(summed_counts_per_celltype[[i]])) {\n",
    "    assay(summed_counts_per_celltype[[i]], \"logcounts\") <- \n",
    "      as.matrix(assay(summed_counts_per_celltype[[i]], \"logcounts\"))\n",
    "  }\n",
    "}\n",
    "\n",
    "# merge supertypes into the main subclass/cell_type and save\n",
    "all_expressed_genes <- unique(unlist(expressed_genes_per_celltype))\n",
    "all_summed_counts <- sce_cbind(\n",
    "  summed_counts_per_celltype,\n",
    "  method = \"union\",\n",
    "  cut_off_batch = 0.00,\n",
    "  cut_off_overall = 0.00,\n",
    "  exprs = c(\"counts\", \"logcounts\"),\n",
    "  colData_names = colnames(colData(summed_counts_per_celltype[[1]])),\n",
    "  batch_names = NULL\n",
    "  )\n",
    "\n",
    "saveRDS(\n",
    "  all_summed_counts,\n",
    "  file.path(\n",
    "    data_dir,\n",
    "    \"counts\",\n",
    "    cell_group,\n",
    "    clean_strings(cell_supertypes[1], preserve_case = TRUE, replace_hyphen=FALSE),\n",
    "    paste0(clean_strings(cell_supertypes[1], preserve_case = TRUE, replace_hyphen=FALSE), \"_\", count_agg_strategy,\"_count_data.rds\")\n",
    "  )\n",
    ")\n",
    "\n",
    "saveRDS(\n",
    "  all_expressed_genes,\n",
    "  file.path(\n",
    "    data_dir,\n",
    "    \"counts\",\n",
    "    cell_group,\n",
    "    clean_strings(cell_supertypes[1], preserve_case = TRUE, replace_hyphen=FALSE),\n",
    "    paste0(clean_strings(cell_supertypes[1], preserve_case = TRUE, replace_hyphen=FALSE), \"_\", count_agg_strategy, \"_expressed_genes.rds\")\n",
    "  )\n",
    ")\n",
    "\n",
    "summed_counts_per_celltype\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scRNA_seq_meta_analysis",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "vscode": {
   "interpreter": {
    "hash": "2f414a000012e1980f5fd1119ed6b9bf8cc73b51e37b52eb52a16ff857effd1c"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
