{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1556cc30",
   "metadata": {},
   "outputs": [],
   "source": [
    "import warnings\n",
    "from warnings import simplefilter\n",
    "warnings.filterwarnings(\"ignore\", category=UserWarning)\n",
    "warnings.filterwarnings('ignore', category=FutureWarning)\n",
    "warnings.filterwarnings(\"ignore\", category=PendingDeprecationWarning)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "62042f97-a76e-4b34-805f-3f1ac81e7288",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import rpy2\n",
    "import logging\n",
    "import warnings\n",
    "import anndata2ri\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import anndata as ad\n",
    "import decoupler as dc\n",
    "import matplotlib.pyplot as plt\n",
    "from gsva_prep import prep_gsva\n",
    "import rpy2.robjects as robjects\n",
    "from itertools import chain\n",
    "from rpy2.robjects import pandas2ri\n",
    "from matplotlib.pyplot import rcParams\n",
    "from functions import helper_functions\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "447b6d1c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-----\n",
      "anndata     0.11.1\n",
      "scanpy      1.9.3\n",
      "-----\n",
      "PIL                         9.5.0\n",
      "anndata2ri                  1.1\n",
      "asttokens                   NA\n",
      "backcall                    0.2.0\n",
      "cffi                        1.15.1\n",
      "comm                        0.1.3\n",
      "cycler                      0.10.0\n",
      "cython_runtime              NA\n",
      "dateutil                    2.8.2\n",
      "debugpy                     1.6.7\n",
      "decorator                   5.1.1\n",
      "decoupler                   1.4.0\n",
      "dot_parser                  NA\n",
      "exceptiongroup              1.1.1\n",
      "executing                   1.2.0\n",
      "functions                   NA\n",
      "google                      NA\n",
      "gsva_prep                   NA\n",
      "h5py                        3.9.0\n",
      "igraph                      0.10.4\n",
      "ipykernel                   6.23.2\n",
      "ipywidgets                  8.0.6\n",
      "jedi                        0.18.2\n",
      "jinja2                      3.1.2\n",
      "joblib                      1.2.0\n",
      "kiwisolver                  1.4.4\n",
      "leidenalg                   0.9.1\n",
      "llvmlite                    0.39.1\n",
      "louvain                     0.8.0\n",
      "markupsafe                  2.1.3\n",
      "matplotlib                  3.7.1\n",
      "mpl_toolkits                NA\n",
      "mpmath                      1.3.0\n",
      "natsort                     8.4.0\n",
      "numba                       0.56.4\n",
      "numexpr                     2.8.4\n",
      "numpy                       1.23.5\n",
      "nvfuser                     NA\n",
      "opt_einsum                  v3.3.0\n",
      "packaging                   23.1\n",
      "pandas                      2.0.2\n",
      "parso                       0.8.3\n",
      "patsy                       0.5.3\n",
      "pexpect                     4.8.0\n",
      "pickleshare                 0.7.5\n",
      "pkg_resources               NA\n",
      "platformdirs                3.7.0\n",
      "plotly                      5.24.1\n",
      "prompt_toolkit              3.0.38\n",
      "psutil                      5.9.5\n",
      "ptyprocess                  0.7.0\n",
      "pure_eval                   0.2.2\n",
      "pyarrow                     12.0.1\n",
      "pycparser                   2.21\n",
      "pydev_ipython               NA\n",
      "pydevconsole                NA\n",
      "pydevd                      2.9.5\n",
      "pydevd_file_utils           NA\n",
      "pydevd_plugins              NA\n",
      "pydevd_tracing              NA\n",
      "pydot                       1.4.2\n",
      "pygments                    2.15.1\n",
      "pyparsing                   3.1.0\n",
      "pytz                        2023.3\n",
      "pytz_deprecation_shim       NA\n",
      "rpy2                        3.5.12\n",
      "scipy                       1.10.1\n",
      "seaborn                     0.12.2\n",
      "session_info                1.0.0\n",
      "setuptools                  68.0.0\n",
      "six                         1.16.0\n",
      "sklearn                     1.2.2\n",
      "sparse                      0.14.0\n",
      "stack_data                  0.6.2\n",
      "statsmodels                 0.14.0\n",
      "sympy                       1.12\n",
      "texttable                   1.6.7\n",
      "threadpoolctl               3.1.0\n",
      "torch                       2.0.1+cu117\n",
      "tornado                     6.3.2\n",
      "tqdm                        4.65.0\n",
      "traitlets                   5.9.0\n",
      "typing_extensions           NA\n",
      "tzlocal                     NA\n",
      "vscode                      NA\n",
      "wcwidth                     0.2.6\n",
      "yaml                        6.0\n",
      "zmq                         25.1.0\n",
      "zoneinfo                    NA\n",
      "-----\n",
      "IPython             8.14.0\n",
      "jupyter_client      8.2.0\n",
      "jupyter_core        5.3.1\n",
      "-----\n",
      "Python 3.10.6 (main, Oct 24 2022, 16:07:47) [GCC 11.2.0]\n",
      "Linux-6.8.0-49-generic-x86_64-with-glibc2.35\n",
      "-----\n",
      "Session information updated at 2025-01-17 22:12\n"
     ]
    }
   ],
   "source": [
    "# # Ignore R warning messages\n",
    "#Note: this can be commented out to get more verbose R output\n",
    "rpy2.rinterface_lib.callbacks.logger.setLevel(logging.ERROR)\n",
    "\n",
    "# Automatically convert rpy2 outputs to pandas dataframes\n",
    "pandas2ri.activate()\n",
    "anndata2ri.activate()\n",
    "%load_ext rpy2.ipython\n",
    "\n",
    "sc.settings.verbosity = 3\n",
    "#sc.set_figure_params(dpi=200, dpi_save=300)\n",
    "sc.logging.print_versions()\n",
    "\n",
    "simplefilter(action=\"ignore\", category=pd.errors.PerformanceWarning)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "048c2e09",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "    WARNING: The R package \"reticulate\" only fixed recently\n",
      "    an issue that caused a segfault when used with rpy2:\n",
      "    https://github.com/rstudio/reticulate/pull/1188\n",
      "    Make sure that you use a version of that package that includes\n",
      "    the fix.\n",
      "    Allowing parallel execution with up to 40 working processes.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "In addition: Warning messages:\n",
       "1: replacing previous import ‘GenomicRanges::intersect’ by ‘SeuratObject::intersect’ when loading ‘hdWGCNA’ \n",
       "2: replacing previous import ‘GenomicRanges::union’ by ‘dplyr::union’ when loading ‘hdWGCNA’ \n",
       "3: replacing previous import ‘GenomicRanges::setdiff’ by ‘dplyr::setdiff’ when loading ‘hdWGCNA’ \n",
       "4: replacing previous import ‘dplyr::as_data_frame’ by ‘igraph::as_data_frame’ when loading ‘hdWGCNA’ \n",
       "5: replacing previous import ‘Seurat::components’ by ‘igraph::components’ when loading ‘hdWGCNA’ \n",
       "6: replacing previous import ‘dplyr::groups’ by ‘igraph::groups’ when loading ‘hdWGCNA’ \n",
       "7: replacing previous import ‘dplyr::union’ by ‘igraph::union’ when loading ‘hdWGCNA’ \n",
       "8: replacing previous import ‘GenomicRanges::subtract’ by ‘magrittr::subtract’ when loading ‘hdWGCNA’ \n",
       "9: replacing previous import ‘Matrix::as.matrix’ by ‘proxy::as.matrix’ when loading ‘hdWGCNA’ \n",
       "10: replacing previous import ‘igraph::groups’ by ‘tidygraph::groups’ when loading ‘hdWGCNA’ \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "suppressPackageStartupMessages({\n",
    "    library(Matrix)\n",
    "    library(viridis)\n",
    "    library(harmony)\n",
    "    library(ggpubr)\n",
    "    library(tictoc)\n",
    "    library(RColorBrewer)\n",
    "    library(Hmisc)\n",
    "    library(corrplot)\n",
    "    library(grid)\n",
    "    library(gridExtra)\n",
    "    library(igraph)\n",
    "    library(ggrepel)\n",
    "    library(readxl)\n",
    "    library(conflicted)\n",
    "    library(dplyr)\n",
    "    library(parallel)\n",
    "    library(stringr)\n",
    "    library(tibble)\n",
    "    library(BiocParallel)\n",
    "\n",
    "    # single-cell analysis package\n",
    "    library(Seurat)\n",
    "    library(zellkonverter)   \n",
    "    library(SingleCellExperiment)\n",
    "    library(tidyr)\n",
    "    library(readxl)\n",
    "    library(GSA)\n",
    "    library(limma)\n",
    "\n",
    "    # plotting and data science packages\n",
    "    library(tidyverse)\n",
    "    library(cowplot)\n",
    "    library(patchwork)\n",
    "    library(ggplot2)\n",
    "\n",
    "    # co-expression network analysis packages:\n",
    "    library(WGCNA)\n",
    "    library(hdWGCNA)\n",
    "\n",
    "    # gene enrichment packages\n",
    "    library(enrichR)\n",
    "    library(GeneOverlap)\n",
    "    library(GSEABase)\n",
    "    library(GSVA) \n",
    "\n",
    "    # cell-cell communication\n",
    "    library(nichenetr)\n",
    "\n",
    "# needs to be run every time you start R and want to use %>%\n",
    "})\n",
    "\n",
    "# using the cowplot theme for ggplot\n",
    "theme_set(theme_cowplot())\n",
    "\n",
    "# set random seed for reproducibility\n",
    "set.seed(12345)\n",
    "\n",
    "# optionally enable multithreading\n",
    "enableWGCNAThreads(nThreads = 40)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a070e8e",
   "metadata": {},
   "source": [
    "## **Data Prep Parameters**\n",
    "\n",
    "- `test_names`: List of the different test names of interest.\n",
    "\n",
    "- `save_prefix`: Preferred prefix for saving critical files. Ideally chosen to be in the format `{source name}_{brain region}`. e.g `mathys_pfc`\n",
    "\n",
    "- `subject_id`: Column name for Subject/Patient ID in both metadata and `.obs`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "53693e9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "save_prefix = 'seaad_mtg'                                       # this takes the format '{StudyName}_{ThreeLetterAccronymForBrainRegion}'\n",
    "subject_id = helper_functions.clean_strings('Donor ID')\n",
    "cell_type_column = 'Subclass'                                   # 'Supertype (non-expanded)', 'Subclass'\n",
    "factor = 'Continuous Pseudo-progression Score'    # pathology.group # Continuous Pseudoprogression Score\n",
    "factor = helper_functions.clean_strings(factor, replace_hyphen=True)\n",
    "test_names = ['late_vs_early']                                  # test categories\n",
    "region_name = save_prefix.split('_')[-1].upper()\n",
    "count_agg_strategy = 'metacell'   # options 'network', 'random', 'pseudobulk', 'smaller_network', 'standardbulk', 'metacell', `blanchardbulk``\n",
    "\n",
    "\n",
    "data_dir = f'/media/tadeoye/Volume1/data/seq/SEA-AD/{region_name}/RNAseq/'\n",
    "save_dir = f'../results/'\n",
    "\n",
    "if not os.path.exists(save_dir):\n",
    "    os.makedirs(save_dir)\n",
    "\n",
    "subclass = {\n",
    "    'excitatory': ['L5 IT', 'L2/3 IT', 'L4 IT', 'L6 IT', 'L6 IT Car3', 'L5/6 NP', 'L6b', 'L6 CT', 'L5 ET'],\n",
    "    'inhibitory': ['Pvalb', 'Sst', 'Lamp5 Lhx6', 'Vip', 'Lamp5', 'Sncg', 'Chandelier', 'Sst Chodl', 'Pax6'],\n",
    "    'astrocyte': ['Astrocyte'],\n",
    "    'microglia': ['Microglia-PVM'],\n",
    "    'opc': ['OPC'],\n",
    "    'oligodendrocyte': ['Oligodendrocyte'],\n",
    "    'endothelial': ['Endothelial'],\n",
    "    'vlmc': ['VLMC'],\n",
    "    }\n",
    "\n",
    "cell_supertypes = list(chain(*list(subclass.values())))\n",
    "Subclass = robjects.ListVector(subclass)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fae59c52",
   "metadata": {},
   "source": [
    "# **Load and Merge Pseudo-bulked Data**\n",
    "\n",
    "We have previously standardized and pseudo-bulked data for each cell sypertype (`00_Standardize_and_Split_RNAseq.ipynb`, `01_pseuobulking.ipynb`). First, we load the data for each cell supertype and merge them into a single seurat object. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "161b7f91",
   "metadata": {},
   "outputs": [],
   "source": [
    "clean_subclass = robjects.ListVector({k: [helper_functions.clean_strings(v, preserve_case=True) for v in val] for k, val in subclass.items()})\n",
    "clean_cell_supertypes = [helper_functions.clean_strings(cell_type, preserve_case=True) for cell_type in cell_supertypes]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b9789076",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "In addition: Warning messages:\n",
       "1: In gzfile(file, \"rb\") :\n",
       "  cannot open compressed file '/media/tadeoye/Volume1/data/seq/SEA-AD/MTG/RNAseq//counts/excitatory/L5_ET/L5_ET_metacell_count_data.rds', probable reason 'No such file or directory'\n",
       "2: In value[[3L]](cond) :\n",
       "  PSEUDOBULKED EXPRESSION DATA NOT FOUND FOR 'L5 ET'. PLEASE CONFIRM THAT THIS IS INTENTIONAL. SKIPPING L5 ET\n",
       "3: In gzfile(file, \"rb\") :\n",
       "  cannot open compressed file '/media/tadeoye/Volume1/data/seq/SEA-AD/MTG/RNAseq//counts/inhibitory/Sst_Chodl/Sst_Chodl_metacell_count_data.rds', probable reason 'No such file or directory'\n",
       "4: In value[[3L]](cond) :\n",
       "  PSEUDOBULKED EXPRESSION DATA NOT FOUND FOR 'Sst Chodl'. PLEASE CONFIRM THAT THIS IS INTENTIONAL. SKIPPING Sst Chodl\n",
       "5: In gzfile(file, \"rb\") :\n",
       "  cannot open compressed file '/media/tadeoye/Volume1/data/seq/SEA-AD/MTG/RNAseq//counts/microglia/Microglia_PVM/Microglia_PVM_metacell_count_data.rds', probable reason 'No such file or directory'\n",
       "6: In value[[3L]](cond) :\n",
       "  PSEUDOBULKED EXPRESSION DATA NOT FOUND FOR 'Microglia-PVM'. PLEASE CONFIRM THAT THIS IS INTENTIONAL. SKIPPING Microglia-PVM\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R -i count_agg_strategy -i data_dir -i subject_id -i Subclass -i cell_supertypes -i clean_cell_supertypes -i clean_subclass -o summed_counts_per_celltype\n",
    "\n",
    "source('../scripts/functions/helper_functions.r')\n",
    "\n",
    "summed_counts_per_celltype <- list()\n",
    "expressed_genes_per_celltype <- list()\n",
    "subclass <- Subclass\n",
    "\n",
    "for (cell_type in seq_along(cell_supertypes)) {\n",
    "  tryCatch({\n",
    "    # Read summed counts and expressed genes\n",
    "    summed_counts_per_celltype[[cell_supertypes[cell_type]]] <- readRDS(\n",
    "    file.path(\n",
    "      data_dir,\n",
    "      \"counts\",\n",
    "      find_parent_key(cell_supertypes[cell_type], subclass),\n",
    "      clean_strings(cell_supertypes[cell_type], preserve_case = TRUE),\n",
    "      paste0(clean_strings(cell_supertypes[cell_type], preserve_case = TRUE), \"_\", count_agg_strategy,\"_count_data.rds\")\n",
    "    )\n",
    "  )\n",
    "  \n",
    "  # Save expressed genes\n",
    "  expressed_genes_per_celltype[[cell_supertypes[cell_type]]] <- readRDS(\n",
    "    file.path(\n",
    "      data_dir,\n",
    "      \"counts\", \n",
    "      find_parent_key(cell_supertypes[cell_type], subclass),\n",
    "      clean_strings(cell_supertypes[cell_type], preserve_case = TRUE),\n",
    "      paste0(clean_strings(cell_supertypes[cell_type], preserve_case = TRUE), \"_\", count_agg_strategy,\"_expressed_genes.rds\")\n",
    "    )\n",
    "  )\n",
    "  },\n",
    "  error = function(e) {\n",
    "    warning(sprintf(\"PSEUDOBULKED EXPRESSION DATA NOT FOUND FOR '%s'. PLEASE CONFIRM THAT THIS IS INTENTIONAL. SKIPPING %s\", \n",
    "                   cell_supertypes[cell_type], cell_supertypes[cell_type]))\n",
    "  })\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "7ad3ee69",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_annot = sc.read_h5ad(os.path.join(data_dir, f'counts/anndata/{save_prefix.upper()}_RNAseq_final-nuclei.2024-02-13.h5ad'))\n",
    "adata_annot.X = adata_annot.layers['UMIs'] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d6f545e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "for cell_type in summed_counts_per_celltype.keys():\n",
    "        \n",
    "        pseudo_genes_detected = helper_functions.update_pseudo_numeric_covariates(\n",
    "            summed_counts_per_celltype[cell_type], \n",
    "            adata_annot = adata_annot, \n",
    "            covariate='Genes detected', \n",
    "            agg_method = 'count'\n",
    "            )\n",
    "\n",
    "        pseudo_number_of_umis = helper_functions.update_pseudo_numeric_covariates(\n",
    "            summed_counts_per_celltype[cell_type], \n",
    "            adata_annot = adata_annot, \n",
    "            covariate='Number of UMIs', \n",
    "            agg_method = 'sum'\n",
    "            )\n",
    "        \n",
    "        summed_counts_per_celltype[cell_type].obs['pseudo_genes_detected'] = (pseudo_genes_detected - pseudo_genes_detected.min())/\\\n",
    "                                                            (pseudo_genes_detected.max() - pseudo_genes_detected.min())\n",
    "        summed_counts_per_celltype[cell_type].obs['pseudo_number_of_umis'] = (pseudo_number_of_umis - pseudo_number_of_umis.min())/\\\n",
    "                                                            (pseudo_number_of_umis.max() - pseudo_number_of_umis.min())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ead7e571",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_merged = ad.concat([summed_counts_per_celltype[key] for key in summed_counts_per_celltype.keys()], join='outer')\n",
    "try:\n",
    "    adata_merged.X = adata_merged.X.toarray()\n",
    "except:\n",
    "    pass\n",
    "\n",
    "mixed_columns = []\n",
    "for col in adata_merged.obs.columns:\n",
    "    if adata_merged.obs[col].dtype == 'object':\n",
    "        mixed_columns.append(col)\n",
    "\n",
    "for col in mixed_columns:\n",
    "    try:\n",
    "        adata_merged.obs[col] = pd.to_numeric(adata_merged.obs[col])\n",
    "    except:\n",
    "        adata_merged.obs[col] = adata_merged.obs[col].astype(str)\n",
    "        \n",
    "adata_merged.obs[cell_type_column] = adata_merged.obs[cell_type_column].astype('category')    \n",
    "\n",
    "# adata_merged.X = csr_matrix(adata_merged.X)\n",
    "\n",
    "# for layer in adata_merged.layers:\n",
    "#     adata_merged.layers[layer] = csr_matrix(adata_merged.layers[layer])\n",
    "    \n",
    "adata_merged.write_h5ad(data_dir+f'counts/anndata/all_subclass_{count_agg_strategy}_anndata.h5ad', compression='gzip')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "208d677e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 111258 × 36601\n",
       "    obs: 'donor_id', 'orig.ident', 'nCount_originalexp', 'nFeature_originalexp', 'cells_merged', 'Supertype', 'ident', 'method', 'Subclass', 'Supertype..non.expanded.', 'continuous_pseudo_progression_score', 'age_at_death_binned_codes', 'sex', 'race_choice_white', 'genes_detected', 'number_of_umis', 'pmi', 'apoe4_status', 'batch', 'pseudo_genes_detected', 'pseudo_number_of_umis'\n",
       "    layers: 'logcounts'"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_merged"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scRNA_seq_meta_analysis",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
