{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "62042f97-a76e-4b34-805f-3f1ac81e7288",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import rpy2\n",
    "import logging\n",
    "import warnings\n",
    "import anndata2ri\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import anndata as ad\n",
    "import decoupler as dc\n",
    "import matplotlib.pyplot as plt\n",
    "from gsva_prep import prep_gsva\n",
    "import rpy2.robjects as robjects\n",
    "from itertools import chain\n",
    "from rpy2.robjects import pandas2ri\n",
    "from matplotlib.pyplot import rcParams\n",
    "from functions import helper_functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "447b6d1c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-----\n",
      "anndata     0.8.0\n",
      "scanpy      1.9.3\n",
      "-----\n",
      "PIL                         9.5.0\n",
      "anndata2ri                  1.1\n",
      "asttokens                   NA\n",
      "backcall                    0.2.0\n",
      "cffi                        1.15.1\n",
      "comm                        0.1.3\n",
      "cycler                      0.10.0\n",
      "cython_runtime              NA\n",
      "dateutil                    2.8.2\n",
      "debugpy                     1.6.7\n",
      "decorator                   5.1.1\n",
      "decoupler                   1.4.0\n",
      "executing                   1.2.0\n",
      "functions                   NA\n",
      "google                      NA\n",
      "gsva_prep                   NA\n",
      "h5py                        3.9.0\n",
      "igraph                      0.10.4\n",
      "ipykernel                   6.23.2\n",
      "ipywidgets                  8.0.6\n",
      "jedi                        0.18.2\n",
      "jinja2                      3.1.2\n",
      "joblib                      1.2.0\n",
      "kiwisolver                  1.4.4\n",
      "leidenalg                   0.9.1\n",
      "llvmlite                    0.39.1\n",
      "louvain                     0.8.0\n",
      "markupsafe                  2.1.3\n",
      "matplotlib                  3.7.1\n",
      "mpl_toolkits                NA\n",
      "natsort                     8.4.0\n",
      "numba                       0.56.4\n",
      "numexpr                     2.8.4\n",
      "numpy                       1.23.5\n",
      "packaging                   23.1\n",
      "pandas                      2.0.2\n",
      "parso                       0.8.3\n",
      "patsy                       0.5.3\n",
      "pexpect                     4.8.0\n",
      "pickleshare                 0.7.5\n",
      "pkg_resources               NA\n",
      "platformdirs                3.7.0\n",
      "prompt_toolkit              3.0.38\n",
      "psutil                      5.9.5\n",
      "ptyprocess                  0.7.0\n",
      "pure_eval                   0.2.2\n",
      "pyarrow                     12.0.1\n",
      "pycparser                   2.21\n",
      "pydev_ipython               NA\n",
      "pydevconsole                NA\n",
      "pydevd                      2.9.5\n",
      "pydevd_file_utils           NA\n",
      "pydevd_plugins              NA\n",
      "pydevd_tracing              NA\n",
      "pygments                    2.15.1\n",
      "pyparsing                   3.1.0\n",
      "pytz                        2023.3\n",
      "pytz_deprecation_shim       NA\n",
      "rpy2                        3.5.12\n",
      "scipy                       1.10.1\n",
      "seaborn                     0.12.2\n",
      "session_info                1.0.0\n",
      "setuptools                  68.0.0\n",
      "six                         1.16.0\n",
      "sklearn                     1.2.2\n",
      "sparse                      0.14.0\n",
      "stack_data                  0.6.2\n",
      "statsmodels                 0.14.0\n",
      "texttable                   1.6.7\n",
      "threadpoolctl               3.1.0\n",
      "tornado                     6.3.2\n",
      "tqdm                        4.65.0\n",
      "traitlets                   5.9.0\n",
      "typing_extensions           NA\n",
      "tzlocal                     NA\n",
      "vscode                      NA\n",
      "wcwidth                     0.2.6\n",
      "yaml                        6.0\n",
      "zmq                         25.1.0\n",
      "zoneinfo                    NA\n",
      "-----\n",
      "IPython             8.14.0\n",
      "jupyter_client      8.2.0\n",
      "jupyter_core        5.3.1\n",
      "-----\n",
      "Python 3.10.6 (main, Oct 24 2022, 16:07:47) [GCC 11.2.0]\n",
      "Linux-6.8.0-48-generic-x86_64-with-glibc2.35\n",
      "-----\n",
      "Session information updated at 2024-11-21 12:38\n"
     ]
    }
   ],
   "source": [
    "# # Ignore R warning messages\n",
    "#Note: this can be commented out to get more verbose R output\n",
    "rpy2.rinterface_lib.callbacks.logger.setLevel(logging.ERROR)\n",
    "\n",
    "warnings.filterwarnings(\"ignore\", category=PendingDeprecationWarning)\n",
    "warnings.filterwarnings(\"ignore\", category=UserWarning)\n",
    "warnings.filterwarnings(\"ignore\", category=FutureWarning)\n",
    "\n",
    "# Automatically convert rpy2 outputs to pandas dataframes\n",
    "pandas2ri.activate()\n",
    "anndata2ri.activate()\n",
    "%load_ext rpy2.ipython\n",
    "\n",
    "sc.settings.verbosity = 3\n",
    "#sc.set_figure_params(dpi=200, dpi_save=300)\n",
    "sc.logging.print_versions()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "048c2e09",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "    WARNING: The R package \"reticulate\" only fixed recently\n",
      "    an issue that caused a segfault when used with rpy2:\n",
      "    https://github.com/rstudio/reticulate/pull/1188\n",
      "    Make sure that you use a version of that package that includes\n",
      "    the fix.\n",
      "    Allowing parallel execution with up to 40 working processes.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "In addition: Warning messages:\n",
       "1: replacing previous import ‘GenomicRanges::intersect’ by ‘SeuratObject::intersect’ when loading ‘hdWGCNA’ \n",
       "2: replacing previous import ‘GenomicRanges::union’ by ‘dplyr::union’ when loading ‘hdWGCNA’ \n",
       "3: replacing previous import ‘GenomicRanges::setdiff’ by ‘dplyr::setdiff’ when loading ‘hdWGCNA’ \n",
       "4: replacing previous import ‘dplyr::as_data_frame’ by ‘igraph::as_data_frame’ when loading ‘hdWGCNA’ \n",
       "5: replacing previous import ‘Seurat::components’ by ‘igraph::components’ when loading ‘hdWGCNA’ \n",
       "6: replacing previous import ‘dplyr::groups’ by ‘igraph::groups’ when loading ‘hdWGCNA’ \n",
       "7: replacing previous import ‘dplyr::union’ by ‘igraph::union’ when loading ‘hdWGCNA’ \n",
       "8: replacing previous import ‘GenomicRanges::subtract’ by ‘magrittr::subtract’ when loading ‘hdWGCNA’ \n",
       "9: replacing previous import ‘Matrix::as.matrix’ by ‘proxy::as.matrix’ when loading ‘hdWGCNA’ \n",
       "10: replacing previous import ‘igraph::groups’ by ‘tidygraph::groups’ when loading ‘hdWGCNA’ \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "suppressPackageStartupMessages({\n",
    "    library(Matrix)\n",
    "    library(viridis)\n",
    "    library(harmony)\n",
    "    library(ggpubr)\n",
    "    library(tictoc)\n",
    "    library(RColorBrewer)\n",
    "    library(Hmisc)\n",
    "    library(corrplot)\n",
    "    library(grid)\n",
    "    library(gridExtra)\n",
    "    library(igraph)\n",
    "    library(ggrepel)\n",
    "    library(readxl)\n",
    "    library(conflicted)\n",
    "    library(dplyr)\n",
    "    library(parallel)\n",
    "    library(stringr)\n",
    "    library(BiocParallel)\n",
    "\n",
    "    # single-cell analysis package\n",
    "    library(Seurat)\n",
    "    library(zellkonverter)   \n",
    "    library(SingleCellExperiment)\n",
    "    library(tidyr)\n",
    "    library(readxl)\n",
    "    library(GSA)\n",
    "    library(limma)\n",
    "\n",
    "    # plotting and data science packages\n",
    "    library(tidyverse)\n",
    "    library(cowplot)\n",
    "    library(patchwork)\n",
    "    library(ggplot2)\n",
    "\n",
    "    # co-expression network analysis packages:\n",
    "    library(WGCNA)\n",
    "    library(hdWGCNA)\n",
    "\n",
    "    # gene enrichment packages\n",
    "    library(enrichR)\n",
    "    library(GeneOverlap)\n",
    "    library(GSEABase)\n",
    "    library(GSVA) \n",
    "\n",
    "    # cell-cell communication\n",
    "    library(nichenetr)\n",
    "\n",
    "# needs to be run every time you start R and want to use %>%\n",
    "})\n",
    "\n",
    "# using the cowplot theme for ggplot\n",
    "theme_set(theme_cowplot())\n",
    "\n",
    "# set random seed for reproducibility\n",
    "set.seed(12345)\n",
    "\n",
    "# optionally enable multithreading\n",
    "enableWGCNAThreads(nThreads = 40)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f21b7795",
   "metadata": {},
   "source": [
    "# **Systematic differential analysis of pathway activity**\n",
    "\n",
    "Pathway activity scores were computed in accordance with the overall protocol outlined in [**Joel W. Blanchard et. al.**](https://doi.org/10.1038/s41586-022-05439-w).\n",
    "\n",
    "- Aggregate gene expression profiles into `pseudo-cells` or `pseudo-bulks` depending on the desired strategy in **[Gazestani. et. al. 2023](https://www.sciencedirect.com/science/article/pii/S0092867423008590?via%3Dihub)** (`pseudo-cells`), [**Joel W. Blanchard et. al.**](https://doi.org/10.1038/s41586-022-05439-w) (`pseudo-bulks`), or [**Morabito et al.**](https://www.sciencedirect.com/science/article/pii/S2667237523001273?via%3Dihub) (`meta-cells`)\n",
    " \n",
    "\n",
    "- Use `PAGODA2`, `GSVA`, or `AUCell`, to obtain pathway activity scores for pathways (`GO Ontology Biological Processes`) obtained from [**Maayan Lab**](https://maayanlab.cloud/Enrichr/#libraries), [**MSigDB**](https://www.gsea-msigdb.org/gsea/msigdb/human/collections.jsp#C5) or custom gene sets.\n",
    "\n",
    "- Perform differential pathway activity analysis using limma"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a070e8e",
   "metadata": {},
   "source": [
    "## **Data Prep Parameters**\n",
    "\n",
    "- `test_names`: List of the different test names of interest.\n",
    "\n",
    "- `save_prefix`: Preferred prefix for saving critical files. Ideally chosen to be in the format `{source name}_{brain region}`. e.g `mathys_pfc`\n",
    "\n",
    "- `subject_id`: Column name for Subject/Patient ID in both metadata and `.obs`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "53693e9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "save_prefix = 'seaad_mtg'                                       # this takes the format '{StudyName}_{ThreeLetterAccronymForBrainRegion}'\n",
    "subject_id = helper_functions.clean_strings('Donor ID')\n",
    "cell_type_column = 'Subclass'                                   # 'Supertype (non-expanded)', 'Subclass'\n",
    "factor = 'pathology.group'                                      # pathology.group # Continuous Pseudoprogression Score\n",
    "factor = factor.replace(\" \", \"\").replace(\"-\", \"\")\n",
    "test_names = ['late_vs_early']                                  # test categories\n",
    "region_name = save_prefix.split('_')[-1].upper()\n",
    "\n",
    "data_dir = f'/media/tadeoye/Volume1/data/seq/SEA-AD/{region_name}/RNAseq/'\n",
    "save_dir = f'../results/'\n",
    "\n",
    "if not os.path.exists(save_dir):\n",
    "    os.makedirs(save_dir)\n",
    "\n",
    "subclass = {\n",
    "    'excitatory': ['L5 IT', 'L2/3 IT', 'L4 IT', 'L6 IT', 'L6 IT Car3', 'L5/6 NP', 'L6b', 'L6 CT', 'L5 ET'],\n",
    "    'inhibitory': ['Pvalb', 'Sst', 'Lamp5 Lhx6', 'Vip', 'Lamp5', 'Sncg', 'Chandelier', 'Sst Chodl', 'Pax6'],\n",
    "    'astrocyte': ['Astrocyte'],\n",
    "    'microglia': ['Microglia-PVM'],\n",
    "    'opc': ['OPC'],\n",
    "    'oligodendrocyte': ['Oligodendrocyte'],\n",
    "    'endothelial': ['Endothelial'],\n",
    "    'vlmc': ['VLMC'],\n",
    "    }\n",
    "\n",
    "cell_supertypes = list(chain(*list(subclass.values())))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fae59c52",
   "metadata": {},
   "source": [
    "# **Load and Merge Pseudo-bulked Data**\n",
    "\n",
    "We have previously standardized and pseudo-bulked data for each cell sypertype (`00_Standardize_and_Split_RNAseq.ipynb`, `01_pseuobulking.ipynb`). First, we load the data for each cell supertype and merge them into a single seurat object. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "161b7f91",
   "metadata": {},
   "outputs": [],
   "source": [
    "clean_subclass = robjects.ListVector({k: [helper_functions.clean_strings(v, preserve_case=True) for v in val] for k, val in subclass.items()})\n",
    "clean_cell_supertypes = [helper_functions.clean_strings(cell_type, preserve_case=True) for cell_type in cell_supertypes]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b9789076",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "In addition: Warning messages:\n",
       "1: In gzfile(file, \"rb\") :\n",
       "  cannot open compressed file '/media/tadeoye/Volume1/data/seq/SEA-AD/MTG/RNAseq//anndata/excitatory/L5_ET/L5_ET_pseudobulked_count_data.rds', probable reason 'No such file or directory'\n",
       "2: In value[[3L]](cond) :\n",
       "  PSEUDOBULKED EXPRESSION DATA NOT FOUND FOR 'L5_ET'. PLEASE CONFIRM THAT THIS IS INTENTIONAL. SKIPPING L5_ET\n",
       "3: In gzfile(file, \"rb\") :\n",
       "  cannot open compressed file '/media/tadeoye/Volume1/data/seq/SEA-AD/MTG/RNAseq//anndata/inhibitory/Sst_Chodl/Sst_Chodl_pseudobulked_count_data.rds', probable reason 'No such file or directory'\n",
       "4: In value[[3L]](cond) :\n",
       "  PSEUDOBULKED EXPRESSION DATA NOT FOUND FOR 'Sst_Chodl'. PLEASE CONFIRM THAT THIS IS INTENTIONAL. SKIPPING Sst_Chodl\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R -i data_dir -i subject_id -i cell_supertypes -i clean_cell_supertypes -i clean_subclass -o summed_counts_per_celltype\n",
    "\n",
    "source('../scripts/functions/helper_functions.r')\n",
    "\n",
    "summed_counts_per_celltype <- list()\n",
    "expressed_genes_per_celltype <- list()\n",
    "\n",
    "for (cell_type in clean_cell_supertypes) {\n",
    "  \n",
    "  tryCatch({\n",
    "    # Read summed counts and expressed genes\n",
    "    summed_counts_per_celltype[[cell_type]] <- readRDS(\n",
    "      file.path(\n",
    "        data_dir,\n",
    "        \"anndata\",\n",
    "        find_parent_key(cell_type, clean_subclass),\n",
    "        cell_type,\n",
    "        paste0(cell_type, \"_pseudobulked_count_data.rds\")\n",
    "      )\n",
    "    )\n",
    "   \n",
    "    expressed_genes_per_celltype[[cell_type]] <- readRDS(\n",
    "      file.path(\n",
    "        data_dir,\n",
    "        \"anndata\",\n",
    "        find_parent_key(cell_type, clean_subclass),\n",
    "        cell_type,\n",
    "        paste0(cell_type, \"_pseudobulked_expressed_genes.rds\")\n",
    "      )\n",
    "    )      \n",
    "  },\n",
    "  error = function(e) {\n",
    "    warning(sprintf(\"PSEUDOBULKED EXPRESSION DATA NOT FOUND FOR '%s'. PLEASE CONFIRM THAT THIS IS INTENTIONAL. SKIPPING %s\", \n",
    "                   cell_type, cell_type))\n",
    "  })\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "77103505",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 150897 × 36601\n",
       "    obs: 'donor_id', 'orig.ident', 'nCount_originalexp', 'nFeature_originalexp', 'cells_merged', 'Subclass', 'ident', 'method', 'continuous_pseudo_progression_score', 'age_at_death_binned_codes', 'sex', 'race_choice_white', 'genes_detected', 'number_of_umis', 'pmi', 'apoe4_status'\n",
       "    layers: 'logcounts'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_merged = ad.concat([summed_counts_per_celltype[key] for key in summed_counts_per_celltype.keys()], join='outer')\n",
    "adata_merged.X = adata_merged.X.toarray()\n",
    "adata_merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "9aec190d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Merged data loaded\"\n",
      "class: SingleCellExperiment \n",
      "dim: 36601 150897 \n",
      "metadata(0):\n",
      "assays(2): X logcounts\n",
      "rownames(36601): MIR1302-2HG FAM138A ... AC007325.4 AC007325.2\n",
      "rowData names(0):\n",
      "colnames(150897): L5 IT#H19.33.004_1 L5 IT#H19.33.004_2 ...\n",
      "  VLMC#H21.33.027_8 VLMC#H21.33.027_9\n",
      "colData names(16): donor_id orig.ident ... pmi apoe4_status\n",
      "reducedDimNames(0):\n",
      "mainExpName: NULL\n",
      "altExpNames(0):\n"
     ]
    }
   ],
   "source": [
    "%%R -i adata_merged\n",
    "\n",
    "print('Merged data loaded')\n",
    "adata_merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ecff6ea1",
   "metadata": {},
   "outputs": [],
   "source": [
    "del adata_merged, summed_counts_per_celltype"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f3fe637e",
   "metadata": {},
   "source": [
    "# **Pathway Database**\n",
    "\n",
    "First, we load pre-defined pathway annotations from established databases and then filter these databases for expressed genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "731970b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "database_dir = data_dir.replace(f\"/seq/SEA-AD/{region_name}/RNAseq/\", \"/pathway_databases\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a712714a",
   "metadata": {},
   "source": [
    "### **Curated Gene Sets Obtained From [Galea et al. 2022](https://doi.org/10.1016/j.nbd.2022.105655)**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "13d75124",
   "metadata": {},
   "outputs": [],
   "source": [
    "galea_curated = pd.read_excel(f'{database_dir}/galea_2022_curated_gene_sets/gene_sets.xlsx', )\n",
    "galea_curated['Gene set'] = galea_curated['Functional categories'] + ': ' + galea_curated['Functional subcategories']\n",
    "galea_curated.rename(columns={'Gene symbol': 'Gene'}, inplace=True)\n",
    "gl_gmt = helper_functions.save_curated_as_gmt(galea_curated, name_col='Gene set', member_col='Gene', \n",
    "                                              tempdir=f'{database_dir}/galea_2022_curated_gene_sets/')\n",
    "gl_gmt_path = f'{database_dir}/galea_2022_curated_gene_sets/gs.gmt'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "690b790b",
   "metadata": {},
   "source": [
    "### **Curated Gene Sets Obtained From [Gabitto et al. 2022](https://doi.org/10.1101/2023.05.08.539485)**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "fcef2e89",
   "metadata": {},
   "outputs": [],
   "source": [
    "gabitto_curated = pd.read_excel(f'{database_dir}/gabitto_2023_curated_gene_sets/gene_sets.xlsx')\n",
    "gb_gmt = helper_functions.save_curated_as_gmt(gabitto_curated, name_col='Gene set', member_col='Gene', \n",
    "                                              tempdir=f'{database_dir}/gabitto_2023_curated_gene_sets/')\n",
    "gb_gmt_path = f'{database_dir}/gabitto_2023_curated_gene_sets/gs.gmt'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c08474c2",
   "metadata": {},
   "source": [
    "### **Curated Synaptic Related Gene Sets Obtained From [Koopmans et al. 2019](https://www.sciencedirect.com/science/article/pii/S0896627319304271?via%3Dihub)**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "1ae4706d",
   "metadata": {},
   "outputs": [],
   "source": [
    "syngo_curated = pd.read_excel(f'{database_dir}/SynGO_bulk_download_release_20231201/syngo_annotations.xlsx')\n",
    "syngo_curated.rename(columns={'go_name': 'Gene set', 'hgnc_symbol': 'Gene'}, inplace=True)\n",
    "syngo_gmt = helper_functions.save_curated_as_gmt(syngo_curated, name_col='Gene set', member_col='Gene', \n",
    "                                                 tempdir=f'{database_dir}/SynGO_bulk_download_release_20231201/')\n",
    "syngo_gmt_path = f'{database_dir}/SynGO_bulk_download_release_20231201/gs.gmt'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc26b19a",
   "metadata": {},
   "source": [
    "### **Curated Gene Sets Obtained From Human Molecular Signatures Database ([MSigDB](https://www.gsea-msigdb.org/gsea/msigdb/human/collections.jsp#C5))**\n",
    "- **[Reactome](https://www.gsea-msigdb.org/gsea/msigdb/human/genesets.jsp?collection=CP:REACTOME)**\n",
    "- **[KEGG](https://www.gsea-msigdb.org/gsea/msigdb/human/genesets.jsp?collection=CP:KEGG_LEGACY)**\n",
    "- **[Gene Ontology: Biological Processes](https://www.gsea-msigdb.org/gsea/msigdb/human/genesets.jsp?collection=GO:BP)**\n",
    "- **[TFT: transcription factor targets](https://www.gsea-msigdb.org/gsea/msigdb/human/genesets.jsp?collection=TFT)**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "7224578f",
   "metadata": {},
   "outputs": [],
   "source": [
    "bp_gmt_path = f'{database_dir}/MSigDB/c5.go.bp.v2024.1.Hs.symbols.gmt'                  # CP: Canonical pathways BP: subset of GO\n",
    "ke_gmt_path = f'{database_dir}/MSigDB/c2.cp.kegg_legacy.v2024.1.Hs.symbols.gmt'         # CP: Canonical pathwaysKEGG_LEGACY subset of CP\n",
    "re_gmt_path = f'{database_dir}/MSigDB/c2.cp.reactome.v2024.1.Hs.symbols.gmt'            # CP: Canonical pathways Reactome subset of CP\n",
    "tft_gtrd_gmt_path = f'{database_dir}/MSigDB/c3.tft.gtrd.v2024.1.Hs.symbols.gmt'         # TFT: transcription factor targets: GTRD subset of TFT\n",
    "tft_legc_gmt_path = f'{database_dir}/MSigDB/c3.tft.tft_legacy.v2024.1.Hs.symbols.gmt'   # TFT: transcription factor targets: TFT_LEGACY subset of TFT"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ebc03930",
   "metadata": {},
   "source": [
    "# **Pathway Activity Scoring**\n",
    "\n",
    "We now implement the pathway activity soring pipeline employed by **[Yaru Zhang et al. 2020](https://www.sciencedirect.com/science/article/pii/S2001037020304293#s0135)**. \n",
    "\n",
    "Briefly, **[Zhang et al](https://www.sciencedirect.com/science/article/pii/S2001037020304293#s0135)** performed a systematic benchmarking study to evaluate the performance of unsupervised pathway activity score(`PAS`) transformation algorithms. Essentially, PAS algorithms tranform the gene-level data into explainable gene sets representing biological processes or pathways to uncover the potential mechanism of cell heterogeneity. \n",
    "\n",
    "The tools benchmarked in the study include : \n",
    "\n",
    "> *Please hover over each method name to view full description*\n",
    "\n",
    "- [`AUCell`](https://www.nature.com/articles/nmeth.4463 \"AUCell is a rank-based enrichment analysis method for quantifying gene set activity in single-cell transcriptomics data. The algorithm employs a recovery-based scoring approach that first ranks genes according to their expression magnitude within individual cells. Subsequently, it computes an area under the recovery curve (AUC) by assessing the fraction of gene set members among the top-ranked genes, thereby generating cell-specific enrichment scores that represent the PAS\"): Area under the ranked gene expression curve.\n",
    "\n",
    "- [`Vision`](https://doi.org/10.1038/s41467-019-12235-0 \"Vision is an annotation toolkit that uses autocorrelation statistics to identify biological variations across cells. Vision starts to identify closest K-nearest neighbors of each cell for generating a cell–cell K-nearest-neighbor (KNN) graph. PASs in Vision are calculated by averaging expressed genes for each gene set. To account for the influence of sample-level metrics (the number of UMIs/reads per cells), PASs are then corrected by their means and standard deviations. Expression data used in Vision could be scaled and normalized, but not log-transformed.\"): Summarizing the noermalized expression of genes in the gene sets.\n",
    "\n",
    "- [`Pagoda2`](https://doi.org/10.1038/nbt.4038.Integrative \"Pathway and gene set overdispersion analysis (Pagoda2) implements a pathway overdispersion analysis method for single-cell data. It first evaluates the variance of individual genes, taking into account technical factors. It then identifies pathways showing coordinated variability by combining the variance patterns of pathway genes using weighted PCA, while controlling for technical aspects and adjusting for multiple hypothesis testing.\"): First principal component of gene sets\n",
    "\n",
    "- [`GSVA`](https://doi.org/10.1186/1471-2105-14-7 \"GSVA implements a non-parametric, unsupervised method for estimating variation in pathway activity across samples. It first estimates a normalized relative expression level per gene across samples. This expression level is then rank-ordered for each sample and aggregated into gene sets by calculating sample-wise enrichment scores using a Kolmogorov–Smirnov-like rank statistic.\"): Kolmogorov-Smirnov-like rank statistic based on kernel estimation of th cumulative density.\n",
    "\n",
    "- [`ssGSEA`](https://doi.org/10.1038/nature08460.Systematic \"Briefly, ssGSEA implements a rank-based method to generate an enrichment score per sample per gene set. It first ranks all genes by their expression value for each sample. Then, for each gene set, it calculates an enrichment score that represents the degree of absolute enrichment of that gene set in each sample based on the integration of the differences between empirical cumulative distribution functions of gene expression ranks.\"): Kolmogorov-Smirnov-like rank statistic based on gene expression of single sample. \n",
    "\n",
    "- [`Combined z-score`](https://doi.org/10.1371/journal.pcbi.1000217 \"This method implements a parametric method for pathway activity estimation. It first standardizes gene expression values across samples to generate z-scores for each gene. These gene-level z-scores are then aggregated for each pathway by calculating their mean or sum, producing a pathway-level score that represents the overall activation state of the pathway in each sample.\"): Combinzed z-score."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd8d762c",
   "metadata": {},
   "source": [
    "## **Method Selection for PAS**\n",
    "\n",
    "**[Zhang et al.](https://www.sciencedirect.com/science/article/pii/S2001037020304293#s0135)** identified `Pagoda2` and `Vision` as the highest performing methods for pathway activity scoring, likely due to their direct estimation of pathway activity scores (PAS) from the expression data.\n",
    "\n",
    "<figure align=\"center\">\n",
    "   <img src=\"../docs/pas_performance_summary.jpg\" alt=\"Pathway Activity Score Performance Summary\" width=\"30%\">\n",
    "   <figcaption><b>Figure 1:</b> Comparative performance evaluation of pathway activity scoring methods across multiple metrics</figcaption>\n",
    "</figure>\n",
    "\n",
    "Based on these findings, our analysis focuses on the three top-performing methods `Pagoda2`, `Vision`, and `AUCell`, inlcuding `GSVA` following our previous work ([**Adeoye et al. 2024**](https://www.aginganddisease.org/EN/10.14336/AD.2024.0429)). However, we here elect to use only `AUCell` as the other listed methods do not scale well with such large data.\n",
    "\n",
    "> **Note**: Users preferring GSVA can still efficiently apply it by first setting `count_agg_strategy = pseudobulk` in `01_pseudobulking.ipynb` to use standard pseudobulking instead of metacells. But we doubt this will yield reliable results.\n",
    "\n",
    "The authors provide the [PASBench](https://github.com/sulab-wmu/PASBench/tree/master) package that contain the seven PAS tools and evaluation metrics."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37ad504c",
   "metadata": {},
   "source": [
    "## **Parameter Selection for [PASBench](https://github.com/sulab-wmu/PASBench/tree/master)**\n",
    "\n",
    "The following parameters control the calculation of pathway activity scores from single-cell gene expression data:\n",
    "\n",
    "#### **Core Parameters**\n",
    "- `counts`: the pseudobulked SingleCellEXperiment object. Values must contain expression `counts` assay/layer.\n",
    "\n",
    "- `gmt_file`: Path to GMT format file containing pathway definitions.\n",
    "\n",
    "- `species`: Species designation for pathway analysis. Options: `\"human\"` or `\"mouse\"`.\n",
    "\n",
    "- `pathway`: Abbreviated name of pathway database (e.g., `KEGG`, `GO`, `REACTOME`). Only called when gmt_file is `none`\n",
    "\n",
    "- `tool`: Specifies the analysis method to use. In this case, any of `Pagoda2` or `AUCell`\n",
    " > *See [method descriptions above](#) for details on each tool*\n",
    "\n",
    "- `filter`: Controls expression-based gene filtering. When `TRUE`, removes genes expressed in <5% of cells. When `FALSE`, retains all genes. We already filtered genes before pseudobulking, so this **[Zhang et al.](https://www.sciencedirect.com/science/article/pii/S2001037020304293#s0135)** will be set to `FALSE`\n",
    "\n",
    "- `normalize`: Specifies data normalization method. This is required to make expression values comparable across cells. The specified method should be appropriate for chosen tool. Options are `log`,`CLR`(Centered Log Ratio in Seurat),`RC` (Relative Counts in Seurat),`scran`,`scTransform`,`none`.\n",
    "\n",
    "The figure below shows the distribution of averaged performance score across normalization strategies from **[Zhang et al.](https://www.sciencedirect.com/science/article/pii/S2001037020304293#s0135)**. \n",
    "\n",
    "<figure align=\"center\">\n",
    "   <img src=\"../docs/performance_by_normalization_strategy.jpg\" alt=\"Averaged Performance Score Across Filtering and Normalization Strategies\" width=\"50%\">\n",
    "   <figcaption><b>Figure 2:</b> Comparative performance evaluation of pathway activity scoring methods across filtering and normalization strategies</figcaption>\n",
    "</figure>\n",
    "\n",
    "This figure shows that:\n",
    "\n",
    "- `Pagoda2` and `Vision` display the same performance irrespective of the chosen normalization method. Hence, we perform no normalization\n",
    "\n",
    "- perfomance of `AUCell` depends on the normalization strategy. So, we choose [scTransform](https://github.com/satijalab/sctransform) normalization, since the other tools all display better performance with this method.\n",
    "\n",
    "- For `GSVA`, we recommend setting the the normalization method to none, as the default `assay` chosen from the `SingleCellExperiment` object is `logcount` appropriate for the `Gaussian` kernel applied to `log CPM` data. User can likewise change parameters like `maxDiff`, `min/max_set_size`, and `kcdf`. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5ee3087",
   "metadata": {},
   "outputs": [],
   "source": [
    "pas_methods = ['AUCell']\n",
    "normalization_methods = robjects.ListVector({'AUCell': 'none'})\n",
    "assay_name = 'logcounts'    # assay name in pseudobulked singlecellexperiment object. \n",
    "                            # If set to `logcounts` then `kcdf` must be set to `Gaussian` and normalization method must be 'none'\n",
    "kcdf = 'Gaussian'           # determines the kernel density used to model gene expression across samples\n",
    "max_set_size = 150          # Upper bound of gene set size\n",
    "min_set_size = 5            # lower bound of gene set size\n",
    "maxDiff = 'TRUE'            # Controls how enrichment score is calculated from the Kolmogorov-Smirnov random walk statistics\n",
    "                            # controls which curated gene set database the statistic is estimated for\n",
    "n_cores = 40\n",
    "\n",
    "curated_gene_set_db = robjects.ListVector({\n",
    "    'gabitto': gb_gmt_path,\n",
    "    'galea': gl_gmt_path,\n",
    "    'syngo': syngo_gmt_path,\n",
    "    'gobp': bp_gmt_path,\n",
    "    'reactome': re_gmt_path,\n",
    "    'kegg': ke_gmt_path, \n",
    "    'gtrd': tft_gtrd_gmt_path,\n",
    "    'tft': tft_legc_gmt_path,\n",
    "    })       "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "371b8792",
   "metadata": {},
   "source": [
    "#### **Output Structure**\n",
    "The implemented algorithms generates two complementary data structures:\n",
    "\n",
    "1. A dictionary containing pathway activity scores for each gene set and method defined in `curated_gene_set_db`\n",
    "2. A SingleCellExperiment object encoding a matrix of pathway activity scores (PAS) across cell types, where:\n",
    "   - Rows represent individual pathways\n",
    "   - Columns represent distinct cell types\n",
    "   - Matrix elements contain non-/normalized PAS\n",
    "\n",
    "The SingleCellExperiment object maintains the relationship between PAS and cellular identities while preserving the original experimental design structure. This allows us to perform downstream analysis on activuty scores like clustering and visualizing PAS in cellular populations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "543d5532",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Estimating Activity Scores of gabitto Pathways using AUCell scoring method\"\n",
      "[1] \"........\"\n",
      "Do not filter genes.\n",
      "[1] \"normalize error\"\n",
      "Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck\n",
      "\n",
      "Number of nodes: 150897\n",
      "Number of edges: 4437850\n",
      "\n",
      "Running Louvain algorithm...\n",
      "Maximum modularity in 10 random starts: 0.9422\n",
      "Number of communities: 25\n",
      "Elapsed time: 120 seconds\n",
      "[1] \"Estimating Activity Scores of galea Pathways using AUCell scoring method\"\n",
      "[1] \"........\"\n",
      "Do not filter genes.\n",
      "[1] \"normalize error\"\n",
      "Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck\n",
      "\n",
      "Number of nodes: 150897\n",
      "Number of edges: 4423655\n",
      "\n",
      "Running Louvain algorithm...\n",
      "Maximum modularity in 10 random starts: 0.9483\n",
      "Number of communities: 32\n",
      "Elapsed time: 127 seconds\n",
      "[1] \"Estimating Activity Scores of syngo Pathways using AUCell scoring method\"\n",
      "[1] \"........\"\n",
      "Do not filter genes.\n",
      "[1] \"normalize error\"\n",
      "Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck\n",
      "\n",
      "Number of nodes: 150897\n",
      "Number of edges: 4501420\n",
      "\n",
      "Running Louvain algorithm...\n",
      "Maximum modularity in 10 random starts: 0.9489\n",
      "Number of communities: 26\n",
      "Elapsed time: 102 seconds\n",
      "[1] \"Estimating Activity Scores of gobp Pathways using AUCell scoring method\"\n",
      "[1] \"........\"\n",
      "Do not filter genes.\n",
      "[1] \"normalize error\"\n",
      "Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck\n",
      "\n",
      "Number of nodes: 150897\n",
      "Number of edges: 4498637\n",
      "\n",
      "Running Louvain algorithm...\n",
      "Maximum modularity in 10 random starts: 0.9496\n",
      "Number of communities: 27\n",
      "Elapsed time: 81 seconds\n",
      "[1] \"Estimating Activity Scores of reactome Pathways using AUCell scoring method\"\n",
      "[1] \"........\"\n",
      "Do not filter genes.\n",
      "[1] \"normalize error\"\n",
      "Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck\n",
      "\n",
      "Number of nodes: 150897\n",
      "Number of edges: 4436751\n",
      "\n",
      "Running Louvain algorithm...\n",
      "Maximum modularity in 10 random starts: 0.9371\n",
      "Number of communities: 27\n",
      "Elapsed time: 78 seconds\n",
      "[1] \"Estimating Activity Scores of kegg Pathways using AUCell scoring method\"\n",
      "[1] \"........\"\n",
      "Do not filter genes.\n",
      "[1] \"normalize error\"\n",
      "Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck\n",
      "\n",
      "Number of nodes: 150897\n",
      "Number of edges: 4414016\n",
      "\n",
      "Running Louvain algorithm...\n",
      "Maximum modularity in 10 random starts: 0.9364\n",
      "Number of communities: 22\n",
      "Elapsed time: 127 seconds\n",
      "[1] \"Estimating Activity Scores of gtrd Pathways using AUCell scoring method\"\n",
      "[1] \"........\"\n",
      "Do not filter genes.\n",
      "[1] \"normalize error\"\n",
      "Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck\n",
      "\n",
      "Number of nodes: 150897\n",
      "Number of edges: 4465619\n",
      "\n",
      "Running Louvain algorithm...\n",
      "Maximum modularity in 10 random starts: 0.9417\n",
      "Number of communities: 21\n",
      "Elapsed time: 117 seconds\n",
      "[1] \"Estimating Activity Scores of tft Pathways using AUCell scoring method\"\n",
      "[1] \"........\"\n",
      "Do not filter genes.\n",
      "[1] \"normalize error\"\n",
      "Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck\n",
      "\n",
      "Number of nodes: 150897\n",
      "Number of edges: 4471133\n",
      "\n",
      "Running Louvain algorithm...\n",
      "Maximum modularity in 10 random starts: 0.9467\n",
      "Number of communities: 27\n",
      "Elapsed time: 113 seconds\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Read 129 records\n",
       "Read 8627 items\n",
       "Warning: Data is of class matrix. Coercing to dgCMatrix.\n",
       "Warning: Feature names cannot have underscores ('_'), replacing with dashes ('-')\n",
       "Warning: Data is of class data.frame. Coercing to dgCMatrix.\n",
       "Centering and scaling data matrix\n",
       "\r  |                                                                            \r  |                                                                      |   0%\r  |                                                                            \r  |======================================================================| 100%\n",
       "Warning: The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric\n",
       "To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'\n",
       "This message will be shown once per session\n",
       "Computing nearest neighbor graph\n",
       "Computing SNN\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Read 215 records\n",
       "Read 3223 items\n",
       "Warning: Data is of class matrix. Coercing to dgCMatrix.\n",
       "Warning: Data is of class data.frame. Coercing to dgCMatrix.\n",
       "Centering and scaling data matrix\n",
       "\r  |                                                                            \r  |                                                                      |   0%\r  |                                                                            \r  |======================================================================| 100%\n",
       "Computing nearest neighbor graph\n",
       "Computing SNN\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Read 267 records\n",
       "Read 4147 items\n",
       "Warning: Data is of class matrix. Coercing to dgCMatrix.\n",
       "Warning in .AUCell_calcAUC(geneSets = geneSets, rankings = rankings, nCores = nCores,  :\n",
       "  The following gene sets will be excluded from the analysis(less than 20% of their genes are available):\n",
       "maintenance of alignment of postsynaptic density and presynaptic active zone (GO:0099559), postsynaptic spectrin-associated cytoskeleton organization (GO:0099190), presynaptic cytoskeleton (GO:0099569), regulation of anterograde synaptic vesicle transport (GO:1903742)\n",
       "Warning: Data is of class data.frame. Coercing to dgCMatrix.\n",
       "Centering and scaling data matrix\n",
       "\r  |                                                                            \r  |                                                                      |   0%\r  |                                                                            \r  |======================================================================| 100%\n",
       "Computing nearest neighbor graph\n",
       "Computing SNN\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Read 7608 records\n",
       "Read 645524 items\n",
       "Warning: Data is of class matrix. Coercing to dgCMatrix.\n",
       "Warning in .AUCell_calcAUC(geneSets = geneSets, rankings = rankings, nCores = nCores,  :\n",
       "  The following gene sets will be excluded from the analysis(less than 20% of their genes are available):\n",
       "GOBP_FORMATION_OF_QUADRUPLE_SL_U4_U5_U6_SNRNP, GOBP_MIRNA_MEDIATED_GENE_SILENCING_BY_INHIBITION_OF_TRANSLATION, GOBP_MIRNA_MEDIATED_GENE_SILENCING_BY_MRNA_DESTABILIZATION, GOBP_NEGATIVE_REGULATION_OF_CELL_CHEMOTAXIS_TO_FIBROBLAST_GROWTH_FACTOR, GOBP_NEGATIVE_REGULATION_OF_INTERLEUKIN_21_PRODUCTION, GOBP_NEGATIVE_REGULATION_OF_INTERLEUKIN_6_MEDIATED_SIGNALING_PATHWAY, GOBP_NEGATIVE_REGULATION_OF_MATRIX_METALLOPEPTIDASE_SECRETION, GOBP_NEGATIVE_REGULATION_OF_VASCULAR_ENDOTHELIAL_CELL_PROLIFERATION, GOBP_NEGATIVE_REGULATION_OF_VASCULAR_ENDOTHELIAL_GROWTH_FACTOR_PRODUCTION, GOBP_PROTEIN_LOCALIZATION_TO_CENP_A_CONTAINING_CHROMATIN, GOBP_REGULATION_OF_CIRCULATING_FIBRINOGEN_LEVELS, GOBP_REGULATION_OF_XENOBIOTIC_DETOXIFICATION_BY_TRANSMEMBRANE_EXPORT_ACROSS_THE_PLASMA_MEMBRANE\n",
       "Warning: Feature names cannot have underscores ('_'), replacing with dashes ('-')\n",
       "Warning: Data is of class data.frame. Coercing to dgCMatrix.\n",
       "Centering and scaling data matrix\n",
       "\r  |                                                                            \r  |                                                                      |   0%\r  |                                                                            \r  |=========                                                             |  12%\r  |                                                                            \r  |==================                                                    |  25%\r  |                                                                            \r  |==========================                                            |  38%\r  |                                                                            \r  |===================================                                   |  50%\r  |                                                                            \r  |============================================                          |  62%\r  |                                                                            \r  |====================================================                  |  75%\r  |                                                                            \r  |=============================================================         |  88%\r  |                                                                            \r  |======================================================================| 100%\n",
       "Computing nearest neighbor graph\n",
       "Computing SNN\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Read 1736 records\n",
       "Read 101062 items\n",
       "Warning: Data is of class matrix. Coercing to dgCMatrix.\n",
       "Warning in .AUCell_calcAUC(geneSets = geneSets, rankings = rankings, nCores = nCores,  :\n",
       "  The following gene sets will be excluded from the analysis(less than 20% of their genes are available):\n",
       "REACTOME_ACTIVATED_PKN1_STIMULATES_TRANSCRIPTION_OF_AR_ANDROGEN_RECEPTOR_REGULATED_GENES_KLK2_AND_KLK3, REACTOME_ASSEMBLY_OF_THE_ORC_COMPLEX_AT_THE_ORIGIN_OF_REPLICATION, REACTOME_CHROMATIN_MODIFICATIONS_DURING_THE_MATERNAL_TO_ZYGOTIC_TRANSITION_MZT, REACTOME_CONDENSATION_OF_PROPHASE_CHROMOSOMES, REACTOME_DNA_METHYLATION, REACTOME_PRC2_METHYLATES_HISTONES_AND_DNA, REACTOME_RECOGNITION_AND_ASSOCIATION_OF_DNA_GLYCOSYLASE_WITH_SITE_CONTAINING_AN_AFFECTED_PURINE, REACTOME_REGULATION_OF_ENDOGENOUS_RETROELEMENTS_BY_THE_HUMAN_SILENCING_HUB_HUSH_COMPLEX, REACTOME_REPLACEMENT_OF_PROTAMINES_BY_NUCLEOSOMES_IN_THE_MALE_PRONUCLEUS, REACTOME_SIRT1_NEGATIVELY_REGULATES_RRNA_EXPRESSION\n",
       "Warning: Feature names cannot have underscores ('_'), replacing with dashes ('-')\n",
       "Warning: Data is of class data.frame. Coercing to dgCMatrix.\n",
       "Centering and scaling data matrix\n",
       "\r  |                                                                            \r  |                                                                      |   0%\r  |                                                                            \r  |===================================                                   |  50%\r  |                                                                            \r  |======================================================================| 100%\n",
       "Computing nearest neighbor graph\n",
       "Computing SNN\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Read 186 records\n",
       "Read 13167 items\n",
       "Warning: Data is of class matrix. Coercing to dgCMatrix.\n",
       "Warning: Feature names cannot have underscores ('_'), replacing with dashes ('-')\n",
       "Warning: Data is of class data.frame. Coercing to dgCMatrix.\n",
       "Centering and scaling data matrix\n",
       "\r  |                                                                            \r  |                                                                      |   0%\r  |                                                                            \r  |======================================================================| 100%\n",
       "Computing nearest neighbor graph\n",
       "Computing SNN\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Read 505 records\n",
       "Read 258113 items\n",
       "Warning: Data is of class matrix. Coercing to dgCMatrix.\n",
       "Warning in .AUCell_calcAUC(geneSets = geneSets, rankings = rankings, nCores = nCores,  :\n",
       "  The following gene sets will be excluded from the analysis(less than 20% of their genes are available):\n",
       "COIL_TARGET_GENES, F2RL1_TARGET_GENES, GTF3A_TARGET_GENES, LAMTOR5_TARGET_GENES, SOD1_TARGET_GENES, TFAM_TARGET_GENES\n",
       "Warning: Feature names cannot have underscores ('_'), replacing with dashes ('-')\n",
       "Warning: Data is of class data.frame. Coercing to dgCMatrix.\n",
       "Centering and scaling data matrix\n",
       "\r  |                                                                            \r  |                                                                      |   0%\r  |                                                                            \r  |======================================================================| 100%\n",
       "Computing nearest neighbor graph\n",
       "Computing SNN\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Read 610 records\n",
       "Read 156411 items\n",
       "Warning: Data is of class matrix. Coercing to dgCMatrix.\n",
       "Warning: Feature names cannot have underscores ('_'), replacing with dashes ('-')\n",
       "Warning: Data is of class data.frame. Coercing to dgCMatrix.\n",
       "Centering and scaling data matrix\n",
       "\r  |                                                                            \r  |                                                                      |   0%\r  |                                                                            \r  |======================================================================| 100%\n",
       "Computing nearest neighbor graph\n",
       "Computing SNN\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "In addition: Warning messages:\n",
       "1: In .AUCell_buildRankings(exprMat = exprMat, featureType = featureType,  :\n",
       "  nCores is no longer used. It will be deprecated in the next AUCell version.\n",
       "2: In .AUCell_buildRankings(exprMat = exprMat, featureType = featureType,  :\n",
       "  nCores is no longer used. It will be deprecated in the next AUCell version.\n",
       "3: In .AUCell_buildRankings(exprMat = exprMat, featureType = featureType,  :\n",
       "  nCores is no longer used. It will be deprecated in the next AUCell version.\n",
       "4: In .AUCell_buildRankings(exprMat = exprMat, featureType = featureType,  :\n",
       "  nCores is no longer used. It will be deprecated in the next AUCell version.\n",
       "5: In .AUCell_buildRankings(exprMat = exprMat, featureType = featureType,  :\n",
       "  nCores is no longer used. It will be deprecated in the next AUCell version.\n",
       "6: In .AUCell_buildRankings(exprMat = exprMat, featureType = featureType,  :\n",
       "  nCores is no longer used. It will be deprecated in the next AUCell version.\n",
       "7: In .AUCell_buildRankings(exprMat = exprMat, featureType = featureType,  :\n",
       "  nCores is no longer used. It will be deprecated in the next AUCell version.\n",
       "8: In .AUCell_buildRankings(exprMat = exprMat, featureType = featureType,  :\n",
       "  nCores is no longer used. It will be deprecated in the next AUCell version.\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R -i max_set_size -i min_set_size -i kcdf -i maxDiff -i curated_gene_set_db -i pas_methods -i n_cores -i normalization_methods -o pas_scores -o pas_object\n",
    "\n",
    "source('../scripts/functions/PASBench/R/main.R')\n",
    "source('../scripts/functions/helper_functions.r')\n",
    "\n",
    "pas_scores = list()\n",
    "pas_object = list()\n",
    "gene_sets <- curated_gene_set_db\n",
    "\n",
    "for (gs in names(gene_sets)){\n",
    "    pas_scores[[gs]] <- list()\n",
    "    pas_object[[gs]] <- list()\n",
    "\n",
    "    all_gs_genes <- unique(unlist(read.geneset(gene_sets[[gs]])))\n",
    "    current_genes <- rownames(adata_merged)\n",
    "    genes_to_keep <- current_genes %in% all_gs_genes\n",
    "\n",
    "    for (method in pas_methods){\n",
    "        \n",
    "        if (method == 'GSVA'){\n",
    "            sce <- assay(adata_merged[genes_to_keep, ], assay_name)\n",
    "        }else{\n",
    "            sce <- assay(adata_merged[genes_to_keep, ], 'X')\n",
    "        }\n",
    "\n",
    "        # run PAS scoring\n",
    "        print(paste0(\"Estimating Activity Scores of \", gs, \" Pathways using \", method, \" scoring method\"))\n",
    "        print(\"........\")\n",
    "\n",
    "        res <- calculate_PAS(\n",
    "            sce,\n",
    "            method,                         # tool name\n",
    "            species = 'none',               # species; human or mouse\n",
    "            pathway = 'none',               # abbreviation for pathway database\n",
    "            gmt_file = gene_sets[[gs]],     # pathways in GMT format\n",
    "            filter = F,                     # whether filtering for genes expressed in less than 5 percent cells\n",
    "            gsvaPar = list(                 # Parameters for when method selected is GSVA\n",
    "                maxDiff = as.logical(maxDiff), \n",
    "                kcdf = kcdf, \n",
    "                minSize = min_set_size, \n",
    "                maxSize = max_set_size\n",
    "                ),\n",
    "            \n",
    "            normalize = normalization_methods[[method]],\n",
    "            n_cores = n_cores,\n",
    "            rand_seed = 12345\n",
    "        )\n",
    "\n",
    "        pas_scores[[gs]][[method]] <- as.data.frame(res)\n",
    "        pas_object[[gs]][[method]] <- as.SingleCellExperiment(prepare_vis(pas_scores[[gs]][[method]]))\n",
    "\n",
    "        colData(pas_object[[gs]][[method]]) <- DataFrame(\n",
    "                merge(as.data.frame(colData(pas_object[[gs]][[method]])), \n",
    "                    as.data.frame(colData(adata_merged)), \n",
    "                    by=0, \n",
    "                    row.names=1)\n",
    "        rownames(colData(pas_object[[gs]][[method]])) <- colData(pas_object[[gs]][[method]])$Row.names\n",
    "\n",
    "        )         \n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "41a09b42",
   "metadata": {},
   "source": [
    "#### **Save Pathway Activity Scores**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "1a82c3c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "if (!dir.exists(paste0(data_dir, \"results/PAS/\"))) {\n",
    "  dir.create(paste0(data_dir, \"results/PAS/\"))\n",
    "}\n",
    "\n",
    "saveRDS(pas_object, paste0(data_dir, \"results/PAS/\", \"all_PAS.rds\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "b8486422",
   "metadata": {},
   "outputs": [],
   "source": [
    "for gs in curated_gene_set_db.names:\n",
    "    for method in pas_methods:\n",
    "        pas_object[gs][method].write_h5ad(os.path.join(data_dir,  \"results\", \"PAS\", f\"{gs}_geneset_{method}_PAS.h5ad\"),\n",
    "                              compression='gzip')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scRNA_seq_meta_analysis",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "vscode": {
   "interpreter": {
    "hash": "2f414a000012e1980f5fd1119ed6b9bf8cc73b51e37b52eb52a16ff857effd1c"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
