{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1556cc30",
   "metadata": {},
   "outputs": [],
   "source": [
    "import warnings\n",
    "from warnings import simplefilter\n",
    "warnings.filterwarnings(\"ignore\", category=UserWarning)\n",
    "warnings.filterwarnings('ignore', category=FutureWarning)\n",
    "warnings.filterwarnings(\"ignore\", category=PendingDeprecationWarning)\n",
    "warnings.filterwarnings('ignore', category=FutureWarning, module='anndata')\n",
    "warnings.filterwarnings('ignore', message='.*reticulate.*')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62042f97-a76e-4b34-805f-3f1ac81e7288",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import rpy2\n",
    "import logging\n",
    "import warnings\n",
    "import anndata2ri\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import anndata as ad\n",
    "from itertools import chain\n",
    "from datetime import datetime\n",
    "import rpy2.robjects as robjects\n",
    "from rpy2.robjects import pandas2ri\n",
    "from scipy.sparse import csr_matrix\n",
    "from joblib import Parallel, delayed\n",
    "from functions import helper_functions\n",
    "from matplotlib.pyplot import rcParams\n",
    "from functions.run_nebula import run_de_tests, build_effect_size_anndata, get_standardized_effects\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "447b6d1c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-----\n",
      "anndata     0.11.1\n",
      "scanpy      1.9.3\n",
      "-----\n",
      "PIL                         9.5.0\n",
      "anndata2ri                  1.1\n",
      "asttokens                   NA\n",
      "backcall                    0.2.0\n",
      "cffi                        1.15.1\n",
      "comm                        0.1.3\n",
      "cycler                      0.10.0\n",
      "cython_runtime              NA\n",
      "dateutil                    2.8.2\n",
      "debugpy                     1.6.7\n",
      "decorator                   5.1.1\n",
      "decoupler                   1.4.0\n",
      "dot_parser                  NA\n",
      "exceptiongroup              1.1.1\n",
      "executing                   1.2.0\n",
      "functions                   NA\n",
      "google                      NA\n",
      "gsva_prep                   NA\n",
      "h5py                        3.9.0\n",
      "igraph                      0.10.4\n",
      "ipykernel                   6.23.2\n",
      "ipywidgets                  8.0.6\n",
      "jedi                        0.18.2\n",
      "jinja2                      3.1.2\n",
      "joblib                      1.2.0\n",
      "kiwisolver                  1.4.4\n",
      "leidenalg                   0.9.1\n",
      "llvmlite                    0.39.1\n",
      "louvain                     0.8.0\n",
      "markupsafe                  2.1.3\n",
      "matplotlib                  3.7.1\n",
      "mpl_toolkits                NA\n",
      "mpmath                      1.3.0\n",
      "natsort                     8.4.0\n",
      "numba                       0.56.4\n",
      "numexpr                     2.8.4\n",
      "numpy                       1.23.5\n",
      "nvfuser                     NA\n",
      "opt_einsum                  v3.3.0\n",
      "packaging                   23.1\n",
      "pandas                      2.0.2\n",
      "parso                       0.8.3\n",
      "patsy                       0.5.3\n",
      "pexpect                     4.8.0\n",
      "pickleshare                 0.7.5\n",
      "pkg_resources               NA\n",
      "platformdirs                3.7.0\n",
      "plotly                      5.24.1\n",
      "prompt_toolkit              3.0.38\n",
      "psutil                      5.9.5\n",
      "ptyprocess                  0.7.0\n",
      "pure_eval                   0.2.2\n",
      "pyarrow                     12.0.1\n",
      "pycparser                   2.21\n",
      "pydev_ipython               NA\n",
      "pydevconsole                NA\n",
      "pydevd                      2.9.5\n",
      "pydevd_file_utils           NA\n",
      "pydevd_plugins              NA\n",
      "pydevd_tracing              NA\n",
      "pydot                       1.4.2\n",
      "pygments                    2.15.1\n",
      "pyparsing                   3.1.0\n",
      "pytz                        2023.3\n",
      "pytz_deprecation_shim       NA\n",
      "rpy2                        3.5.12\n",
      "scipy                       1.10.1\n",
      "seaborn                     0.12.2\n",
      "session_info                1.0.0\n",
      "setuptools                  68.0.0\n",
      "six                         1.16.0\n",
      "sklearn                     1.2.2\n",
      "sparse                      0.14.0\n",
      "stack_data                  0.6.2\n",
      "statsmodels                 0.14.0\n",
      "sympy                       1.12\n",
      "texttable                   1.6.7\n",
      "threadpoolctl               3.1.0\n",
      "torch                       2.0.1+cu117\n",
      "tornado                     6.3.2\n",
      "tqdm                        4.65.0\n",
      "traitlets                   5.9.0\n",
      "typing_extensions           NA\n",
      "tzlocal                     NA\n",
      "vscode                      NA\n",
      "wcwidth                     0.2.6\n",
      "yaml                        6.0\n",
      "zmq                         25.1.0\n",
      "zoneinfo                    NA\n",
      "-----\n",
      "IPython             8.14.0\n",
      "jupyter_client      8.2.0\n",
      "jupyter_core        5.3.1\n",
      "-----\n",
      "Python 3.10.6 (main, Oct 24 2022, 16:07:47) [GCC 11.2.0]\n",
      "Linux-6.8.0-52-generic-x86_64-with-glibc2.35\n",
      "-----\n",
      "Session information updated at 2025-04-02 17:59\n"
     ]
    }
   ],
   "source": [
    "# # Ignore R warning messages\n",
    "#Note: this can be commented out to get more verbose R output\n",
    "rpy2.rinterface_lib.callbacks.logger.setLevel(logging.ERROR)\n",
    "\n",
    "# Automatically convert rpy2 outputs to pandas dataframes\n",
    "pandas2ri.activate()\n",
    "anndata2ri.activate()\n",
    "%load_ext rpy2.ipython\n",
    "\n",
    "sc.settings.verbosity = 3\n",
    "#sc.set_figure_params(dpi=200, dpi_save=300)\n",
    "sc.logging.print_versions()\n",
    "\n",
    "simplefilter(action=\"ignore\", category=pd.errors.PerformanceWarning)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "048c2e09",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "    WARNING: The R package \"reticulate\" only fixed recently\n",
      "    an issue that caused a segfault when used with rpy2:\n",
      "    https://github.com/rstudio/reticulate/pull/1188\n",
      "    Make sure that you use a version of that package that includes\n",
      "    the fix.\n",
      "    Allowing parallel execution with up to 40 working processes.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "In addition: Warning messages:\n",
       "1: replacing previous import ‘GenomicRanges::intersect’ by ‘SeuratObject::intersect’ when loading ‘hdWGCNA’ \n",
       "2: replacing previous import ‘GenomicRanges::union’ by ‘dplyr::union’ when loading ‘hdWGCNA’ \n",
       "3: replacing previous import ‘GenomicRanges::setdiff’ by ‘dplyr::setdiff’ when loading ‘hdWGCNA’ \n",
       "4: replacing previous import ‘dplyr::as_data_frame’ by ‘igraph::as_data_frame’ when loading ‘hdWGCNA’ \n",
       "5: replacing previous import ‘Seurat::components’ by ‘igraph::components’ when loading ‘hdWGCNA’ \n",
       "6: replacing previous import ‘dplyr::groups’ by ‘igraph::groups’ when loading ‘hdWGCNA’ \n",
       "7: replacing previous import ‘dplyr::union’ by ‘igraph::union’ when loading ‘hdWGCNA’ \n",
       "8: replacing previous import ‘GenomicRanges::subtract’ by ‘magrittr::subtract’ when loading ‘hdWGCNA’ \n",
       "9: replacing previous import ‘Matrix::as.matrix’ by ‘proxy::as.matrix’ when loading ‘hdWGCNA’ \n",
       "10: replacing previous import ‘igraph::groups’ by ‘tidygraph::groups’ when loading ‘hdWGCNA’ \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "suppressPackageStartupMessages({\n",
    "    library(Matrix)\n",
    "    library(viridis)\n",
    "    library(harmony)\n",
    "    library(ggpubr)\n",
    "    library(tictoc)\n",
    "    library(RColorBrewer)\n",
    "    library(Hmisc)\n",
    "    library(corrplot)\n",
    "    library(grid)\n",
    "    library(gridExtra)\n",
    "    library(igraph)\n",
    "    library(ggrepel)\n",
    "    library(readxl)\n",
    "    library(conflicted)\n",
    "    library(dplyr)\n",
    "    library(parallel)\n",
    "    library(stringr)\n",
    "    library(tibble)\n",
    "    library(BiocParallel)\n",
    "\n",
    "    # single-cell analysis package\n",
    "    library(Seurat)\n",
    "    library(zellkonverter)   \n",
    "    library(SingleCellExperiment)\n",
    "    library(tidyr)\n",
    "    library(readxl)\n",
    "    library(GSA)\n",
    "    library(limma)\n",
    "\n",
    "    # plotting and data science packages\n",
    "    library(tidyverse)\n",
    "    library(cowplot)\n",
    "    library(patchwork)\n",
    "    library(ggplot2)\n",
    "\n",
    "    # co-expression network analysis packages:\n",
    "    library(WGCNA)\n",
    "    library(hdWGCNA)\n",
    "\n",
    "    # gene enrichment packages\n",
    "    library(enrichR)\n",
    "    library(GeneOverlap)\n",
    "    library(GSEABase)\n",
    "    library(GSVA) \n",
    "\n",
    "    # cell-cell communication\n",
    "    library(nichenetr)\n",
    "\n",
    "# needs to be run every time you start R and want to use %>%\n",
    "})\n",
    "\n",
    "# using the cowplot theme for ggplot\n",
    "theme_set(theme_cowplot())\n",
    "\n",
    "# set random seed for reproducibility\n",
    "set.seed(12345)\n",
    "\n",
    "# optionally enable multithreading\n",
    "enableWGCNAThreads(nThreads = 40)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a070e8e",
   "metadata": {},
   "source": [
    "## **Data Prep Parameters**\n",
    "\n",
    "- `test_names`: List of the different test names of interest.\n",
    "\n",
    "- `save_prefix`: Preferred prefix for saving critical files. Ideally chosen to be in the format `{source name}_{brain region}`. e.g `mathys_pfc`\n",
    "\n",
    "- `subject_id`: Column name for Subject/Patient ID in both metadata and `.obs`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "53693e9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "save_prefix = 'seaad_mtg'                                       # this takes the format '{StudyName}_{ThreeLetterAccronymForBrainRegion}'\n",
    "subject_id = helper_functions.clean_strings('Donor ID')\n",
    "cell_type_column = 'Subclass'                       # 'Supertype (non-expanded)', 'Subclass'\n",
    "pseudobulk_level = 'Supertype'                      # Level at which cells are aggregated into pseudobulks,\n",
    "factor = 'Continuous Pseudo-progression Score'    # pathology.group # Continuous Pseudoprogression Score\n",
    "factor = helper_functions.clean_strings(factor, replace_hyphen=True)\n",
    "test_names = ['late_vs_early']                                  # test categories\n",
    "region_name = save_prefix.split('_')[-1].upper()\n",
    "count_agg_strategy = 'metacell'   # options 'network', 'random', 'pseudobulk', 'smaller_network', 'standardbulk', 'metacell', `blanchardbulk``\n",
    "\n",
    "\n",
    "data_dir = f'/media/tadeoye/Volume1/data/seq/SEA-AD/{region_name}/RNAseq/'\n",
    "save_dir = f'../results/'\n",
    "\n",
    "if not os.path.exists(save_dir):\n",
    "    os.makedirs(save_dir)\n",
    "\n",
    "subclass = {\n",
    "    'excitatory': ['L5 IT', 'L2/3 IT', 'L4 IT', 'L6 IT', 'L6 IT Car3', 'L5/6 NP', 'L6b', 'L6 CT', 'L5 ET'],\n",
    "    'inhibitory': ['Pvalb', 'Sst', 'Lamp5 Lhx6', 'Vip', 'Lamp5', 'Sncg', 'Chandelier', 'Sst Chodl', 'Pax6'],\n",
    "    'astrocyte': ['Astrocyte'],\n",
    "    'microglia': ['Microglia-PVM'],\n",
    "    'opc': ['OPC'],\n",
    "    'oligodendrocyte': ['Oligodendrocyte'],\n",
    "    'endothelial': ['Endothelial'],\n",
    "    'vlmc': ['VLMC'],\n",
    "    }\n",
    "\n",
    "cell_supertypes = list(chain(*list(subclass.values())))\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fae59c52",
   "metadata": {},
   "source": [
    "# **Load Merged Pseudo-bulked Data**\n",
    "\n",
    "We have merged the pseduobulked data for all cell sypertypes (`02_merge_pseudobulks.ipynb`). Now, we load the merged data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "161b7f91",
   "metadata": {},
   "outputs": [],
   "source": [
    "clean_subclass = robjects.ListVector({k: [helper_functions.clean_strings(v, preserve_case=True) for v in val] for k, val in subclass.items()})\n",
    "clean_cell_supertypes = [helper_functions.clean_strings(cell_type, preserve_case=True) for cell_type in cell_supertypes]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "df627c04",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_merged = sc.read_h5ad(data_dir+f'counts/anndata/all_subclass_{count_agg_strategy}_anndata.h5ad')\n",
    "adata_merged.X = csr_matrix(adata_merged.X)\n",
    "\n",
    "adata_annot = sc.read_h5ad(os.path.join(data_dir, f'counts/anndata/{save_prefix.upper()}_RNAseq_final-nuclei.2024-02-13.h5ad'))\n",
    "adata_annot.X = adata_annot.layers['UMIs'] \n",
    "\n",
    "Subclass = robjects.ListVector(subclass)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b9789076",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Merged data loaded\"\n",
      "class: SingleCellExperiment \n",
      "dim: 36601 111258 \n",
      "metadata(0):\n",
      "assays(2): X logcounts\n",
      "rownames(36601): MIR1302-2HG FAM138A ... AC007325.4 AC007325.2\n",
      "rowData names(0):\n",
      "colnames(111258): L5 IT_1#H19.33.004_1 L5 IT_1#H19.33.004_2 ...\n",
      "  VLMC_1#H21.33.027_5 VLMC_1#H21.33.027_6\n",
      "colData names(21): donor_id orig.ident ... pseudo_genes_detected\n",
      "  pseudo_number_of_umis\n",
      "reducedDimNames(0):\n",
      "mainExpName: NULL\n",
      "altExpNames(0):\n"
     ]
    }
   ],
   "source": [
    "%%R -i adata_merged -i count_agg_strategy -i data_dir -i subject_id -i Subclass -i cell_supertypes -i clean_cell_supertypes -i clean_subclass\n",
    "\n",
    "source('../scripts/functions/helper_functions.r')\n",
    "\n",
    "print('Merged data loaded')\n",
    "adata_merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "88eec09d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# del adata_merged, summed_counts_per_celltype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "e1bb460e",
   "metadata": {},
   "outputs": [],
   "source": [
    "n_cores = 40\n",
    "\n",
    "covariates = [\n",
    "    'Continuous Pseudo-progression Score',\n",
    "    'Age at Death binned codes',\n",
    "    'Sex',\n",
    "    'APOE4_Status',\n",
    "    'Pseudo Genes Detected',\n",
    "    'Pseudo Number of UMIs',\n",
    "    # 'PMI',\n",
    "    'method',\n",
    "    'Race (choice=White)',\n",
    "]\n",
    "\n",
    "covariates = helper_functions.clean_strings(covariates, replace_hyphen=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b39695ba",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_csv from `anndata` is deprecated. Import anndata.io.read_csv instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_text from `anndata` is deprecated. Import anndata.io.read_text instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_excel from `anndata` is deprecated. Import anndata.io.read_excel instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_mtx from `anndata` is deprecated. Import anndata.io.read_mtx instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_loom from `anndata` is deprecated. Import anndata.io.read_loom instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_hdf from `anndata` is deprecated. Import anndata.io.read_hdf instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_umi_tools from `anndata` is deprecated. Import anndata.io.read_umi_tools instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_csv from `anndata` is deprecated. Import anndata.io.read_csv instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_text from `anndata` is deprecated. Import anndata.io.read_text instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_excel from `anndata` is deprecated. Import anndata.io.read_excel instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_mtx from `anndata` is deprecated. Import anndata.io.read_mtx instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_loom from `anndata` is deprecated. Import anndata.io.read_loom instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_hdf from `anndata` is deprecated. Import anndata.io.read_hdf instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_umi_tools from `anndata` is deprecated. Import anndata.io.read_umi_tools instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_csv from `anndata` is deprecated. Import anndata.io.read_csv instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_text from `anndata` is deprecated. Import anndata.io.read_text instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_excel from `anndata` is deprecated. Import anndata.io.read_excel instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_mtx from `anndata` is deprecated. Import anndata.io.read_mtx instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_loom from `anndata` is deprecated. Import anndata.io.read_loom instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_hdf from `anndata` is deprecated. Import anndata.io.read_hdf instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_umi_tools from `anndata` is deprecated. Import anndata.io.read_umi_tools instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_csv from `anndata` is deprecated. Import anndata.io.read_csv instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_text from `anndata` is deprecated. Import anndata.io.read_text instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_excel from `anndata` is deprecated. Import anndata.io.read_excel instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_mtx from `anndata` is deprecated. Import anndata.io.read_mtx instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_loom from `anndata` is deprecated. Import anndata.io.read_loom instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_hdf from `anndata` is deprecated. Import anndata.io.read_hdf instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_umi_tools from `anndata` is deprecated. Import anndata.io.read_umi_tools instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_csv from `anndata` is deprecated. Import anndata.io.read_csv instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_text from `anndata` is deprecated. Import anndata.io.read_text instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_excel from `anndata` is deprecated. Import anndata.io.read_excel instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_mtx from `anndata` is deprecated. Import anndata.io.read_mtx instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_loom from `anndata` is deprecated. Import anndata.io.read_loom instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_hdf from `anndata` is deprecated. Import anndata.io.read_hdf instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_umi_tools from `anndata` is deprecated. Import anndata.io.read_umi_tools instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_csv from `anndata` is deprecated. Import anndata.io.read_csv instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_text from `anndata` is deprecated. Import anndata.io.read_text instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_excel from `anndata` is deprecated. Import anndata.io.read_excel instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_mtx from `anndata` is deprecated. Import anndata.io.read_mtx instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_loom from `anndata` is deprecated. Import anndata.io.read_loom instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_hdf from `anndata` is deprecated. Import anndata.io.read_hdf instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_umi_tools from `anndata` is deprecated. Import anndata.io.read_umi_tools instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_csv from `anndata` is deprecated. Import anndata.io.read_csv instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_text from `anndata` is deprecated. Import anndata.io.read_text instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_excel from `anndata` is deprecated. Import anndata.io.read_excel instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_mtx from `anndata` is deprecated. Import anndata.io.read_mtx instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_loom from `anndata` is deprecated. Import anndata.io.read_loom instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_hdf from `anndata` is deprecated. Import anndata.io.read_hdf instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_umi_tools from `anndata` is deprecated. Import anndata.io.read_umi_tools instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_csv from `anndata` is deprecated. Import anndata.io.read_csv instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_text from `anndata` is deprecated. Import anndata.io.read_text instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_excel from `anndata` is deprecated. Import anndata.io.read_excel instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_mtx from `anndata` is deprecated. Import anndata.io.read_mtx instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_loom from `anndata` is deprecated. Import anndata.io.read_loom instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_hdf from `anndata` is deprecated. Import anndata.io.read_hdf instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_umi_tools from `anndata` is deprecated. Import anndata.io.read_umi_tools instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_csv from `anndata` is deprecated. Import anndata.io.read_csv instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_text from `anndata` is deprecated. Import anndata.io.read_text instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_excel from `anndata` is deprecated. Import anndata.io.read_excel instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_mtx from `anndata` is deprecated. Import anndata.io.read_mtx instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_loom from `anndata` is deprecated. Import anndata.io.read_loom instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_hdf from `anndata` is deprecated. Import anndata.io.read_hdf instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_umi_tools from `anndata` is deprecated. Import anndata.io.read_umi_tools instead.\n",
      "  warnings.warn(msg, FutureWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "    WARNING: The R package \"reticulate\" only fixed recently\n",
      "    an issue that caused a segfault when used with rpy2:\n",
      "    https://github.com/rstudio/reticulate/pull/1188\n",
      "    Make sure that you use a version of that package that includes\n",
      "    the fix.\n",
      "    \n",
      "    WARNING: The R package \"reticulate\" only fixed recently\n",
      "    an issue that caused a segfault when used with rpy2:\n",
      "    https://github.com/rstudio/reticulate/pull/1188\n",
      "    Make sure that you use a version of that package that includes\n",
      "    the fix.\n",
      "    "
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_csv from `anndata` is deprecated. Import anndata.io.read_csv instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_text from `anndata` is deprecated. Import anndata.io.read_text instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_excel from `anndata` is deprecated. Import anndata.io.read_excel instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_mtx from `anndata` is deprecated. Import anndata.io.read_mtx instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_loom from `anndata` is deprecated. Import anndata.io.read_loom instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_hdf from `anndata` is deprecated. Import anndata.io.read_hdf instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_umi_tools from `anndata` is deprecated. Import anndata.io.read_umi_tools instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_csv from `anndata` is deprecated. Import anndata.io.read_csv instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_text from `anndata` is deprecated. Import anndata.io.read_text instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_excel from `anndata` is deprecated. Import anndata.io.read_excel instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_mtx from `anndata` is deprecated. Import anndata.io.read_mtx instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_loom from `anndata` is deprecated. Import anndata.io.read_loom instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_hdf from `anndata` is deprecated. Import anndata.io.read_hdf instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_umi_tools from `anndata` is deprecated. Import anndata.io.read_umi_tools instead.\n",
      "  warnings.warn(msg, FutureWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "    WARNING: The R package \"reticulate\" only fixed recently\n",
      "    an issue that caused a segfault when used with rpy2:\n",
      "    https://github.com/rstudio/reticulate/pull/1188\n",
      "    Make sure that you use a version of that package that includes\n",
      "    the fix.\n",
      "    "
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_csv from `anndata` is deprecated. Import anndata.io.read_csv instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_text from `anndata` is deprecated. Import anndata.io.read_text instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_excel from `anndata` is deprecated. Import anndata.io.read_excel instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_mtx from `anndata` is deprecated. Import anndata.io.read_mtx instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_loom from `anndata` is deprecated. Import anndata.io.read_loom instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_hdf from `anndata` is deprecated. Import anndata.io.read_hdf instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_umi_tools from `anndata` is deprecated. Import anndata.io.read_umi_tools instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_csv from `anndata` is deprecated. Import anndata.io.read_csv instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_text from `anndata` is deprecated. Import anndata.io.read_text instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_excel from `anndata` is deprecated. Import anndata.io.read_excel instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_mtx from `anndata` is deprecated. Import anndata.io.read_mtx instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_loom from `anndata` is deprecated. Import anndata.io.read_loom instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_hdf from `anndata` is deprecated. Import anndata.io.read_hdf instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_umi_tools from `anndata` is deprecated. Import anndata.io.read_umi_tools instead.\n",
      "  warnings.warn(msg, FutureWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "    WARNING: The R package \"reticulate\" only fixed recently\n",
      "    an issue that caused a segfault when used with rpy2:\n",
      "    https://github.com/rstudio/reticulate/pull/1188\n",
      "    Make sure that you use a version of that package that includes\n",
      "    the fix.\n",
      "    "
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_csv from `anndata` is deprecated. Import anndata.io.read_csv instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_text from `anndata` is deprecated. Import anndata.io.read_text instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_excel from `anndata` is deprecated. Import anndata.io.read_excel instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_mtx from `anndata` is deprecated. Import anndata.io.read_mtx instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_loom from `anndata` is deprecated. Import anndata.io.read_loom instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_hdf from `anndata` is deprecated. Import anndata.io.read_hdf instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_umi_tools from `anndata` is deprecated. Import anndata.io.read_umi_tools instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_csv from `anndata` is deprecated. Import anndata.io.read_csv instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_text from `anndata` is deprecated. Import anndata.io.read_text instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_excel from `anndata` is deprecated. Import anndata.io.read_excel instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_mtx from `anndata` is deprecated. Import anndata.io.read_mtx instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_loom from `anndata` is deprecated. Import anndata.io.read_loom instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_hdf from `anndata` is deprecated. Import anndata.io.read_hdf instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_umi_tools from `anndata` is deprecated. Import anndata.io.read_umi_tools instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_csv from `anndata` is deprecated. Import anndata.io.read_csv instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_text from `anndata` is deprecated. Import anndata.io.read_text instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_excel from `anndata` is deprecated. Import anndata.io.read_excel instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_mtx from `anndata` is deprecated. Import anndata.io.read_mtx instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_loom from `anndata` is deprecated. Import anndata.io.read_loom instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_hdf from `anndata` is deprecated. Import anndata.io.read_hdf instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_umi_tools from `anndata` is deprecated. Import anndata.io.read_umi_tools instead.\n",
      "  warnings.warn(msg, FutureWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "    WARNING: The R package \"reticulate\" only fixed recently\n",
      "    an issue that caused a segfault when used with rpy2:\n",
      "    https://github.com/rstudio/reticulate/pull/1188\n",
      "    Make sure that you use a version of that package that includes\n",
      "    the fix.\n",
      "    \n",
      "    WARNING: The R package \"reticulate\" only fixed recently\n",
      "    an issue that caused a segfault when used with rpy2:\n",
      "    https://github.com/rstudio/reticulate/pull/1188\n",
      "    Make sure that you use a version of that package that includes\n",
      "    the fix.\n",
      "    "
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_csv from `anndata` is deprecated. Import anndata.io.read_csv instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_text from `anndata` is deprecated. Import anndata.io.read_text instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_excel from `anndata` is deprecated. Import anndata.io.read_excel instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_mtx from `anndata` is deprecated. Import anndata.io.read_mtx instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_loom from `anndata` is deprecated. Import anndata.io.read_loom instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_hdf from `anndata` is deprecated. Import anndata.io.read_hdf instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_umi_tools from `anndata` is deprecated. Import anndata.io.read_umi_tools instead.\n",
      "  warnings.warn(msg, FutureWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "    WARNING: The R package \"reticulate\" only fixed recently\n",
      "    an issue that caused a segfault when used with rpy2:\n",
      "    https://github.com/rstudio/reticulate/pull/1188\n",
      "    Make sure that you use a version of that package that includes\n",
      "    the fix.\n",
      "    "
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_csv from `anndata` is deprecated. Import anndata.io.read_csv instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_text from `anndata` is deprecated. Import anndata.io.read_text instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_excel from `anndata` is deprecated. Import anndata.io.read_excel instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_mtx from `anndata` is deprecated. Import anndata.io.read_mtx instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_loom from `anndata` is deprecated. Import anndata.io.read_loom instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_hdf from `anndata` is deprecated. Import anndata.io.read_hdf instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_umi_tools from `anndata` is deprecated. Import anndata.io.read_umi_tools instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_csv from `anndata` is deprecated. Import anndata.io.read_csv instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_text from `anndata` is deprecated. Import anndata.io.read_text instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_excel from `anndata` is deprecated. Import anndata.io.read_excel instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_mtx from `anndata` is deprecated. Import anndata.io.read_mtx instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_loom from `anndata` is deprecated. Import anndata.io.read_loom instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_hdf from `anndata` is deprecated. Import anndata.io.read_hdf instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_umi_tools from `anndata` is deprecated. Import anndata.io.read_umi_tools instead.\n",
      "  warnings.warn(msg, FutureWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "    WARNING: The R package \"reticulate\" only fixed recently\n",
      "    an issue that caused a segfault when used with rpy2:\n",
      "    https://github.com/rstudio/reticulate/pull/1188\n",
      "    Make sure that you use a version of that package that includes\n",
      "    the fix.\n",
      "    "
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: here() starts at /home/tadeoye/Documents/research_codes/cell_cell_communication\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-21 13:27:48.718252 -- No outgroup specified, running analysis for all groups and tests\n",
      "\n",
      "    WARNING: The R package \"reticulate\" only fixed recently\n",
      "    an issue that caused a segfault when used with rpy2:\n",
      "    https://github.com/rstudio/reticulate/pull/1188\n",
      "    Make sure that you use a version of that package that includes\n",
      "    the fix.\n",
      "    "
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_csv from `anndata` is deprecated. Import anndata.io.read_csv instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_text from `anndata` is deprecated. Import anndata.io.read_text instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_excel from `anndata` is deprecated. Import anndata.io.read_excel instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_mtx from `anndata` is deprecated. Import anndata.io.read_mtx instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_loom from `anndata` is deprecated. Import anndata.io.read_loom instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_hdf from `anndata` is deprecated. Import anndata.io.read_hdf instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_umi_tools from `anndata` is deprecated. Import anndata.io.read_umi_tools instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_csv from `anndata` is deprecated. Import anndata.io.read_csv instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_text from `anndata` is deprecated. Import anndata.io.read_text instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_excel from `anndata` is deprecated. Import anndata.io.read_excel instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_mtx from `anndata` is deprecated. Import anndata.io.read_mtx instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_loom from `anndata` is deprecated. Import anndata.io.read_loom instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_hdf from `anndata` is deprecated. Import anndata.io.read_hdf instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "R[write to console]: here() starts at /home/tadeoye/Documents/research_codes/cell_cell_communication\n",
      "\n",
      "/home/tadeoye/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_umi_tools from `anndata` is deprecated. Import anndata.io.read_umi_tools instead.\n",
      "  warnings.warn(msg, FutureWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-21 13:27:50.040343 -- No outgroup specified, running analysis for all groups and tests\n",
      "\n",
      "    WARNING: The R package \"reticulate\" only fixed recently\n",
      "    an issue that caused a segfault when used with rpy2:\n",
      "    https://github.com/rstudio/reticulate/pull/1188\n",
      "    Make sure that you use a version of that package that includes\n",
      "    the fix.\n",
      "    2025-01-21 13:27:50.713590 -- Removing 3697 features from L5_IT_1 for low numbers of counts per cell\n",
      "2025-01-21 13:27:50.848823 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:27:50.850709 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:27:50.909977 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.25 for 1476 genes in L5_IT_1 because all values are 0 for that covariate\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: here() starts at /home/tadeoye/Documents/research_codes/cell_cell_communication\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-21 13:27:51.736980 -- No outgroup specified, running analysis for all groups and tests\n",
      "\n",
      "    WARNING: The R package \"reticulate\" only fixed recently\n",
      "    an issue that caused a segfault when used with rpy2:\n",
      "    https://github.com/rstudio/reticulate/pull/1188\n",
      "    Make sure that you use a version of that package that includes\n",
      "    the fix.\n",
      "    \n",
      "    WARNING: The R package \"reticulate\" only fixed recently\n",
      "    an issue that caused a segfault when used with rpy2:\n",
      "    https://github.com/rstudio/reticulate/pull/1188\n",
      "    Make sure that you use a version of that package that includes\n",
      "    the fix.\n",
      "    2025-01-21 13:27:52.565351 -- Removing 4218 features from L4_IT_1 for low numbers of counts per cell\n",
      "2025-01-21 13:27:52.706466 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:27:52.707825 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:27:52.759539 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.25 for 885 genes in L4_IT_1 because all values are 0 for that covariate\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: here() starts at /home/tadeoye/Documents/research_codes/cell_cell_communication\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-21 13:27:53.079388 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 13:27:53.134722 -- Removing 7505 features from L6_IT_Car3_1 for low numbers of counts per cell\n",
      "2025-01-21 13:27:53.143022 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:27:53.159398 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.0 for 4357 genes in L6_IT_Car3_1 because all values are 0 for that covariate\n",
      "\n",
      "    WARNING: The R package \"reticulate\" only fixed recently\n",
      "    an issue that caused a segfault when used with rpy2:\n",
      "    https://github.com/rstudio/reticulate/pull/1188\n",
      "    Make sure that you use a version of that package that includes\n",
      "    the fix.\n",
      "    \n",
      "    WARNING: The R package \"reticulate\" only fixed recently\n",
      "    an issue that caused a segfault when used with rpy2:\n",
      "    https://github.com/rstudio/reticulate/pull/1188\n",
      "    Make sure that you use a version of that package that includes\n",
      "    the fix.\n",
      "    2025-01-21 13:27:54.342556 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 13:27:54.438180 -- No outgroup specified, running analysis for all groups and tests\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: here() starts at /home/tadeoye/Documents/research_codes/cell_cell_communication\n",
      "\n",
      "R[write to console]: here() starts at /home/tadeoye/Documents/research_codes/cell_cell_communication\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-21 13:27:54.707096 -- Removing 4471 features from L5_6_NP_1 for low numbers of counts per cell\n",
      "2025-01-21 13:27:54.739760 -- Removing 4169 features from L6_IT_1 for low numbers of counts per cell\n",
      "2025-01-21 13:27:54.766883 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:27:54.768220 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:27:54.798369 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:27:54.800265 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:27:54.799585 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.25 for 2053 genes in L5_6_NP_1 because all values are 0 for that covariate\n",
      "2025-01-21 13:27:54.834197 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.25 for 1528 genes in L6_IT_1 because all values are 0 for that covariate\n",
      "\n",
      "    WARNING: The R package \"reticulate\" only fixed recently\n",
      "    an issue that caused a segfault when used with rpy2:\n",
      "    https://github.com/rstudio/reticulate/pull/1188\n",
      "    Make sure that you use a version of that package that includes\n",
      "    the fix.\n",
      "    \n",
      "    WARNING: The R package \"reticulate\" only fixed recently\n",
      "    an issue that caused a segfault when used with rpy2:\n",
      "    https://github.com/rstudio/reticulate/pull/1188\n",
      "    Make sure that you use a version of that package that includes\n",
      "    the fix.\n",
      "    "
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: here() starts at /home/tadeoye/Documents/research_codes/cell_cell_communication\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-21 13:27:56.936526 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 13:27:57.198544 -- Removing 5968 features from Pvalb_1 for low numbers of counts per cell\n",
      "2025-01-21 13:27:57.243447 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:27:57.292248 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.8095932638239676 for 2905 genes in Pvalb_1 because all values are 0 for that covariate\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: here() starts at /home/tadeoye/Documents/research_codes/cell_cell_communication\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-21 13:27:57.775866 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 13:27:57.900479 -- Removing 4622 features from L6_CT_1 for low numbers of counts per cell\n",
      "2025-01-21 13:27:57.915561 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:27:57.916952 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:27:57.935106 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.25 for 2299 genes in L6_CT_1 because all values are 0 for that covariate\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: here() starts at /home/tadeoye/Documents/research_codes/cell_cell_communication\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-21 13:27:58.238245 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 13:27:58.279042 -- Removing 11698 features from L6b_2 for low numbers of counts per cell\n",
      "2025-01-21 13:27:58.282031 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:27:58.282619 -- Removing continuous_pseudo_progression_score from the covariate formula\n",
      "2025-01-21 13:27:58.283791 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:27:58.284199 -- Removing age_at_death_binned_codes from the covariate formula\n",
      "2025-01-21 13:27:58.284741 -- Removing sex from the covariate formula\n",
      "2025-01-21 13:27:58.285022 -- Removing apoe4_status from the covariate formula\n",
      "2025-01-21 13:27:58.286112 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:27:58.299699 -- Adding 3 pseudocounts to pseudo_genes_detected=0.0 for 5399 genes in L6b_2 because all values are 0 for that covariate\n",
      "\n",
      "    WARNING: The R package \"reticulate\" only fixed recently\n",
      "    an issue that caused a segfault when used with rpy2:\n",
      "    https://github.com/rstudio/reticulate/pull/1188\n",
      "    Make sure that you use a version of that package that includes\n",
      "    the fix.\n",
      "    "
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: here() starts at /home/tadeoye/Documents/research_codes/cell_cell_communication\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-21 13:27:59.504843 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 13:27:59.533555 -- Removing 16021 features from Sst_2 for low numbers of counts per cell\n",
      "2025-01-21 13:27:59.536573 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:27:59.536959 -- Removing continuous_pseudo_progression_score from the covariate formula\n",
      "2025-01-21 13:27:59.537990 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:27:59.538230 -- Removing age_at_death_binned_codes from the covariate formula\n",
      "2025-01-21 13:27:59.538592 -- Removing sex from the covariate formula\n",
      "2025-01-21 13:27:59.538835 -- Removing apoe4_status from the covariate formula\n",
      "2025-01-21 13:27:59.539876 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:27:59.553788 -- Adding 3 pseudocounts to pseudo_genes_detected=0.2540227806906527 for 3420 genes in Sst_2 because all values are 0 for that covariate\n",
      "\n",
      "    WARNING: The R package \"reticulate\" only fixed recently\n",
      "    an issue that caused a segfault when used with rpy2:\n",
      "    https://github.com/rstudio/reticulate/pull/1188\n",
      "    Make sure that you use a version of that package that includes\n",
      "    the fix.\n",
      "    \n",
      "    WARNING: The R package \"reticulate\" only fixed recently\n",
      "    an issue that caused a segfault when used with rpy2:\n",
      "    https://github.com/rstudio/reticulate/pull/1188\n",
      "    Make sure that you use a version of that package that includes\n",
      "    the fix.\n",
      "    "
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: here() starts at /home/tadeoye/Documents/research_codes/cell_cell_communication\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-21 13:28:00.520167 -- No outgroup specified, running analysis for all groups and tests\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: here() starts at /home/tadeoye/Documents/research_codes/cell_cell_communication\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-21 13:28:00.922758 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 13:28:01.299262 -- Removing 5228 features from Vip_1 for low numbers of counts per cell\n",
      "2025-01-21 13:28:01.366303 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:28:01.367593 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:28:01.393096 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.25 for 1636 genes in Vip_1 because all values are 0 for that covariate\n",
      "2025-01-21 13:28:01.501952 -- Removing 4728 features from Lamp5_Lhx6_1 for low numbers of counts per cell\n",
      "2025-01-21 13:28:01.702774 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:28:01.704348 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:28:01.780744 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.25 for 1116 genes in Lamp5_Lhx6_1 because all values are 0 for that covariate\n",
      "\n",
      "    WARNING: The R package \"reticulate\" only fixed recently\n",
      "    an issue that caused a segfault when used with rpy2:\n",
      "    https://github.com/rstudio/reticulate/pull/1188\n",
      "    Make sure that you use a version of that package that includes\n",
      "    the fix.\n",
      "    \n",
      "    WARNING: The R package \"reticulate\" only fixed recently\n",
      "    an issue that caused a segfault when used with rpy2:\n",
      "    https://github.com/rstudio/reticulate/pull/1188\n",
      "    Make sure that you use a version of that package that includes\n",
      "    the fix.\n",
      "    "
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: here() starts at /home/tadeoye/Documents/research_codes/cell_cell_communication\n",
      "\n",
      "R[write to console]: here() starts at /home/tadeoye/Documents/research_codes/cell_cell_communication\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-21 13:28:02.880584 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 13:28:02.913042 -- Removing 11945 features from Sncg_1 for low numbers of counts per cell\n",
      "2025-01-21 13:28:02.916973 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:28:02.930314 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.7107407372495411 for 5044 genes in Sncg_1 because all values are 0 for that covariate\n",
      "2025-01-21 13:28:03.065324 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 13:28:03.112941 -- Removing 10503 features from Lamp5_1 for low numbers of counts per cell\n",
      "2025-01-21 13:28:03.117155 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:28:03.131686 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.0 for 7536 genes in Lamp5_1 because all values are 0 for that covariate\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: here() starts at /home/tadeoye/Documents/research_codes/cell_cell_communication\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-21 13:28:03.616681 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 13:28:03.691375 -- Removing 8227 features from Pax6_1 for low numbers of counts per cell\n",
      "2025-01-21 13:28:03.700126 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:28:03.716642 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.103825014087543 for 7178 genes in Pax6_1 because all values are 0 for that covariate\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: here() starts at /home/tadeoye/Documents/research_codes/cell_cell_communication\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-21 13:28:05.181442 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 13:28:05.370962 -- Removing 6615 features from Chandelier_1 for low numbers of counts per cell\n",
      "2025-01-21 13:28:05.393980 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:28:05.395264 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:28:05.414968 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.25 for 2566 genes in Chandelier_1 because all values are 0 for that covariate\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: here() starts at /home/tadeoye/Documents/research_codes/cell_cell_communication\n",
      "\n",
      "R[write to console]: here() starts at /home/tadeoye/Documents/research_codes/cell_cell_communication\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-21 13:28:08.068388 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 13:28:08.267820 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 13:28:08.300938 -- Removing 19813 features from Endo_2 for low numbers of counts per cell\n",
      "2025-01-21 13:28:08.303917 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:28:08.304327 -- Removing continuous_pseudo_progression_score from the covariate formula\n",
      "2025-01-21 13:28:08.305423 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:28:08.305662 -- Removing age_at_death_binned_codes from the covariate formula\n",
      "2025-01-21 13:28:08.306013 -- Removing sex from the covariate formula\n",
      "2025-01-21 13:28:08.306230 -- Removing apoe4_status from the covariate formula\n",
      "2025-01-21 13:28:08.307325 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:28:08.321196 -- Adding 3 pseudocounts to pseudo_genes_detected=0.7076335877862595 for 3586 genes in Endo_2 because all values are 0 for that covariate\n",
      "2025-01-21 13:28:08.496988 -- Removing 7219 features from Oligo_1 for low numbers of counts per cell\n",
      "2025-01-21 13:28:08.571889 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:28:08.573347 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:28:08.601538 -- Adding 3 pseudocounts to age_at_death_binned_codes=1.0 for 1807 genes in Oligo_1 because all values are 0 for that covariate\n",
      "2025-01-21 13:28:08.886876 -- Removing 4004 features from L2_3_IT_1 for low numbers of counts per cell\n",
      "2025-01-21 13:28:09.035920 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 13:28:09.066520 -- Removing 15747 features from VLMC_1 for low numbers of counts per cell\n",
      "2025-01-21 13:28:09.069584 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: here() starts at /home/tadeoye/Documents/research_codes/cell_cell_communication\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-21 13:28:09.082551 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.2331134537922132 for 2817 genes in VLMC_1 because all values are 0 for that covariate\n",
      "2025-01-21 13:28:12.906560 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:28:12.908974 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:28:13.717519 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.25 for 9 genes in L2_3_IT_1 because all values are 0 for that covariate\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: here() starts at /home/tadeoye/Documents/research_codes/cell_cell_communication\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-21 13:28:14.710580 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 13:28:14.897154 -- Removing 6155 features from Astro_1 for low numbers of counts per cell\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: here() starts at /home/tadeoye/Documents/research_codes/cell_cell_communication\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-21 13:28:14.927143 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:28:14.928847 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:28:14.932982 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 13:28:14.969681 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.25 for 1963 genes in Astro_1 because all values are 0 for that covariate\n",
      "2025-01-21 13:28:16.848134 -- Removing 5840 features from OPC_2 for low numbers of counts per cell\n",
      "2025-01-21 13:28:17.230761 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:28:17.232653 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:28:17.372132 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.25 for 444 genes in OPC_2 because all values are 0 for that covariate\n",
      "2025-01-21 13:28:37.103181 -- Adding 3 pseudocounts to pseudo_genes_detected=0.07882842162357624 for 4802 genes in Sst_2 because all values are 0 for that covariate\n",
      "2025-01-21 13:28:45.756041 -- Adding 3 pseudocounts to pseudo_genes_detected=0.3015267175572519 for 4118 genes in Endo_2 because all values are 0 for that covariate\n",
      "2025-01-21 13:28:57.556101 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.0 for 2053 genes in VLMC_1 because all values are 0 for that covariate\n",
      "2025-01-21 13:29:06.885968 -- Adding 3 pseudocounts to pseudo_genes_detected=0.31306110628439043 for 3483 genes in L6b_2 because all values are 0 for that covariate\n",
      "2025-01-21 13:29:28.545319 -- Adding 3 pseudocounts to pseudo_genes_detected=1.0 for 3203 genes in Endo_2 because all values are 0 for that covariate\n",
      "2025-01-21 13:29:30.024728 -- Adding 3 pseudocounts to pseudo_genes_detected=0.0 for 5769 genes in Sst_2 because all values are 0 for that covariate\n",
      "2025-01-21 13:29:37.968852 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:29:38.015511 -- Removing apoe4_status from the covariate formula\n",
      "2025-01-21 13:29:38.016443 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:29:38.017315 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:29:38.017722 -- Removing method from the covariate formula\n",
      "2025-01-21 13:29:38.017935 -- Removing race_choice_white from the covariate formula\n",
      "2025-01-21 13:29:38.071736 -- Skipping VLMC_1 along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-21 13:29:46.009095 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.7155889517143134 for 4069 genes in Sncg_1 because all values are 0 for that covariate\n",
      "2025-01-21 13:29:50.492334 -- Adding 3 pseudocounts to pseudo_genes_detected=0.25079640891977995 for 3694 genes in L6b_2 because all values are 0 for that covariate\n",
      "2025-01-21 13:30:02.172876 -- Adding 3 pseudocounts to pseudo_genes_detected=0.0 for 4513 genes in Endo_2 because all values are 0 for that covariate\n",
      "2025-01-21 13:30:35.844646 -- Adding 3 pseudocounts to pseudo_genes_detected=0.30356174290363414 for 4408 genes in Sst_2 because all values are 0 for that covariate\n",
      "2025-01-21 13:30:36.930401 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.6626252004619664 for 6622 genes in L6_IT_Car3_1 because all values are 0 for that covariate\n",
      "2025-01-21 13:30:38.350561 -- Adding 3 pseudocounts to pseudo_genes_detected=0.20909354184766876 for 4635 genes in L6b_2 because all values are 0 for that covariate\n",
      "2025-01-21 13:30:50.828049 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:30:50.869828 -- Removing method from the covariate formula\n",
      "2025-01-21 13:30:50.869942 -- Removing race_choice_white from the covariate formula\n",
      "2025-01-21 13:30:50.915594 -- Starting Endo_2 across continuous_pseudo_progression_score\n",
      "The cells are already grouped.2025-01-21 13:30:50.974336 -- Skipping Endo_2 because some predictors have zero variation or a zero vector\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: Error in nebula(counts, obs$donor_id, pred = df, offset = obs$pseudo_number_of_umis,  : \n",
      "  Some predictors have zero variation or a zero vector.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-21 13:31:10.046384 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.0 for 3075 genes in Sncg_1 because all values are 0 for that covariate\n",
      "2025-01-21 13:31:26.951521 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:31:26.991080 -- Removing method from the covariate formula\n",
      "2025-01-21 13:31:26.991195 -- Removing race_choice_white from the covariate formula\n",
      "2025-01-21 13:31:27.032148 -- Starting Sst_2 across continuous_pseudo_progression_score\n",
      "The cells are already grouped.2025-01-21 13:31:27.098640 -- Skipping Sst_2 because some predictors have zero variation or a zero vector\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: Error in nebula(counts, obs$donor_id, pred = df, offset = obs$pseudo_number_of_umis,  : \n",
      "  Some predictors have zero variation or a zero vector.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-21 13:31:28.230507 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 13:31:28.532481 -- Removing 5772 features from Sst_3 for low numbers of counts per cell\n",
      "2025-01-21 13:31:28.583692 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:31:28.585210 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:31:28.612314 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.25 for 1686 genes in Sst_3 because all values are 0 for that covariate\n",
      "2025-01-21 13:31:32.941771 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.3327136446937277 for 8447 genes in Lamp5_1 because all values are 0 for that covariate\n",
      "2025-01-21 13:31:38.433136 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:31:38.474615 -- Removing method from the covariate formula\n",
      "2025-01-21 13:31:38.474744 -- Removing race_choice_white from the covariate formula\n",
      "2025-01-21 13:31:38.517644 -- Starting L6b_2 across continuous_pseudo_progression_score\n",
      "The cells are already grouped.2025-01-21 13:31:38.580403 -- Skipping L6b_2 because some predictors have zero variation or a zero vector\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: Error in nebula(counts, obs$donor_id, pred = df, offset = obs$pseudo_number_of_umis,  : \n",
      "  Some predictors have zero variation or a zero vector.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-21 13:31:40.349449 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 13:31:40.613446 -- Removing 4321 features from L6b_6 for low numbers of counts per cell\n",
      "2025-01-21 13:31:40.655458 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:31:40.657016 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:31:40.692940 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.0 for 1113 genes in L6b_6 because all values are 0 for that covariate\n",
      "2025-01-21 13:32:16.497030 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.0 for 36 genes in L2_3_IT_1 because all values are 0 for that covariate\n",
      "2025-01-21 13:32:23.621066 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:32:23.719479 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:32:23.720609 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:32:23.721079 -- Removing method from the covariate formula\n",
      "2025-01-21 13:32:23.721274 -- Removing race_choice_white from the covariate formula\n",
      "2025-01-21 13:32:23.811665 -- Skipping Sncg_1 along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-21 13:32:25.491084 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 13:32:25.615821 -- Removing 6813 features from Sncg_4 for low numbers of counts per cell\n",
      "2025-01-21 13:32:25.631125 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:32:25.632426 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:32:25.653853 -- Adding 3 pseudocounts to age_at_death_binned_codes=1.0 for 2940 genes in Sncg_4 because all values are 0 for that covariate\n",
      "2025-01-21 13:32:38.053116 -- Adding 3 pseudocounts to age_at_death_binned_codes=1.0 for 2077 genes in L6_CT_1 because all values are 0 for that covariate\n",
      "2025-01-21 13:33:34.678111 -- Adding 3 pseudocounts to age_at_death_binned_codes=1.0 for 3934 genes in Astro_1 because all values are 0 for that covariate\n",
      "2025-01-21 13:34:42.324109 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.07717490666535043 for 2700 genes in L6_IT_Car3_1 because all values are 0 for that covariate\n",
      "2025-01-21 13:34:59.578745 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.5 for 1385 genes in L6_IT_1 because all values are 0 for that covariate\n",
      "2025-01-21 13:35:18.889292 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.0 for 7064 genes in Pax6_1 because all values are 0 for that covariate\n",
      "2025-01-21 13:35:27.305727 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.5938649624864023 for 2635 genes in Lamp5_1 because all values are 0 for that covariate\n",
      "2025-01-21 13:36:19.587815 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.5 for 1953 genes in Chandelier_1 because all values are 0 for that covariate\n",
      "2025-01-21 13:36:24.437206 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.3051331342656901 for 4207 genes in L6_IT_Car3_1 because all values are 0 for that covariate\n",
      "2025-01-21 13:36:26.256532 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.75 for 1268 genes in L6b_6 because all values are 0 for that covariate\n",
      "2025-01-21 13:36:42.034792 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.6321304159502532 for 4268 genes in Lamp5_1 because all values are 0 for that covariate\n",
      "2025-01-21 13:36:47.926076 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.75 for 2095 genes in L6_CT_1 because all values are 0 for that covariate\n",
      "2025-01-21 13:38:40.371250 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.75 for 3018 genes in Sncg_4 because all values are 0 for that covariate\n",
      "2025-01-21 13:38:45.295779 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:38:45.352815 -- Removing apoe4_status from the covariate formula\n",
      "2025-01-21 13:38:45.353778 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:38:45.354653 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:38:45.355086 -- Removing method from the covariate formula\n",
      "2025-01-21 13:38:45.355286 -- Removing race_choice_white from the covariate formula\n",
      "2025-01-21 13:38:45.432835 -- Skipping Lamp5_1 along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-21 13:38:46.492937 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 13:38:47.520014 -- Removing 5368 features from Lamp5_2 for low numbers of counts per cell\n",
      "2025-01-21 13:38:47.728483 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:38:47.729658 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:38:47.802638 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.25 for 859 genes in Lamp5_2 because all values are 0 for that covariate\n",
      "2025-01-21 13:39:04.519142 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:39:04.620228 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:39:04.621051 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:39:04.621391 -- Removing method from the covariate formula\n",
      "2025-01-21 13:39:04.621520 -- Removing race_choice_white from the covariate formula\n",
      "2025-01-21 13:39:04.713754 -- Skipping L6_IT_Car3_1 along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-21 13:39:05.843271 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 13:39:05.913768 -- Removing 5137 features from L6_IT_Car3_2 for low numbers of counts per cell\n",
      "2025-01-21 13:39:05.928345 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:39:05.929544 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:39:05.946074 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.0 for 2289 genes in L6_IT_Car3_2 because all values are 0 for that covariate\n",
      "2025-01-21 13:39:39.877977 -- Adding 3 pseudocounts to age_at_death_binned_codes=1.0 for 2773 genes in Sst_3 because all values are 0 for that covariate\n",
      "2025-01-21 13:39:50.345331 -- Adding 3 pseudocounts to age_at_death_binned_codes=1.0 for 642 genes in L4_IT_1 because all values are 0 for that covariate\n",
      "2025-01-21 13:40:36.205482 -- Adding 3 pseudocounts to age_at_death_binned_codes=1.0 for 1661 genes in L5_6_NP_1 because all values are 0 for that covariate\n",
      "2025-01-21 13:40:59.086400 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.5 for 10419 genes in L6_CT_1 because all values are 0 for that covariate\n",
      "2025-01-21 13:41:21.402031 -- Adding 3 pseudocounts to age_at_death_binned_codes=1.0 for 1027 genes in L6_IT_1 because all values are 0 for that covariate\n",
      "2025-01-21 13:41:45.125995 -- Adding 3 pseudocounts to age_at_death_binned_codes=1.0 for 1210 genes in Vip_1 because all values are 0 for that covariate\n",
      "2025-01-21 13:41:47.154860 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.5 for 1325 genes in L6b_6 because all values are 0 for that covariate\n",
      "2025-01-21 13:41:59.469596 -- Adding 3 pseudocounts to age_at_death_binned_codes=1.0 for 270 genes in OPC_2 because all values are 0 for that covariate\n",
      "2025-01-21 13:42:01.037951 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.75 for 5846 genes in Astro_1 because all values are 0 for that covariate\n",
      "2025-01-21 13:42:03.618343 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.5 for 1739 genes in Oligo_1 because all values are 0 for that covariate\n",
      "2025-01-21 13:42:12.793578 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.0 for 763 genes in Pvalb_1 because all values are 0 for that covariate\n",
      "2025-01-21 13:42:20.474311 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.75 for 2628 genes in L6_IT_Car3_2 because all values are 0 for that covariate\n",
      "2025-01-21 13:42:30.213098 -- Adding 3 pseudocounts to age_at_death_binned_codes=1.0 for 643 genes in L5_IT_1 because all values are 0 for that covariate\n",
      "2025-01-21 13:42:36.787112 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.6008881402401076 for 8867 genes in Pax6_1 because all values are 0 for that covariate\n",
      "2025-01-21 13:42:36.863657 -- Adding 3 pseudocounts to age_at_death_binned_codes=1.0 for 3409 genes in Chandelier_1 because all values are 0 for that covariate\n",
      "2025-01-21 13:45:07.015502 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.5 for 5826 genes in Sncg_4 because all values are 0 for that covariate\n",
      "2025-01-21 13:46:01.095132 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:46:01.172746 -- Removing sex from the covariate formula\n",
      "2025-01-21 13:46:01.240738 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:46:01.241796 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:46:01.242194 -- Removing method from the covariate formula\n",
      "2025-01-21 13:46:01.242334 -- Removing race_choice_white from the covariate formula\n",
      "2025-01-21 13:46:03.356140 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.75 for 1169 genes in L6_IT_1 because all values are 0 for that covariate\n",
      "2025-01-21 13:46:04.136590 -- Skipping Pvalb_1 along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-21 13:46:04.509430 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.25 for 4023 genes in L6_IT_Car3_2 because all values are 0 for that covariate\n",
      "2025-01-21 13:46:05.595746 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 13:46:07.271365 -- Removing 4765 features from Pvalb_2 for low numbers of counts per cell\n",
      "2025-01-21 13:46:07.587123 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:46:07.588743 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:46:07.737113 -- Adding 3 pseudocounts to age_at_death_binned_codes=1.0 for 539 genes in Pvalb_2 because all values are 0 for that covariate\n",
      "2025-01-21 13:47:24.970844 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.25 for 3875 genes in L6b_6 because all values are 0 for that covariate\n",
      "2025-01-21 13:47:57.621608 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:47:57.624213 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:48:03.747373 -- Adding 3 pseudocounts to method=10xMulti for 70 genes in L2_3_IT_1 because all values are 0 for that covariate\n",
      "2025-01-21 13:48:37.233474 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.5 for 904 genes in L4_IT_1 because all values are 0 for that covariate\n",
      "2025-01-21 13:48:53.797737 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.5 for 836 genes in L5_IT_1 because all values are 0 for that covariate\n",
      "2025-01-21 13:49:40.407838 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.5 for 419 genes in OPC_2 because all values are 0 for that covariate\n",
      "2025-01-21 13:49:43.424538 -- Adding 3 pseudocounts to age_at_death_binned_codes=1.0 for 745 genes in Lamp5_Lhx6_1 because all values are 0 for that covariate\n",
      "2025-01-21 13:50:05.702295 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.5 for 1173 genes in Vip_1 because all values are 0 for that covariate\n",
      "2025-01-21 13:50:45.479935 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.5 for 1424 genes in L5_6_NP_1 because all values are 0 for that covariate\n",
      "2025-01-21 13:51:25.065758 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.0 for 7239 genes in L6_IT_1 because all values are 0 for that covariate\n",
      "2025-01-21 13:51:52.659279 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.5 for 3339 genes in L6_IT_Car3_2 because all values are 0 for that covariate\n",
      "2025-01-21 13:51:54.454056 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.6287889247940761 for 6696 genes in Pax6_1 because all values are 0 for that covariate\n",
      "2025-01-21 13:53:29.646721 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.75 for 1246 genes in Chandelier_1 because all values are 0 for that covariate\n",
      "2025-01-21 13:54:35.340565 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.5 for 5086 genes in Astro_1 because all values are 0 for that covariate\n",
      "2025-01-21 13:55:42.121733 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.75 for 722 genes in Oligo_1 because all values are 0 for that covariate\n",
      "2025-01-21 13:56:14.972185 -- Adding 3 pseudocounts to age_at_death_binned_codes=1.0 for 416 genes in Lamp5_2 because all values are 0 for that covariate\n",
      "2025-01-21 13:56:42.908838 -- Adding 3 pseudocounts to sex=Male for 168 genes in L6_IT_Car3_2 because all values are 0 for that covariate\n",
      "2025-01-21 13:56:57.741157 -- Adding 3 pseudocounts to apoe4_status=Y for 101 genes in L6_IT_Car3_2 because all values are 0 for that covariate\n",
      "2025-01-21 13:57:06.528691 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:57:06.530008 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:57:06.530363 -- Removing method from the covariate formula\n",
      "2025-01-21 13:57:06.530490 -- Removing race_choice_white from the covariate formula\n",
      "2025-01-21 13:57:07.635247 -- Skipping L6_IT_Car3_2 along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-21 13:57:09.092704 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 13:57:09.786626 -- Removing 4063 features from L6_IT_Car3_3 for low numbers of counts per cell\n",
      "2025-01-21 13:57:09.984062 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:57:09.985625 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 13:57:10.041332 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.25 for 511 genes in L6_IT_Car3_3 because all values are 0 for that covariate\n",
      "2025-01-21 13:57:13.498949 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.75 for 304 genes in L5_IT_1 because all values are 0 for that covariate\n",
      "2025-01-21 13:57:32.545102 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.0 for 4823 genes in Chandelier_1 because all values are 0 for that covariate\n",
      "2025-01-21 13:57:34.997436 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.25 for 4189 genes in Sncg_4 because all values are 0 for that covariate\n",
      "2025-01-21 13:58:11.534466 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.75 for 653 genes in Vip_1 because all values are 0 for that covariate\n",
      "2025-01-21 13:58:12.888036 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.75 for 1740 genes in Sst_3 because all values are 0 for that covariate\n",
      "2025-01-21 13:59:10.085904 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.7760627731036908 for 6130 genes in Pax6_1 because all values are 0 for that covariate\n",
      "2025-01-21 13:59:26.321852 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.75 for 1000 genes in L5_6_NP_1 because all values are 0 for that covariate\n",
      "2025-01-21 14:00:13.591560 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.0 for 2807 genes in L5_IT_1 because all values are 0 for that covariate\n",
      "2025-01-21 14:00:21.326578 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.5 for 801 genes in Pvalb_2 because all values are 0 for that covariate\n",
      "2025-01-21 14:01:05.162862 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.75 for 179 genes in L4_IT_1 because all values are 0 for that covariate\n",
      "2025-01-21 14:01:21.117622 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.25 for 1481 genes in Oligo_1 because all values are 0 for that covariate\n",
      "2025-01-21 14:01:34.454012 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.75 for 116 genes in OPC_2 because all values are 0 for that covariate\n",
      "2025-01-21 14:02:04.595314 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.0 for 5917 genes in L6_CT_1 because all values are 0 for that covariate\n",
      "2025-01-21 14:02:42.930618 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.0 for 2506 genes in Vip_1 because all values are 0 for that covariate\n",
      "2025-01-21 14:03:29.914026 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.0 for 2331 genes in L4_IT_1 because all values are 0 for that covariate\n",
      "2025-01-21 14:03:48.434387 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.0 for 9480 genes in Sncg_4 because all values are 0 for that covariate\n",
      "2025-01-21 14:03:55.327708 -- Adding 3 pseudocounts to sex=Female for 4 genes in L6b_6 because all values are 0 for that covariate\n",
      "2025-01-21 14:03:56.412850 -- Adding 3 pseudocounts to sex=Male for 53 genes in L6b_6 because all values are 0 for that covariate\n",
      "2025-01-21 14:04:10.183175 -- Adding 3 pseudocounts to apoe4_status=Y for 54 genes in L6b_6 because all values are 0 for that covariate\n",
      "2025-01-21 14:04:24.041838 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 14:04:24.042713 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 14:04:24.105768 -- Adding 3 pseudocounts to method=10xMulti for 431 genes in L6b_6 because all values are 0 for that covariate\n",
      "2025-01-21 14:04:42.570611 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.5 for 519 genes in Lamp5_2 because all values are 0 for that covariate\n",
      "2025-01-21 14:04:52.473581 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.0 for 1865 genes in OPC_2 because all values are 0 for that covariate\n",
      "2025-01-21 14:05:31.645926 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.0 for 5551 genes in L5_6_NP_1 because all values are 0 for that covariate\n",
      "2025-01-21 14:05:34.454744 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.0 for 5708 genes in Astro_1 because all values are 0 for that covariate\n",
      "2025-01-21 14:05:54.969707 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.7050133460488578 for 7621 genes in Pax6_1 because all values are 0 for that covariate\n",
      "2025-01-21 14:05:59.418669 -- Adding 3 pseudocounts to age_at_death_binned_codes=1.0 for 889 genes in L6_IT_Car3_3 because all values are 0 for that covariate\n",
      "2025-01-21 14:06:15.552343 -- Adding 3 pseudocounts to race_choice_white=Unchecked for 5560 genes in L6b_6 because all values are 0 for that covariate\n",
      "2025-01-21 14:06:39.978861 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.5 for 1759 genes in Sst_3 because all values are 0 for that covariate\n",
      "2025-01-21 14:07:32.957342 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.5 for 846 genes in Lamp5_Lhx6_1 because all values are 0 for that covariate\n",
      "2025-01-21 14:12:43.993238 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.0 for 8818 genes in Oligo_1 because all values are 0 for that covariate\n",
      "2025-01-21 14:13:18.466964 -- Adding 3 pseudocounts to sex=Female for 4 genes in Chandelier_1 because all values are 0 for that covariate\n",
      "2025-01-21 14:13:19.277856 -- Adding 3 pseudocounts to sex=Male for 12 genes in Chandelier_1 because all values are 0 for that covariate\n",
      "2025-01-21 14:13:21.802108 -- Adding 3 pseudocounts to apoe4_status=Y for 210 genes in Chandelier_1 because all values are 0 for that covariate\n",
      "2025-01-21 14:14:02.593125 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 14:14:02.594271 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 14:14:02.648232 -- Adding 3 pseudocounts to method=10xMulti for 3290 genes in Chandelier_1 because all values are 0 for that covariate\n",
      "2025-01-21 14:14:21.411286 -- Adding 3 pseudocounts to sex=Female for 32 genes in L6_CT_1 because all values are 0 for that covariate\n",
      "2025-01-21 14:14:24.134221 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.6847115721583495 for 7490 genes in Pax6_1 because all values are 0 for that covariate\n",
      "2025-01-21 14:14:25.423689 -- Adding 3 pseudocounts to apoe4_status=Y for 460 genes in L6_CT_1 because all values are 0 for that covariate\n",
      "2025-01-21 14:15:13.792885 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.0 for 5678 genes in Sst_3 because all values are 0 for that covariate\n",
      "2025-01-21 14:15:18.208300 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.75 for 194 genes in Lamp5_2 because all values are 0 for that covariate\n",
      "2025-01-21 14:15:22.694935 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 14:15:22.696605 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 14:15:22.745643 -- Adding 3 pseudocounts to method=10xMulti for 1331 genes in L6_CT_1 because all values are 0 for that covariate\n",
      "2025-01-21 14:18:01.833139 -- Adding 3 pseudocounts to sex=Female for 8 genes in Astro_1 because all values are 0 for that covariate\n",
      "2025-01-21 14:18:02.896537 -- Adding 3 pseudocounts to sex=Male for 5 genes in Astro_1 because all values are 0 for that covariate\n",
      "2025-01-21 14:18:03.640557 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 14:18:03.641449 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 14:18:03.688700 -- Adding 3 pseudocounts to method=10xMulti for 2599 genes in Astro_1 because all values are 0 for that covariate\n",
      "2025-01-21 14:18:06.528371 -- Adding 3 pseudocounts to sex=Male for 569 genes in Sncg_4 because all values are 0 for that covariate\n",
      "2025-01-21 14:18:06.600314 -- Removing race_choice_white from the covariate formula\n",
      "2025-01-21 14:18:07.829965 -- Skipping L6_CT_1 along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-21 14:18:08.845966 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 14:18:09.180235 -- Removing 3849 features from L6_CT_2 for low numbers of counts per cell\n",
      "2025-01-21 14:18:09.250311 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 14:18:09.251739 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 14:18:09.279498 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.25 for 1321 genes in L6_CT_2 because all values are 0 for that covariate\n",
      "2025-01-21 14:18:29.230106 -- Adding 3 pseudocounts to race_choice_white=Unchecked for 602 genes in L2_3_IT_1 because all values are 0 for that covariate\n",
      "2025-01-21 14:18:58.480907 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 14:18:58.481950 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 14:18:58.482378 -- Removing method from the covariate formula\n",
      "2025-01-21 14:18:58.482548 -- Removing race_choice_white from the covariate formula\n",
      "2025-01-21 14:18:59.661472 -- Skipping Sncg_4 along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-21 14:19:00.673258 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 14:19:00.695392 -- Removing 17283 features from Sncg_6 for low numbers of counts per cell\n",
      "2025-01-21 14:19:00.697060 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 14:19:00.697244 -- Removing continuous_pseudo_progression_score from the covariate formula\n",
      "2025-01-21 14:19:00.697919 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 14:19:00.698058 -- Removing age_at_death_binned_codes from the covariate formula\n",
      "2025-01-21 14:19:00.698213 -- Removing sex from the covariate formula\n",
      "2025-01-21 14:19:00.698335 -- Removing apoe4_status from the covariate formula\n",
      "2025-01-21 14:19:00.698968 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 14:19:00.710542 -- Adding 3 pseudocounts to pseudo_genes_detected=0.0 for 3360 genes in Sncg_6 because all values are 0 for that covariate\n",
      "2025-01-21 14:19:15.484907 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.0 for 2120 genes in Lamp5_2 because all values are 0 for that covariate\n",
      "2025-01-21 14:19:34.057411 -- Adding 3 pseudocounts to pseudo_genes_detected=0.14745392064979698 for 4061 genes in Sncg_6 because all values are 0 for that covariate\n",
      "2025-01-21 14:20:05.092238 -- Adding 3 pseudocounts to sex=Male for 17 genes in Vip_1 because all values are 0 for that covariate\n",
      "2025-01-21 14:20:12.320875 -- Adding 3 pseudocounts to apoe4_status=Y for 186 genes in Vip_1 because all values are 0 for that covariate\n",
      "2025-01-21 14:20:16.163018 -- Adding 3 pseudocounts to pseudo_genes_detected=0.058731646360512355 for 4177 genes in Sncg_6 because all values are 0 for that covariate\n",
      "2025-01-21 14:21:00.284457 -- Adding 3 pseudocounts to pseudo_genes_detected=0.22149328334895346 for 5111 genes in Sncg_6 because all values are 0 for that covariate\n",
      "2025-01-21 14:21:21.298979 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.5 for 923 genes in L6_IT_Car3_3 because all values are 0 for that covariate\n",
      "2025-01-21 14:21:24.498619 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.25 for 5706 genes in Pvalb_2 because all values are 0 for that covariate\n",
      "2025-01-21 14:21:29.810073 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 14:21:29.811475 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 14:21:29.927573 -- Adding 3 pseudocounts to method=10xMulti for 1235 genes in Vip_1 because all values are 0 for that covariate\n",
      "2025-01-21 14:21:56.851027 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 14:21:56.893461 -- Removing method from the covariate formula\n",
      "2025-01-21 14:21:56.893633 -- Removing race_choice_white from the covariate formula\n",
      "2025-01-21 14:21:56.918517 -- Starting Sncg_6 across continuous_pseudo_progression_score\n",
      "The cells are already grouped.2025-01-21 14:21:56.982166 -- Skipping Sncg_6 because some predictors have zero variation or a zero vector\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: Error in nebula(counts, obs$donor_id, pred = df, offset = obs$pseudo_number_of_umis,  : \n",
      "  Some predictors have zero variation or a zero vector.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-21 14:21:58.030243 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 14:21:58.057616 -- Removing 11316 features from Sncg_8 for low numbers of counts per cell\n",
      "2025-01-21 14:21:58.060636 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 14:21:58.073602 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.7142033482533316 for 5673 genes in Sncg_8 because all values are 0 for that covariate\n",
      "2025-01-21 14:22:53.491066 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.632943265190392 for 6894 genes in Pax6_1 because all values are 0 for that covariate\n",
      "2025-01-21 14:23:46.137206 -- Adding 3 pseudocounts to race_choice_white=Unchecked for 7387 genes in Astro_1 because all values are 0 for that covariate\n",
      "2025-01-21 14:24:04.019304 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.27633850362408346 for 4919 genes in Sncg_8 because all values are 0 for that covariate\n",
      "2025-01-21 14:24:42.233503 -- Adding 3 pseudocounts to apoe4_status=Y for 62 genes in L6_IT_1 because all values are 0 for that covariate\n",
      "2025-01-21 14:24:49.839209 -- Adding 3 pseudocounts to race_choice_white=Unchecked for 3183 genes in Chandelier_1 because all values are 0 for that covariate\n",
      "2025-01-21 14:24:59.109236 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 14:24:59.110134 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 14:24:59.192778 -- Adding 3 pseudocounts to method=10xMulti for 2977 genes in L6_IT_1 because all values are 0 for that covariate\n",
      "2025-01-21 14:25:56.789003 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.0 for 2346 genes in Sncg_8 because all values are 0 for that covariate\n",
      "2025-01-21 14:26:51.659840 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 14:26:51.734769 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 14:26:51.735599 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 14:26:51.735973 -- Removing method from the covariate formula\n",
      "2025-01-21 14:26:51.736130 -- Removing race_choice_white from the covariate formula\n",
      "2025-01-21 14:26:51.777461 -- Starting Sncg_8 across continuous_pseudo_progression_score\n",
      "The cells are already grouped.The average number of cells per subject ( 5.33 ) is less than 30. The 'method' is set as 'HL'.\n",
      "Remove  919  genes having low expression.\n",
      "Analyzing  24366  genes with  3  subjects and  16  cells.\n",
      "2025-01-21 14:27:58.949936 -- Adding 3 pseudocounts to age_at_death_binned_codes=1.0 for 1025 genes in L6_CT_2 because all values are 0 for that covariate\n",
      "2025-01-21 14:28:12.943661 -- Adding 3 pseudocounts to sex=Female for 3 genes in L5_IT_1 because all values are 0 for that covariate\n",
      "2025-01-21 14:28:14.984522 -- Adding 3 pseudocounts to apoe4_status=Y for 87 genes in L5_IT_1 because all values are 0 for that covariate\n",
      "2025-01-21 14:29:06.707984 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 14:29:06.709671 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 14:29:06.872121 -- Adding 3 pseudocounts to method=10xMulti for 1810 genes in L5_IT_1 because all values are 0 for that covariate\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: In addition: \n",
      "R[write to console]: Warning message:\n",
      "\n",
      "R[write to console]: In nebula(counts, obs$donor_id, pred = df, offset = obs$pseudo_number_of_umis,  :\n",
      "R[write to console]: \n",
      " \n",
      "R[write to console]:  Some predictors in the design matrix are collinear or linearly dependent. The effects of these predictors are not identifiable.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-21 14:29:36.038454 -- Sncg_8 along continuous_pseudo_progression_score was written to disk\n",
      "2025-01-21 14:30:04.282567 -- Starting L6b_6 across continuous_pseudo_progression_score\n",
      "The cells are already grouped.The average number of cells per subject ( 10.05 ) is less than 30. The 'method' is set as 'HL'.\n",
      "Remove  1  genes having low expression.\n",
      "Analyzing  32279  genes with  20  subjects and  201  cells.\n",
      "2025-01-21 14:30:11.695864 -- Adding 3 pseudocounts to race_choice_white=Unchecked for 5343 genes in Vip_1 because all values are 0 for that covariate\n",
      "2025-01-21 14:30:18.558800 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.75 for 324 genes in Lamp5_Lhx6_1 because all values are 0 for that covariate\n",
      "2025-01-21 14:31:00.906824 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.6670367899703017 for 8394 genes in Pax6_1 because all values are 0 for that covariate\n",
      "2025-01-21 14:34:11.023319 -- L6b_6 along continuous_pseudo_progression_score was written to disk\n",
      "2025-01-21 14:34:13.033413 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.75 for 384 genes in L6_IT_Car3_3 because all values are 0 for that covariate\n",
      "2025-01-21 14:35:18.422035 -- Skipping Chandelier_1 along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-21 14:35:19.414496 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 14:35:19.435096 -- Removing 17386 features from Chandelier_2 for low numbers of counts per cell\n",
      "2025-01-21 14:35:19.437068 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 14:35:19.437280 -- Removing continuous_pseudo_progression_score from the covariate formula\n",
      "2025-01-21 14:35:19.437940 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 14:35:19.438071 -- Removing age_at_death_binned_codes from the covariate formula\n",
      "2025-01-21 14:35:19.438277 -- Removing sex from the covariate formula\n",
      "2025-01-21 14:35:19.438405 -- Removing apoe4_status from the covariate formula\n",
      "2025-01-21 14:35:19.439046 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 14:35:19.451061 -- Adding 3 pseudocounts to pseudo_genes_detected=0.0 for 5155 genes in Chandelier_2 because all values are 0 for that covariate\n",
      "2025-01-21 14:35:23.029274 -- Adding 3 pseudocounts to sex=Female for 1 genes in L4_IT_1 because all values are 0 for that covariate\n",
      "2025-01-21 14:35:23.944111 -- Adding 3 pseudocounts to sex=Male for 4 genes in L4_IT_1 because all values are 0 for that covariate\n",
      "2025-01-21 14:35:27.419405 -- Adding 3 pseudocounts to apoe4_status=Y for 68 genes in L4_IT_1 because all values are 0 for that covariate\n",
      "2025-01-21 14:35:42.233742 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.5 for 1847 genes in L6_CT_2 because all values are 0 for that covariate\n",
      "2025-01-21 14:36:15.242974 -- Adding 3 pseudocounts to pseudo_genes_detected=0.3751342642320085 for 3832 genes in Chandelier_2 because all values are 0 for that covariate\n",
      "2025-01-21 14:36:22.502488 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 14:36:22.503832 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 14:36:22.736541 -- Adding 3 pseudocounts to method=10xMulti for 211 genes in L4_IT_1 because all values are 0 for that covariate\n",
      "2025-01-21 14:36:34.737364 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.0 for 2469 genes in Lamp5_Lhx6_1 because all values are 0 for that covariate\n",
      "2025-01-21 14:36:58.849450 -- Adding 3 pseudocounts to pseudo_genes_detected=0.9100429645542428 for 3756 genes in Chandelier_2 because all values are 0 for that covariate\n",
      "2025-01-21 14:37:40.844280 -- Adding 3 pseudocounts to pseudo_genes_detected=0.2696025778732546 for 3842 genes in Chandelier_2 because all values are 0 for that covariate\n",
      "2025-01-21 14:38:23.721833 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 14:38:23.779478 -- Removing method from the covariate formula\n",
      "2025-01-21 14:38:23.779608 -- Removing race_choice_white from the covariate formula\n",
      "2025-01-21 14:38:23.814369 -- Starting Chandelier_2 across continuous_pseudo_progression_score\n",
      "The cells are already grouped.2025-01-21 14:38:23.879650 -- Skipping Chandelier_2 because some predictors have zero variation or a zero vector\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: Error in nebula(counts, obs$donor_id, pred = df, offset = obs$pseudo_number_of_umis,  : \n",
      "  Some predictors have zero variation or a zero vector.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-21 14:38:54.638028 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.0 for 2732 genes in L6_IT_Car3_3 because all values are 0 for that covariate\n",
      "2025-01-21 14:39:03.942511 -- Removing race_choice_white from the covariate formula\n",
      "2025-01-21 14:39:06.735275 -- Skipping L6_IT_1 along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-21 14:39:07.689679 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 14:39:09.892593 -- Adding 3 pseudocounts to race_choice_white=Unchecked for 4974 genes in L4_IT_1 because all values are 0 for that covariate\n",
      "2025-01-21 14:39:10.165621 -- Removing 3536 features from L6_IT_2 for low numbers of counts per cell\n",
      "2025-01-21 14:39:10.709512 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 14:39:10.710736 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 14:39:10.884571 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.25 for 88 genes in L6_IT_2 because all values are 0 for that covariate\n",
      "2025-01-21 14:39:36.870962 -- Adding 3 pseudocounts to sex=Female for 9 genes in L5_6_NP_1 because all values are 0 for that covariate\n",
      "2025-01-21 14:39:40.431937 -- Adding 3 pseudocounts to apoe4_status=Y for 197 genes in L5_6_NP_1 because all values are 0 for that covariate\n",
      "2025-01-21 14:40:42.288520 -- Skipping Astro_1 along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-21 14:40:43.214859 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 14:40:46.202101 -- Removing 5617 features from Astro_2 for low numbers of counts per cell\n",
      "2025-01-21 14:40:46.810929 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 14:40:46.812280 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 14:40:47.014899 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.25 for 132 genes in Astro_2 because all values are 0 for that covariate\n",
      "2025-01-21 14:40:48.633418 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 14:40:48.769320 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 14:40:48.770170 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 14:40:48.810230 -- Removing race_choice_white from the covariate formula\n",
      "2025-01-21 14:40:50.001830 -- Skipping Pax6_1 along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-21 14:40:53.253892 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 14:40:53.254985 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 14:40:53.368416 -- Adding 3 pseudocounts to method=10xMulti for 895 genes in L5_6_NP_1 because all values are 0 for that covariate\n",
      "2025-01-21 14:43:09.529288 -- Adding 3 pseudocounts to sex=Female for 1 genes in Sst_3 because all values are 0 for that covariate\n",
      "2025-01-21 14:43:09.885005 -- Adding 3 pseudocounts to sex=Male for 7 genes in Sst_3 because all values are 0 for that covariate\n",
      "2025-01-21 14:43:11.991830 -- Adding 3 pseudocounts to apoe4_status=Y for 202 genes in Sst_3 because all values are 0 for that covariate\n",
      "2025-01-21 14:44:08.434932 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 14:44:08.436129 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 14:44:08.528475 -- Adding 3 pseudocounts to method=10xMulti for 5083 genes in Sst_3 because all values are 0 for that covariate\n",
      "2025-01-21 14:44:09.529783 -- Adding 3 pseudocounts to age_at_death_binned_codes=1.0 for 126 genes in L6_IT_2 because all values are 0 for that covariate\n",
      "2025-01-21 14:46:19.949585 -- Adding 3 pseudocounts to race_choice_white=Unchecked for 9507 genes in L5_6_NP_1 because all values are 0 for that covariate\n",
      "2025-01-21 14:47:03.081361 -- Adding 3 pseudocounts to race_choice_white=Unchecked for 8954 genes in L5_IT_1 because all values are 0 for that covariate\n",
      "2025-01-21 14:48:40.002484 -- Adding 3 pseudocounts to age_at_death_binned_codes=1.0 for 38 genes in Astro_2 because all values are 0 for that covariate\n",
      "2025-01-21 14:48:47.877183 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.75 for 1015 genes in L6_CT_2 because all values are 0 for that covariate\n",
      "2025-01-21 14:50:56.453335 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.5 for 153 genes in Astro_2 because all values are 0 for that covariate\n",
      "2025-01-21 14:51:14.860786 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.5 for 98 genes in L6_IT_2 because all values are 0 for that covariate\n",
      "2025-01-21 14:55:53.288081 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.0 for 3776 genes in L6_CT_2 because all values are 0 for that covariate\n",
      "2025-01-21 14:56:45.768958 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.75 for 22 genes in L6_IT_2 because all values are 0 for that covariate\n",
      "2025-01-21 14:57:28.683760 -- Adding 3 pseudocounts to apoe4_status=Y for 30 genes in OPC_2 because all values are 0 for that covariate\n",
      "2025-01-21 14:58:00.190405 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.0 for 772 genes in L6_IT_2 because all values are 0 for that covariate\n",
      "2025-01-21 14:58:18.385436 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 14:58:18.386946 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 14:58:18.826996 -- Adding 3 pseudocounts to method=10xMulti for 2543 genes in OPC_2 because all values are 0 for that covariate\n",
      "2025-01-21 15:00:08.449615 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.75 for 30 genes in Astro_2 because all values are 0 for that covariate\n",
      "2025-01-21 15:01:40.437265 -- Adding 3 pseudocounts to sex=Female for 1 genes in Lamp5_2 because all values are 0 for that covariate\n",
      "2025-01-21 15:01:41.796326 -- Adding 3 pseudocounts to sex=Male for 6 genes in Lamp5_2 because all values are 0 for that covariate\n",
      "2025-01-21 15:01:49.489199 -- Adding 3 pseudocounts to apoe4_status=Y for 22 genes in Lamp5_2 because all values are 0 for that covariate\n",
      "2025-01-21 15:01:57.077009 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.0 for 312 genes in Astro_2 because all values are 0 for that covariate\n",
      "2025-01-21 15:02:15.674619 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 15:02:15.675872 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 15:02:15.987972 -- Adding 3 pseudocounts to method=10xMulti for 465 genes in Lamp5_2 because all values are 0 for that covariate\n",
      "2025-01-21 15:06:25.613612 -- Skipping Vip_1 along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-21 15:06:27.291667 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 15:06:27.356774 -- Removing 5752 features from Vip_2 for low numbers of counts per cell\n",
      "2025-01-21 15:06:27.370330 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 15:06:27.371313 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 15:06:27.388151 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.25 for 4667 genes in Vip_2 because all values are 0 for that covariate\n",
      "2025-01-21 15:07:52.463902 -- Adding 3 pseudocounts to race_choice_white=Unchecked for 4797 genes in Sst_3 because all values are 0 for that covariate\n",
      "2025-01-21 15:10:58.443991 -- Adding 3 pseudocounts to sex=Female for 2 genes in L6_IT_Car3_3 because all values are 0 for that covariate\n",
      "2025-01-21 15:10:59.941408 -- Adding 3 pseudocounts to sex=Male for 5 genes in L6_IT_Car3_3 because all values are 0 for that covariate\n",
      "2025-01-21 15:11:03.741990 -- Adding 3 pseudocounts to apoe4_status=Y for 68 genes in L6_IT_Car3_3 because all values are 0 for that covariate\n",
      "2025-01-21 15:11:27.643740 -- Adding 3 pseudocounts to race_choice_white=Unchecked for 3699 genes in Lamp5_2 because all values are 0 for that covariate\n",
      "2025-01-21 15:11:52.109043 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 15:11:52.110348 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 15:11:52.301671 -- Adding 3 pseudocounts to method=10xMulti for 3172 genes in L6_IT_Car3_3 because all values are 0 for that covariate\n",
      "2025-01-21 15:13:39.582642 -- Adding 3 pseudocounts to age_at_death_binned_codes=1.0 for 2894 genes in Vip_2 because all values are 0 for that covariate\n",
      "2025-01-21 15:18:08.921726 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.5 for 3551 genes in Vip_2 because all values are 0 for that covariate\n",
      "2025-01-21 15:18:44.272779 -- Adding 3 pseudocounts to sex=Female for 7 genes in Oligo_1 because all values are 0 for that covariate\n",
      "2025-01-21 15:18:47.788957 -- Adding 3 pseudocounts to apoe4_status=Y for 181 genes in Oligo_1 because all values are 0 for that covariate\n",
      "2025-01-21 15:20:06.937384 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 15:20:06.938710 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 15:20:07.198049 -- Adding 3 pseudocounts to race_choice_white=Unchecked for 8293 genes in Oligo_1 because all values are 0 for that covariate\n",
      "2025-01-21 15:20:36.283407 -- Adding 3 pseudocounts to apoe4_status=Y for 20 genes in Astro_2 because all values are 0 for that covariate\n",
      "2025-01-21 15:21:47.273690 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 15:21:47.275244 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 15:21:48.231770 -- Adding 3 pseudocounts to method=10xMulti for 258 genes in Astro_2 because all values are 0 for that covariate\n",
      "2025-01-21 15:22:48.809119 -- Adding 3 pseudocounts to sex=Female for 1 genes in L6_CT_2 because all values are 0 for that covariate\n",
      "2025-01-21 15:22:49.393286 -- Adding 3 pseudocounts to apoe4_status=Y for 152 genes in L6_CT_2 because all values are 0 for that covariate\n",
      "2025-01-21 15:23:06.911710 -- Adding 3 pseudocounts to sex=Female for 10 genes in Lamp5_Lhx6_1 because all values are 0 for that covariate\n",
      "2025-01-21 15:23:18.328289 -- Adding 3 pseudocounts to sex=Male for 4 genes in Lamp5_Lhx6_1 because all values are 0 for that covariate\n",
      "2025-01-21 15:23:23.384467 -- Adding 3 pseudocounts to apoe4_status=Y for 58 genes in Lamp5_Lhx6_1 because all values are 0 for that covariate\n",
      "2025-01-21 15:23:41.174109 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.0 for 8465 genes in Vip_2 because all values are 0 for that covariate\n",
      "2025-01-21 15:23:54.106536 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 15:23:54.107516 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 15:23:54.233641 -- Adding 3 pseudocounts to method=10xMulti for 55 genes in L6_CT_2 because all values are 0 for that covariate\n",
      "2025-01-21 15:24:17.887902 -- Adding 3 pseudocounts to race_choice_white=Unchecked for 4670 genes in L6_CT_2 because all values are 0 for that covariate\n",
      "2025-01-21 15:24:28.875628 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 15:24:28.876843 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 15:24:29.171005 -- Adding 3 pseudocounts to method=10xMulti for 355 genes in Lamp5_Lhx6_1 because all values are 0 for that covariate\n",
      "2025-01-21 15:30:29.733761 -- Starting Sst_3 across continuous_pseudo_progression_score\n",
      "The cells are already grouped.The average number of cells per subject ( 10.77 ) is less than 30. The 'method' is set as 'HL'.\n",
      "Remove  0  genes having low expression.\n",
      "Analyzing  30829  genes with  30  subjects and  323  cells.\n",
      "2025-01-21 15:31:11.282019 -- Adding 3 pseudocounts to race_choice_white=Unchecked for 5521 genes in Lamp5_Lhx6_1 because all values are 0 for that covariate\n",
      "2025-01-21 15:35:07.138777 -- Sst_3 along continuous_pseudo_progression_score was written to disk\n",
      "2025-01-21 15:35:08.177759 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 15:35:08.201537 -- Removing 17270 features from Sst_5 for low numbers of counts per cell\n",
      "2025-01-21 15:35:08.203867 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 15:35:08.204132 -- Removing continuous_pseudo_progression_score from the covariate formula\n",
      "2025-01-21 15:35:08.204826 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 15:35:08.204971 -- Removing age_at_death_binned_codes from the covariate formula\n",
      "2025-01-21 15:35:08.205217 -- Removing sex from the covariate formula\n",
      "2025-01-21 15:35:08.205348 -- Removing apoe4_status from the covariate formula\n",
      "2025-01-21 15:35:08.206030 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 15:35:08.218213 -- Adding 3 pseudocounts to pseudo_genes_detected=0.26439263097294186 for 5053 genes in Sst_5 because all values are 0 for that covariate\n",
      "2025-01-21 15:36:02.467150 -- Adding 3 pseudocounts to pseudo_genes_detected=0.0 for 6413 genes in Sst_5 because all values are 0 for that covariate\n",
      "2025-01-21 15:37:00.237785 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.75 for 3137 genes in Vip_2 because all values are 0 for that covariate\n",
      "2025-01-21 15:37:08.458124 -- Adding 3 pseudocounts to race_choice_white=Unchecked for 3321 genes in Astro_2 because all values are 0 for that covariate\n",
      "2025-01-21 15:37:13.184520 -- Adding 3 pseudocounts to pseudo_genes_detected=0.3395221646516984 for 4823 genes in Sst_5 because all values are 0 for that covariate\n",
      "2025-01-21 15:38:08.059201 -- Adding 3 pseudocounts to pseudo_genes_detected=0.14838802533103057 for 5393 genes in Sst_5 because all values are 0 for that covariate\n",
      "2025-01-21 15:39:11.323573 -- Adding 3 pseudocounts to pseudo_genes_detected=0.2517271157167531 for 3742 genes in Sst_5 because all values are 0 for that covariate\n",
      "2025-01-21 15:39:56.454182 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 15:39:56.505184 -- Removing method from the covariate formula\n",
      "2025-01-21 15:39:56.505298 -- Removing race_choice_white from the covariate formula\n",
      "2025-01-21 15:39:56.531616 -- Starting Sst_5 across continuous_pseudo_progression_score\n",
      "The cells are already grouped.2025-01-21 15:39:56.559687 -- Skipping Sst_5 because some predictors have zero variation or a zero vector\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: Error in nebula(counts, obs$donor_id, pred = df, offset = obs$pseudo_number_of_umis,  : \n",
      "  Some predictors have zero variation or a zero vector.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-21 15:39:57.646669 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 15:39:57.739524 -- Removing 6693 features from Sst_13 for low numbers of counts per cell\n",
      "2025-01-21 15:39:57.756032 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 15:39:57.757016 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 15:39:57.775453 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.25 for 2935 genes in Sst_13 because all values are 0 for that covariate\n",
      "2025-01-21 15:41:48.550832 -- Adding 3 pseudocounts to apoe4_status=Y for 14 genes in L6_IT_2 because all values are 0 for that covariate\n",
      "2025-01-21 15:41:59.318426 -- Adding 3 pseudocounts to sex=Male for 258 genes in Vip_2 because all values are 0 for that covariate\n",
      "2025-01-21 15:42:23.935158 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 15:42:23.936000 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 15:42:23.972247 -- Adding 3 pseudocounts to method=10xMulti for 2670 genes in Vip_2 because all values are 0 for that covariate\n",
      "2025-01-21 15:42:36.063048 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 15:42:36.064511 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 15:42:36.929991 -- Adding 3 pseudocounts to method=10xMulti for 67 genes in L6_IT_2 because all values are 0 for that covariate\n",
      "2025-01-21 15:43:37.321123 -- Skipping L4_IT_1 along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-21 15:43:38.356748 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 15:43:43.270661 -- Skipping L5_6_NP_1 along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-21 15:43:45.218412 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 15:43:45.281085 -- Removing 5530 features from L5_6_NP_2 for low numbers of counts per cell\n",
      "2025-01-21 15:43:45.292666 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 15:43:45.293625 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 15:43:45.310016 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.25 for 2990 genes in L5_6_NP_2 because all values are 0 for that covariate\n",
      "2025-01-21 15:43:53.213190 -- Removing 4159 features from L4_IT_2 for low numbers of counts per cell\n",
      "2025-01-21 15:43:56.202623 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 15:43:56.205104 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 15:43:56.942901 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.25 for 6 genes in L4_IT_2 because all values are 0 for that covariate\n",
      "2025-01-21 15:43:59.129977 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.0 for 8479 genes in Sst_13 because all values are 0 for that covariate\n",
      "2025-01-21 15:45:47.208799 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.5 for 2 genes in L4_IT_2 because all values are 0 for that covariate\n",
      "2025-01-21 15:46:25.252018 -- Adding 3 pseudocounts to race_choice_white=Unchecked for 1924 genes in L6_IT_2 because all values are 0 for that covariate\n",
      "2025-01-21 15:46:25.625631 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.75 for 1 genes in L4_IT_2 because all values are 0 for that covariate\n",
      "2025-01-21 15:46:42.166171 -- Adding 3 pseudocounts to race_choice_white=Unchecked for 5823 genes in Vip_2 because all values are 0 for that covariate\n",
      "2025-01-21 15:46:44.272866 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.0 for 182 genes in L4_IT_2 because all values are 0 for that covariate\n",
      "2025-01-21 15:48:20.454826 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.5 for 2341 genes in L5_6_NP_2 because all values are 0 for that covariate\n",
      "2025-01-21 15:49:04.924676 -- Adding 3 pseudocounts to race_choice_white=Unchecked for 5718 genes in L6_IT_Car3_3 because all values are 0 for that covariate\n",
      "2025-01-21 15:51:55.022200 -- Adding 3 pseudocounts to age_at_death_binned_codes=1.0 for 3380 genes in L5_6_NP_2 because all values are 0 for that covariate\n",
      "2025-01-21 15:55:34.081790 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.75 for 3592 genes in Sst_13 because all values are 0 for that covariate\n",
      "2025-01-21 15:56:06.453426 -- Skipping Vip_2 along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-21 15:56:07.451719 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 15:56:08.077295 -- Removing 4698 features from Vip_4 for low numbers of counts per cell\n",
      "2025-01-21 15:56:08.199088 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 15:56:08.200328 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 15:56:08.239639 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.25 for 966 genes in Vip_4 because all values are 0 for that covariate\n",
      "2025-01-21 15:57:07.549518 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.75 for 4693 genes in L5_6_NP_2 because all values are 0 for that covariate\n",
      "2025-01-21 15:57:31.690568 -- Skipping L6_CT_2 along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-21 16:00:31.419644 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.5 for 3782 genes in Sst_13 because all values are 0 for that covariate\n",
      "2025-01-21 16:04:21.269561 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.0 for 9075 genes in L5_6_NP_2 because all values are 0 for that covariate\n",
      "2025-01-21 16:05:45.624725 -- Adding 3 pseudocounts to age_at_death_binned_codes=1.0 for 6397 genes in Sst_13 because all values are 0 for that covariate\n",
      "2025-01-21 16:07:46.042732 -- Adding 3 pseudocounts to age_at_death_binned_codes=1.0 for 663 genes in Vip_4 because all values are 0 for that covariate\n",
      "2025-01-21 16:08:20.498422 -- Adding 3 pseudocounts to race_choice_white=Unchecked for 4724 genes in OPC_2 because all values are 0 for that covariate\n",
      "2025-01-21 16:11:09.931180 -- Skipping L5_IT_1 along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-21 16:11:10.848514 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 16:11:14.400689 -- Removing 3613 features from L5_IT_2 for low numbers of counts per cell\n",
      "2025-01-21 16:11:15.163450 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 16:11:15.165433 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 16:11:15.365975 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.25 for 153 genes in L5_IT_2 because all values are 0 for that covariate\n",
      "2025-01-21 16:14:42.397710 -- Adding 3 pseudocounts to sex=Female for 56 genes in Sst_13 because all values are 0 for that covariate\n",
      "2025-01-21 16:14:47.354748 -- Adding 3 pseudocounts to apoe4_status=Y for 1006 genes in Sst_13 because all values are 0 for that covariate\n",
      "2025-01-21 16:15:44.365592 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.5 for 748 genes in Vip_4 because all values are 0 for that covariate\n",
      "2025-01-21 16:16:13.004424 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 16:16:13.005408 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 16:16:13.041792 -- Adding 3 pseudocounts to method=10xMulti for 882 genes in Sst_13 because all values are 0 for that covariate\n",
      "2025-01-21 16:17:27.969567 -- Removing race_choice_white from the covariate formula\n",
      "2025-01-21 16:17:28.281578 -- Starting Sst_13 across continuous_pseudo_progression_score\n",
      "The cells are already grouped.The average number of cells per subject ( 5.29 ) is less than 30. The 'method' is set as 'HL'.\n",
      "Remove  0  genes having low expression.\n",
      "Analyzing  29908  genes with  17  subjects and  90  cells.\n",
      "2025-01-21 16:18:29.185238 -- Adding 3 pseudocounts to apoe4_status=Y for 652 genes in L5_6_NP_2 because all values are 0 for that covariate\n",
      "2025-01-21 16:19:31.225332 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 16:19:31.226366 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 16:19:31.268832 -- Adding 3 pseudocounts to method=10xMulti for 6922 genes in L5_6_NP_2 because all values are 0 for that covariate\n",
      "2025-01-21 16:20:34.040819 -- Sst_13 along continuous_pseudo_progression_score was written to disk\n",
      "2025-01-21 16:20:35.267987 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 16:20:35.417011 -- Removing 5558 features from Sst_19 for low numbers of counts per cell\n",
      "2025-01-21 16:20:35.442481 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 16:20:35.443500 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 16:20:35.468713 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.25 for 1249 genes in Sst_19 because all values are 0 for that covariate\n",
      "2025-01-21 16:21:28.413574 -- Skipping Oligo_1 along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-21 16:21:30.316634 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 16:21:35.130635 -- Removing 7241 features from Oligo_2 for low numbers of counts per cell\n",
      "2025-01-21 16:21:35.993263 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 16:21:35.994910 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 16:21:36.110023 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.25 for 154 genes in Oligo_2 because all values are 0 for that covariate\n",
      "2025-01-21 16:22:25.866956 -- Adding 3 pseudocounts to age_at_death_binned_codes=1.0 for 13 genes in L5_IT_2 because all values are 0 for that covariate\n",
      "2025-01-21 16:23:22.174453 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.5 for 33 genes in L5_IT_2 because all values are 0 for that covariate\n",
      "2025-01-21 16:23:28.546345 -- Adding 3 pseudocounts to age_at_death_binned_codes=1.0 for 6058 genes in Sst_19 because all values are 0 for that covariate\n",
      "2025-01-21 16:24:48.358318 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.75 for 368 genes in Vip_4 because all values are 0 for that covariate\n",
      "2025-01-21 16:25:02.117153 -- Skipping Lamp5_2 along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-21 16:25:03.039391 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 16:25:03.122796 -- Removing 6214 features from Lamp5_3 for low numbers of counts per cell\n",
      "2025-01-21 16:25:03.138327 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 16:25:03.139256 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 16:25:03.154461 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.25 for 5741 genes in Lamp5_3 because all values are 0 for that covariate\n",
      "2025-01-21 16:25:47.385006 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.75 for 2 genes in L5_IT_2 because all values are 0 for that covariate\n",
      "2025-01-21 16:25:56.307514 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.0 for 417 genes in L5_IT_2 because all values are 0 for that covariate\n",
      "2025-01-21 16:29:12.865323 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.0 for 2251 genes in Vip_4 because all values are 0 for that covariate\n",
      "2025-01-21 16:30:28.777997 -- Removing race_choice_white from the covariate formula\n",
      "2025-01-21 16:30:28.951757 -- Skipping L5_6_NP_2 along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-21 16:34:47.826573 -- Adding 3 pseudocounts to age_at_death_binned_codes=1.0 for 2420 genes in Lamp5_3 because all values are 0 for that covariate\n",
      "2025-01-21 16:35:19.252875 -- Adding 3 pseudocounts to age_at_death_binned_codes=1.0 for 17 genes in Oligo_2 because all values are 0 for that covariate\n",
      "2025-01-21 16:36:51.093384 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.5 for 46 genes in Oligo_2 because all values are 0 for that covariate\n",
      "2025-01-21 16:37:06.385164 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.75 for 3935 genes in Sst_19 because all values are 0 for that covariate\n",
      "2025-01-21 16:38:51.534217 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.5 for 4179 genes in Lamp5_3 because all values are 0 for that covariate\n",
      "2025-01-21 16:41:00.682810 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.0 for 1145 genes in Oligo_2 because all values are 0 for that covariate\n",
      "2025-01-21 16:41:37.567799 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 16:41:37.569861 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 16:41:42.075786 -- Adding 3 pseudocounts to method=10xMulti for 2 genes in L4_IT_2 because all values are 0 for that covariate\n",
      "2025-01-21 16:42:22.341667 -- Adding 3 pseudocounts to race_choice_white=Unchecked for 1217 genes in L4_IT_2 because all values are 0 for that covariate\n",
      "2025-01-21 16:43:21.486747 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.75 for 825 genes in Pvalb_2 because all values are 0 for that covariate\n",
      "2025-01-21 16:45:51.129319 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.5 for 4781 genes in Sst_19 because all values are 0 for that covariate\n",
      "2025-01-21 16:45:53.808185 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.75 for 3700 genes in Lamp5_3 because all values are 0 for that covariate\n",
      "2025-01-21 16:52:08.510487 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.0 for 6972 genes in Lamp5_3 because all values are 0 for that covariate\n",
      "2025-01-21 16:55:05.610278 -- Skipping L6_IT_Car3_3 along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-21 16:55:22.208147 -- Adding 3 pseudocounts to sex=Male for 6 genes in Vip_4 because all values are 0 for that covariate\n",
      "2025-01-21 16:55:26.680219 -- Adding 3 pseudocounts to apoe4_status=Y for 97 genes in Vip_4 because all values are 0 for that covariate\n",
      "2025-01-21 16:56:10.085639 -- Adding 3 pseudocounts to apoe4_status=Y for 7 genes in L5_IT_2 because all values are 0 for that covariate\n",
      "2025-01-21 16:56:28.493352 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.0 for 10031 genes in Sst_19 because all values are 0 for that covariate\n",
      "2025-01-21 16:56:33.526098 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 16:56:33.527365 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 16:56:33.719005 -- Adding 3 pseudocounts to method=10xMulti for 958 genes in Vip_4 because all values are 0 for that covariate\n",
      "2025-01-21 16:56:40.503810 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 16:56:40.505785 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 16:56:41.639269 -- Adding 3 pseudocounts to method=10xMulti for 204 genes in L5_IT_2 because all values are 0 for that covariate\n",
      "2025-01-21 17:03:17.216917 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.0 for 1277 genes in Pvalb_2 because all values are 0 for that covariate\n",
      "2025-01-21 17:03:53.603213 -- Adding 3 pseudocounts to sex=Male for 47 genes in Lamp5_3 because all values are 0 for that covariate\n",
      "2025-01-21 17:03:58.532484 -- Adding 3 pseudocounts to apoe4_status=Y for 106 genes in Lamp5_3 because all values are 0 for that covariate\n",
      "2025-01-21 17:04:09.391920 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 17:04:09.392790 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 17:04:09.429952 -- Adding 3 pseudocounts to method=10xMulti for 871 genes in Lamp5_3 because all values are 0 for that covariate\n",
      "2025-01-21 17:05:37.647626 -- Removing race_choice_white from the covariate formula\n",
      "2025-01-21 17:05:37.863760 -- Skipping Lamp5_3 along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-21 17:05:38.818357 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 17:05:38.931598 -- Removing 5827 features from Lamp5_5 for low numbers of counts per cell\n",
      "2025-01-21 17:05:38.948315 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 17:05:38.949398 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 17:05:38.965312 -- Adding 3 pseudocounts to age_at_death_binned_codes=1.0 for 3899 genes in Lamp5_5 because all values are 0 for that covariate\n",
      "2025-01-21 17:07:34.070224 -- Adding 3 pseudocounts to race_choice_white=Unchecked for 4863 genes in Vip_4 because all values are 0 for that covariate\n",
      "2025-01-21 17:11:18.042010 -- Adding 3 pseudocounts to race_choice_white=Unchecked for 2227 genes in L5_IT_2 because all values are 0 for that covariate\n",
      "2025-01-21 17:12:32.631999 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.75 for 2953 genes in Lamp5_5 because all values are 0 for that covariate\n",
      "2025-01-21 17:13:04.155896 -- Skipping Lamp5_Lhx6_1 along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-21 17:17:46.207242 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.5 for 4121 genes in Lamp5_5 because all values are 0 for that covariate\n",
      "2025-01-21 17:18:37.650135 -- Adding 3 pseudocounts to sex=Female for 9 genes in Sst_19 because all values are 0 for that covariate\n",
      "2025-01-21 17:18:38.930548 -- Adding 3 pseudocounts to apoe4_status=Y for 1556 genes in Sst_19 because all values are 0 for that covariate\n",
      "2025-01-21 17:22:04.526480 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 17:22:04.527636 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 17:22:04.581079 -- Adding 3 pseudocounts to method=10xMulti for 733 genes in Sst_19 because all values are 0 for that covariate\n",
      "2025-01-21 17:23:43.090014 -- Adding 3 pseudocounts to race_choice_white=Unchecked for 5908 genes in Sst_19 because all values are 0 for that covariate\n",
      "2025-01-21 17:25:03.850100 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.25 for 4052 genes in Lamp5_5 because all values are 0 for that covariate\n",
      "2025-01-21 17:31:43.580534 -- Skipping L6_IT_2 along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-21 17:32:15.449691 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.0 for 5003 genes in Lamp5_5 because all values are 0 for that covariate\n",
      "2025-01-21 17:34:13.746446 -- Adding 3 pseudocounts to sex=Female for 9 genes in Pvalb_2 because all values are 0 for that covariate\n",
      "2025-01-21 17:34:27.125994 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 17:34:27.127067 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 17:34:27.516089 -- Removing race_choice_white from the covariate formula\n",
      "2025-01-21 17:34:33.980944 -- Skipping Pvalb_2 along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-21 17:34:34.894310 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 17:34:35.489021 -- Removing 5043 features from Pvalb_3 for low numbers of counts per cell\n",
      "2025-01-21 17:34:35.595923 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 17:34:35.632042 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.2003973100826735 for 2292 genes in Pvalb_3 because all values are 0 for that covariate\n",
      "2025-01-21 17:36:51.433393 -- Starting Sst_19 across continuous_pseudo_progression_score\n",
      "The cells are already grouped.The average number of cells per subject ( 8.26 ) is less than 30. The 'method' is set as 'HL'.\n",
      "Remove  0  genes having low expression.\n",
      "Analyzing  31043  genes with  19  subjects and  157  cells.\n",
      "2025-01-21 17:40:18.465117 -- Sst_19 along continuous_pseudo_progression_score was written to disk\n",
      "2025-01-21 17:40:19.538894 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 17:40:19.564182 -- Removing 13540 features from Sst_20 for low numbers of counts per cell\n",
      "2025-01-21 17:40:19.566611 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 17:40:19.579479 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.0 for 2289 genes in Sst_20 because all values are 0 for that covariate\n",
      "2025-01-21 17:40:48.814505 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.3343978735301191 for 2910 genes in Sst_20 because all values are 0 for that covariate\n",
      "2025-01-21 17:41:14.110969 -- Adding 3 pseudocounts to sex=Female for 3 genes in Lamp5_5 because all values are 0 for that covariate\n",
      "2025-01-21 17:41:14.456287 -- Adding 3 pseudocounts to sex=Male for 26 genes in Lamp5_5 because all values are 0 for that covariate\n",
      "2025-01-21 17:41:17.437170 -- Adding 3 pseudocounts to apoe4_status=Y for 22 genes in Lamp5_5 because all values are 0 for that covariate\n",
      "2025-01-21 17:41:19.927038 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 17:41:19.928121 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 17:41:19.967491 -- Adding 3 pseudocounts to method=10xMulti for 537 genes in Lamp5_5 because all values are 0 for that covariate\n",
      "2025-01-21 17:41:27.165447 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 17:41:27.189416 -- Removing sex from the covariate formula\n",
      "2025-01-21 17:41:27.189530 -- Removing apoe4_status from the covariate formula\n",
      "2025-01-21 17:41:27.190218 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 17:41:27.200571 -- Adding 3 pseudocounts to pseudo_genes_detected=0.271875 for 5069 genes in Sst_20 because all values are 0 for that covariate\n",
      "2025-01-21 17:42:17.698792 -- Adding 3 pseudocounts to race_choice_white=Unchecked for 7780 genes in Lamp5_5 because all values are 0 for that covariate\n",
      "2025-01-21 17:42:29.721468 -- Adding 3 pseudocounts to pseudo_genes_detected=0.0 for 6191 genes in Sst_20 because all values are 0 for that covariate\n",
      "2025-01-21 17:43:48.293667 -- Adding 3 pseudocounts to pseudo_genes_detected=0.2569711538461539 for 6671 genes in Sst_20 because all values are 0 for that covariate\n",
      "2025-01-21 17:45:14.570108 -- Adding 3 pseudocounts to pseudo_genes_detected=0.03725961538461535 for 5218 genes in Sst_20 because all values are 0 for that covariate\n",
      "2025-01-21 17:46:24.564798 -- Adding 3 pseudocounts to pseudo_genes_detected=0.019711538461538447 for 4140 genes in Sst_20 because all values are 0 for that covariate\n",
      "2025-01-21 17:47:21.399732 -- Adding 3 pseudocounts to pseudo_genes_detected=0.20024038461538463 for 6087 genes in Sst_20 because all values are 0 for that covariate\n",
      "2025-01-21 17:48:46.634042 -- Adding 3 pseudocounts to pseudo_genes_detected=0.2860576923076923 for 4629 genes in Sst_20 because all values are 0 for that covariate\n",
      "2025-01-21 17:49:53.669245 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 17:49:53.739441 -- Removing method from the covariate formula\n",
      "2025-01-21 17:49:53.739554 -- Removing race_choice_white from the covariate formula\n",
      "2025-01-21 17:49:53.766733 -- Starting Sst_20 across continuous_pseudo_progression_score\n",
      "The cells are already grouped.The average number of cells per subject ( 3.5 ) is less than 30. The 'method' is set as 'HL'.\n",
      "Remove  0  genes having low expression.\n",
      "Analyzing  23061  genes with  2  subjects and  7  cells.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: In addition: \n",
      "R[write to console]: Warning message:\n",
      "\n",
      "R[write to console]: In nebula(counts, obs$donor_id, pred = df, offset = obs$pseudo_number_of_umis,  :\n",
      "R[write to console]: \n",
      " \n",
      "R[write to console]:  Some predictors in the design matrix are collinear or linearly dependent. The effects of these predictors are not identifiable.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-21 17:51:28.779778 -- Sst_20 along continuous_pseudo_progression_score was written to disk\n",
      "2025-01-21 17:51:29.746049 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 17:51:29.780253 -- Removing 12305 features from Sst_23 for low numbers of counts per cell\n",
      "2025-01-21 17:51:29.783878 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 17:51:29.795730 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.3944035688622841 for 4315 genes in Sst_23 because all values are 0 for that covariate\n",
      "2025-01-21 17:52:42.862800 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.0 for 3436 genes in Sst_23 because all values are 0 for that covariate\n",
      "2025-01-21 17:53:41.080438 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.5400110102823146 for 4399 genes in Sst_23 because all values are 0 for that covariate\n",
      "2025-01-21 17:55:01.150296 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 17:55:01.183639 -- Removing sex from the covariate formula\n",
      "2025-01-21 17:55:01.207446 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 17:55:01.208257 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 17:55:01.208610 -- Removing method from the covariate formula\n",
      "2025-01-21 17:55:01.267157 -- Starting Sst_23 across continuous_pseudo_progression_score\n",
      "The cells are already grouped.The average number of cells per subject ( 4 ) is less than 30. The 'method' is set as 'HL'.\n",
      "Remove  1868  genes having low expression.\n",
      "Analyzing  22428  genes with  3  subjects and  12  cells.\n",
      "2025-01-21 17:56:26.322190 -- Skipping Lamp5_5 along continuous_pseudo_progression_score (already exists)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: In addition: \n",
      "R[write to console]: Warning message:\n",
      "\n",
      "R[write to console]: In nebula(counts, obs$donor_id, pred = df, offset = obs$pseudo_number_of_umis,  :\n",
      "R[write to console]: \n",
      " \n",
      "R[write to console]:  Some predictors in the design matrix are collinear or linearly dependent. The effects of these predictors are not identifiable.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-21 17:56:43.816906 -- Sst_23 along continuous_pseudo_progression_score was written to disk\n",
      "2025-01-21 17:56:44.736710 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 17:56:44.777621 -- Removing 9774 features from Sst_25 for low numbers of counts per cell\n",
      "2025-01-21 17:56:44.784146 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 17:56:44.795557 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.2381996399997909 for 4202 genes in Sst_25 because all values are 0 for that covariate\n",
      "2025-01-21 17:58:47.858554 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.7221901108281324 for 3026 genes in Sst_25 because all values are 0 for that covariate\n",
      "2025-01-21 18:00:17.758875 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.0 for 3422 genes in Sst_25 because all values are 0 for that covariate\n",
      "2025-01-21 18:00:22.883192 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.44256170877410667 for 1404 genes in Pvalb_3 because all values are 0 for that covariate\n",
      "2025-01-21 18:02:01.356039 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 18:02:01.437753 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 18:02:01.438662 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 18:02:01.439254 -- Removing method from the covariate formula\n",
      "2025-01-21 18:02:01.520733 -- Starting Sst_25 across continuous_pseudo_progression_score\n",
      "The cells are already grouped.The average number of cells per subject ( 10 ) is less than 30. The 'method' is set as 'HL'.\n",
      "Remove  1000  genes having low expression.\n",
      "Analyzing  25827  genes with  3  subjects and  30  cells.\n",
      "2025-01-21 18:03:08.291693 -- Skipping Vip_4 along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-21 18:03:09.807709 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 18:03:09.850349 -- Removing 9891 features from Vip_5 for low numbers of counts per cell\n",
      "2025-01-21 18:03:09.856016 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 18:03:09.869218 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.0 for 5987 genes in Vip_5 because all values are 0 for that covariate\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: In addition: \n",
      "R[write to console]: Warning message:\n",
      "\n",
      "R[write to console]: In nebula(counts, obs$donor_id, pred = df, offset = obs$pseudo_number_of_umis,  :\n",
      "R[write to console]: \n",
      " \n",
      "R[write to console]:  Some predictors in the design matrix are collinear or linearly dependent. The effects of these predictors are not identifiable.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-21 18:04:07.098163 -- Sst_25 along continuous_pseudo_progression_score was written to disk\n",
      "2025-01-21 18:05:37.143941 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.5116409139749925 for 3699 genes in Vip_5 because all values are 0 for that covariate\n",
      "2025-01-21 18:07:09.576108 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.6626252004619664 for 4394 genes in Vip_5 because all values are 0 for that covariate\n",
      "2025-01-21 18:09:02.080120 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.5446361465002717 for 5758 genes in Vip_5 because all values are 0 for that covariate\n",
      "2025-01-21 18:11:34.693100 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 18:11:34.731747 -- Removing sex from the covariate formula\n",
      "2025-01-21 18:11:34.756519 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 18:11:34.757248 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 18:11:34.781972 -- Removing race_choice_white from the covariate formula\n",
      "2025-01-21 18:11:34.827866 -- Skipping Vip_5 along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-21 18:11:35.823612 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 18:11:35.849363 -- Removing 11073 features from Vip_6 for low numbers of counts per cell\n",
      "2025-01-21 18:11:35.852087 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 18:11:35.864042 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.09311581412241624 for 2758 genes in Vip_6 because all values are 0 for that covariate\n",
      "2025-01-21 18:12:27.338722 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.1305249105440453 for 4749 genes in Vip_6 because all values are 0 for that covariate\n",
      "2025-01-21 18:12:37.044267 -- Skipping OPC_2 along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-21 18:12:37.934670 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 18:12:37.967559 -- Removing 9544 features from OPC_2_1-SEAAD for low numbers of counts per cell\n",
      "2025-01-21 18:12:37.972816 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 18:12:37.985005 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.0 for 8839 genes in OPC_2_1-SEAAD because all values are 0 for that covariate\n",
      "2025-01-21 18:13:58.499544 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.0 for 4745 genes in Vip_6 because all values are 0 for that covariate\n",
      "2025-01-21 18:15:32.083937 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 18:15:32.156531 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 18:15:32.157332 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 18:15:32.187836 -- Removing race_choice_white from the covariate formula\n",
      "2025-01-21 18:15:32.223315 -- Skipping Vip_6 along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-21 18:15:33.138126 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 18:15:33.288011 -- Removing 4621 features from Vip_9 for low numbers of counts per cell\n",
      "2025-01-21 18:15:33.310549 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 18:15:33.311678 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 18:15:33.329403 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.25 for 3959 genes in Vip_9 because all values are 0 for that covariate\n",
      "2025-01-21 18:15:46.806748 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.44634186825168054 for 3365 genes in Pvalb_3 because all values are 0 for that covariate\n",
      "2025-01-21 18:17:36.203663 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.2610991781329547 for 5955 genes in OPC_2_1-SEAAD because all values are 0 for that covariate\n",
      "2025-01-21 18:21:04.391449 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.5426955661336851 for 3931 genes in OPC_2_1-SEAAD because all values are 0 for that covariate\n",
      "2025-01-21 18:21:18.538272 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 18:21:18.540224 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 18:21:19.852697 -- Adding 3 pseudocounts to method=10xMulti for 35 genes in Oligo_2 because all values are 0 for that covariate\n",
      "2025-01-21 18:23:11.145212 -- Skipping L2_3_IT_1 along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-21 18:23:12.147225 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 18:23:12.305154 -- Removing 3969 features from L2_3_IT_3 for low numbers of counts per cell\n",
      "2025-01-21 18:23:12.327718 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 18:23:12.328662 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 18:23:12.344681 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.25 for 4230 genes in L2_3_IT_3 because all values are 0 for that covariate\n",
      "2025-01-21 18:23:24.822766 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.4555814716010037 for 3410 genes in OPC_2_1-SEAAD because all values are 0 for that covariate\n",
      "2025-01-21 18:24:19.206744 -- Adding 3 pseudocounts to race_choice_white=Unchecked for 1766 genes in Oligo_2 because all values are 0 for that covariate\n",
      "2025-01-21 18:25:28.831849 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.6421478368254953 for 6814 genes in OPC_2_1-SEAAD because all values are 0 for that covariate\n",
      "2025-01-21 18:25:55.355340 -- Adding 3 pseudocounts to age_at_death_binned_codes=1.0 for 2853 genes in Vip_9 because all values are 0 for that covariate\n",
      "2025-01-21 18:29:38.945913 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 18:29:39.045012 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 18:29:39.045791 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 18:29:39.046123 -- Removing method from the covariate formula\n",
      "2025-01-21 18:29:39.046254 -- Removing race_choice_white from the covariate formula\n",
      "2025-01-21 18:29:39.125942 -- Skipping OPC_2_1-SEAAD along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-21 18:29:40.101887 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 18:29:40.155087 -- Removing 8742 features from OPC_2_2-SEAAD for low numbers of counts per cell\n",
      "2025-01-21 18:29:40.163126 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 18:29:40.179558 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.6805884842827156 for 3903 genes in OPC_2_2-SEAAD because all values are 0 for that covariate\n",
      "2025-01-21 18:32:28.495060 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.5921040215808836 for 3433 genes in OPC_2_2-SEAAD because all values are 0 for that covariate\n",
      "2025-01-21 18:33:24.153854 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.5 for 3084 genes in Vip_9 because all values are 0 for that covariate\n",
      "2025-01-21 18:34:39.480907 -- Adding 3 pseudocounts to age_at_death_binned_codes=1.0 for 2861 genes in L2_3_IT_3 because all values are 0 for that covariate\n",
      "2025-01-21 18:34:58.385717 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.24235273105826066 for 5869 genes in OPC_2_2-SEAAD because all values are 0 for that covariate\n",
      "2025-01-21 18:39:15.382355 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.0 for 6974 genes in OPC_2_2-SEAAD because all values are 0 for that covariate\n",
      "2025-01-21 18:41:31.314812 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.75 for 2542 genes in Vip_9 because all values are 0 for that covariate\n",
      "2025-01-21 18:42:30.942761 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.5 for 2563 genes in L2_3_IT_3 because all values are 0 for that covariate\n",
      "2025-01-21 18:44:29.757726 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.8239942281019594 for 4978 genes in OPC_2_2-SEAAD because all values are 0 for that covariate\n",
      "2025-01-21 18:46:46.109508 -- Skipping Astro_2 along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-21 18:46:46.999549 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 18:46:47.293644 -- Removing 6216 features from Astro_3 for low numbers of counts per cell\n",
      "2025-01-21 18:46:47.351352 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 18:46:47.352519 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 18:46:47.389925 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.25 for 850 genes in Astro_3 because all values are 0 for that covariate\n",
      "2025-01-21 18:48:13.255901 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.0 for 4956 genes in Vip_9 because all values are 0 for that covariate\n",
      "2025-01-21 18:48:17.803166 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 18:48:17.889508 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 18:48:17.890427 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 18:48:17.920665 -- Removing race_choice_white from the covariate formula\n",
      "2025-01-21 18:48:18.009881 -- Skipping OPC_2_2-SEAAD along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-21 18:49:40.797727 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.75 for 1855 genes in L2_3_IT_3 because all values are 0 for that covariate\n",
      "2025-01-21 18:51:30.824518 -- Adding 3 pseudocounts to age_at_death_binned_codes=1.0 for 1722 genes in Astro_3 because all values are 0 for that covariate\n",
      "2025-01-21 18:52:48.920893 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.0 for 2410 genes in Pvalb_3 because all values are 0 for that covariate\n",
      "2025-01-21 18:54:57.248831 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.0 for 3209 genes in L2_3_IT_3 because all values are 0 for that covariate\n",
      "2025-01-21 19:01:00.527879 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.5 for 1361 genes in Astro_3 because all values are 0 for that covariate\n",
      "2025-01-21 19:01:08.813743 -- Adding 3 pseudocounts to sex=Female for 1 genes in Vip_9 because all values are 0 for that covariate\n",
      "2025-01-21 19:01:09.001865 -- Adding 3 pseudocounts to sex=Male for 24 genes in Vip_9 because all values are 0 for that covariate\n",
      "2025-01-21 19:01:13.093841 -- Adding 3 pseudocounts to apoe4_status=Y for 25 genes in Vip_9 because all values are 0 for that covariate\n",
      "2025-01-21 19:01:17.071566 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 19:01:17.072541 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 19:01:17.118606 -- Adding 3 pseudocounts to method=10xMulti for 912 genes in Vip_9 because all values are 0 for that covariate\n",
      "2025-01-21 19:03:42.277880 -- Adding 3 pseudocounts to race_choice_white=Unchecked for 9770 genes in Vip_9 because all values are 0 for that covariate\n",
      "2025-01-21 19:04:03.170032 -- Adding 3 pseudocounts to sex=Female for 27 genes in L2_3_IT_3 because all values are 0 for that covariate\n",
      "2025-01-21 19:04:07.821307 -- Adding 3 pseudocounts to sex=Male for 2 genes in L2_3_IT_3 because all values are 0 for that covariate\n",
      "2025-01-21 19:04:08.241906 -- Adding 3 pseudocounts to apoe4_status=Y for 51 genes in L2_3_IT_3 because all values are 0 for that covariate\n",
      "2025-01-21 19:04:17.009682 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 19:04:17.010662 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 19:04:17.063358 -- Adding 3 pseudocounts to method=10xMulti for 5300 genes in L2_3_IT_3 because all values are 0 for that covariate\n",
      "2025-01-21 19:08:35.531283 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.75 for 997 genes in Astro_3 because all values are 0 for that covariate\n",
      "2025-01-21 19:14:08.001852 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.0 for 4961 genes in Astro_3 because all values are 0 for that covariate\n",
      "2025-01-21 19:18:58.885441 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.5739030546125884 for 2206 genes in Pvalb_3 because all values are 0 for that covariate\n",
      "2025-01-21 19:19:10.993623 -- Adding 3 pseudocounts to race_choice_white=Unchecked for 5080 genes in L2_3_IT_3 because all values are 0 for that covariate\n",
      "2025-01-21 19:29:25.891692 -- Skipping Vip_9 along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-21 19:29:26.727180 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 19:29:27.027346 -- Removing 4625 features from Vip_11 for low numbers of counts per cell\n",
      "2025-01-21 19:29:27.081762 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 19:29:27.082946 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 19:29:27.104933 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.25 for 1887 genes in Vip_11 because all values are 0 for that covariate\n",
      "2025-01-21 19:33:33.853867 -- Skipping L2_3_IT_3 along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-21 19:33:34.776317 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 19:33:36.726725 -- Removing 3967 features from L2_3_IT_5 for low numbers of counts per cell\n",
      "2025-01-21 19:33:37.211869 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 19:33:37.213481 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 19:33:37.327462 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.25 for 269 genes in L2_3_IT_5 because all values are 0 for that covariate\n",
      "2025-01-21 19:39:55.231264 -- Adding 3 pseudocounts to age_at_death_binned_codes=1.0 for 1498 genes in Vip_11 because all values are 0 for that covariate\n",
      "2025-01-21 19:41:39.634742 -- Adding 3 pseudocounts to sex=Female for 23 genes in Astro_3 because all values are 0 for that covariate\n",
      "2025-01-21 19:41:47.509033 -- Adding 3 pseudocounts to apoe4_status=Y for 91 genes in Astro_3 because all values are 0 for that covariate\n",
      "2025-01-21 19:42:18.307112 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 19:42:18.308081 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 19:42:18.420195 -- Adding 3 pseudocounts to method=10xMulti for 202 genes in Astro_3 because all values are 0 for that covariate\n",
      "2025-01-21 19:42:59.958260 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 19:43:00.497315 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 19:43:00.498379 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 19:43:00.498794 -- Removing method from the covariate formula\n",
      "2025-01-21 19:43:00.498924 -- Removing race_choice_white from the covariate formula\n",
      "2025-01-21 19:43:03.003443 -- Skipping Pvalb_3 along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-21 19:43:03.899240 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 19:43:04.111564 -- Removing 6776 features from Pvalb_5 for low numbers of counts per cell\n",
      "2025-01-21 19:43:04.149342 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 19:43:04.185393 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.1305623567943546 for 936 genes in Pvalb_5 because all values are 0 for that covariate\n",
      "2025-01-21 19:43:25.515011 -- Adding 3 pseudocounts to race_choice_white=Unchecked for 2483 genes in Astro_3 because all values are 0 for that covariate\n",
      "2025-01-21 19:43:53.834503 -- Adding 3 pseudocounts to age_at_death_binned_codes=1.0 for 89 genes in L2_3_IT_5 because all values are 0 for that covariate\n",
      "2025-01-21 19:46:32.682186 -- Skipping L5_IT_2 along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-21 19:46:33.541335 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 19:46:33.645971 -- Removing 4146 features from L5_IT_3 for low numbers of counts per cell\n",
      "2025-01-21 19:46:33.661304 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 19:46:33.662455 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 19:46:33.679993 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.25 for 3021 genes in L5_IT_3 because all values are 0 for that covariate\n",
      "2025-01-21 19:46:44.864302 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.0 for 2245 genes in Pvalb_5 because all values are 0 for that covariate\n",
      "2025-01-21 19:47:18.174590 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.5 for 73 genes in L2_3_IT_5 because all values are 0 for that covariate\n",
      "2025-01-21 19:48:12.037986 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.5 for 1779 genes in Vip_11 because all values are 0 for that covariate\n",
      "2025-01-21 19:50:04.711770 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.75 for 10 genes in L2_3_IT_5 because all values are 0 for that covariate\n",
      "2025-01-21 19:50:27.508844 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.0 for 853 genes in L2_3_IT_5 because all values are 0 for that covariate\n",
      "2025-01-21 19:52:45.870734 -- Adding 3 pseudocounts to age_at_death_binned_codes=1.0 for 3799 genes in L5_IT_3 because all values are 0 for that covariate\n",
      "2025-01-21 19:55:35.224539 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 19:55:35.407088 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 19:55:35.408029 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 19:55:35.408396 -- Removing method from the covariate formula\n",
      "2025-01-21 19:55:35.408509 -- Removing race_choice_white from the covariate formula\n",
      "2025-01-21 19:55:35.944555 -- Skipping Pvalb_5 along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-21 19:55:36.843532 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 19:55:37.461096 -- Removing 5010 features from Pvalb_6 for low numbers of counts per cell\n",
      "2025-01-21 19:55:37.578783 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 19:55:37.609871 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.0 for 4500 genes in Pvalb_6 because all values are 0 for that covariate\n",
      "2025-01-21 19:57:12.633280 -- Skipping Astro_3 along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-21 19:57:13.626115 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 19:57:13.645762 -- Removing 16395 features from Astro_4 for low numbers of counts per cell\n",
      "2025-01-21 19:57:13.649002 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 19:57:13.649247 -- Removing continuous_pseudo_progression_score from the covariate formula\n",
      "2025-01-21 19:57:13.649847 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 19:57:13.649963 -- Removing age_at_death_binned_codes from the covariate formula\n",
      "2025-01-21 19:57:13.650140 -- Removing sex from the covariate formula\n",
      "2025-01-21 19:57:13.650244 -- Removing apoe4_status from the covariate formula\n",
      "2025-01-21 19:57:13.650815 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 19:57:13.661042 -- Adding 3 pseudocounts to pseudo_genes_detected=0.5627002288329519 for 4250 genes in Astro_4 because all values are 0 for that covariate\n",
      "2025-01-21 19:57:51.406298 -- Adding 3 pseudocounts to pseudo_genes_detected=0.27505720823798635 for 5830 genes in Astro_4 because all values are 0 for that covariate\n",
      "2025-01-21 19:58:02.411822 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.75 for 1236 genes in Vip_11 because all values are 0 for that covariate\n",
      "2025-01-21 19:58:45.229574 -- Adding 3 pseudocounts to pseudo_genes_detected=0.5274599542334095 for 4042 genes in Astro_4 because all values are 0 for that covariate\n",
      "2025-01-21 19:59:23.659712 -- Adding 3 pseudocounts to pseudo_genes_detected=0.0 for 4875 genes in Astro_4 because all values are 0 for that covariate\n",
      "2025-01-21 20:00:11.179527 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 20:00:11.214190 -- Removing method from the covariate formula\n",
      "2025-01-21 20:00:11.214308 -- Removing race_choice_white from the covariate formula\n",
      "2025-01-21 20:00:11.236977 -- Starting Astro_4 across continuous_pseudo_progression_score\n",
      "The cells are already grouped.2025-01-21 20:00:11.285042 -- Skipping Astro_4 because some predictors have zero variation or a zero vector\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: Error in nebula(counts, obs$donor_id, pred = df, offset = obs$pseudo_number_of_umis,  : \n",
      "  Some predictors have zero variation or a zero vector.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-21 20:00:12.179688 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 20:00:12.218427 -- Removing 9462 features from Astro_5 for low numbers of counts per cell\n",
      "2025-01-21 20:00:12.225681 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 20:00:12.237025 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.5772646320229491 for 6673 genes in Astro_5 because all values are 0 for that covariate\n",
      "2025-01-21 20:00:34.062399 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.5 for 2880 genes in L5_IT_3 because all values are 0 for that covariate\n",
      "2025-01-21 20:03:29.392642 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.7579101799462293 for 5258 genes in Astro_5 because all values are 0 for that covariate\n",
      "2025-01-21 20:04:53.421954 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.0 for 3045 genes in Vip_11 because all values are 0 for that covariate\n",
      "2025-01-21 20:06:12.629632 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.4980047339953148 for 4648 genes in Astro_5 because all values are 0 for that covariate\n",
      "2025-01-21 20:06:33.750665 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.75 for 1543 genes in L5_IT_3 because all values are 0 for that covariate\n",
      "2025-01-21 20:08:40.007483 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.08119837912323202 for 4674 genes in Astro_5 because all values are 0 for that covariate\n",
      "2025-01-21 20:09:44.689206 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.0 for 6128 genes in L5_IT_3 because all values are 0 for that covariate\n",
      "2025-01-21 20:11:12.937632 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.0 for 6294 genes in Astro_5 because all values are 0 for that covariate\n",
      "2025-01-21 20:14:42.224240 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 20:14:42.311266 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 20:14:42.312014 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 20:14:42.312309 -- Removing method from the covariate formula\n",
      "2025-01-21 20:14:42.312423 -- Removing race_choice_white from the covariate formula\n",
      "2025-01-21 20:14:42.380548 -- Starting Astro_5 across continuous_pseudo_progression_score\n",
      "The cells are already grouped.The average number of cells per subject ( 5.8 ) is less than 30. The 'method' is set as 'HL'.\n",
      "Remove  0  genes having low expression.\n",
      "Analyzing  27139  genes with  5  subjects and  29  cells.\n",
      "2025-01-21 20:16:48.170671 -- Astro_5 along continuous_pseudo_progression_score was written to disk\n",
      "2025-01-21 20:16:49.248419 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 20:16:49.320506 -- Removing 7469 features from Astro_6-SEAAD for low numbers of counts per cell\n",
      "2025-01-21 20:16:49.330700 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 20:16:49.342664 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.4391105805341134 for 10092 genes in Astro_6-SEAAD because all values are 0 for that covariate\n",
      "2025-01-21 20:21:59.790896 -- Adding 3 pseudocounts to sex=Female for 1 genes in Vip_11 because all values are 0 for that covariate\n",
      "2025-01-21 20:22:00.172671 -- Adding 3 pseudocounts to sex=Male for 10 genes in Vip_11 because all values are 0 for that covariate\n",
      "2025-01-21 20:22:03.707918 -- Adding 3 pseudocounts to apoe4_status=Y for 120 genes in Vip_11 because all values are 0 for that covariate\n",
      "2025-01-21 20:22:26.345545 -- Adding 3 pseudocounts to apoe4_status=Y for 120 genes in L5_IT_3 because all values are 0 for that covariate\n",
      "2025-01-21 20:22:41.433023 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 20:22:41.433910 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 20:22:41.476619 -- Adding 3 pseudocounts to method=10xMulti for 1158 genes in L5_IT_3 because all values are 0 for that covariate\n",
      "2025-01-21 20:22:44.306864 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 20:22:44.308081 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 20:22:44.415797 -- Adding 3 pseudocounts to method=10xMulti for 1028 genes in Vip_11 because all values are 0 for that covariate\n",
      "2025-01-21 20:22:48.924476 -- Adding 3 pseudocounts to sex=Female for 1 genes in L2_3_IT_5 because all values are 0 for that covariate\n",
      "2025-01-21 20:22:52.149366 -- Adding 3 pseudocounts to apoe4_status=Y for 9 genes in L2_3_IT_5 because all values are 0 for that covariate\n",
      "2025-01-21 20:23:12.683017 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 20:23:12.684642 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 20:23:13.309836 -- Adding 3 pseudocounts to method=10xMulti for 1532 genes in L2_3_IT_5 because all values are 0 for that covariate\n",
      "2025-01-21 20:25:06.263079 -- Removing race_choice_white from the covariate formula\n",
      "2025-01-21 20:25:06.525195 -- Skipping L5_IT_3 along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-21 20:25:07.431505 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 20:25:08.908537 -- Removing 3693 features from L5_IT_5 for low numbers of counts per cell\n",
      "2025-01-21 20:25:09.250740 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 20:25:09.251950 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 20:25:09.334905 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.25 for 499 genes in L5_IT_5 because all values are 0 for that covariate\n",
      "2025-01-21 20:28:31.534997 -- Adding 3 pseudocounts to race_choice_white=Unchecked for 6726 genes in Vip_11 because all values are 0 for that covariate\n",
      "2025-01-21 20:28:39.308730 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.6805884842827156 for 6509 genes in Astro_6-SEAAD because all values are 0 for that covariate\n",
      "2025-01-21 20:36:23.162815 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.5921040215808836 for 7370 genes in Astro_6-SEAAD because all values are 0 for that covariate\n",
      "2025-01-21 20:40:27.305593 -- Adding 3 pseudocounts to age_at_death_binned_codes=1.0 for 45 genes in L5_IT_5 because all values are 0 for that covariate\n",
      "2025-01-21 20:41:50.187396 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.5 for 230 genes in L5_IT_5 because all values are 0 for that covariate\n",
      "2025-01-21 20:45:17.047103 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.0 for 5705 genes in Astro_6-SEAAD because all values are 0 for that covariate\n",
      "2025-01-21 20:46:39.937755 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.3866247141842833 for 1858 genes in Pvalb_6 because all values are 0 for that covariate\n",
      "2025-01-21 20:48:51.544974 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.75 for 17 genes in L5_IT_5 because all values are 0 for that covariate\n",
      "2025-01-21 20:49:22.720466 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.0 for 1120 genes in L5_IT_5 because all values are 0 for that covariate\n",
      "2025-01-21 20:52:14.580204 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.4327433703086429 for 6091 genes in Astro_6-SEAAD because all values are 0 for that covariate\n",
      "2025-01-21 20:53:25.097651 -- Skipping Oligo_2 along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-21 20:53:25.955796 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 20:53:26.174512 -- Removing 7046 features from Oligo_2_1-SEAAD for low numbers of counts per cell\n",
      "2025-01-21 20:53:26.208345 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 20:53:26.209588 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 20:53:26.250624 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.75 for 407 genes in Oligo_2_1-SEAAD because all values are 0 for that covariate\n",
      "2025-01-21 20:55:14.522534 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.0 for 8529 genes in Oligo_2_1-SEAAD because all values are 0 for that covariate\n",
      "2025-01-21 20:59:43.882098 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.4333650298931525 for 3160 genes in Astro_6-SEAAD because all values are 0 for that covariate\n",
      "2025-01-21 21:03:40.142199 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.8239942281019594 for 7549 genes in Astro_6-SEAAD because all values are 0 for that covariate\n",
      "2025-01-21 21:06:17.779685 -- Skipping Vip_11 along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-21 21:06:18.781103 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 21:06:18.841230 -- Removing 6900 features from Vip_12 for low numbers of counts per cell\n",
      "2025-01-21 21:06:18.848488 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 21:06:18.849435 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 21:06:18.862697 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.25 for 5459 genes in Vip_12 because all values are 0 for that covariate\n",
      "2025-01-21 21:07:46.872816 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.0011244734255105116 for 1677 genes in Pvalb_6 because all values are 0 for that covariate\n",
      "2025-01-21 21:11:08.444637 -- Adding 3 pseudocounts to age_at_death_binned_codes=1.0 for 3810 genes in Vip_12 because all values are 0 for that covariate\n",
      "2025-01-21 21:13:05.345990 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.43612000773889137 for 8674 genes in Astro_6-SEAAD because all values are 0 for that covariate\n",
      "2025-01-21 21:14:33.758628 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.75 for 4021 genes in Vip_12 because all values are 0 for that covariate\n",
      "2025-01-21 21:18:11.757610 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.5 for 4891 genes in Vip_12 because all values are 0 for that covariate\n",
      "2025-01-21 21:21:47.208938 -- Adding 3 pseudocounts to race_choice_white=Unchecked for 2181 genes in L2_3_IT_5 because all values are 0 for that covariate\n",
      "2025-01-21 21:22:40.646473 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.0 for 6251 genes in Vip_12 because all values are 0 for that covariate\n",
      "2025-01-21 21:23:38.704769 -- Adding 3 pseudocounts to apoe4_status=Y for 31 genes in L5_IT_5 because all values are 0 for that covariate\n",
      "2025-01-21 21:24:02.701407 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 21:24:02.825756 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 21:24:02.826567 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 21:24:02.856388 -- Removing race_choice_white from the covariate formula\n",
      "2025-01-21 21:24:03.008712 -- Starting Astro_6-SEAAD across continuous_pseudo_progression_score\n",
      "The cells are already grouped.The average number of cells per subject ( 10.38 ) is less than 30. The 'method' is set as 'HL'.\n",
      "Remove  0  genes having low expression.\n",
      "Analyzing  29132  genes with  8  subjects and  83  cells.\n",
      "2025-01-21 21:24:35.969448 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 21:24:35.970720 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 21:24:36.444469 -- Adding 3 pseudocounts to method=10xMulti for 440 genes in L5_IT_5 because all values are 0 for that covariate\n",
      "2025-01-21 21:26:33.240908 -- Astro_6-SEAAD along continuous_pseudo_progression_score was written to disk\n",
      "2025-01-21 21:26:49.153604 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.4471274380159813 for 3540 genes in Pvalb_6 because all values are 0 for that covariate\n",
      "2025-01-21 21:28:29.839888 -- Adding 3 pseudocounts to apoe4_status=Y for 1617 genes in Vip_12 because all values are 0 for that covariate\n",
      "2025-01-21 21:29:59.688839 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 21:29:59.689863 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 21:29:59.719105 -- Adding 3 pseudocounts to method=10xMulti for 1491 genes in Vip_12 because all values are 0 for that covariate\n",
      "2025-01-21 21:31:23.372617 -- Adding 3 pseudocounts to race_choice_white=Unchecked for 5475 genes in Vip_12 because all values are 0 for that covariate\n",
      "2025-01-21 21:33:29.897637 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.25 for 4647 genes in Oligo_2_1-SEAAD because all values are 0 for that covariate\n",
      "2025-01-21 21:36:34.228228 -- Skipping Vip_12 along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-21 21:36:35.158634 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 21:36:35.240661 -- Removing 5167 features from Vip_13 for low numbers of counts per cell\n",
      "2025-01-21 21:36:35.256208 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 21:36:35.257616 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 21:36:35.273136 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.25 for 4494 genes in Vip_13 because all values are 0 for that covariate\n",
      "2025-01-21 21:38:01.836522 -- Adding 3 pseudocounts to race_choice_white=Unchecked for 3764 genes in L5_IT_5 because all values are 0 for that covariate\n",
      "2025-01-21 21:44:24.241715 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.5 for 4589 genes in Vip_13 because all values are 0 for that covariate\n",
      "2025-01-21 21:52:26.511637 -- Adding 3 pseudocounts to age_at_death_binned_codes=1.0 for 4358 genes in Vip_13 because all values are 0 for that covariate\n",
      "2025-01-21 21:54:12.502066 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.5 for 2168 genes in Oligo_2_1-SEAAD because all values are 0 for that covariate\n",
      "2025-01-21 22:00:06.619961 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.75 for 1800 genes in Vip_13 because all values are 0 for that covariate\n",
      "2025-01-21 22:03:16.572140 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.0 for 6232 genes in Vip_13 because all values are 0 for that covariate\n",
      "2025-01-21 22:03:51.887207 -- Adding 3 pseudocounts to age_at_death_binned_codes=1.0 for 7702 genes in Oligo_2_1-SEAAD because all values are 0 for that covariate\n",
      "2025-01-21 22:06:30.108285 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.7395815668730475 for 3916 genes in Pvalb_6 because all values are 0 for that covariate\n",
      "2025-01-21 22:14:21.828584 -- Adding 3 pseudocounts to sex=Female for 25 genes in Vip_13 because all values are 0 for that covariate\n",
      "2025-01-21 22:14:24.642867 -- Adding 3 pseudocounts to apoe4_status=Y for 127 genes in Vip_13 because all values are 0 for that covariate\n",
      "2025-01-21 22:14:38.302479 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 22:14:38.303343 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 22:14:38.349442 -- Adding 3 pseudocounts to method=10xMulti for 187 genes in Vip_13 because all values are 0 for that covariate\n",
      "2025-01-21 22:14:58.247434 -- Adding 3 pseudocounts to race_choice_white=Unchecked for 5671 genes in Vip_13 because all values are 0 for that covariate\n",
      "2025-01-21 22:25:05.404485 -- Skipping Vip_13 along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-21 22:25:06.362077 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 22:25:06.474878 -- Removing 4746 features from Vip_15 for low numbers of counts per cell\n",
      "2025-01-21 22:25:06.496382 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 22:25:06.497378 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 22:25:06.514258 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.25 for 3401 genes in Vip_15 because all values are 0 for that covariate\n",
      "2025-01-21 22:32:25.058911 -- Skipping L4_IT_2 along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-21 22:32:26.047292 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 22:32:26.127253 -- Removing 5367 features from L4_IT_3 for low numbers of counts per cell\n",
      "2025-01-21 22:32:26.138097 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 22:32:26.139236 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 22:32:26.156147 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.25 for 1574 genes in L4_IT_3 because all values are 0 for that covariate\n",
      "2025-01-21 22:32:29.707959 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.5 for 3073 genes in Vip_15 because all values are 0 for that covariate\n",
      "2025-01-21 22:34:35.094585 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.5 for 3946 genes in L4_IT_3 because all values are 0 for that covariate\n",
      "2025-01-21 22:38:14.337299 -- Adding 3 pseudocounts to sex=Female for 7 genes in Oligo_2_1-SEAAD because all values are 0 for that covariate\n",
      "2025-01-21 22:38:16.308640 -- Adding 3 pseudocounts to apoe4_status=Y for 4 genes in Oligo_2_1-SEAAD because all values are 0 for that covariate\n",
      "2025-01-21 22:38:17.470114 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 22:38:17.471444 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 22:38:17.610788 -- Adding 3 pseudocounts to race_choice_white=Unchecked for 6097 genes in Oligo_2_1-SEAAD because all values are 0 for that covariate\n",
      "2025-01-21 22:39:11.830155 -- Adding 3 pseudocounts to age_at_death_binned_codes=1.0 for 4506 genes in Vip_15 because all values are 0 for that covariate\n",
      "2025-01-21 22:40:00.163143 -- Adding 3 pseudocounts to age_at_death_binned_codes=1.0 for 5381 genes in L4_IT_3 because all values are 0 for that covariate\n",
      "2025-01-21 22:45:03.093677 -- Skipping L2_3_IT_5 along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-21 22:45:04.016714 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 22:45:13.294074 -- Removing 4007 features from L2_3_IT_6 for low numbers of counts per cell\n",
      "2025-01-21 22:45:15.210501 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 22:45:15.212764 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 22:45:15.644360 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.25 for 36 genes in L2_3_IT_6 because all values are 0 for that covariate\n",
      "2025-01-21 22:47:25.121816 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.75 for 3986 genes in L4_IT_3 because all values are 0 for that covariate\n",
      "2025-01-21 22:49:09.630869 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.75 for 2511 genes in Vip_15 because all values are 0 for that covariate\n",
      "2025-01-21 22:50:20.579916 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.15756432628475384 for 1941 genes in Pvalb_6 because all values are 0 for that covariate\n",
      "2025-01-21 22:52:02.246187 -- Adding 3 pseudocounts to age_at_death_binned_codes=1.0 for 3 genes in L2_3_IT_6 because all values are 0 for that covariate\n",
      "2025-01-21 22:52:36.473080 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.5 for 2 genes in L2_3_IT_6 because all values are 0 for that covariate\n",
      "2025-01-21 22:53:00.194058 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.0 for 56 genes in L2_3_IT_6 because all values are 0 for that covariate\n",
      "2025-01-21 22:53:01.681520 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.0 for 9490 genes in L4_IT_3 because all values are 0 for that covariate\n",
      "2025-01-21 22:54:41.290960 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.0 for 4324 genes in Vip_15 because all values are 0 for that covariate\n",
      "2025-01-21 23:03:43.860616 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 23:03:43.862510 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 23:03:46.801420 -- Adding 3 pseudocounts to method=10xMulti for 307 genes in L2_3_IT_6 because all values are 0 for that covariate\n",
      "2025-01-21 23:04:13.938879 -- Adding 3 pseudocounts to sex=Female for 1 genes in Vip_15 because all values are 0 for that covariate\n",
      "2025-01-21 23:04:14.113645 -- Adding 3 pseudocounts to sex=Male for 24 genes in Vip_15 because all values are 0 for that covariate\n",
      "2025-01-21 23:04:17.406835 -- Adding 3 pseudocounts to apoe4_status=Y for 140 genes in Vip_15 because all values are 0 for that covariate\n",
      "2025-01-21 23:04:36.133587 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 23:04:36.134452 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 23:04:36.178662 -- Adding 3 pseudocounts to method=10xMulti for 1182 genes in Vip_15 because all values are 0 for that covariate\n",
      "2025-01-21 23:05:33.393998 -- Skipping Oligo_2_1-SEAAD along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-21 23:05:34.305108 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 23:05:37.663270 -- Removing 7340 features from Oligo_4 for low numbers of counts per cell\n",
      "2025-01-21 23:05:38.278545 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 23:05:38.279861 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 23:05:38.391357 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.25 for 408 genes in Oligo_4 because all values are 0 for that covariate\n",
      "2025-01-21 23:06:31.054246 -- Adding 3 pseudocounts to sex=Female for 13 genes in L4_IT_3 because all values are 0 for that covariate\n",
      "2025-01-21 23:06:32.255733 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 23:06:32.256570 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 23:06:32.294136 -- Adding 3 pseudocounts to method=10xMulti for 245 genes in L4_IT_3 because all values are 0 for that covariate\n",
      "2025-01-21 23:06:53.682907 -- Removing race_choice_white from the covariate formula\n",
      "2025-01-21 23:06:53.861745 -- Skipping L4_IT_3 along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-21 23:06:54.807308 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 23:06:55.777010 -- Removing 4158 features from L4_IT_4 for low numbers of counts per cell\n",
      "2025-01-21 23:06:55.986941 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 23:06:55.988143 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 23:06:56.052442 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.25 for 500 genes in L4_IT_4 because all values are 0 for that covariate\n",
      "2025-01-21 23:07:14.137901 -- Removing race_choice_white from the covariate formula\n",
      "2025-01-21 23:07:15.330686 -- Skipping Vip_15 along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-21 23:07:16.198208 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 23:07:16.217847 -- Removing 15304 features from Vip_18 for low numbers of counts per cell\n",
      "2025-01-21 23:07:16.219718 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 23:07:16.219896 -- Removing continuous_pseudo_progression_score from the covariate formula\n",
      "2025-01-21 23:07:16.220541 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 23:07:16.220666 -- Removing age_at_death_binned_codes from the covariate formula\n",
      "2025-01-21 23:07:16.220827 -- Removing sex from the covariate formula\n",
      "2025-01-21 23:07:16.220948 -- Removing apoe4_status from the covariate formula\n",
      "2025-01-21 23:07:16.221559 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 23:07:16.232690 -- Adding 3 pseudocounts to pseudo_genes_detected=0.0 for 3574 genes in Vip_18 because all values are 0 for that covariate\n",
      "2025-01-21 23:07:47.280421 -- Adding 3 pseudocounts to pseudo_genes_detected=0.30771175726927946 for 3024 genes in Vip_18 because all values are 0 for that covariate\n",
      "2025-01-21 23:08:13.563595 -- Adding 3 pseudocounts to pseudo_genes_detected=0.25638432364096087 for 4323 genes in Vip_18 because all values are 0 for that covariate\n",
      "2025-01-21 23:08:52.581311 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 23:08:52.608181 -- Removing method from the covariate formula\n",
      "2025-01-21 23:08:52.608281 -- Removing race_choice_white from the covariate formula\n",
      "2025-01-21 23:08:52.629426 -- Starting Vip_18 across continuous_pseudo_progression_score\n",
      "The cells are already grouped.2025-01-21 23:08:52.684003 -- Skipping Vip_18 because some predictors have zero variation or a zero vector\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: Error in nebula(counts, obs$donor_id, pred = df, offset = obs$pseudo_number_of_umis,  : \n",
      "  Some predictors have zero variation or a zero vector.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-21 23:08:53.681901 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 23:08:53.808325 -- Removing 5156 features from Vip_19 for low numbers of counts per cell\n",
      "2025-01-21 23:08:53.824934 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 23:08:53.825832 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 23:08:53.841966 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.25 for 3727 genes in Vip_19 because all values are 0 for that covariate\n",
      "2025-01-21 23:12:02.550272 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 23:12:03.146231 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 23:12:03.147293 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 23:12:03.147691 -- Removing method from the covariate formula\n",
      "2025-01-21 23:12:05.344444 -- Skipping Pvalb_6 along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-21 23:12:06.711450 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 23:12:06.864553 -- Removing 5642 features from Pvalb_7 for low numbers of counts per cell\n",
      "2025-01-21 23:12:06.899935 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 23:12:06.939796 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.0 for 610 genes in Pvalb_7 because all values are 0 for that covariate\n",
      "2025-01-21 23:14:12.883336 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.6043731048194847 for 2413 genes in Pvalb_7 because all values are 0 for that covariate\n",
      "2025-01-21 23:16:06.951506 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.5 for 3140 genes in Vip_19 because all values are 0 for that covariate\n",
      "2025-01-21 23:16:16.655666 -- Adding 3 pseudocounts to age_at_death_binned_codes=1.0 for 418 genes in L4_IT_4 because all values are 0 for that covariate\n",
      "2025-01-21 23:22:14.787385 -- Adding 3 pseudocounts to age_at_death_binned_codes=1.0 for 3562 genes in Vip_19 because all values are 0 for that covariate\n",
      "2025-01-21 23:22:32.834868 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 23:22:32.898211 -- Removing sex from the covariate formula\n",
      "2025-01-21 23:22:32.898322 -- Removing apoe4_status from the covariate formula\n",
      "2025-01-21 23:22:32.899295 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 23:22:32.900051 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 23:22:32.900388 -- Removing method from the covariate formula\n",
      "2025-01-21 23:22:32.900556 -- Removing race_choice_white from the covariate formula\n",
      "2025-01-21 23:22:33.417857 -- Skipping Pvalb_7 along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-21 23:22:34.309975 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 23:22:34.392428 -- Removing 7276 features from Pvalb_8 for low numbers of counts per cell\n",
      "2025-01-21 23:22:34.406549 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 23:22:34.406766 -- Removing continuous_pseudo_progression_score from the covariate formula\n",
      "2025-01-21 23:22:34.407439 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 23:22:34.407566 -- Removing age_at_death_binned_codes from the covariate formula\n",
      "2025-01-21 23:22:34.407708 -- Removing sex from the covariate formula\n",
      "2025-01-21 23:22:34.407819 -- Removing apoe4_status from the covariate formula\n",
      "2025-01-21 23:22:34.408437 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 23:22:34.409125 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 23:22:34.409382 -- Removing method from the covariate formula\n",
      "2025-01-21 23:22:34.409485 -- Removing race_choice_white from the covariate formula\n",
      "2025-01-21 23:22:34.611033 -- Starting Pvalb_8 across continuous_pseudo_progression_score\n",
      "The cells are already grouped.2025-01-21 23:22:34.664463 -- Skipping Pvalb_8 because some predictors have zero variation or a zero vector\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: Error in nebula(counts, obs$donor_id, pred = df, offset = obs$pseudo_number_of_umis,  : \n",
      "  Some predictors have zero variation or a zero vector.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-21 23:22:35.698484 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 23:22:36.183310 -- Removing 5620 features from Pvalb_9 for low numbers of counts per cell\n",
      "2025-01-21 23:22:36.276298 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 23:22:36.325673 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.3774144952899409 for 2427 genes in Pvalb_9 because all values are 0 for that covariate\n",
      "2025-01-21 23:24:04.055271 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.5 for 390 genes in L4_IT_4 because all values are 0 for that covariate\n",
      "2025-01-21 23:29:15.849709 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.75 for 3411 genes in Vip_19 because all values are 0 for that covariate\n",
      "2025-01-21 23:29:54.366836 -- Adding 3 pseudocounts to age_at_death_binned_codes=1.0 for 56 genes in Oligo_4 because all values are 0 for that covariate\n",
      "2025-01-21 23:31:21.004792 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.75 for 86 genes in L4_IT_4 because all values are 0 for that covariate\n",
      "2025-01-21 23:32:36.775165 -- Skipping L5_IT_5 along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-21 23:32:37.725447 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-21 23:32:37.963195 -- Removing 4111 features from L5_IT_6 for low numbers of counts per cell\n",
      "2025-01-21 23:32:38.007053 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-21 23:32:38.008490 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-21 23:32:38.036845 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.25 for 1388 genes in L5_IT_6 because all values are 0 for that covariate\n",
      "2025-01-21 23:32:57.838204 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.0 for 1911 genes in L4_IT_4 because all values are 0 for that covariate\n",
      "2025-01-21 23:33:15.401510 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.5 for 10 genes in Oligo_4 because all values are 0 for that covariate\n",
      "2025-01-21 23:33:51.822413 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.75 for 11 genes in Oligo_4 because all values are 0 for that covariate\n",
      "2025-01-21 23:34:31.223297 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.0 for 3393 genes in Oligo_4 because all values are 0 for that covariate\n",
      "2025-01-21 23:35:57.439527 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.0 for 4026 genes in Vip_19 because all values are 0 for that covariate\n",
      "2025-01-21 23:38:33.507262 -- Adding 3 pseudocounts to age_at_death_binned_codes=1.0 for 4128 genes in L5_IT_6 because all values are 0 for that covariate\n",
      "2025-01-21 23:42:59.922524 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.20218260643368402 for 2694 genes in Pvalb_9 because all values are 0 for that covariate\n",
      "2025-01-21 23:43:56.801419 -- Adding 3 pseudocounts to sex=Male for 43 genes in Vip_19 because all values are 0 for that covariate\n",
      "2025-01-21 23:44:01.954393 -- Adding 3 pseudocounts to apoe4_status=Y for 406 genes in Vip_19 because all values are 0 for that covariate\n",
      "2025-01-21 23:44:50.549065 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-21 23:44:50.549926 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-21 23:44:50.611077 -- Adding 3 pseudocounts to method=10xMulti for 1131 genes in Vip_19 because all values are 0 for that covariate\n",
      "2025-01-21 23:47:05.639063 -- Adding 3 pseudocounts to race_choice_white=Unchecked for 6048 genes in Vip_19 because all values are 0 for that covariate\n",
      "2025-01-21 23:55:56.986515 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.5 for 2001 genes in L5_IT_6 because all values are 0 for that covariate\n",
      "2025-01-21 23:59:10.106037 -- Starting Vip_19 across continuous_pseudo_progression_score\n",
      "The cells are already grouped.The average number of cells per subject ( 5.56 ) is less than 30. The 'method' is set as 'HL'.\n",
      "Remove  0  genes having low expression.\n",
      "Analyzing  31445  genes with  25  subjects and  139  cells.\n",
      "2025-01-22 00:00:58.409165 -- Adding 3 pseudocounts to race_choice_white=Unchecked for 1579 genes in L2_3_IT_6 because all values are 0 for that covariate\n",
      "2025-01-22 00:02:20.219460 -- Vip_19 along continuous_pseudo_progression_score was written to disk\n",
      "2025-01-22 00:02:21.280120 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-22 00:02:21.433056 -- Removing 4763 features from Vip_23 for low numbers of counts per cell\n",
      "2025-01-22 00:02:21.452460 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-22 00:02:21.453650 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-22 00:02:21.470891 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.25 for 3496 genes in Vip_23 because all values are 0 for that covariate\n",
      "2025-01-22 00:04:24.422186 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.75 for 427 genes in L5_IT_6 because all values are 0 for that covariate\n",
      "2025-01-22 00:05:33.717931 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.4763411065681484 for 9472 genes in Pvalb_9 because all values are 0 for that covariate\n",
      "2025-01-22 00:06:12.772123 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.0 for 3367 genes in L5_IT_6 because all values are 0 for that covariate\n",
      "2025-01-22 00:08:40.776766 -- Adding 3 pseudocounts to sex=Female for 1 genes in L4_IT_4 because all values are 0 for that covariate\n",
      "2025-01-22 00:08:42.125033 -- Adding 3 pseudocounts to sex=Male for 1 genes in L4_IT_4 because all values are 0 for that covariate\n",
      "2025-01-22 00:08:43.746318 -- Adding 3 pseudocounts to apoe4_status=Y for 43 genes in L4_IT_4 because all values are 0 for that covariate\n",
      "2025-01-22 00:09:33.719399 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-22 00:09:33.721094 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-22 00:09:34.045753 -- Adding 3 pseudocounts to method=10xMulti for 139 genes in L4_IT_4 because all values are 0 for that covariate\n",
      "2025-01-22 00:10:34.668230 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.5 for 3047 genes in Vip_23 because all values are 0 for that covariate\n",
      "2025-01-22 00:12:17.212008 -- Adding 3 pseudocounts to race_choice_white=Unchecked for 5575 genes in L4_IT_4 because all values are 0 for that covariate\n",
      "2025-01-22 00:17:43.690430 -- Adding 3 pseudocounts to age_at_death_binned_codes=1.0 for 3580 genes in Vip_23 because all values are 0 for that covariate\n",
      "2025-01-22 00:20:33.192422 -- Adding 3 pseudocounts to sex=Female for 3 genes in L5_IT_6 because all values are 0 for that covariate\n",
      "2025-01-22 00:20:34.046433 -- Adding 3 pseudocounts to apoe4_status=Y for 51 genes in L5_IT_6 because all values are 0 for that covariate\n",
      "2025-01-22 00:20:47.129891 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-22 00:20:47.130741 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-22 00:20:47.196654 -- Adding 3 pseudocounts to method=10xMulti for 1140 genes in L5_IT_6 because all values are 0 for that covariate\n",
      "2025-01-22 00:25:38.641406 -- Adding 3 pseudocounts to race_choice_white=Unchecked for 5117 genes in L5_IT_6 because all values are 0 for that covariate\n",
      "2025-01-22 00:26:10.991186 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.75 for 2636 genes in Vip_23 because all values are 0 for that covariate\n",
      "2025-01-22 00:32:23.792644 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.0 for 4683 genes in Vip_23 because all values are 0 for that covariate\n",
      "2025-01-22 00:43:30.247792 -- Adding 3 pseudocounts to sex=Female for 3 genes in Vip_23 because all values are 0 for that covariate\n",
      "2025-01-22 00:43:30.721302 -- Adding 3 pseudocounts to sex=Male for 15 genes in Vip_23 because all values are 0 for that covariate\n",
      "2025-01-22 00:43:32.920531 -- Adding 3 pseudocounts to apoe4_status=Y for 365 genes in Vip_23 because all values are 0 for that covariate\n",
      "2025-01-22 00:44:24.411642 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-22 00:44:24.412464 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-22 00:44:24.461060 -- Adding 3 pseudocounts to method=10xMulti for 1210 genes in Vip_23 because all values are 0 for that covariate\n",
      "2025-01-22 00:47:18.040605 -- Adding 3 pseudocounts to race_choice_white=Unchecked for 9402 genes in Vip_23 because all values are 0 for that covariate\n",
      "2025-01-22 00:47:30.439444 -- Skipping L5_IT_6 along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-22 00:47:31.386646 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-22 00:47:31.964867 -- Removing 3809 features from L5_IT_7 for low numbers of counts per cell\n",
      "2025-01-22 00:47:32.078533 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-22 00:47:32.079738 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-22 00:47:32.122572 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.25 for 897 genes in L5_IT_7 because all values are 0 for that covariate\n",
      "2025-01-22 00:57:45.946514 -- Adding 3 pseudocounts to age_at_death_binned_codes=1.0 for 387 genes in L5_IT_7 because all values are 0 for that covariate\n",
      "2025-01-22 01:02:10.369280 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.5 for 939 genes in L5_IT_7 because all values are 0 for that covariate\n",
      "2025-01-22 01:09:56.862692 -- Starting Vip_23 across continuous_pseudo_progression_score\n",
      "The cells are already grouped.The average number of cells per subject ( 5.48 ) is less than 30. The 'method' is set as 'HL'.\n",
      "Remove  0  genes having low expression.\n",
      "Analyzing  31838  genes with  31  subjects and  170  cells.\n",
      "2025-01-22 01:12:53.250675 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.75 for 193 genes in L5_IT_7 because all values are 0 for that covariate\n",
      "2025-01-22 01:13:21.143650 -- Vip_23 along continuous_pseudo_progression_score was written to disk\n",
      "2025-01-22 01:15:05.264212 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.0 for 2439 genes in L5_IT_7 because all values are 0 for that covariate\n",
      "2025-01-22 01:24:41.164837 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.0 for 1110 genes in Pvalb_9 because all values are 0 for that covariate\n",
      "2025-01-22 01:33:50.665014 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.4124328412522276 for 3146 genes in Pvalb_9 because all values are 0 for that covariate\n",
      "2025-01-22 01:42:49.538605 -- Adding 3 pseudocounts to sex=Female for 1 genes in L5_IT_7 because all values are 0 for that covariate\n",
      "2025-01-22 01:42:50.484900 -- Adding 3 pseudocounts to apoe4_status=Y for 97 genes in L5_IT_7 because all values are 0 for that covariate\n",
      "2025-01-22 01:43:56.994619 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-22 01:43:56.995846 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-22 01:43:57.197174 -- Adding 3 pseudocounts to method=10xMulti for 1539 genes in L5_IT_7 because all values are 0 for that covariate\n",
      "2025-01-22 01:57:15.446178 -- Skipping L4_IT_4 along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-22 01:59:51.229002 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-22 01:59:51.628631 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-22 01:59:51.629625 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-22 01:59:53.129093 -- Starting Pvalb_9 across continuous_pseudo_progression_score\n",
      "The cells are already grouped.Remove  0  genes having low expression.\n",
      "Analyzing  30981  genes with  5  subjects and  605  cells.\n",
      "2025-01-22 02:01:26.350092 -- Adding 3 pseudocounts to race_choice_white=Unchecked for 4286 genes in L5_IT_7 because all values are 0 for that covariate\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: In addition: \n",
      "R[write to console]: Warning message:\n",
      "\n",
      "R[write to console]: In nebula(counts, obs$donor_id, pred = df, offset = obs$pseudo_number_of_umis,  :\n",
      "R[write to console]: \n",
      " \n",
      "R[write to console]:  Some predictors in the design matrix are collinear or linearly dependent. The effects of these predictors are not identifiable.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-22 02:02:18.459190 -- Pvalb_9 along continuous_pseudo_progression_score was written to disk\n",
      "2025-01-22 02:02:19.352399 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-22 02:02:19.487096 -- Removing 5899 features from Pvalb_10 for low numbers of counts per cell\n",
      "2025-01-22 02:02:19.515304 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-22 02:02:19.542853 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.707882151477128 for 2460 genes in Pvalb_10 because all values are 0 for that covariate\n",
      "2025-01-22 02:11:04.181711 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.0 for 779 genes in Pvalb_10 because all values are 0 for that covariate\n",
      "2025-01-22 02:13:50.984269 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-22 02:13:51.049825 -- Removing sex from the covariate formula\n",
      "2025-01-22 02:13:51.049937 -- Removing apoe4_status from the covariate formula\n",
      "2025-01-22 02:13:51.051027 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-22 02:13:51.052519 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-22 02:13:51.109824 -- Removing race_choice_white from the covariate formula\n",
      "2025-01-22 02:13:51.551947 -- Starting Pvalb_10 across continuous_pseudo_progression_score\n",
      "The cells are already grouped.Remove  195  genes having low expression.\n",
      "Analyzing  30507  genes with  2  subjects and  245  cells.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: In addition: \n",
      "R[write to console]: Warning message:\n",
      "\n",
      "R[write to console]: In nebula(counts, obs$donor_id, pred = df, offset = obs$pseudo_number_of_umis,  :\n",
      "R[write to console]: \n",
      " \n",
      "R[write to console]:  Some predictors in the design matrix are collinear or linearly dependent. The effects of these predictors are not identifiable.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-22 02:15:38.011543 -- Pvalb_10 along continuous_pseudo_progression_score was written to disk\n",
      "2025-01-22 02:15:38.962814 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-22 02:15:39.583940 -- Removing 5376 features from Pvalb_12 for low numbers of counts per cell\n",
      "2025-01-22 02:15:39.706340 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-22 02:15:39.788640 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.0 for 1667 genes in Pvalb_12 because all values are 0 for that covariate\n",
      "2025-01-22 02:32:51.799059 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.4145860536076113 for 1313 genes in Pvalb_12 because all values are 0 for that covariate\n",
      "2025-01-22 02:46:22.137995 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.6666314547467305 for 3811 genes in Pvalb_12 because all values are 0 for that covariate\n",
      "2025-01-22 02:49:20.826863 -- Skipping L5_IT_7 along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-22 02:56:06.432352 -- Adding 3 pseudocounts to sex=Male for 1 genes in Oligo_4 because all values are 0 for that covariate\n",
      "2025-01-22 02:56:10.970162 -- Adding 3 pseudocounts to apoe4_status=Y for 1 genes in Oligo_4 because all values are 0 for that covariate\n",
      "2025-01-22 02:56:14.501207 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-22 02:56:14.502558 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-22 02:56:15.456708 -- Adding 3 pseudocounts to method=10xMulti for 393 genes in Oligo_4 because all values are 0 for that covariate\n",
      "2025-01-22 03:19:13.858087 -- Adding 3 pseudocounts to race_choice_white=Unchecked for 9017 genes in Oligo_4 because all values are 0 for that covariate\n",
      "2025-01-22 03:25:17.147588 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.07279917361352964 for 2288 genes in Pvalb_12 because all values are 0 for that covariate\n",
      "2025-01-22 03:48:35.976226 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-22 03:48:36.176718 -- Removing sex from the covariate formula\n",
      "2025-01-22 03:48:36.317889 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-22 03:48:36.319042 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-22 03:48:36.319464 -- Removing method from the covariate formula\n",
      "2025-01-22 03:48:36.319576 -- Removing race_choice_white from the covariate formula\n",
      "2025-01-22 03:48:38.811403 -- Starting Pvalb_12 across continuous_pseudo_progression_score\n",
      "The cells are already grouped.Remove  0  genes having low expression.\n",
      "Analyzing  31225  genes with  4  subjects and  767  cells.\n",
      "2025-01-22 03:51:01.718570 -- Pvalb_12 along continuous_pseudo_progression_score was written to disk\n",
      "2025-01-22 03:51:03.037997 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-22 03:51:03.360062 -- Removing 5485 features from Pvalb_14 for low numbers of counts per cell\n",
      "2025-01-22 03:51:03.403819 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-22 03:51:03.432492 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.5750832273486403 for 2736 genes in Pvalb_14 because all values are 0 for that covariate\n",
      "2025-01-22 04:04:37.687388 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.6980818916251965 for 3367 genes in Pvalb_14 because all values are 0 for that covariate\n",
      "2025-01-22 04:21:20.593411 -- Adding 3 pseudocounts to continuous_pseudo_progression_score=0.0 for 1102 genes in Pvalb_14 because all values are 0 for that covariate\n",
      "2025-01-22 04:26:50.901645 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-22 04:26:51.176321 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-22 04:26:51.177413 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-22 04:26:51.178017 -- Removing method from the covariate formula\n",
      "2025-01-22 04:26:51.178284 -- Removing race_choice_white from the covariate formula\n",
      "2025-01-22 04:26:51.940693 -- Starting Pvalb_14 across continuous_pseudo_progression_score\n",
      "The cells are already grouped.Remove  5  genes having low expression.\n",
      "Analyzing  31111  genes with  3  subjects and  357  cells.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: In addition: \n",
      "R[write to console]: Warning message:\n",
      "\n",
      "R[write to console]: In nebula(counts, obs$donor_id, pred = df, offset = obs$pseudo_number_of_umis,  :\n",
      "R[write to console]: \n",
      " \n",
      "R[write to console]:  Some predictors in the design matrix are collinear or linearly dependent. The effects of these predictors are not identifiable.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-22 04:28:47.188093 -- Pvalb_14 along continuous_pseudo_progression_score was written to disk\n",
      "2025-01-22 04:28:48.020374 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-22 04:28:50.500477 -- Removing 4636 features from Pvalb_15 for low numbers of counts per cell\n",
      "2025-01-22 04:28:50.981817 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-22 04:28:50.983095 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-22 04:28:51.334298 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.75 for 75 genes in Pvalb_15 because all values are 0 for that covariate\n",
      "2025-01-22 04:32:19.825550 -- Adding 3 pseudocounts to age_at_death_binned_codes=1.0 for 330 genes in Pvalb_15 because all values are 0 for that covariate\n",
      "2025-01-22 04:47:43.233803 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.25 for 734 genes in Pvalb_15 because all values are 0 for that covariate\n",
      "2025-01-22 04:49:57.172146 -- Skipping L2_3_IT_6 along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-22 04:49:58.102094 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-22 04:49:58.269378 -- Removing 4185 features from L2_3_IT_7 for low numbers of counts per cell\n",
      "2025-01-22 04:49:58.291342 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-22 04:49:58.292568 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-22 04:49:58.317673 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.25 for 3764 genes in L2_3_IT_7 because all values are 0 for that covariate\n",
      "2025-01-22 04:59:44.874296 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.5 for 2827 genes in L2_3_IT_7 because all values are 0 for that covariate\n",
      "2025-01-22 05:07:09.243170 -- Adding 3 pseudocounts to age_at_death_binned_codes=1.0 for 3732 genes in L2_3_IT_7 because all values are 0 for that covariate\n",
      "2025-01-22 05:16:56.549744 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.75 for 1411 genes in L2_3_IT_7 because all values are 0 for that covariate\n",
      "2025-01-22 05:20:38.223548 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.0 for 5064 genes in L2_3_IT_7 because all values are 0 for that covariate\n",
      "2025-01-22 05:21:51.179938 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.0 for 1529 genes in Pvalb_15 because all values are 0 for that covariate\n",
      "2025-01-22 05:34:06.851800 -- Adding 3 pseudocounts to sex=Female for 7 genes in L2_3_IT_7 because all values are 0 for that covariate\n",
      "2025-01-22 05:34:07.996314 -- Adding 3 pseudocounts to sex=Male for 3 genes in L2_3_IT_7 because all values are 0 for that covariate\n",
      "2025-01-22 05:34:08.603595 -- Adding 3 pseudocounts to apoe4_status=Y for 12 genes in L2_3_IT_7 because all values are 0 for that covariate\n",
      "2025-01-22 05:34:10.572983 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-22 05:34:10.574052 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-22 05:34:10.624194 -- Adding 3 pseudocounts to method=10xMulti for 2732 genes in L2_3_IT_7 because all values are 0 for that covariate\n",
      "2025-01-22 05:41:24.627849 -- Removing race_choice_white from the covariate formula\n",
      "2025-01-22 05:41:24.937769 -- Skipping L2_3_IT_7 along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-22 05:41:25.821244 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-22 05:41:26.053885 -- Removing 4502 features from L2_3_IT_8 for low numbers of counts per cell\n",
      "2025-01-22 05:41:26.099598 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-22 05:41:26.100719 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-22 05:41:26.123242 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.25 for 2478 genes in L2_3_IT_8 because all values are 0 for that covariate\n",
      "2025-01-22 05:52:19.663901 -- Adding 3 pseudocounts to age_at_death_binned_codes=1.0 for 1280 genes in L2_3_IT_8 because all values are 0 for that covariate\n",
      "2025-01-22 05:57:56.122754 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.5 for 1417 genes in L2_3_IT_8 because all values are 0 for that covariate\n",
      "2025-01-22 06:04:08.811717 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.75 for 899 genes in L2_3_IT_8 because all values are 0 for that covariate\n",
      "2025-01-22 06:08:07.313043 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.0 for 6092 genes in L2_3_IT_8 because all values are 0 for that covariate\n",
      "2025-01-22 06:32:46.795371 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.5 for 901 genes in Pvalb_15 because all values are 0 for that covariate\n",
      "2025-01-22 06:34:37.435909 -- Adding 3 pseudocounts to sex=Female for 37 genes in L2_3_IT_8 because all values are 0 for that covariate\n",
      "2025-01-22 06:34:47.331294 -- Adding 3 pseudocounts to apoe4_status=Y for 255 genes in L2_3_IT_8 because all values are 0 for that covariate\n",
      "2025-01-22 06:35:54.623763 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-22 06:35:54.624830 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-22 06:35:54.706023 -- Adding 3 pseudocounts to method=10xMulti for 5258 genes in L2_3_IT_8 because all values are 0 for that covariate\n",
      "2025-01-22 06:58:53.534912 -- Adding 3 pseudocounts to race_choice_white=Unchecked for 6709 genes in L2_3_IT_8 because all values are 0 for that covariate\n",
      "2025-01-22 07:14:40.579326 -- Adding 3 pseudocounts to sex=Male for 8 genes in Pvalb_15 because all values are 0 for that covariate\n",
      "2025-01-22 07:15:03.260637 -- Adding 3 pseudocounts to apoe4_status=Y for 27 genes in Pvalb_15 because all values are 0 for that covariate\n",
      "2025-01-22 07:16:20.316419 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-22 07:16:20.317908 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-22 07:16:20.338065 -- Adding 3 pseudocounts to method=10xMulti for 5527 genes in Pvalb_15 because all values are 0 for that covariate\n",
      "2025-01-22 07:28:10.085702 -- Skipping L2_3_IT_8 along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-22 07:28:11.002091 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-22 07:28:12.415044 -- Removing 3977 features from L2_3_IT_10 for low numbers of counts per cell\n",
      "2025-01-22 07:28:12.723360 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-22 07:28:12.724989 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-22 07:28:12.774698 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.25 for 978 genes in L2_3_IT_10 because all values are 0 for that covariate\n",
      "2025-01-22 07:55:02.784252 -- Adding 3 pseudocounts to age_at_death_binned_codes=1.0 for 110 genes in L2_3_IT_10 because all values are 0 for that covariate\n",
      "2025-01-22 07:58:04.657210 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.5 for 66 genes in L2_3_IT_10 because all values are 0 for that covariate\n",
      "2025-01-22 07:59:53.743984 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.75 for 42 genes in L2_3_IT_10 because all values are 0 for that covariate\n",
      "2025-01-22 08:01:03.109658 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.0 for 1742 genes in L2_3_IT_10 because all values are 0 for that covariate\n",
      "2025-01-22 08:48:52.449511 -- Adding 3 pseudocounts to sex=Female for 2 genes in L2_3_IT_10 because all values are 0 for that covariate\n",
      "2025-01-22 08:48:56.609472 -- Adding 3 pseudocounts to apoe4_status=Y for 65 genes in L2_3_IT_10 because all values are 0 for that covariate\n",
      "2025-01-22 08:50:44.093828 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-22 08:50:44.095138 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-22 08:50:44.572408 -- Adding 3 pseudocounts to method=10xMulti for 2116 genes in L2_3_IT_10 because all values are 0 for that covariate\n",
      "2025-01-22 09:48:44.013112 -- Adding 3 pseudocounts to race_choice_white=Unchecked for 3799 genes in L2_3_IT_10 because all values are 0 for that covariate\n",
      "2025-01-22 11:32:33.579970 -- Skipping L2_3_IT_10 along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-22 11:32:34.502683 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-22 11:32:34.943989 -- Removing 4080 features from L2_3_IT_12 for low numbers of counts per cell\n",
      "2025-01-22 11:32:35.024733 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-22 11:32:35.026160 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-22 11:32:35.050617 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.25 for 1525 genes in L2_3_IT_12 because all values are 0 for that covariate\n",
      "2025-01-22 11:33:22.694103 -- Removing race_choice_white from the covariate formula\n",
      "2025-01-22 11:33:32.430021 -- Starting Pvalb_15 across continuous_pseudo_progression_score\n",
      "The cells are already grouped.Remove  0  genes having low expression.\n",
      "Analyzing  31965  genes with  27  subjects and  3259  cells.\n",
      "2025-01-22 11:37:48.172356 -- Pvalb_15 along continuous_pseudo_progression_score was written to disk\n",
      "2025-01-22 11:44:50.524580 -- Adding 3 pseudocounts to age_at_death_binned_codes=1.0 for 946 genes in L2_3_IT_12 because all values are 0 for that covariate\n",
      "2025-01-22 11:52:20.350998 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.5 for 1101 genes in L2_3_IT_12 because all values are 0 for that covariate\n",
      "2025-01-22 12:01:04.785947 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.75 for 201 genes in L2_3_IT_12 because all values are 0 for that covariate\n",
      "2025-01-22 12:02:40.139762 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.0 for 4620 genes in L2_3_IT_12 because all values are 0 for that covariate\n",
      "2025-01-22 12:06:28.945331 -- Skipping Oligo_4 along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-22 12:38:44.984335 -- Adding 3 pseudocounts to sex=Female for 4 genes in L2_3_IT_12 because all values are 0 for that covariate\n",
      "2025-01-22 12:38:47.063209 -- Adding 3 pseudocounts to apoe4_status=Y for 70 genes in L2_3_IT_12 because all values are 0 for that covariate\n",
      "2025-01-22 12:39:19.945660 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-22 12:39:19.946928 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-22 12:39:20.078496 -- Adding 3 pseudocounts to method=10xMulti for 2628 genes in L2_3_IT_12 because all values are 0 for that covariate\n",
      "2025-01-22 12:59:55.605558 -- Adding 3 pseudocounts to race_choice_white=Unchecked for 8083 genes in L2_3_IT_12 because all values are 0 for that covariate\n",
      "2025-01-22 14:03:10.570568 -- Skipping L2_3_IT_12 along continuous_pseudo_progression_score (already exists)\n",
      "2025-01-22 14:03:12.657128 -- No outgroup specified, running analysis for all groups and tests\n",
      "2025-01-22 14:03:14.614152 -- Removing 4118 features from L2_3_IT_13 for low numbers of counts per cell\n",
      "2025-01-22 14:03:15.015424 -- Detected continuous_pseudo_progression_score as an integer or float, applying a min-max normalization\n",
      "2025-01-22 14:03:15.016751 -- Detected age_at_death_binned_codes as an integer or float, applying a min-max normalization\n",
      "2025-01-22 14:03:15.108949 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.25 for 265 genes in L2_3_IT_13 because all values are 0 for that covariate\n",
      "2025-01-22 14:13:04.403709 -- Adding 3 pseudocounts to age_at_death_binned_codes=1.0 for 100 genes in L2_3_IT_13 because all values are 0 for that covariate\n",
      "2025-01-22 14:16:43.813047 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.5 for 52 genes in L2_3_IT_13 because all values are 0 for that covariate\n",
      "2025-01-22 14:18:37.836929 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.75 for 8 genes in L2_3_IT_13 because all values are 0 for that covariate\n",
      "2025-01-22 14:18:55.640170 -- Adding 3 pseudocounts to age_at_death_binned_codes=0.0 for 1476 genes in L2_3_IT_13 because all values are 0 for that covariate\n",
      "2025-01-22 15:13:10.138993 -- Adding 3 pseudocounts to apoe4_status=Y for 7 genes in L2_3_IT_13 because all values are 0 for that covariate\n",
      "2025-01-22 15:13:25.619573 -- Detected pseudo_genes_detected as an integer or float, applying a min-max normalization\n",
      "2025-01-22 15:13:25.621297 -- Detected pseudo_number_of_umis as an integer or float, applying a min-max normalization\n",
      "2025-01-22 15:13:26.225798 -- Adding 3 pseudocounts to method=10xMulti for 1133 genes in L2_3_IT_13 because all values are 0 for that covariate\n",
      "2025-01-22 15:55:03.813523 -- Adding 3 pseudocounts to race_choice_white=Unchecked for 2328 genes in L2_3_IT_13 because all values are 0 for that covariate\n",
      "2025-01-22 17:20:27.664842 -- Skipping L2_3_IT_13 along continuous_pseudo_progression_score (already exists)\n"
     ]
    }
   ],
   "source": [
    "\n",
    "if run_differential_expression:\n",
    "    # Set up parameters\n",
    "    params = {\n",
    "        'region': 'MTG',\n",
    "        'target': '',\n",
    "        'split_key': pseudobulk_level,\n",
    "        'outgroup_de': False, \n",
    "        'outgroup': '', \n",
    "        'covariates': covariates,\n",
    "        'random_effect': subject_id,\n",
    "        'covariate_formula': ' + '.join(covariates[1:]) + \" + \",\n",
    "        'tests': [factor],\n",
    "        'layer': 'X',  \n",
    "        'offset_variable': helper_functions.clean_strings('Pseudo Number of UMIs'),\n",
    "        'group_cell': False,\n",
    "        'save_dir': data_dir+'results/differential_gene_expression'  \n",
    "    }\n",
    "\n",
    "    adata_merged.obs[params['split_key']] = adata_merged.obs[params['split_key']].astype('category')\n",
    "\n",
    "    os.makedirs(params['save_dir'], exist_ok=True)\n",
    "\n",
    "    def process_target(target, adata_merged, params, cell_type_column):\n",
    "        adata = adata_merged[adata_merged.obs[cell_type_column] == target]\n",
    "        results = []\n",
    "        for b in adata.obs[params['split_key']].cat.categories:\n",
    "            try:\n",
    "                result = run_de_tests(\n",
    "                    adata_merged,\n",
    "                    params['region'],\n",
    "                    helper_functions.clean_strings(target,preserve_case=True,replace_hyphen=False),\n",
    "                    params['split_key'],\n",
    "                    params['outgroup_de'],\n",
    "                    b,\n",
    "                    params['covariates'],\n",
    "                    params['random_effect'],\n",
    "                    params['covariate_formula'],\n",
    "                    params['tests'],\n",
    "                    params['layer'],\n",
    "                    params['offset_variable'],\n",
    "                    params['group_cell'],\n",
    "                    params['save_dir']\n",
    "                )\n",
    "                results.append(result)\n",
    "            except Exception as e:\n",
    "                error_msg = str(e)\n",
    "                if \"Some predictors have zero variation or a zero vector\" in error_msg:\n",
    "                    print(f\"{datetime.now()} -- Skipping {b} because some predictors have zero variation or a zero vector\")\n",
    "                else:\n",
    "                    print(f\"{datetime.now()} -- Error processing {b}: {str(e)}\") \n",
    "                continue\n",
    "                \n",
    "        return results\n",
    "\n",
    "    # Parallelize at target level\n",
    "    targets = adata_merged.obs[cell_type_column].unique()\n",
    "    results = Parallel(n_jobs=n_cores)(\n",
    "        delayed(process_target)(target, adata_merged, params, cell_type_column) \n",
    "        for target in targets\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "id": "0f97a6a3",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'build_effect_size_anndata' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[173], line 5\u001b[0m\n\u001b[1;32m      1\u001b[0m blacklisted_genes \u001b[38;5;241m=\u001b[39m [\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mMTRNR2L12\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mTTTY14\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mUSP9Y\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mNLGN4Y\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mUTY\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mDDX3Y\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mKDM5D\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mTTTY10\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n\u001b[1;32m      2\u001b[0m blacklisted_genes\u001b[38;5;241m.\u001b[39mextend(adata_merged\u001b[38;5;241m.\u001b[39mvar_names[adata_merged\u001b[38;5;241m.\u001b[39mvar_names\u001b[38;5;241m.\u001b[39mstr\u001b[38;5;241m.\u001b[39mstartswith(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mMT-\u001b[39m\u001b[38;5;124m\"\u001b[39m)]\u001b[38;5;241m.\u001b[39mto_list())\n\u001b[0;32m----> 5\u001b[0m effect_sizes, pvalues, std_errors  \u001b[38;5;241m=\u001b[39m \u001b[43mbuild_effect_size_anndata\u001b[49m(\n\u001b[1;32m      6\u001b[0m     results_dir\u001b[38;5;241m=\u001b[39mos\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mjoin(data_dir\u001b[38;5;241m+\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mresults/differential_gene_expression\u001b[39m\u001b[38;5;124m'\u001b[39m,),\n\u001b[1;32m      7\u001b[0m     glob_pattern\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m*.csv\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m      8\u001b[0m     file_pattern\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m_across_continuous_pseudo_progression_score_DE.csv\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m      9\u001b[0m     test\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcontinuous_pseudo_progression_score\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m     10\u001b[0m     adata\u001b[38;5;241m=\u001b[39madata_merged,\n\u001b[1;32m     11\u001b[0m     subclass\u001b[38;5;241m=\u001b[39mcell_type_column,\n\u001b[1;32m     12\u001b[0m     celltype\u001b[38;5;241m=\u001b[39mcell_type_column,\n\u001b[1;32m     13\u001b[0m     blacklisted_genes\u001b[38;5;241m=\u001b[39mblacklisted_genes,\n\u001b[1;32m     14\u001b[0m )\n",
      "\u001b[0;31mNameError\u001b[0m: name 'build_effect_size_anndata' is not defined"
     ]
    }
   ],
   "source": [
    "blacklisted_genes = [\"MTRNR2L12\", \"TTTY14\", \"USP9Y\", \"NLGN4Y\", \"UTY\", \"DDX3Y\", \"KDM5D\", \"TTTY10\"]\n",
    "blacklisted_genes.extend(adata_merged.var_names[adata_merged.var_names.str.startswith(\"MT-\")].to_list())\n",
    "\n",
    "\n",
    "effect_sizes, pvalues, std_errors  = build_effect_size_anndata(\n",
    "    results_dir=os.path.join(data_dir+'results/differential_gene_expression',),\n",
    "    glob_pattern=\"*.csv\",\n",
    "    file_pattern=\"_across_continuous_pseudo_progression_score_DE.csv\",\n",
    "    test=\"continuous_pseudo_progression_score\",\n",
    "    adata=adata_merged,\n",
    "    subclass=cell_type_column,\n",
    "    celltype=cell_type_column,\n",
    "    blacklisted_genes=blacklisted_genes,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "id": "0935965f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Taxonomy Level</th>\n",
       "      <th>Population</th>\n",
       "      <th>Effect size across all of pseudoprogression</th>\n",
       "      <th>Effect size across early pseudoprogression</th>\n",
       "      <th>Effect size across late pseudoprogression</th>\n",
       "      <th>Mean expression (natural log UMIs per 10k plus 1)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>Supertype</td>\n",
       "      <td>Lamp5_Lhx6_1</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>Supertype</td>\n",
       "      <td>Lamp5_Lhx6_1</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>Supertype</td>\n",
       "      <td>Lamp5_Lhx6_1</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>AL627309.1</td>\n",
       "      <td>Supertype</td>\n",
       "      <td>Lamp5_Lhx6_1</td>\n",
       "      <td>-2.17</td>\n",
       "      <td>-0.79</td>\n",
       "      <td>0.28</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>AL627309.3</td>\n",
       "      <td>Supertype</td>\n",
       "      <td>Lamp5_Lhx6_1</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6072275</th>\n",
       "      <td>6072275</td>\n",
       "      <td>AC141272.1</td>\n",
       "      <td>Class</td>\n",
       "      <td>Non-neuronal and non-neural</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6072276</th>\n",
       "      <td>6072276</td>\n",
       "      <td>AC023491.2</td>\n",
       "      <td>Class</td>\n",
       "      <td>Non-neuronal and non-neural</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6072277</th>\n",
       "      <td>6072277</td>\n",
       "      <td>AC007325.1</td>\n",
       "      <td>Class</td>\n",
       "      <td>Non-neuronal and non-neural</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6072278</th>\n",
       "      <td>6072278</td>\n",
       "      <td>AC007325.4</td>\n",
       "      <td>Class</td>\n",
       "      <td>Non-neuronal and non-neural</td>\n",
       "      <td>0.47</td>\n",
       "      <td>-0.04</td>\n",
       "      <td>0.29</td>\n",
       "      <td>0.02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6072279</th>\n",
       "      <td>6072279</td>\n",
       "      <td>AC007325.2</td>\n",
       "      <td>Class</td>\n",
       "      <td>Non-neuronal and non-neural</td>\n",
       "      <td>0.42</td>\n",
       "      <td>-0.35</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.03</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6072280 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Unnamed: 0         Gene Taxonomy Level                   Population  \\\n",
       "0                 0  MIR1302-2HG      Supertype                 Lamp5_Lhx6_1   \n",
       "1                 1      FAM138A      Supertype                 Lamp5_Lhx6_1   \n",
       "2                 2        OR4F5      Supertype                 Lamp5_Lhx6_1   \n",
       "3                 3   AL627309.1      Supertype                 Lamp5_Lhx6_1   \n",
       "4                 4   AL627309.3      Supertype                 Lamp5_Lhx6_1   \n",
       "...             ...          ...            ...                          ...   \n",
       "6072275     6072275   AC141272.1          Class  Non-neuronal and non-neural   \n",
       "6072276     6072276   AC023491.2          Class  Non-neuronal and non-neural   \n",
       "6072277     6072277   AC007325.1          Class  Non-neuronal and non-neural   \n",
       "6072278     6072278   AC007325.4          Class  Non-neuronal and non-neural   \n",
       "6072279     6072279   AC007325.2          Class  Non-neuronal and non-neural   \n",
       "\n",
       "         Effect size across all of pseudoprogression  \\\n",
       "0                                               0.00   \n",
       "1                                               0.00   \n",
       "2                                               0.00   \n",
       "3                                              -2.17   \n",
       "4                                               0.00   \n",
       "...                                              ...   \n",
       "6072275                                         0.00   \n",
       "6072276                                         0.00   \n",
       "6072277                                         0.03   \n",
       "6072278                                         0.47   \n",
       "6072279                                         0.42   \n",
       "\n",
       "         Effect size across early pseudoprogression  \\\n",
       "0                                              0.00   \n",
       "1                                              0.00   \n",
       "2                                              0.00   \n",
       "3                                             -0.79   \n",
       "4                                              0.00   \n",
       "...                                             ...   \n",
       "6072275                                        0.00   \n",
       "6072276                                        0.00   \n",
       "6072277                                        0.05   \n",
       "6072278                                       -0.04   \n",
       "6072279                                       -0.35   \n",
       "\n",
       "         Effect size across late pseudoprogression  \\\n",
       "0                                             0.00   \n",
       "1                                             0.00   \n",
       "2                                             0.00   \n",
       "3                                             0.28   \n",
       "4                                             0.00   \n",
       "...                                            ...   \n",
       "6072275                                       0.00   \n",
       "6072276                                       0.00   \n",
       "6072277                                       0.04   \n",
       "6072278                                       0.29   \n",
       "6072279                                       0.11   \n",
       "\n",
       "         Mean expression (natural log UMIs per 10k plus 1)  \n",
       "0                                                     0.00  \n",
       "1                                                     0.00  \n",
       "2                                                     0.00  \n",
       "3                                                     0.01  \n",
       "4                                                     0.00  \n",
       "...                                                    ...  \n",
       "6072275                                               0.00  \n",
       "6072276                                               0.00  \n",
       "6072277                                               0.00  \n",
       "6072278                                               0.02  \n",
       "6072279                                               0.03  \n",
       "\n",
       "[6072280 rows x 8 columns]"
      ]
     },
     "execution_count": 174,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "effect_size_table = pd.read_csv('/media/tadeoye/Volume1/data/seq/SEA-AD/MTG/supplementary/Nebula Results/effect_size_table.csv')\n",
    "effect_size_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c9cfa09",
   "metadata": {},
   "outputs": [],
   "source": [
    "results_dir = os.path.join(data_dir+'results/differential_gene_expression',\n",
    "                             subclass.capitalize(),\n",
    "                             factor, \n",
    "                             f'{subclass.capitalize()}_{subclass.capitalize()}_across_{factor}_DE.csv')\n",
    "\n",
    "effect_sizes = get_standardized_effects(results_dir, factor, filter_genes=False, thr = 0.01)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scRNA_seq_meta_analysis",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
