{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "62042f97-a76e-4b34-805f-3f1ac81e7288",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import rpy2\n",
    "import logging\n",
    "import warnings\n",
    "import anndata2ri\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import decoupler as dc\n",
    "import matplotlib.pyplot as plt\n",
    "from gsva_prep import prep_gsva\n",
    "import rpy2.robjects as robjects\n",
    "from rpy2.robjects import pandas2ri\n",
    "from matplotlib.pyplot import rcParams\n",
    "from functions import pathway_analyses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "447b6d1c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-----\n",
      "anndata     0.8.0\n",
      "scanpy      1.9.3\n",
      "-----\n",
      "PIL                         9.5.0\n",
      "anndata2ri                  1.1\n",
      "appnope                     0.1.3\n",
      "asttokens                   NA\n",
      "backcall                    0.2.0\n",
      "cffi                        1.15.1\n",
      "comm                        0.1.3\n",
      "cycler                      0.10.0\n",
      "cython_runtime              NA\n",
      "dateutil                    2.8.2\n",
      "debugpy                     1.6.7\n",
      "decorator                   5.1.1\n",
      "decoupler                   1.4.0\n",
      "executing                   1.2.0\n",
      "functions                   NA\n",
      "google                      NA\n",
      "gsva_prep                   NA\n",
      "h5py                        3.9.0\n",
      "igraph                      0.10.4\n",
      "ipykernel                   6.23.2\n",
      "ipywidgets                  8.0.6\n",
      "jedi                        0.18.2\n",
      "jinja2                      3.1.2\n",
      "joblib                      1.2.0\n",
      "kiwisolver                  1.4.4\n",
      "leidenalg                   0.9.1\n",
      "llvmlite                    0.39.1\n",
      "louvain                     0.8.0\n",
      "markupsafe                  2.1.3\n",
      "matplotlib                  3.7.1\n",
      "mpl_toolkits                NA\n",
      "natsort                     8.4.0\n",
      "numba                       0.56.4\n",
      "numexpr                     2.8.4\n",
      "numpy                       1.23.5\n",
      "packaging                   23.1\n",
      "pandas                      2.0.2\n",
      "parso                       0.8.3\n",
      "patsy                       0.5.3\n",
      "pexpect                     4.8.0\n",
      "pickleshare                 0.7.5\n",
      "pkg_resources               NA\n",
      "platformdirs                3.7.0\n",
      "prompt_toolkit              3.0.38\n",
      "psutil                      5.9.5\n",
      "ptyprocess                  0.7.0\n",
      "pure_eval                   0.2.2\n",
      "pyarrow                     12.0.1\n",
      "pycparser                   2.21\n",
      "pydev_ipython               NA\n",
      "pydevconsole                NA\n",
      "pydevd                      2.9.5\n",
      "pydevd_file_utils           NA\n",
      "pydevd_plugins              NA\n",
      "pydevd_tracing              NA\n",
      "pygments                    2.15.1\n",
      "pyparsing                   3.1.0\n",
      "pytz                        2023.3\n",
      "pytz_deprecation_shim       NA\n",
      "rpy2                        3.5.12\n",
      "scipy                       1.10.1\n",
      "seaborn                     0.12.2\n",
      "session_info                1.0.0\n",
      "setuptools                  68.0.0\n",
      "six                         1.16.0\n",
      "sklearn                     1.2.2\n",
      "sparse                      0.14.0\n",
      "stack_data                  0.6.2\n",
      "statsmodels                 0.14.0\n",
      "texttable                   1.6.7\n",
      "threadpoolctl               3.1.0\n",
      "tornado                     6.3.2\n",
      "tqdm                        4.65.0\n",
      "traitlets                   5.9.0\n",
      "typing_extensions           NA\n",
      "tzlocal                     NA\n",
      "vscode                      NA\n",
      "wcwidth                     0.2.6\n",
      "yaml                        6.0\n",
      "zmq                         25.1.0\n",
      "zoneinfo                    NA\n",
      "-----\n",
      "IPython             8.14.0\n",
      "jupyter_client      8.2.0\n",
      "jupyter_core        5.3.1\n",
      "-----\n",
      "Python 3.10.6 (main, Oct 24 2022, 11:04:34) [Clang 12.0.0 ]\n",
      "macOS-10.16-x86_64-i386-64bit\n",
      "-----\n",
      "Session information updated at 2024-04-17 14:42\n"
     ]
    }
   ],
   "source": [
    "# # Ignore R warning messages\n",
    "#Note: this can be commented out to get more verbose R output\n",
    "rpy2.rinterface_lib.callbacks.logger.setLevel(logging.ERROR)\n",
    "\n",
    "warnings.filterwarnings(\"ignore\", category=PendingDeprecationWarning)\n",
    "warnings.filterwarnings(\"ignore\", category=UserWarning)\n",
    "warnings.filterwarnings(\"ignore\", category=FutureWarning)\n",
    "\n",
    "# Automatically convert rpy2 outputs to pandas dataframes\n",
    "pandas2ri.activate()\n",
    "anndata2ri.activate()\n",
    "%load_ext rpy2.ipython\n",
    "\n",
    "sc.settings.verbosity = 3\n",
    "#sc.set_figure_params(dpi=200, dpi_save=300)\n",
    "sc.logging.print_versions()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "048c2e09",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "    WARNING: The R package \"reticulate\" only fixed recently\n",
      "    an issue that caused a segfault when used with rpy2:\n",
      "    https://github.com/rstudio/reticulate/pull/1188\n",
      "    Make sure that you use a version of that package that includes\n",
      "    the fix.\n",
      "    "
     ]
    },
    {
     "data": {
      "text/plain": [
       "In addition: Warning messages:\n",
       "1: package ‘MatrixGenerics’ was built under R version 4.3.1 \n",
       "2: package ‘S4Vectors’ was built under R version 4.3.1 \n",
       "3: package ‘IRanges’ was built under R version 4.3.1 \n",
       "4: package ‘GenomeInfoDb’ was built under R version 4.3.1 \n",
       "5: package ‘AnnotationDbi’ was built under R version 4.3.1 \n",
       "6: package ‘GSVA’ was built under R version 4.3.1 \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "suppressPackageStartupMessages({\n",
    "    library(Matrix)\n",
    "    library(viridis)\n",
    "    library(harmony)\n",
    "    library(ggpubr)\n",
    "    library(tictoc)\n",
    "    library(RColorBrewer)\n",
    "    library(Hmisc)\n",
    "    library(corrplot)\n",
    "    library(grid)\n",
    "    library(gridExtra)\n",
    "    library(igraph)\n",
    "    library(ggrepel)\n",
    "    library(readxl)\n",
    "    library(conflicted)\n",
    "    library(dplyr)\n",
    "    library(parallel)\n",
    "    library(stringr)\n",
    "\n",
    "\n",
    "\n",
    "    # single-cell analysis package\n",
    "    library(Seurat)\n",
    "    library(SingleCellExperiment)\n",
    "    library(tidyr)\n",
    "    library(readxl)\n",
    "    library(GSA)\n",
    "    library(limma)\n",
    "\n",
    "    # plotting and data science packages\n",
    "    library(tidyverse)\n",
    "    library(cowplot)\n",
    "    library(patchwork)\n",
    "    library(ggplot2)\n",
    "\n",
    "    # co-expression network analysis packages:\n",
    "    library(WGCNA)\n",
    "    library(hdWGCNA)\n",
    "\n",
    "    # gene enrichment packages\n",
    "    library(enrichR)\n",
    "    library(GeneOverlap)\n",
    "    library(GSEABase)\n",
    "    library(GSVA) \n",
    "\n",
    "# needs to be run every time you start R and want to use %>%\n",
    "})\n",
    "\n",
    "# using the cowplot theme for ggplot\n",
    "theme_set(theme_cowplot())\n",
    "\n",
    "# set random seed for reproducibility\n",
    "set.seed(12345)\n",
    "\n",
    "# optionally enable multithreading\n",
    "# enableWGCNAThreads(nThreads = 4)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f21b7795",
   "metadata": {},
   "source": [
    "# **Systematic differential analysis of pathway activity**\n",
    "\n",
    "Pathway activity scores were computed in accordance with the overall protocol outlined in [**Joel W. Blanchard et. al.**](https://doi.org/10.1038/s41586-022-05439-w).\n",
    "\n",
    "- Aggregate gene expression profiles into `pseudo-cells` or `pseudo-bulks` depending to the desired strategy in **[Gazestani. et. al. 2023](https://www.sciencedirect.com/science/article/pii/S0092867423008590?via%3Dihub)**\n",
    "\n",
    "- Use GSVA to obtain pathway activity scores for pathways obtained from [**Gene Ontology biological processes (2021) edition**](https://maayanlab.cloud/Enrichr/#libraries).\n",
    "\n",
    "- Perform differential pathway activity analysis using limma "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a070e8e",
   "metadata": {},
   "source": [
    "\n",
    "### **Data Prep Parameters**\n",
    "\n",
    "\n",
    "- `metadata`: Path to metadata. It must include a `pathology.group` column with unique groups being `no`, `early`, and `late`.\n",
    "\n",
    "- `map_meta`: Indicates if metadata mapping is required for `pathology.group`. If set to False, `pathology.group` must exist in `adata.obs`\n",
    "\n",
    "- `test_names`: List of the different test names of interest.\n",
    "\n",
    "- `save_prefix`: Preferred prefix for saving critical files. Ideally chosen to be in the format `{source name}_{brain region}`. e.g `mathys_pfc`\n",
    "\n",
    "- `subject_id`: Column name for Subject/Patient ID in both metadata and `.obs`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "53693e9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "save_prefix = 'gazestani_pfc' # this takes the format '{StudyName}_{ThreeLetterAccronymForBrainRegion}'\n",
    "cell_group = 'endothelial'\n",
    "cell_type_column = 'Subclass' # 'Supertype (non-expanded)', 'Subclass'\n",
    "control_group = 'no'     # name of the control group in metadata \n",
    "factor = 'pathology.group'  # pathology.group # Continuous Pseudoprogression Score\n",
    "factor = factor.replace(\" \", \"\").replace(\"-\", \"\")\n",
    "\n",
    "subclass = {'seaad_mtg': {'excitatory': ['L5 IT', 'L2/3 IT', 'L4 IT', 'L6 IT', 'L6 IT Car3', 'L5/6 NP', 'L6b', 'L6 CT', 'L5 ET'],\n",
    "                          'inhibitory': ['Pvalb', 'Sst', 'Lamp5 Lhx6', 'Vip', 'Lamp5', 'Sncg', 'Chandelier', 'Sst Chodl', 'Pax6'],\n",
    "                          'astrocyte': ['Astrocyte'],\n",
    "                          'microglia': ['Microglia-PVM'],\n",
    "                          'opc': ['OPC'],\n",
    "                          'oligodendrocyte': ['Oligodendrocyte'],\n",
    "                          'endothelial': ['Endothelial'],\n",
    "                           'vlmc': ['VLMC'],\n",
    "                         },\n",
    "            \n",
    "            'gazestani_pfc': {'excitatory': ['FEZF2', 'RORB', 'THEMIS', 'LINC00507', 'CTGF'],\n",
    "                              'inhibitory': ['PVALB', 'SST', 'LHX6', 'LAMP5', 'VIP', 'NDNF'],\n",
    "                              'astrocyte': ['WIF1', 'CHI3L1', 'PTCSC3', 'GRIA1'],\n",
    "                              'microglia': ['Myeloid', 'CX3CR1', 'GPNMB', 'Prolif', 'IRM', 'Macrophage', 'CRM'],\n",
    "                              'opc': ['PBX3', 'ANKRD55', 'BRCA2', 'OPC', 'PLP1'],\n",
    "                              'oligodendrocyte': ['GRIK2', 'BACE2', 'PLXDC2', 'SLC38A1'],\n",
    "                              'endothelial': ['ABCC9', 'TMEM45B', 'GPR126', 'C7', 'HRK', 'IGFBP6', 'DOCK8', 'G0S2', 'APLN', 'COL8A1'],\n",
    "                        }\n",
    "            } \n",
    "\n",
    "cell_supertype = subclass[save_prefix][cell_group][9]\n",
    "\n",
    "map_meta = True\n",
    "\n",
    "get_cell_types = False   # whether to reformat cell_type annotation\n",
    "\n",
    "subject_ids_for_study = {'leng_sfg': 'PatientID',\n",
    "                        'leng_etc': 'PatientID',\n",
    "                        'seaad_mtg': 'Donor ID', \n",
    "                        'gazestani_pfc': 'individualID'}\n",
    "\n",
    "subject_id = subject_ids_for_study[save_prefix]     # for leng this is `PatientID` for mathys is 'Subject', and allen is 'individualID'\n",
    "metadata = f'../data/raw/{save_prefix}/{save_prefix}_metadata.csv' # Metatdata location\n",
    "meta = pd.read_csv(metadata, encoding_errors='ignore')\n",
    "test_names = ['early_vs_no', 'late_vs_early', 'late_vs_no', 'ad_vs_no'] # test categories\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "76c62004",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for ctype in subclass[save_prefix].keys():\n",
    "\n",
    "#     sce_obj = sc.read_h5ad(f'../data/raw/{save_prefix}/anndata/rna/{ctype}_raw_anndata.h5ad')\n",
    "#     sce_obj.obs['Subclass'] = sce_obj.obs['anno_ctype'].apply(lambda x: x.split('_')[0])\n",
    "#     sce_obj.obs['Supertype'] = sce_obj.obs['anno_ctype']\n",
    "#     print(ctype)\n",
    "#     sce_obj.write_h5ad(f'../data/raw/{save_prefix}/anndata/rna/{ctype}_raw_anndata.h5ad', compression='gzip')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fae59c52",
   "metadata": {},
   "source": [
    "### **Load and Prep Data**\n",
    "\n",
    "load the preprocessed and annotated data for differential pathway analysis.\n",
    "\n",
    "Make sure the anndata has the count data in either `.layers` or in `.X`. if count data is not in `.layers` it will be assumed that `.X ` contains the counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "3888255a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\"counts\" not in layers...\n",
      "analysis requires unnormalized count data...\n",
      "fetching count data from \"adata.X\"...\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "View of AnnData object with n_obs × n_vars = 426 × 32373\n",
       "    obs: 'nUMI', 'nGene', 'dataset', 'status', 'status2', 'clusters', 'QC_Gene_total_count.x', 'QC_Gene_unique_count.x', 'QC_MT.pct.x', 'QC_IEG.pct.x', 'QC_top50_pct.x', 'anno_batch.x', 'cell_barcode', 'sample', 'anno_organism', 'ds_batch', 'anno_batch.y', 'anno_orig_cellState', 'anno_age', 'anno_sex', 'anno_First_author', 'anno_pmid', 'anno_region', 'anno_ctype', 'anno_class', 'anno_braak_score', 'anno_condition', 'anno_RNAseq_profiling_method', 'anno_RNAseq_platform', 'QC_MT.pct.y', 'QC_IEG.pct.y', 'QC_Gene_total_count.y', 'QC_Gene_unique_count.y', 'QC_top50_pct.y', 'UMAP_1', 'UMAP_2', 'cell_id', 'class_label', 'class_name', 'class_softmax_probability', 'subclass_label', 'subclass_name', 'subclass_softmax_probability', 'supertype_label', 'supertype_name', 'supertype_softmax_probability', 'individualID', 'Subclass', 'Supertype', 'pathology.group'\n",
       "    layers: 'counts'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_annot = sc.read_h5ad(f'../data/raw/{save_prefix}/anndata/rna/{cell_group}_raw_anndata.h5ad')\n",
    "adata_annot.obs_names_make_unique()\n",
    "adata_annot.var_names_make_unique()\n",
    "\n",
    "if 'counts' not in adata_annot.layers.keys():\n",
    "    print('\"counts\" not in layers...')\n",
    "    print('analysis requires unnormalized count data...')\n",
    "    print('fetching count data from \"adata.X\"...')\n",
    "    adata_annot.layers['counts'] = adata_annot.X.copy()\n",
    "\n",
    "else:\n",
    "    adata_annot.X = adata_annot.layers['counts'].copy()\n",
    "\n",
    "layer_keys = list(adata_annot.layers.keys())\n",
    "for layer in layer_keys:\n",
    "    if layer != 'counts':\n",
    "        del adata_annot.layers[layer]\n",
    "\n",
    "del adata_annot.obsm, adata_annot.varm, adata_annot.uns, adata_annot.obsp, adata_annot.X, adata_annot.var \n",
    "\n",
    "\n",
    "celltypes = list(adata_annot.obs[cell_type_column].unique())\n",
    "\n",
    "# map the pathology group to teh subject id in .obs\n",
    "\n",
    "meta = pd.read_csv(metadata, encoding_errors='ignore')\n",
    "meta = meta.astype(str)\n",
    "mapping = dict(zip(meta[subject_id], meta['pathology.group']))\n",
    "adata_annot.obs[subject_id] = adata_annot.obs[subject_id].astype(\"category\")\n",
    "adata_annot.obs['pathology.group'] = adata_annot.obs[subject_id].map(mapping)\n",
    "\n",
    "\n",
    "adata_annot = adata_annot[adata_annot.obs.Subclass == cell_supertype]\n",
    "\n",
    "if save_prefix in ['seaad_mtg', 'seaad_pfc']:\n",
    "    adata_annot = adata_annot[adata_annot.obs['Neurotypical reference'] == 'False']\n",
    "    adata_annot = adata_annot[adata_annot.obs['Severely Affected Donor'] == 'N']\n",
    "\n",
    "\n",
    "adata_annot"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "95e3f93b",
   "metadata": {},
   "source": [
    "### **Count Aggregation Parameters**\n",
    "\n",
    "- `filter_genes`: Setermines if genes should be filtered using `gene_celltype_threshold` before DE tests.\n",
    "\n",
    "- `pseudobulking_strategies`: Defines the pseudobulking strategy and includes options such as `network`, `random`, `smaller_network`, `standardbulk`, and `blanchardbulk`.\n",
    "\n",
    "- `normalization_methods`: Specifies normalization methods for data, options are \"TMM\", \"VST\", \"rmTop50\", and \"none\" (default: \"CPM\").\n",
    "\n",
    "- `gene_selection`: Required only if `pseudobulking_strategy=='metacell'`, indicating the gene selection method for setting up a Seurat object for WGCNA.\n",
    "\n",
    "- `geneSet`: Required only if `pseudobulking_strategy=='metacell'` and `gene_selction='custom'`, specifying a custom gene set.\n",
    "\n",
    "- `gene_celltype_threshold`: Determines the number of cells a gene must be expressed in.\n",
    "\n",
    "- `pathway_gene_threshold`: Determines the number of genes that must be present in a pathway.\n",
    "\n",
    "- `min_size_limit`: Specifies the minimum acceptable size of the pseudobulk data, with pseudobulks below this cell threshold excluded from the analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "fb37e8af",
   "metadata": {},
   "outputs": [],
   "source": [
    "filter_genes = \"TRUE\"\n",
    "count_agg_strategy = 'network'    # options 'network', 'random', 'smaller_network', 'standardbulk', 'metacell', `blanchardbulk``\n",
    "normalization_methods = [\"CPM\"]   # options: \"TMM\", \"VST\", \"rmTop50\", \"none\" default: \"CPM\"\n",
    "gene_selection = 'fraction'       # Only required if `pseudobulking_strategy=='metacell'``. specifies the gene selection method when setting up seurat object for WGCNA. \n",
    "geneSet = ''                      # Only required if `pseudobulking_strategy=='metacell'`` and `gene_selction='custom'`\n",
    "gene_celltype_threshold = 0.10    # determines number of cells the gene must be expressed in\n",
    "pathway_gene_threshold = 0.33     # determines number of genes that must be present in that pathway\n",
    "min_size_limit = 30               # min_size_limit: the minimum acceptable size of the pseudobulk data.\n",
    "                                  # min_size_limit: pseudobulks with cells less than this threshold are excluded from the analysis "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "63debebe",
   "metadata": {},
   "source": [
    "### **Differential Pathway Activity Testsing Parameters:**\n",
    "\n",
    "The primary differential pathway activity analaysis:\n",
    "\n",
    "- `covariates`: List of covariates, including (`including pathology.group`, ) e.g `Sex`, `Sample Batch`, and `Age`. Covariates can be factors that are not of primary interest but might have an effect on the pathological status. For no additional confounders, set `covaraites = ['pathology.group']`.\n",
    "\n",
    "    - `Note`: Ensure that continuous covariates are similarly scaled.\n",
    "\n",
    "    - `Standardization`: Maintain uniformity in categorical data. E.g., Avoid having 'M', 'Male', and 'male' in the same dataset.\n",
    "\n",
    "    - `Impact Assessment`: Regularly assess the impact of the covariates included the differential expression (DE) results by excluding them one by one and checking how the DE patterns change. If a covariate unduly influences results, exercise caution.\n",
    "\n",
    "- `random_effect`: Column name(s) for the random effect(s) (in this case `subject_id` or `Sample_batch`). For pseudobulk data, set do not include `subject_id`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "687feb23",
   "metadata": {},
   "outputs": [],
   "source": [
    "######################### Differential expression arguments #########################\n",
    "\n",
    "random_effect = \"\".join(subject_id.split(\" \"))      # Technical covariate to be included as random effect not of interest (regressed out by duplicatecorrection)\n",
    "duplicate_correction = 'False'                      # whether to run duplicate corretion in Limma to block out technical_covariate\n",
    "jack_knifing = False                                # False if adata_annot.obs[subject_id].nunique()<=10 else True\n",
    "\n",
    "covaraites_for_study = {'leng_sfg': ['pathology.group', 'QC_Gene_unique_log', 'QC_MT.pct'],  # ['ageDeath.cat',],\n",
    "                        'leng_etc': ['pathology.group', 'QC_Gene_unique_log', 'QC_MT.pct'],  # ['ageDeath.cat',]\n",
    "                        'seaad_mtg': ['Continuous Pseudo-progression Score', 'Age at Death', 'Gender', 'QC_Gene_unique_log', 'QC_MT.pct'],\n",
    "                        'gazestani_pfc': ['pathology.group', 'anno_sex', 'anno_age', 'QC_Gene_unique_log', 'QC_MT.pct']}\n",
    "\n",
    "numeric_covariates_for_study = {'gazestani_pfc': ['anno_age',],\n",
    "                                'seaad_mtg': ['Age at Death'],\n",
    "                                'leng_etc': 'None'}\n",
    "\n",
    "\n",
    "covariates = covaraites_for_study[save_prefix]  # list of covariates to be accounted for in regression.True\n",
    "numeric_covariates = numeric_covariates_for_study[save_prefix]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "fe85fb47",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 426 × 32373\n",
       "    obs: 'anno_age', 'anno_sex', 'individualID', 'Subclass', 'pathology.group'\n",
       "    layers: 'counts'"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "if \" \" in subject_id:\n",
    "    subject_id2  = \"\".join(subject_id.split(\" \"))\n",
    "    subject_id2 = subject_id2.replace(\"-\", \"\")\n",
    "    adata_annot.obs[subject_id2] = adata_annot.obs[subject_id].copy()\n",
    "    meta.rename(columns={subject_id: subject_id2}, inplace=True)\n",
    "    del adata_annot.obs[subject_id]\n",
    "    subject_id = subject_id2\n",
    "    \n",
    "    \n",
    "for col in adata_annot.obs.keys():\n",
    "    if col not in covariates and col != cell_type_column and col!=subject_id:\n",
    "            try:\n",
    "                del adata_annot.obs[col]\n",
    "            except KeyError:\n",
    "                continue\n",
    "\n",
    "for col in covariates:\n",
    "    if \" \" in col:\n",
    "        col2  = \"\".join(col.split(\" \"))\n",
    "        col2 = col2.replace(\"-\", \"\")\n",
    "        adata_annot.obs.rename(columns={col: col2}, inplace=True)\n",
    "        meta.rename(columns={col: col2}, inplace=True)\n",
    "        covariates[covariates.index(col)] = col2\n",
    "        \n",
    "        if col in numeric_covariates:\n",
    "            numeric_covariates[numeric_covariates.index(col)] = col2\n",
    "\n",
    "\n",
    "adata_annot.obs[cell_type_column] =  adata_annot.obs[cell_type_column].astype(\"category\")  \n",
    "adata_annot.obs.rename(columns={cell_type_column: ''.join(char for char in cell_type_column if char.isalnum())}, inplace=True)\n",
    "cell_type_column = ''.join(char for char in cell_type_column if char.isalnum())\n",
    "\n",
    "adata_annot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "738c2a7b",
   "metadata": {},
   "source": [
    "### **Send parameters to R interface with rpy2**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "2641b9ec",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "class: SingleCellExperiment \n",
      "dim: 32373 426 \n",
      "metadata(0):\n",
      "assays(1): counts\n",
      "rownames(32373): A1BG A1BG-AS1 ... RP11-110H1.8 PRSS58\n",
      "rowData names(0):\n",
      "colnames(426): 89 123 ... 22303 22380\n",
      "colData names(5): anno_age anno_sex individualID Subclass\n",
      "  pathology.group\n",
      "reducedDimNames(0):\n",
      "mainExpName: NULL\n",
      "altExpNames(0):\n",
      "[1] \"loaded data into memory for recursive use\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i adata_annot -i subject_id -i cell_type_column -i cell_group -i geneSet -i gene_celltype_threshold -i celltypes -i test_names -i save_prefix -i gene_selection -i celltypes -i control_group -i metadata -i subject_id -i covariates -i random_effect -i min_size_limit -i count_agg_strategy -i numeric_covariates \n",
    "\n",
    "print(adata_annot)\n",
    "print('loaded data into memory for recursive use')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53e506d2",
   "metadata": {},
   "source": [
    "# **Count Aggregation**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d1d8668",
   "metadata": {},
   "source": [
    "## **Standard Pseudobulk Approach**\n",
    "\n",
    "To create pseudobulk data per donor and cell_type"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "64ff536c",
   "metadata": {},
   "source": [
    "### **Pseudo-bulk With Decoupler**\n",
    "\n",
    "#### **Requied Parameters when using decoupler**\n",
    "\n",
    "\n",
    "- `min_counts`: Filter to remove samples by a minimum number of summed counts in a sample-group pair when using standard pseudobulk method\n",
    "\n",
    "- `min_count`: Minimum count requiered per gene for at least some samples.\n",
    "\n",
    "- `min_total_count`: Minimum total count required per gene across all samples.\n",
    "\n",
    "- `large_n`: Number of samples per group that is considered to be \"large\".\n",
    "    \n",
    "- `min_prop`: Minimum proportion of samples in the smallest group that express the gene."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "23402d2c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get pseudo-bulk profile\n",
    "\n",
    "min_counts = 1000       #    Filter to remove samples by a minimum number of summed counts in a sample-group pair when using standard pseudobulk method\n",
    "min_count = 10          #    Minimum count requiered per gene for at least some samples.\n",
    "min_total_count = 15    #    Minimum total count required per gene across all samples.\n",
    "large_n = 1             #    Number of samples per group that is considered to be \"large\".\n",
    "min_prop = 1            #    Minimum proportion of samples in the smallest group that express the gene.\n",
    "\n",
    "if count_agg_strategy == 'standardbulk':\n",
    "    \n",
    "    pdata = dc.get_pseudobulk(\n",
    "        adata_annot,\n",
    "        sample_col=subject_id,\n",
    "        groups_col=cell_type_column,\n",
    "        layer='counts',\n",
    "        mode='sum',\n",
    "        min_cells=min_size_limit, \n",
    "        min_counts=min_counts,\n",
    "    )\n",
    "\n",
    "    # Sample-cell-type QC visualization\n",
    "    dc.plot_psbulk_samples(pdata, \n",
    "                           groupby=[subject_id, cell_type_column], \n",
    "                           figsize=(12, 6))\n",
    "    \n",
    "    summed_counts_per_celltype = {}\n",
    "    expressed_genes_per_celltype = {}\n",
    "\n",
    "    fig, axs = plt.subplots(ncols=5, nrows=len(pdata.obs[cell_type_column].unique())//5+1, figsize=(14, 8))\n",
    "    axs = axs.flatten()\n",
    "    for i, cell_type in enumerate(pdata.obs[cell_type_column].unique()):\n",
    "        summed_counts_per_celltype[cell_type] = pdata[pdata.obs[cell_type_column]==cell_type].copy()\n",
    "        \n",
    "        # gene QC visualization\n",
    "\n",
    "        dc.plot_filter_by_expr(summed_counts_per_celltype[cell_type], \n",
    "                               group='pathology.group', \n",
    "                               min_count=min_count,\n",
    "                               min_total_count=min_total_count, \n",
    "                               large_n=large_n, \n",
    "                               min_prop=min_prop, \n",
    "                               ax=axs[i],\n",
    "                               )\n",
    "        \n",
    "        axs[i].set_title(cell_type)\n",
    "\n",
    "        \n",
    "        expressed_genes_per_celltype[cell_type] = dc.filter_by_expr(summed_counts_per_celltype[cell_type],\n",
    "                                                                    group='pathology.group', \n",
    "                                                                    min_count=min_count,\n",
    "                                                                    min_total_count=min_total_count, \n",
    "                                                                    large_n=large_n, \n",
    "                                                                    min_prop=min_prop)\n",
    "    plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "8e6539da",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "\n",
    "if count_agg_strategy == 'standardbulk':\n",
    "    summed_counts_per_celltype =  robjects.ListVector( {\n",
    "                                                        cell_type: summed_counts_per_celltype[cell_type]\n",
    "                                                    \n",
    "                                                        for cell_type in summed_counts_per_celltype.keys()\n",
    "                                                        }\n",
    "                                                    )\n",
    "\n",
    "    expressed_genes_per_celltype =  robjects.ListVector( {\n",
    "                                                        cell_type: expressed_genes_per_celltype[cell_type]\n",
    "                                                    \n",
    "                                                        for cell_type in expressed_genes_per_celltype.keys()\n",
    "                                                        }\n",
    "                                                    )\n",
    "else:\n",
    "    summed_counts_per_celltype = 'NULL'\n",
    "    expressed_genes_per_celltype = 'NULL'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "5aa5a888",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Pseudobulk method not implemented.\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i summed_counts_per_celltype -i expressed_genes_per_celltype\n",
    "\n",
    "if (summed_counts_per_celltype == 'NULL'){\n",
    "    print('Pseudobulk method not implemented.')\n",
    "} else {\n",
    "    print('done')\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b31d3038",
   "metadata": {},
   "source": [
    "### **Pseuobulking With Strategy From [Joel W. Blanchard1](https://doi.org/10.1038/s41586-022-05439-w)**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "03b62009",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "if (count_agg_strategy == 'blanchardbulk'){\n",
    "    \n",
    "    conflicts_prefer(base::unname)\n",
    "    \n",
    "    source('../scripts/functions/qc_and_annotation_aux_functions.r')\n",
    "\n",
    "    library(SingleCellExperiment)\n",
    "    library(Matrix)\n",
    "    library(tidyr)\n",
    "\n",
    "    sce <- adata_annot\n",
    "\n",
    "    # Update sce with QC_Gene_total_count and QC_Gene_unique_count\n",
    "    for (id in unique(colData(sce)$subject_id)){\n",
    "        mask <- colData(sce)[, subject_id] == id\n",
    "        QC_Gene_total_count <- apply(counts(sce[, mask]), 1, sum)\n",
    "        \n",
    "        sce$QC_Gene_total_count[mask] <- sum(QC_Gene_total_count)\n",
    "        sce$QC_Gene_unique_count[mask] <- sum(QC_Gene_total_count != 0)\n",
    "        \n",
    "        sce$QC_Gene_total_log[mask] <- log2(sce$QC_Gene_total_count[mask])\n",
    "        sce$QC_Gene_unique_log[mask] <- log2(sce$QC_Gene_unique_count[mask])\n",
    "    }\n",
    "\n",
    "    print('normalizing counts')\n",
    "    sce = normalize.default(sce)\n",
    "\n",
    "    # standardize subject IDs\n",
    "    meta = colData(sce)\n",
    "    meta[, subject_id] <- gsub(\"\\\\.\", \"_\", meta[, subject_id])\n",
    "    cell_labels = rownames(meta)\n",
    "\n",
    "    # use matrix multiplication to summarize (sum) across counts per cell (including all individuals)\n",
    "    print('sum counts by individual')\n",
    "    labels = as.data.frame(as.character(interaction(meta[, cell_type_column], meta[, subject_id])))\n",
    "    summed_logcounts_cellxind = sum_counts(logcounts(sce), labels, cell_labels)\n",
    "    \n",
    "    # get averages corresponding to both count matrices\n",
    "    avs_logcounts_cellxind = t(apply(summed_logcounts_cellxind$summed_counts, 1, function(x){x/summed_logcounts_cellxind$ncells}))\n",
    "\n",
    "    # split the averages by celltype\n",
    "    x = (strsplit(colnames(avs_logcounts_cellxind), '[.]'))\n",
    "    celltype = unlist(lapply(1:length(x), function(i) x[[i]][[1]]))\n",
    "    individual = unlist(lapply(1:length(x), function(i) x[[i]][[2]]))\n",
    "    celltype_unique = unique(celltype)\n",
    "\n",
    "    avs_by_ind_out = list()\n",
    "    summed_counts_per_celltype = list()\n",
    "    expressed_genes_per_celltype = list()\n",
    "\n",
    "    for(i in celltype_unique){\n",
    "\n",
    "        index = celltype==i\n",
    "        df = avs_logcounts_cellxind[, index]\n",
    "        colnames(df) = individual[index]\n",
    "        avs_by_ind_out[[i]] = df\n",
    "\n",
    "        # create SingleCellExperiment object for summed counts\n",
    "        summed_counts_per_celltype[[i]] = SingleCellExperiment(assays = list(logcounts = as.matrix(df)))\n",
    "        rownames(colData(summed_counts_per_celltype[[i]])) = colnames(df)\n",
    "        \n",
    "        metadata <- meta[!duplicated(meta[, subject_id]), ]\n",
    "        colData(summed_counts_per_celltype[[i]]) <- merge(colData(summed_counts_per_celltype[[i]]), \n",
    "                                                    metadata, \n",
    "                                                    by = NULL,\n",
    "                                                    by.x = \"row.names\",\n",
    "                                                    by.y = subject_id, \n",
    "                                                    all.x = TRUE)\n",
    "\n",
    "        rownames(rowData(summed_counts_per_celltype[[i]])) <- rownames(df)  \n",
    "\n",
    "        colData(summed_counts_per_celltype[[i]])[, subject_id] = colData(summed_counts_per_celltype[[i]])$Row.names\n",
    "        \n",
    "        # in how many cells per celltype is each gene detected?\n",
    "        print('getting nonzero counts')\n",
    "\n",
    "        mask <- colData(sce)[, cell_type_column] == i\n",
    "        sce_temp <- sce[, mask]\n",
    "        meta_temp <- as.data.frame(colData(sce_temp)[, cell_type_column])\n",
    "\n",
    "        counts_nonzero = counts(sce_temp)>0\n",
    "        \n",
    "        # determine expressed genes based on threshold\n",
    "        detected_genes_cell = sum_counts(counts_nonzero, label = meta_temp, cell_labels)\n",
    "        fraction_detected_genes_cell = t(apply(detected_genes_cell$summed_counts, 1, function(x){x/detected_genes_cell$ncells}))\n",
    "\n",
    "        # get expression list\n",
    "        fraction_detected_genes_cell_binary = fraction_detected_genes_cell>gene_celltype_threshold\n",
    "        genes = rownames(fraction_detected_genes_cell)\n",
    "        expressed = lapply(colnames(fraction_detected_genes_cell_binary), function(x) genes[unname(fraction_detected_genes_cell_binary[,x])])\n",
    "        names(expressed) = colnames(fraction_detected_genes_cell_binary)\n",
    "\n",
    "        expressed_genes_per_celltype[[i]] <- names(expressed)\n",
    "            \n",
    "    }\n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c8b11465",
   "metadata": {},
   "source": [
    "## **Pseudocell approach [Gazestani. et. al. 2023](https://www.sciencedirect.com/science/article/pii/S0092867423008590?via%3Dihub)**\n",
    "\n",
    "### **Pseudocell construction** \n",
    "\n",
    "\n",
    "Existing pathway enrichment techniques, initially designed for bulk RNA-seq data, typically employ either cell-based or pseudobulk-based approaches. While cell-based methods are time-consuming and sensitive to single-cell data dropouts, pseudobulk methods, which aggregate cell expressions into a single sample, address these issues but may lead to a loss of statistical power, neglecting variations in confidence levels between individuals with different cell-type sampling sizes.\n",
    "\n",
    "To address these challenges pseudocell/metacell methods have been introduced this method lies between the cell-based and pseudobulk-based approaches, so it can take advantage of each method while remedying the issues attributed to them.\n",
    "\n",
    "According to **[Gazestani. et. al. 2023](https://www.sciencedirect.com/science/article/pii/S0092867423008590?via%3Dihub)**, we first aggregate the raw UMI count of, on average, every 30 cells per subject and cell type. \n",
    "\n",
    "We construct one pseudocell for cell types that have between `15 to 45 cells in a donor` and exclude cell types that had less than `15 cells`. This pseudocell-based analysis reduces the impact of dropout and technical variability, while ameliorating low statistical power and high variation in sample size issues attributed to the pseudobulk approaches. \n",
    "\n",
    "**[Gazestani. et. al. 2023](https://www.sciencedirect.com/science/article/pii/S0092867423008590?via%3Dihub)**, provide functions:\n",
    "\n",
    "- `To generate pseudocells that are composed of, on average, 20 cells or higher. This function can also generate pseudobulk samples.`\n",
    "\n",
    "- `To generate pseudocells that are on average composed of 10 cells`"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5d3c5a72",
   "metadata": {},
   "source": [
    "### **Available Aggregration Methods**\n",
    "\n",
    "\n",
    "1. #### **Pseudocells from low-dim network embedding**\n",
    "\n",
    "The main arguments for running the function:\n",
    "\n",
    "- `parsing.col.names`: a vector of column names to be used to parse the data. Usually the columns related to the donor id and the cell type annotation (see above figure).\n",
    "\n",
    "- `pseudocell.size`: the average size of pseudcoells. If NULL, generates pseudobulk data.\n",
    "\n",
    "- `inputExpData`: Input single cell data object in format.\n",
    "\n",
    "- `min_size_limit`: minimum acceptable size of the pseudocells. usually 10 or 15.\n",
    "\n",
    "- `inputPhenoData`: the meta data on the cells that matches in order with the input PC space to use to create the similarity network. It's highly recommended to be used if. defualt: NULL.\n",
    "\n",
    "- `nPCs`: Number of pcs to use to create the similarity network. Used only if human or mouse\n",
    "\n",
    "- `rand_pseudobulk_mod`: To randomly assign cells to pseudocells in each cell type from each individual; or create a similarity network between these cells and use that similarity net to group cells with highly similar expression patterns to each other. `default: TRUE, but Here we set to FALSE``\n",
    "\n",
    "The authors reiterate that it's highly recommended to provide if the function will generate the embedding for each cell type in each individual. it's possible that such embedding is driven by the quality of the cells in the group as opposed to biological variation, and hence quality of the pseudocells and inference on them would be limited.\n",
    "\n",
    "\n",
    "2. #### **Pseudocells with random cell assignment**\n",
    "\n",
    "Alternatively, we can create pseudocells using a method that randomly assign cells to pseudocells in each cell type from each individual by setting `rand_pseudobulk_mod=TRUE`. Rather than the previous method which creates a similarity network between these cells and uses that similarity network to group cells with highly similar expression patterns to each other.\n",
    "\n",
    "\n",
    "3. #### **Pseudocells with smaller size**\n",
    "\n",
    "We can also generate pseudocells using that allows us to have pseudocells of smaller size. The main arguments for this function:\n",
    "\n",
    "- `inputExpData`: Input single cell data object in SingleCellExperiment format\n",
    "\n",
    "- `embeddings`: Input PC space to use to create the similarity network. It can't be set to NULL\n",
    "\n",
    "- `pseudobulk_split_col`: The column name to parse the single cell data.\n",
    "\n",
    "- `min_dataset_size`: Minimum number of acceptable pseudocell size. default: 4.\n",
    "\n",
    "\n",
    "#### **Gene Filtering**\n",
    "\n",
    "We employ a rigorous filtering strategy to focus on biologically relevant gene programs. \n",
    "\n",
    "Genes were initially filtered based on their expression across samples, with those exhibiting low expression levels in more than 99% of the dataset excluded. \n",
    "\n",
    "- First, a counts-per-million (CPM) threshold was applied, retaining genes with CPM values surpassing the higher of 1% of the total samples or the minimum between 15 and one-third of the total samples. This strategy aimed to ensure the inclusion of genes with sufficient expression depth and magnitude. \n",
    "\n",
    "Moreover, to enhance the specificity of the retained genes, we implement a secondary filtering step based on individual sample expression. \n",
    "\n",
    "- Genes with lower than a count of 1 between maximum of 1% of the total samples or the minimum between 10 and one-third of the total samples were excluded. This approach, which includes both global and sample-specific considerations, facilitates the identification of a robust set of genes for downstream analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "b144d494",
   "metadata": {},
   "outputs": [],
   "source": [
    "############## PARAMETERS FOR NETWORK AND RANDOM #####################\n",
    "\n",
    "pseudocell_size = 30    # the average size of pseudcoells. If NULL, generates pseudobulk data.\n",
    "nPCs = 30               # Number of pcs to use to create the similarity network. Used only if human or mouse\n",
    "organism = 'Human' \n",
    "\n",
    "############### PARAMETERS FOR 'SMALL NETWORK' #####################\n",
    "\n",
    "pseudobulk_split_col = \"lib_anno\"\n",
    "min_dataset_size = 4 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "442d4b83",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[conflicted] Will prefer GSEABase::setdiff over any other package.\n",
      "[conflicted] Will prefer SingleCellExperiment::counts over any other package.\n",
      "Main Functions:\n",
      ".myRead10X()\n",
      ".myRead10X_h5()\n",
      ".myLigerToExpSet()\n",
      ".mycBindFn()\n",
      ".myExpSetCreatorFn()\n",
      ".myIntegrative_oneline()\n",
      ".myFindAllMarkers()\n",
      ".myAnnotateFn()\n",
      ".my2dPlot()\n",
      ".myPseudoCellfn_v2()\n",
      ".myLabelTransfer_harmony()\n",
      ".myLabelTransfer_liger()\n",
      ".myMapToHuman()\n",
      ".myRiverPlotFn()\n",
      ".myClusteringOptimizerFn()\n",
      ".myMarkerBasedAnalysisFn()\n",
      ".mycellAssignHeatmap()\n",
      ".myMetaMarkerFn()\n",
      ".myFindNeighbors()\n",
      ".myVlnPlot()\n",
      ".myFeaturePlot()\n",
      ".myheatmap.3()\n",
      ".myEvalMarkers()\n",
      ".myReadGMT()\n",
      ".mySplitObject()\n",
      ".myRTNgsea(two-sided GSEA implementation)\n",
      "[1] \"batch information is in the anno_batch variable\"\n",
      "[1] \"Number of MT genes in the dataset: 30 / 13\"\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Loading required package: GenomicFeatures\n",
       "Loading required package: AnnotationFilter\n",
       "------------------------------------------------------------------------------\n",
       "You have loaded plyr after dplyr - this is likely to cause problems.\n",
       "If you need functions from both plyr and dplyr, please load plyr first, then dplyr:\n",
       "library(plyr); library(dplyr)\n",
       "------------------------------------------------------------------------------\n",
       "qs 0.25.5\n",
       "Warning: Feature names cannot have underscores ('_'), replacing with dashes ('-')\n",
       "Performing log-normalization\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Calculating gene variances\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Calculating feature variances of standardized and clipped values\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Centering and scaling data matrix\n",
       "\r  |                                                                            \r  |                                                                      |   0%\r  |                                                                            \r  |===================================                                   |  50%\r  |                                                                            \r  |======================================================================| 100%\n",
       "In addition: Warning messages:\n",
       "1: package ‘scuttle’ was built under R version 4.3.1 \n",
       "2: package ‘ensembldb’ was built under R version 4.3.1 \n",
       "3: package ‘GenomicFeatures’ was built under R version 4.3.1 \n",
       "4: package 'scran' was built under R version 4.3.1 \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R -i organism -i nPCs -i pseudocell_size -i pseudobulk_split_col -i min_dataset_size\n",
    "\n",
    "  \n",
    "library(scuttle)\n",
    "library(Matrix)\n",
    "library(ensembldb)\n",
    "library(EnsDb.Hsapiens.v86)\n",
    "\n",
    "conflicts_prefer(GSEABase::setdiff)\n",
    "conflicts_prefer(SingleCellExperiment::counts)\n",
    "\n",
    "\n",
    "source(\"../scripts/functions/deg_functions/sconline_code.R\")\n",
    "\n",
    "# generating the embedding space\n",
    "\n",
    "exp_seurat = .extraExport2SeuratFn(adata_annot) %>%\n",
    "Seurat::NormalizeData() %>%\n",
    "    FindVariableFeatures() %>% \n",
    "    ScaleData() %>% \n",
    "    RunPCA(verbose=F)\n",
    "\n",
    "embedding_data = exp_seurat@reductions$pca@cell.embeddings[,1:30]\n",
    "  \n",
    "if (count_agg_strategy == 'network'){\n",
    "\n",
    "  # generating the pseudocells\n",
    "    \n",
    "  # Function adds/modifies three annotation columns: pseuodcell_size, QC_Gene_total_count, QC_Gene_unique_count\n",
    "  # QC_Gene_total_count: equivalant to nUMI for the pseudobulk samples\n",
    "  # QC_Gene_unique_count: equivalant to nGene for the pseudobulk samples\n",
    "  # use scale(tst$QC_Gene_total_count) and scale(tst$pseudocell_size) as additional covariates for the DE analysis\n",
    "\n",
    "  res = suppressWarnings(.sconline.PseudobulkGeneration(argList = NULL, \n",
    "                                  # The columns in the pheonData that will be used to parse the expression data \n",
    "                                  # and generate the pseudocell/pseudobulk data\n",
    "                                    parsing.col.names = c(subject_id, cell_type_column), \n",
    "                                  # average pseudocell size.\n",
    "                                    pseudocell.size = pseudocell_size,\n",
    "                                    inputExpData = adata_annot,\n",
    "                                  # minimum acceptable size (ie, #cells) for each pseudobulk\n",
    "                                    min_size_limit = min_size_limit,\n",
    "                                  # in case we want to run the function outside sconline space\n",
    "                                    inputPhenoData = as.data.frame(colData(adata_annot)),\n",
    "                                  # the embedding space to be used for the generation of the pseudobulk.\n",
    "                                  # only needed when pseudocell.size is not null\n",
    "                                    inputEmbedding = embedding_data, \n",
    "                                  # the dimension of the embedding space for the construction of pseudobulk data\n",
    "                                    nPCs = nPCs, \n",
    "                                    ncores = 1,\n",
    "                                    rand_pseudobulk_mod = F,\n",
    "                                  # used to identify and estimate Mitochondrial Genes\n",
    "                                    organism = organism))\n",
    "                                    \n",
    "  res$QC_Gene_total_log=log2(res$QC_Gene_total_count)\n",
    "  res$QC_Gene_unique_log=log2(res$QC_Gene_unique_count)\n",
    "  res$QC_MT_log=log2(res$QC_MT.pct)\n",
    "\n",
    "  summed_counts_per_celltype <- lapply(unique(colData(res)[, cell_type_column]), function(cell_type) res[, colData(res)[, cell_type_column] == cell_type,])\n",
    "  names(summed_counts_per_celltype) <- unique(colData(res)[, cell_type_column])\n",
    "\n",
    "} else  if (count_agg_strategy == 'random'){\n",
    "\n",
    "  res = suppressWarnings(.sconline.PseudobulkGeneration(argList=NULL, \n",
    "                                  # The columns in the pheonData that will be used to parse the expression data \n",
    "                                  # and generate the pseudocell/pseudobulk data\n",
    "                                    parsing.col.names = c(subject_id, cell_type_column), \n",
    "                                  # average pseudocell size.\n",
    "                                    pseudocell.size = pseudocell_size,\n",
    "                                    inputExpData = adata_annot,\n",
    "                                  # minimum acceptable size (ie, #cells) for each pseudobul\n",
    "                                    min_size_limit = min_size_limit,\n",
    "                                  # in case we want to run the function outside sconline space\n",
    "                                    inputPhenoData = as.data.frame(colData(adata_annot)),\n",
    "                                    ncores = 1,\n",
    "                                    rand_pseudobulk_mod = T,\n",
    "                                  # used to identify and estimate Mitochondrial Genes\n",
    "                                    organism = organism))\n",
    "\n",
    "  res$QC_Gene_total_log=log2(res$QC_Gene_total_count)\n",
    "  res$QC_Gene_unique_log=log2(res$QC_Gene_unique_count)\n",
    "  res$QC_MT_log=log2(res$QC_MT.pct)\n",
    "\n",
    "\n",
    "  summed_counts_per_celltype <- lapply(unique(colData(res)[, cell_type_column]), function(cell_type) res[, colData(res)[, cell_type_column] == cell_type,])\n",
    "  names(summed_counts_per_celltype) <- unique(colData(res)[, cell_type_column])      \n",
    "                                \n",
    "}  else if (count_agg_strategy == 'pseudobulk'){\n",
    "\n",
    "  res = suppressWarnings(.sconline.PseudobulkGeneration(argList = NULL, \n",
    "                                  # The columns in the pheonData that will be used to parse the expression data \n",
    "                                  # and generate the pseudocell/pseudobulk data\n",
    "                                    parsing.col.names = c(subject_id, cell_type_column),  \n",
    "                                  # average pseudocell size.\n",
    "                                    pseudocell.size = NULL,\n",
    "                                    inputExpData = adata_annot,\n",
    "                                  # minimum acceptable size (ie, #cells) for each pseudobulk\n",
    "                                    min_size_limit = min_size_limit,\n",
    "                                  # in case we want to run the function outside sconline space\n",
    "                                    inputPhenoData = as.data.frame(colData(adata_annot)),\n",
    "                                  # the embedding space to be used for the generation of the pseudobulk.\n",
    "                                  # only needed when pseudocell.size is not null\n",
    "                                    inputEmbedding = NULL, \n",
    "                                  # the dimension of the embedding space for the construction of pseudobulk data\n",
    "                                    nPCs = NULL, \n",
    "                                    ncores = 1,\n",
    "                                    rand_pseudobulk_mod = F,\n",
    "                                    organism = organism))\n",
    "\n",
    "  res$QC_Gene_total_log=log2(res$QC_Gene_total_count)\n",
    "  res$QC_Gene_unique_log=log2(res$QC_Gene_unique_count)\n",
    "  res$QC_MT_log=log2(res$QC_MT.pct)\n",
    "\n",
    "  summed_counts_per_celltype <- lapply(unique(colData(res)[, cell_type_column]), function(cell_type) res[, colData(res)[, cell_type_column] == cell_type,])\n",
    "  names(summed_counts_per_celltype) <- unique(colData(res)[, cell_type_column])  \n",
    "\n",
    "} else  if (count_agg_strategy == 'smaller_network') {   \n",
    "\n",
    "  adata_annot$lib_anno = paste0(adata_annot[[subject_id]], \"_\", adata_annot[[cell_type_column]])\n",
    "\n",
    "  # creates pseudobulk samples of median size 10\n",
    "  res = .sconline.Pseudobulk10(inputExpData=adata_annot,\n",
    "                                      embeddings = embedding_data,\n",
    "                                      pseudobulk_split_col = \"lib_anno\",\n",
    "                                      min_dataset_size = min_dataset_siz,\n",
    "                                      organism = organism)   \n",
    "\n",
    "  res$QC_Gene_total_log=log2(res$QC_Gene_total_count)\n",
    "  res$QC_Gene_unique_log=log2(res$QC_Gene_unique_count)\n",
    "\n",
    "\n",
    "  summed_counts_per_celltype <- lapply(unique(colData(res)[, cell_type_column]), function(cell_type) res[, colData(res)[, cell_type_column] == cell_type,])\n",
    "  names(summed_counts_per_celltype) <- unique(colData(res)[, cell_type_column])  \n",
    "\n",
    "} else {\n",
    "\n",
    "  summed_counts_per_celltype <- NULL\n",
    "  names(summed_counts_per_celltype) <- NULL\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "77efec55",
   "metadata": {},
   "source": [
    "Filter genes and perform log normalization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "59b74372",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[conflicted] Will prefer base::intersect over any other package.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Number of expressed genes: 7713\"\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Loading required package: edgeR\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "conflicts_prefer(base::intersect)\n",
    "\n",
    "if (count_agg_strategy != 'standardbulk') {\n",
    "    expressed_genes_per_celltype = list()\n",
    "    for (cell_type in names(summed_counts_per_celltype)) {\n",
    "        \n",
    "        # Apply gene expression filtering for each cell type\n",
    "        res <- tryCatch(\n",
    "        {\n",
    "            .sconline.FilterGenesByExpr(summed_counts_per_celltype[[cell_type]],\n",
    "                                        raw_sce = adata_annot[, colData(adata_annot)[, cell_type_column] == cell_type],\n",
    "                                        threshold = gene_celltype_threshold,\n",
    "                                        filter_from = 'raw',\n",
    "                                        TMMnorm = TRUE,\n",
    "                                        quantile.norm = FALSE,\n",
    "                                        prior.count = 0.25)\n",
    "        },\n",
    "        error = function(e) {\n",
    "            cat(cell_type, \": An error occurred\", conditionMessage(e), \"\\n\")\n",
    "            return(NULL)\n",
    "        })\n",
    "\n",
    "        if (!is.null(res)){\n",
    "        \n",
    "            # Subset the counts data to include only selected genes\n",
    "            summed_counts_per_celltype[[cell_type]] <- summed_counts_per_celltype[[cell_type]][rownames(res$E), ]\n",
    "            expressed_genes_per_celltype[[cell_type]] <- as.vector(rownames(res$E))\n",
    "\n",
    "            # Update logcounts assay with the filtered expression values\n",
    "            assays(summed_counts_per_celltype[[cell_type]])[[\"logcounts\"]] <- res$E\n",
    "            \n",
    "        }else{\n",
    "            summed_counts_per_celltype[[cell_type]] <- NULL\n",
    "            expressed_genes_per_celltype[[cell_type]] <- NULL\n",
    "        }\n",
    "    }\n",
    "} else if (count_agg_strategy == 'standardbulk'){\n",
    "\n",
    "    for (cell_type in names(summed_counts_per_celltype)) {\n",
    "        \n",
    "        inputExpData <-  summed_counts_per_celltype[[cell_type]]\n",
    "        tmpexp <- counts(inputExpData)[keep, ]\n",
    "        dge <- DGEList(tmpexp)\n",
    "        dge <- calcNormFactors(dge)\n",
    "        \n",
    "        # Create logCPM object for log-transformed expression values\n",
    "        logCPM <- new(\"EList\")\n",
    "        logCPM$E <- edgeR::cpm(dge, normalized.lib.sizes = TMMnorm, log = TRUE, prior.count = 0.25)\n",
    "        \n",
    "        assays(summed_counts_per_celltype[[cell_type]])[[\"logcounts\"]]  <- logCPM$E\n",
    "\n",
    "    }\n",
    "\n",
    "}"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "c568575c",
   "metadata": {},
   "source": [
    "## **Load and Filter Pathways of Interest Using Expressed Genes**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "bdb0193c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def save_curated_as_gmt(df, name_col, member_col, tempdir):\n",
    "    \"\"\"\n",
    "    Save a DataFrame as a GMT file.\n",
    "\n",
    "    Parameters:\n",
    "    - gmt_df (DataFrame): DataFrame containing the gene set information.\n",
    "    - name_col (str): Name of the column containing gene names.\n",
    "    - tempdir (str): Path to the temporary directory where the GMT file will be saved.\n",
    "    \"\"\"\n",
    "\n",
    "    gmt_df = df.groupby(name_col)[member_col].unique().reset_index()\n",
    "    gmt_df[member_col] = gmt_df[member_col].apply(list)\n",
    "    gmt_df['description'] = gmt_df[name_col]\n",
    "\n",
    "    with open(os.path.join(tempdir, \"gs.gmt\"), 'w') as of:\n",
    "        for _, row in gmt_df.iterrows():\n",
    "            name = row[name_col]                \n",
    "            # description = row['description']\n",
    "            genes = '\\t'.join(row[member_col])\n",
    "            of.write(f\"{name}\\t{genes}\\n\")\n",
    "    return gmt_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a0a3c54a",
   "metadata": {},
   "source": [
    "### **Curated Gene Sets Obtained From [Galea et al. 2022](https://doi.org/10.1016/j.nbd.2022.105655)**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "34125a5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "galea_curated = pd.read_excel('../data/pathway_databases/galea_2022_curated_gene_sets/gene_sets.xlsx', )\n",
    "galea_curated['Gene set'] = galea_curated['Functional categories'] + ': ' + galea_curated['Functional subcategories']\n",
    "galea_curated.rename(columns={'Gene symbol': 'Gene'}, inplace=True)\n",
    "gl_gmt = save_curated_as_gmt(galea_curated, name_col='Gene set', member_col='Gene', tempdir=f'../data/pathway_databases/galea_2022_curated_gene_sets/')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "03693dc8",
   "metadata": {},
   "source": [
    "### **Curated Gene Sets Obtained From [Gabitto et al. 2022](https://doi.org/10.1101/2023.05.08.539485)**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "c61b469d",
   "metadata": {},
   "outputs": [],
   "source": [
    "gabitto_curated = pd.read_excel('../data/pathway_databases/gabitto_2023_curated_gene_sets/gene_sets.xlsx')\n",
    "gb_gmt = save_curated_as_gmt(gabitto_curated, name_col='Gene set', member_col='Gene', tempdir=f'../data/pathway_databases/gabitto_2023_curated_gene_sets/')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d00c6780",
   "metadata": {},
   "source": [
    "### **Curated Synaptic Related Gene Sets Obtained From [Koopmans et al. 2019](https://www.sciencedirect.com/science/article/pii/S0896627319304271?via%3Dihub)**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "158e5ea3",
   "metadata": {},
   "outputs": [],
   "source": [
    "syngo_curated = pd.read_excel('../data/pathway_databases/SynGO_bulk_download_release_20231201/syngo_annotations.xlsx')\n",
    "syngo_curated.rename(columns={'go_name': 'Gene set', 'hgnc_symbol': 'Gene'}, inplace=True)\n",
    "syngo_gmt = save_curated_as_gmt(syngo_curated, name_col='Gene set', member_col='Gene', tempdir=f'../data/pathway_databases/SynGO_bulk_download_release_20231201/')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "631bb641",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[conflicted] Removing existing preference.\n",
      "[conflicted] Will prefer base::intersect over any other package.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"|| getting pathway databases... ||\"\n",
      "1234567891011121314151617181920212223242526272829303132333435363738394041424344454647484950515253545556575859606162636465666768697071727374757677787980818283848586878889909192939495969798991001011021031041051061071081091101111121131141151161171181191201211221231241251261271281291\n",
      "2\n",
      "3\n",
      "4\n",
      "5\n",
      "6\n",
      "7\n",
      "8\n",
      "9\n",
      "10\n",
      "11\n",
      "12\n",
      "13\n",
      "14\n",
      "15\n",
      "16\n",
      "17\n",
      "18\n",
      "19\n",
      "20\n",
      "21\n",
      "22\n",
      "23\n",
      "24\n",
      "25\n",
      "26\n",
      "27\n",
      "28\n",
      "29\n",
      "30\n",
      "31\n",
      "32\n",
      "33\n",
      "34\n",
      "35\n",
      "36\n",
      "37\n",
      "38\n",
      "39\n",
      "40\n",
      "41\n",
      "42\n",
      "43\n",
      "44\n",
      "45\n",
      "46\n",
      "47\n",
      "48\n",
      "49\n",
      "50\n",
      "51\n",
      "52\n",
      "53\n",
      "54\n",
      "55\n",
      "56\n",
      "57\n",
      "58\n",
      "59\n",
      "60\n",
      "61\n",
      "62\n",
      "63\n",
      "64\n",
      "65\n",
      "66\n",
      "67\n",
      "68\n",
      "69\n",
      "70\n",
      "71\n",
      "72\n",
      "73\n",
      "74\n",
      "75\n",
      "76\n",
      "77\n",
      "78\n",
      "79\n",
      "80\n",
      "81\n",
      "82\n",
      "83\n",
      "84\n",
      "85\n",
      "86\n",
      "87\n",
      "88\n",
      "89\n",
      "90\n",
      "91\n",
      "92\n",
      "93\n",
      "94\n",
      "95\n",
      "96\n",
      "97\n",
      "98\n",
      "99\n",
      "100\n",
      "101\n",
      "102\n",
      "103\n",
      "104\n",
      "105\n",
      "106\n",
      "107\n",
      "108\n",
      "109\n",
      "110\n",
      "111\n",
      "112\n",
      "113\n",
      "114\n",
      "115\n",
      "116\n",
      "117\n",
      "118\n",
      "119\n",
      "120\n",
      "121\n",
      "122\n",
      "123\n",
      "124\n",
      "125\n",
      "126\n",
      "127\n",
      "128\n",
      "1234567891011121314151617181920212223242526272829303132333435363738394041424344454647484950515253545556575859606162636465666768697071727374757677787980818283848586878889909192939495969798991001011021031041051061071081091101111121131141151161171181191201211221231241251261271281291301311321331341351361371381391401411421431441451461471481491501511521531541551561571581591601611621631641651661671681691701711721731741751761771781791801811821831841851861871881891901911921931941951961971981992002012022032042052062072082092102112122132142151\n",
      "2\n",
      "3\n",
      "4\n",
      "5\n",
      "6\n",
      "7\n",
      "8\n",
      "9\n",
      "10\n",
      "11\n",
      "12\n",
      "13\n",
      "14\n",
      "15\n",
      "16\n",
      "17\n",
      "18\n",
      "19\n",
      "20\n",
      "21\n",
      "22\n",
      "23\n",
      "24\n",
      "25\n",
      "26\n",
      "27\n",
      "28\n",
      "29\n",
      "30\n",
      "31\n",
      "32\n",
      "33\n",
      "34\n",
      "35\n",
      "36\n",
      "37\n",
      "38\n",
      "39\n",
      "40\n",
      "41\n",
      "42\n",
      "43\n",
      "44\n",
      "45\n",
      "46\n",
      "47\n",
      "48\n",
      "49\n",
      "50\n",
      "51\n",
      "52\n",
      "53\n",
      "54\n",
      "55\n",
      "56\n",
      "57\n",
      "58\n",
      "59\n",
      "60\n",
      "61\n",
      "62\n",
      "63\n",
      "64\n",
      "65\n",
      "66\n",
      "67\n",
      "68\n",
      "69\n",
      "70\n",
      "71\n",
      "72\n",
      "73\n",
      "74\n",
      "75\n",
      "76\n",
      "77\n",
      "78\n",
      "79\n",
      "80\n",
      "81\n",
      "82\n",
      "83\n",
      "84\n",
      "85\n",
      "86\n",
      "87\n",
      "88\n",
      "89\n",
      "90\n",
      "91\n",
      "92\n",
      "93\n",
      "94\n",
      "95\n",
      "96\n",
      "97\n",
      "98\n",
      "99\n",
      "100\n",
      "101\n",
      "102\n",
      "103\n",
      "104\n",
      "105\n",
      "106\n",
      "107\n",
      "108\n",
      "109\n",
      "110\n",
      "111\n",
      "112\n",
      "113\n",
      "114\n",
      "115\n",
      "116\n",
      "117\n",
      "118\n",
      "119\n",
      "120\n",
      "121\n",
      "122\n",
      "123\n",
      "124\n",
      "125\n",
      "126\n",
      "127\n",
      "128\n",
      "129\n",
      "130\n",
      "131\n",
      "132\n",
      "133\n",
      "134\n",
      "135\n",
      "136\n",
      "137\n",
      "138\n",
      "139\n",
      "140\n",
      "141\n",
      "142\n",
      "143\n",
      "144\n",
      "145\n",
      "146\n",
      "147\n",
      "148\n",
      "149\n",
      "150\n",
      "151\n",
      "152\n",
      "153\n",
      "154\n",
      "155\n",
      "156\n",
      "157\n",
      "158\n",
      "159\n",
      "160\n",
      "161\n",
      "162\n",
      "163\n",
      "164\n",
      "165\n",
      "166\n",
      "167\n",
      "168\n",
      "169\n",
      "170\n",
      "171\n",
      "172\n",
      "173\n",
      "174\n",
      "175\n",
      "176\n",
      "177\n",
      "178\n",
      "179\n",
      "180\n",
      "181\n",
      "182\n",
      "183\n",
      "184\n",
      "185\n",
      "186\n",
      "187\n",
      "188\n",
      "189\n",
      "190\n",
      "191\n",
      "192\n",
      "193\n",
      "194\n",
      "195\n",
      "196\n",
      "197\n",
      "198\n",
      "199\n",
      "200\n",
      "201\n",
      "202\n",
      "203\n",
      "204\n",
      "205\n",
      "206\n",
      "207\n",
      "208\n",
      "209\n",
      "210\n",
      "211\n",
      "212\n",
      "213\n",
      "214\n",
      "1234567891011121314151617181920212223242526272829303132333435363738394041424344454647484950515253545556575859606162636465666768697071727374757677787980818283848586878889909192939495969798991001011021031041051061071081091101111121131141151161171181191201211221231241251261271281291301311321331341351361371381391401411421431441451461471481491501511521531541551561571581591601611621631641651661671681691701711721731741751761771781791801811821831841851861871881891901911921931941951961971981992002012022032042052062072082092102112122132142152162172182192202212222232242252262272282292302312322332342352362372382392402412422432442452462472482492502512522532542552562572582592602612622632642652662671\n",
      "2\n",
      "3\n",
      "4\n",
      "5\n",
      "6\n",
      "7\n",
      "8\n",
      "9\n",
      "10\n",
      "11\n",
      "12\n",
      "13\n",
      "14\n",
      "15\n",
      "16\n",
      "17\n",
      "18\n",
      "19\n",
      "20\n",
      "21\n",
      "22\n",
      "23\n",
      "24\n",
      "25\n",
      "26\n",
      "27\n",
      "28\n",
      "29\n",
      "30\n",
      "31\n",
      "32\n",
      "33\n",
      "34\n",
      "35\n",
      "36\n",
      "37\n",
      "38\n",
      "39\n",
      "40\n",
      "41\n",
      "42\n",
      "43\n",
      "44\n",
      "45\n",
      "46\n",
      "47\n",
      "48\n",
      "49\n",
      "50\n",
      "51\n",
      "52\n",
      "53\n",
      "54\n",
      "55\n",
      "56\n",
      "57\n",
      "58\n",
      "59\n",
      "60\n",
      "61\n",
      "62\n",
      "63\n",
      "64\n",
      "65\n",
      "66\n",
      "67\n",
      "68\n",
      "69\n",
      "70\n",
      "71\n",
      "72\n",
      "73\n",
      "74\n",
      "75\n",
      "76\n",
      "77\n",
      "78\n",
      "79\n",
      "80\n",
      "81\n",
      "82\n",
      "83\n",
      "84\n",
      "85\n",
      "86\n",
      "87\n",
      "88\n",
      "89\n",
      "90\n",
      "91\n",
      "92\n",
      "93\n",
      "94\n",
      "95\n",
      "96\n",
      "97\n",
      "98\n",
      "99\n",
      "100\n",
      "101\n",
      "102\n",
      "103\n",
      "104\n",
      "105\n",
      "106\n",
      "107\n",
      "108\n",
      "109\n",
      "110\n",
      "111\n",
      "112\n",
      "113\n",
      "114\n",
      "115\n",
      "116\n",
      "117\n",
      "118\n",
      "119\n",
      "120\n",
      "121\n",
      "122\n",
      "123\n",
      "124\n",
      "125\n",
      "126\n",
      "127\n",
      "128\n",
      "129\n",
      "130\n",
      "131\n",
      "132\n",
      "133\n",
      "134\n",
      "135\n",
      "136\n",
      "137\n",
      "138\n",
      "139\n",
      "140\n",
      "141\n",
      "142\n",
      "143\n",
      "144\n",
      "145\n",
      "146\n",
      "147\n",
      "148\n",
      "149\n",
      "150\n",
      "151\n",
      "152\n",
      "153\n",
      "154\n",
      "155\n",
      "156\n",
      "157\n",
      "158\n",
      "159\n",
      "160\n",
      "161\n",
      "162\n",
      "163\n",
      "164\n",
      "165\n",
      "166\n",
      "167\n",
      "168\n",
      "169\n",
      "170\n",
      "171\n",
      "172\n",
      "173\n",
      "174\n",
      "175\n",
      "176\n",
      "177\n",
      "178\n",
      "179\n",
      "180\n",
      "181\n",
      "182\n",
      "183\n",
      "184\n",
      "185\n",
      "186\n",
      "187\n",
      "188\n",
      "189\n",
      "190\n",
      "191\n",
      "192\n",
      "193\n",
      "194\n",
      "195\n",
      "196\n",
      "197\n",
      "198\n",
      "199\n",
      "200\n",
      "201\n",
      "202\n",
      "203\n",
      "204\n",
      "205\n",
      "206\n",
      "207\n",
      "208\n",
      "209\n",
      "210\n",
      "211\n",
      "212\n",
      "213\n",
      "214\n",
      "215\n",
      "216\n",
      "217\n",
      "218\n",
      "219\n",
      "220\n",
      "221\n",
      "222\n",
      "223\n",
      "224\n",
      "225\n",
      "226\n",
      "227\n",
      "228\n",
      "229\n",
      "230\n",
      "231\n",
      "232\n",
      "233\n",
      "234\n",
      "235\n",
      "236\n",
      "237\n",
      "238\n",
      "239\n",
      "240\n",
      "241\n",
      "242\n",
      "243\n",
      "244\n",
      "245\n",
      "246\n",
      "247\n",
      "248\n",
      "249\n",
      "250\n",
      "251\n",
      "252\n",
      "253\n",
      "254\n",
      "255\n",
      "256\n",
      "257\n",
      "258\n",
      "259\n",
      "260\n",
      "261\n",
      "262\n",
      "263\n",
      "264\n",
      "265\n",
      "266\n",
      "1234567891011121314151617181920212223242526272829303132333435363738394041424344454647484950515253545556575859606162636465666768697071727374757677787980818283848586878889909192939495969798991001011021031041051061071081091101111121131141151161171181191201211221231241251261271281291301311321331341351361371381391401411421431441451461471481491501511521531541551561571581591601611621631641651661671681691701711721731741751761771781791801811821831841851861871881891901911921931941951961971981992002012022032042052062072082092102112122132142152162172182192202212222232242252262272282292302312322332342352362372382392402412422432442452462472482492502512522532542552562572582592602612622632642652662672682692702712722732742752762772782792802812822832842852862872882892902912922932942952962972982993003013023033043053063073083093103113123133143153163173183193203213223233243253263273283293303313323333343353363373383393403413423433443453463473483493503513523533543553563573583593603613623633643653663673683693703713723733743753763773783793803813823833843853863873883893903913923933943953963973983994004014024034044054064074084094104114124134144154164174184194204214224234244254264274284294304314324334344354364374384394404414424434444454464474484494504514524534544554564574584594604614624634644654664674684694704714724734744754764774784794804814824834844854864874884894904914924934944954964974984995005015025035045055065075085095105115125135145155165175185195205215225235245255265275285295305315325335345355365375385395405415425435445455465475485495505515525535545555565575585595605615625635645655665675685695705715725735745755765775785795805815825835845855865875885895905915925935945955965975985996006016026036046056066076086096106116126136146156166176186196206216226236246256266276286296306316326336346356366376386396406416426436446456466476486496506516526536546556566576586596606616626636646656666676686696706716726736746756766776786796806816826836846856866876886896906916926936946956966976986997007017027037047057067077087097107117127137147157167177187197207217227237247257267277287297307317327337347357367377387397407417427437447457467477487497507517527537547557567577587597607617627637647657667677687697707717727737747757767777787797807817827837847857867877887897907917927937947957967977987998008018028038048058068078088098108118128138148158168178188198208218228238248258268278288298308318328338348358368378388398408418428438448458468478488498508518528538548558568578588598608618628638648658668678688698708718728738748758768778788798808818828838848858868878888898908918928938948958968978988999009019029039049059069079089099109119129139149159169179189199209219229239249259269279289299309319329339349359369379389399409419429439449459469479489499509519529539549559569579589599609619629639649659669679689699709719729739749759769779789799809819829839849859869879889899909919929939949959969979989991000100110021003100410051006100710081009101010111012101310141015101610171018101910201021102210231024102510261027102810291030103110321033103410351036103710381039104010411042104310441045104610471048104910501051105210531054105510561057105810591060106110621063106410651066106710681069107010711072107310741075107610771078107910801081108210831084108510861087108810891090109110921093109410951096109710981099110011011102110311041105110611071108110911101111111211131114111511161117111811191120112111221123112411251126112711281129113011311132113311341135113611371138113911401141114211431144114511461147114811491150115111521153115411551156115711581159116011611162116311641165116611671168116911701171117211731174117511761177117811791180118111821183118411851186118711881189119011911192119311941195119611971198119912001201120212031204120512061207120812091210121112121213121412151216121712181219122012211222122312241225122612271228122912301231123212331234123512361237123812391240124112421243124412451246124712481249125012511252125312541255125612571258125912601261126212631264126512661267126812691270127112721273127412751276127712781279128012811282128312841285128612871288128912901291129212931294129512961297129812991300130113021303130413051306130713081309131013111312131313141315131613171318131913201321132213231324132513261327132813291330133113321333133413351336133713381339134013411342134313441345134613471348134913501351135213531354135513561357135813591360136113621363136413651366136713681369137013711372137313741375137613771378137913801381138213831384138513861387138813891390139113921393139413951396139713981399140014011402140314041405140614071408140914101411141214131414141514161417141814191420142114221423142414251426142714281429143014311432143314341435143614371438143914401441144214431444144514461447144814491450145114521453145414551456145714581459146014611462146314641465146614671468146914701471147214731474147514761477147814791480148114821483148414851486148714881489149014911492149314941495149614971498149915001501150215031504150515061507150815091510151115121513151415151516151715181519152015211522152315241525152615271528152915301531153215331534153515361537153815391540154115421543154415451546154715481549155015511552155315541555155615571558155915601561156215631564156515661567156815691570157115721573157415751576157715781579158015811582158315841585158615871588158915901591159215931594159515961597159815991600160116021603160416051606160716081609161016111612161316141615161616171618161916201621162216231624162516261627162816291630163116321633163416351636163716381639164016411642164316441645164616471648164916501651165216531654165516561657165816591660166116621663166416651666166716681669167016711672167316741675167616771678167916801681168216831684168516861687168816891690169116921693169416951696169716981699170017011702170317041705170617071708170917101711171217131714171517161717171817191720172117221723172417251726172717281729173017311732173317341735173617371738173917401741174217431744174517461747174817491750175117521753175417551756175717581759176017611762176317641765176617671768176917701771177217731774177517761777177817791780178117821783178417851786178717881789179017911792179317941795179617971798179918001801180218031804180518061807180818091810181118121813181418151816181718181819182018211822182318241825182618271828182918301831183218331834183518361837183818391840184118421843184418451846184718481849185018511852185318541855185618571858185918601861186218631864186518661867186818691870187118721873187418751876187718781879188018811882188318841885188618871888188918901891189218931894189518961897189818991900190119021903190419051906190719081909191019111912191319141915191619171918191919201921192219231924192519261927192819291930193119321933193419351936193719381939194019411942194319441945194619471948194919501951195219531954195519561957195819591960196119621963196419651966196719681969197019711972197319741975197619771978197919801981198219831984198519861987198819891990199119921993199419951996199719981999200020012002200320042005200620072008200920102011201220132014201520162017201820192020202120222023202420252026202720282029203020312032203320342035203620372038203920402041204220432044204520462047204820492050205120522053205420552056205720582059206020612062206320642065206620672068206920702071207220732074207520762077207820792080208120822083208420852086208720882089209020912092209320942095209620972098209921002101210221032104210521062107210821092110211121122113211421152116211721182119212021212122212321242125212621272128212921302131213221332134213521362137213821392140214121422143214421452146214721482149215021512152215321542155215621572158215921602161216221632164216521662167216821692170217121722173217421752176217721782179218021812182218321842185218621872188218921902191219221932194219521962197219821992200220122022203220422052206220722082209221022112212221322142215221622172218221922202221222222232224222522262227222822292230223122322233223422352236223722382239224022412242224322442245224622472248224922502251225222532254225522562257225822592260226122622263226422652266226722682269227022712272227322742275227622772278227922802281228222832284228522862287228822892290229122922293229422952296229722982299230023012302230323042305230623072308230923102311231223132314231523162317231823192320232123222323232423252326232723282329233023312332233323342335233623372338233923402341234223432344234523462347234823492350235123522353235423552356235723582359236023612362236323642365236623672368236923702371237223732374237523762377237823792380238123822383238423852386238723882389239023912392239323942395239623972398239924002401240224032404240524062407240824092410241124122413241424152416241724182419242024212422242324242425242624272428242924302431243224332434243524362437243824392440244124422443244424452446244724482449245024512452245324542455245624572458245924602461246224632464246524662467246824692470247124722473247424752476247724782479248024812482248324842485248624872488248924902491249224932494249524962497249824992500250125022503250425052506250725082509251025112512251325142515251625172518251925202521252225232524252525262527252825292530253125322533253425352536253725382539254025412542254325442545254625472548254925502551255225532554255525562557255825592560256125622563256425652566256725682569257025712572257325742575257625772578257925802581258225832584258525862587258825892590259125922593259425952596259725982599260026012602260326042605260626072608260926102611261226132614261526162617261826192620262126222623262426252626262726282629263026312632263326342635263626372638263926402641264226432644264526462647264826492650265126522653265426552656265726582659266026612662266326642665266626672668266926702671267226732674267526762677267826792680268126822683268426852686268726882689269026912692269326942695269626972698269927002701270227032704270527062707270827092710271127122713271427152716271727182719272027212722272327242725272627272728272927302731273227332734273527362737273827392740274127422743274427452746274727482749275027512752275327542755275627572758275927602761276227632764276527662767276827692770277127722773277427752776277727782779278027812782278327842785278627872788278927902791279227932794279527962797279827992800280128022803280428052806280728082809281028112812281328142815281628172818281928202821282228232824282528262827282828292830283128322833283428352836283728382839284028412842284328442845284628472848284928502851285228532854285528562857285828592860286128622863286428652866286728682869287028712872287328742875287628772878287928802881288228832884288528862887288828892890289128922893289428952896289728982899290029012902290329042905290629072908290929102911291229132914291529162917291829192920292129222923292429252926292729282929293029312932293329342935293629372938293929402941294229432944294529462947294829492950295129522953295429552956295729582959296029612962296329642965296629672968296929702971297229732974297529762977297829792980298129822983298429852986298729882989299029912992299329942995299629972998299930003001300230033004300530063007300830093010301130123013301430153016301730183019302030213022302330243025302630273028302930303031303230333034303530363037303830393040304130423043304430453046304730483049305030513052305330543055305630573058305930603061306230633064306530663067306830693070307130723073307430753076307730783079308030813082308330843085308630873088308930903091309230933094309530963097309830993100310131023103310431053106310731083109311031113112311331143115311631173118311931203121312231233124312531263127312831293130313131323133313431353136313731383139314031413142314331443145314631473148314931503151315231533154315531563157315831593160316131623163316431653166316731683169317031713172317331743175317631773178317931803181318231833184318531863187318831893190319131923193319431953196319731983199320032013202320332043205320632073208320932103211321232133214321532163217321832193220322132223223322432253226322732283229323032313232323332343235323632373238323932403241324232433244324532463247324832493250325132523253325432553256325732583259326032613262326332643265326632673268326932703271327232733274327532763277327832793280328132823283328432853286328732883289329032913292329332943295329632973298329933003301330233033304330533063307330833093310331133123313331433153316331733183319332033213322332333243325332633273328332933303331333233333334333533363337333833393340334133423343334433453346334733483349335033513352335333543355335633573358335933603361336233633364336533663367336833693370337133723373337433753376337733783379338033813382338333843385338633873388338933903391339233933394339533963397339833993400340134023403340434053406340734083409341034113412341334143415341634173418341934203421342234233424342534263427342834293430343134323433343434353436343734383439344034413442344334443445344634473448344934503451345234533454345534563457345834593460346134623463346434653466346734683469347034713472347334743475347634773478347934803481348234833484348534863487348834893490349134923493349434953496349734983499350035013502350335043505350635073508350935103511351235133514351535163517351835193520352135223523352435253526352735283529353035313532353335343535353635373538353935403541354235433544354535463547354835493550355135523553355435553556355735583559356035613562356335643565356635673568356935703571357235733574357535763577357835793580358135823583358435853586358735883589359035913592359335943595359635973598359936003601360236033604360536063607360836093610361136123613361436153616361736183619362036213622362336243625362636273628362936303631363236333634363536363637363836393640364136423643364436453646364736483649365036513652365336543655365636573658365936603661366236633664366536663667366836693670367136723673367436753676367736783679368036813682368336843685368636873688368936903691369236933694369536963697369836993700370137023703370437053706370737083709371037113712371337143715371637173718371937203721372237233724372537263727372837293730373137323733373437353736373737383739374037413742374337443745374637473748374937503751375237533754375537563757375837593760376137623763376437653766376737683769377037713772377337743775377637773778377937803781378237833784378537863787378837893790379137923793379437953796379737983799380038013802380338043805380638073808380938103811381238133814381538163817381838193820382138223823382438253826382738283829383038313832383338343835383638373838383938403841384238433844384538463847384838493850385138523853385438553856385738583859386038613862386338643865386638673868386938703871387238733874387538763877387838793880388138823883388438853886388738883889389038913892389338943895389638973898389939003901390239033904390539063907390839093910391139123913391439153916391739183919392039213922392339243925392639273928392939303931393239333934393539363937393839393940394139423943394439453946394739483949395039513952395339543955395639573958395939603961396239633964396539663967396839693970397139723973397439753976397739783979398039813982398339843985398639873988398939903991399239933994399539963997399839994000400140024003400440054006400740084009401040114012401340144015401640174018401940204021402240234024402540264027402840294030403140324033403440354036403740384039404040414042404340444045404640474048404940504051405240534054405540564057405840594060406140624063406440654066406740684069407040714072407340744075407640774078407940804081408240834084408540864087408840894090409140924093409440954096409740984099410041014102410341044105410641074108410941104111411241134114411541164117411841194120412141224123412441254126412741284129413041314132413341344135413641374138413941404141414241434144414541464147414841494150415141524153415441554156415741584159416041614162416341644165416641674168416941704171417241734174417541764177417841794180418141824183418441854186418741884189419041914192419341944195419641974198419942004201420242034204420542064207420842094210421142124213421442154216421742184219422042214222422342244225422642274228422942304231423242334234423542364237423842394240424142424243424442454246424742484249425042514252425342544255425642574258425942604261426242634264426542664267426842694270427142724273427442754276427742784279428042814282428342844285428642874288428942904291429242934294429542964297429842994300430143024303430443054306430743084309431043114312431343144315431643174318431943204321432243234324432543264327432843294330433143324333433443354336433743384339434043414342434343444345434643474348434943504351435243534354435543564357435843594360436143624363436443654366436743684369437043714372437343744375437643774378437943804381438243834384438543864387438843894390439143924393439443954396439743984399440044014402440344044405440644074408440944104411441244134414441544164417441844194420442144224423442444254426442744284429443044314432443344344435443644374438443944404441444244434444444544464447444844494450445144524453445444554456445744584459446044614462446344644465446644674468446944704471447244734474447544764477447844794480448144824483448444854486448744884489449044914492449344944495449644974498449945004501450245034504450545064507450845094510451145124513451445154516451745184519452045214522452345244525452645274528452945304531453245334534453545364537453845394540454145424543454445454546454745484549455045514552455345544555455645574558455945604561456245634564456545664567456845694570457145724573457445754576457745784579458045814582458345844585458645874588458945904591459245934594459545964597459845994600460146024603460446054606460746084609461046114612461346144615461646174618461946204621462246234624462546264627462846294630463146324633463446354636463746384639464046414642464346444645464646474648464946504651465246534654465546564657465846594660466146624663466446654666466746684669467046714672467346744675467646774678467946804681468246834684468546864687468846894690469146924693469446954696469746984699470047014702470347044705470647074708470947104711471247134714471547164717471847194720472147224723472447254726472747284729473047314732473347344735473647374738473947404741474247434744474547464747474847494750475147524753475447554756475747584759476047614762476347644765476647674768476947704771477247734774477547764777477847794780478147824783478447854786478747884789479047914792479347944795479647974798479948004801480248034804480548064807480848094810481148124813481448154816481748184819482048214822482348244825482648274828482948304831483248334834483548364837483848394840484148424843484448454846484748484849485048514852485348544855485648574858485948604861486248634864486548664867486848694870487148724873487448754876487748784879488048814882488348844885488648874888488948904891489248934894489548964897489848994900490149024903490449054906490749084909491049114912491349144915491649174918491949204921492249234924492549264927492849294930493149324933493449354936493749384939494049414942494349444945494649474948494949504951495249534954495549564957495849594960496149624963496449654966496749684969497049714972497349744975497649774978497949804981498249834984498549864987498849894990499149924993499449954996499749984999500050015002500350045005500650075008500950105011501250135014501550165017501850195020502150225023502450255026502750285029503050315032503350345035503650375038503950405041504250435044504550465047504850495050505150525053505450555056505750585059506050615062506350645065506650675068506950705071507250735074507550765077507850795080508150825083508450855086508750885089509050915092509350945095509650975098509951005101510251035104510551065107510851095110511151125113511451155116511751185119512051215122512351245125512651275128512951305131513251335134513551365137513851395140514151425143514451455146514751485149515051515152515351545155515651575158515951605161516251635164516551665167516851695170517151725173517451755176517751785179518051815182518351845185518651875188518951905191519251935194519551965197519851995200520152025203520452055206520752085209521052115212521352145215521652175218521952205221522252235224522552265227522852295230523152325233523452355236523752385239524052415242524352445245524652475248524952505251525252535254525552565257525852595260526152625263526452655266526752685269527052715272527352745275527652775278527952805281528252835284528552865287528852895290529152925293529452955296529752985299530053015302530353045305530653075308530953105311531253135314531553165317531853195320532153225323532453255326532753285329533053315332533353345335533653375338533953405341534253435344534553465347534853495350535153525353535453555356535753585359536053615362536353645365536653675368536953705371537253735374537553765377537853795380538153825383538453855386538753885389539053915392539353945395539653975398539954005401540254035404540554065407540854095410541154125413541454155416541754185419542054215422542354245425542654275428542954305431543254335434543554365437543854395440544154425443544454455446544754485449545054515452545354545455545654575458545954605461546254635464546554665467546854695470547154725473547454755476547754785479548054815482548354845485548654875488548954905491549254935494549554965497549854995500550155025503550455055506550755085509551055115512551355145515551655175518551955205521552255235524552555265527552855295530553155325533553455355536553755385539554055415542554355445545554655475548554955505551555255535554555555565557555855595560556155625563556455655566556755685569557055715572557355745575557655775578557955805581558255835584558555865587558855895590559155925593559455955596559755985599560056015602560356045605560656075608560956105611561256135614561556165617561856195620562156225623562456255626562756285629563056315632563356345635563656375638563956405641564256435644564556465647564856495650565156525653565456555656565756585659566056615662566356645665566656675668566956705671567256735674567556765677567856795680568156825683568456855686568756885689569056915692569356945695569656975698569957005701570257035704570557065707570857095710571157125713571457155716571757185719572057215722572357245725572657275728572957305731573257335734573557365737573857395740574157425743574457455746574757485749575057515752575357545755575657575758575957605761576257635764576557665767576857695770577157725773577457755776577757785779578057815782578357845785578657875788578957905791579257935794579557965797579857995800580158025803580458055806580758085809581058115812581358145815581658175818581958205821582258235824582558265827582858295830583158325833583458355836583758385839584058415842584358445845584658475848584958505851585258535854585558565857585858595860586158625863586458655866586758685869587058715872587358745875587658775878587958805881588258835884588558865887588858895890589158925893589458955896589758985899590059015902590359045905590659075908590959105911591259135914591559165917591859195920592159225923592459255926592759285929593059315932593359345935593659375938593959405941594259435944594559465947594859495950595159525953595459555956595759585959596059615962596359645965596659675968596959705971597259735974597559765977597859795980598159825983598459855986598759885989599059915992599359945995599659975998599960006001600260036004600560066007600860096010601160126013601460156016601760186019602060216022602360246025602660276028602960306031603260336034603560361\n",
      "2\n",
      "3\n",
      "4\n",
      "5\n",
      "6\n",
      "7\n",
      "8\n",
      "9\n",
      "10\n",
      "11\n",
      "12\n",
      "13\n",
      "14\n",
      "15\n",
      "16\n",
      "17\n",
      "18\n",
      "19\n",
      "20\n",
      "21\n",
      "22\n",
      "23\n",
      "24\n",
      "25\n",
      "26\n",
      "27\n",
      "28\n",
      "29\n",
      "30\n",
      "31\n",
      "32\n",
      "33\n",
      "34\n",
      "35\n",
      "36\n",
      "37\n",
      "38\n",
      "39\n",
      "40\n",
      "41\n",
      "42\n",
      "43\n",
      "44\n",
      "45\n",
      "46\n",
      "47\n",
      "48\n",
      "49\n",
      "50\n",
      "51\n",
      "52\n",
      "53\n",
      "54\n",
      "55\n",
      "56\n",
      "57\n",
      "58\n",
      "59\n",
      "60\n",
      "61\n",
      "62\n",
      "63\n",
      "64\n",
      "65\n",
      "66\n",
      "67\n",
      "68\n",
      "69\n",
      "70\n",
      "71\n",
      "72\n",
      "73\n",
      "74\n",
      "75\n",
      "76\n",
      "77\n",
      "78\n",
      "79\n",
      "80\n",
      "81\n",
      "82\n",
      "83\n",
      "84\n",
      "85\n",
      "86\n",
      "87\n",
      "88\n",
      "89\n",
      "90\n",
      "91\n",
      "92\n",
      "93\n",
      "94\n",
      "95\n",
      "96\n",
      "97\n",
      "98\n",
      "99\n",
      "100\n",
      "101\n",
      "102\n",
      "103\n",
      "104\n",
      "105\n",
      "106\n",
      "107\n",
      "108\n",
      "109\n",
      "110\n",
      "111\n",
      "112\n",
      "113\n",
      "114\n",
      "115\n",
      "116\n",
      "117\n",
      "118\n",
      "119\n",
      "120\n",
      "121\n",
      "122\n",
      "123\n",
      "124\n",
      "125\n",
      "126\n",
      "127\n",
      "128\n",
      "129\n",
      "130\n",
      "131\n",
      "132\n",
      "133\n",
      "134\n",
      "135\n",
      "136\n",
      "137\n",
      "138\n",
      "139\n",
      "140\n",
      "141\n",
      "142\n",
      "143\n",
      "144\n",
      "145\n",
      "146\n",
      "147\n",
      "148\n",
      "149\n",
      "150\n",
      "151\n",
      "152\n",
      "153\n",
      "154\n",
      "155\n",
      "156\n",
      "157\n",
      "158\n",
      "159\n",
      "160\n",
      "161\n",
      "162\n",
      "163\n",
      "164\n",
      "165\n",
      "166\n",
      "167\n",
      "168\n",
      "169\n",
      "170\n",
      "171\n",
      "172\n",
      "173\n",
      "174\n",
      "175\n",
      "176\n",
      "177\n",
      "178\n",
      "179\n",
      "180\n",
      "181\n",
      "182\n",
      "183\n",
      "184\n",
      "185\n",
      "186\n",
      "187\n",
      "188\n",
      "189\n",
      "190\n",
      "191\n",
      "192\n",
      "193\n",
      "194\n",
      "195\n",
      "196\n",
      "197\n",
      "198\n",
      "199\n",
      "200\n",
      "201\n",
      "202\n",
      "203\n",
      "204\n",
      "205\n",
      "206\n",
      "207\n",
      "208\n",
      "209\n",
      "210\n",
      "211\n",
      "212\n",
      "213\n",
      "214\n",
      "215\n",
      "216\n",
      "217\n",
      "218\n",
      "219\n",
      "220\n",
      "221\n",
      "222\n",
      "223\n",
      "224\n",
      "225\n",
      "226\n",
      "227\n",
      "228\n",
      "229\n",
      "230\n",
      "231\n",
      "232\n",
      "233\n",
      "234\n",
      "235\n",
      "236\n",
      "237\n",
      "238\n",
      "239\n",
      "240\n",
      "241\n",
      "242\n",
      "243\n",
      "244\n",
      "245\n",
      "246\n",
      "247\n",
      "248\n",
      "249\n",
      "250\n",
      "251\n",
      "252\n",
      "253\n",
      "254\n",
      "255\n",
      "256\n",
      "257\n",
      "258\n",
      "259\n",
      "260\n",
      "261\n",
      "262\n",
      "263\n",
      "264\n",
      "265\n",
      "266\n",
      "267\n",
      "268\n",
      "269\n",
      "270\n",
      "271\n",
      "272\n",
      "273\n",
      "274\n",
      "275\n",
      "276\n",
      "277\n",
      "278\n",
      "279\n",
      "280\n",
      "281\n",
      "282\n",
      "283\n",
      "284\n",
      "285\n",
      "286\n",
      "287\n",
      "288\n",
      "289\n",
      "290\n",
      "291\n",
      "292\n",
      "293\n",
      "294\n",
      "295\n",
      "296\n",
      "297\n",
      "298\n",
      "299\n",
      "300\n",
      "301\n",
      "302\n",
      "303\n",
      "304\n",
      "305\n",
      "306\n",
      "307\n",
      "308\n",
      "309\n",
      "310\n",
      "311\n",
      "312\n",
      "313\n",
      "314\n",
      "315\n",
      "316\n",
      "317\n",
      "318\n",
      "319\n",
      "320\n",
      "321\n",
      "322\n",
      "323\n",
      "324\n",
      "325\n",
      "326\n",
      "327\n",
      "328\n",
      "329\n",
      "330\n",
      "331\n",
      "332\n",
      "333\n",
      "334\n",
      "335\n",
      "336\n",
      "337\n",
      "338\n",
      "339\n",
      "340\n",
      "341\n",
      "342\n",
      "343\n",
      "344\n",
      "345\n",
      "346\n",
      "347\n",
      "348\n",
      "349\n",
      "350\n",
      "351\n",
      "352\n",
      "353\n",
      "354\n",
      "355\n",
      "356\n",
      "357\n",
      "358\n",
      "359\n",
      "360\n",
      "361\n",
      "362\n",
      "363\n",
      "364\n",
      "365\n",
      "366\n",
      "367\n",
      "368\n",
      "369\n",
      "370\n",
      "371\n",
      "372\n",
      "373\n",
      "374\n",
      "375\n",
      "376\n",
      "377\n",
      "378\n",
      "379\n",
      "380\n",
      "381\n",
      "382\n",
      "383\n",
      "384\n",
      "385\n",
      "386\n",
      "387\n",
      "388\n",
      "389\n",
      "390\n",
      "391\n",
      "392\n",
      "393\n",
      "394\n",
      "395\n",
      "396\n",
      "397\n",
      "398\n",
      "399\n",
      "400\n",
      "401\n",
      "402\n",
      "403\n",
      "404\n",
      "405\n",
      "406\n",
      "407\n",
      "408\n",
      "409\n",
      "410\n",
      "411\n",
      "412\n",
      "413\n",
      "414\n",
      "415\n",
      "416\n",
      "417\n",
      "418\n",
      "419\n",
      "420\n",
      "421\n",
      "422\n",
      "423\n",
      "424\n",
      "425\n",
      "426\n",
      "427\n",
      "428\n",
      "429\n",
      "430\n",
      "431\n",
      "432\n",
      "433\n",
      "434\n",
      "435\n",
      "436\n",
      "437\n",
      "438\n",
      "439\n",
      "440\n",
      "441\n",
      "442\n",
      "443\n",
      "444\n",
      "445\n",
      "446\n",
      "447\n",
      "448\n",
      "449\n",
      "450\n",
      "451\n",
      "452\n",
      "453\n",
      "454\n",
      "455\n",
      "456\n",
      "457\n",
      "458\n",
      "459\n",
      "460\n",
      "461\n",
      "462\n",
      "463\n",
      "464\n",
      "465\n",
      "466\n",
      "467\n",
      "468\n",
      "469\n",
      "470\n",
      "471\n",
      "472\n",
      "473\n",
      "474\n",
      "475\n",
      "476\n",
      "477\n",
      "478\n",
      "479\n",
      "480\n",
      "481\n",
      "482\n",
      "483\n",
      "484\n",
      "485\n",
      "486\n",
      "487\n",
      "488\n",
      "489\n",
      "490\n",
      "491\n",
      "492\n",
      "493\n",
      "494\n",
      "495\n",
      "496\n",
      "497\n",
      "498\n",
      "499\n",
      "500\n",
      "501\n",
      "502\n",
      "503\n",
      "504\n",
      "505\n",
      "506\n",
      "507\n",
      "508\n",
      "509\n",
      "510\n",
      "511\n",
      "512\n",
      "513\n",
      "514\n",
      "515\n",
      "516\n",
      "517\n",
      "518\n",
      "519\n",
      "520\n",
      "521\n",
      "522\n",
      "523\n",
      "524\n",
      "525\n",
      "526\n",
      "527\n",
      "528\n",
      "529\n",
      "530\n",
      "531\n",
      "532\n",
      "533\n",
      "534\n",
      "535\n",
      "536\n",
      "537\n",
      "538\n",
      "539\n",
      "540\n",
      "541\n",
      "542\n",
      "543\n",
      "544\n",
      "545\n",
      "546\n",
      "547\n",
      "548\n",
      "549\n",
      "550\n",
      "551\n",
      "552\n",
      "553\n",
      "554\n",
      "555\n",
      "556\n",
      "557\n",
      "558\n",
      "559\n",
      "560\n",
      "561\n",
      "562\n",
      "563\n",
      "564\n",
      "565\n",
      "566\n",
      "567\n",
      "568\n",
      "569\n",
      "570\n",
      "571\n",
      "572\n",
      "573\n",
      "574\n",
      "575\n",
      "576\n",
      "577\n",
      "578\n",
      "579\n",
      "580\n",
      "581\n",
      "582\n",
      "583\n",
      "584\n",
      "585\n",
      "586\n",
      "587\n",
      "588\n",
      "589\n",
      "590\n",
      "591\n",
      "592\n",
      "593\n",
      "594\n",
      "595\n",
      "596\n",
      "597\n",
      "598\n",
      "599\n",
      "600\n",
      "601\n",
      "602\n",
      "603\n",
      "604\n",
      "605\n",
      "606\n",
      "607\n",
      "608\n",
      "609\n",
      "610\n",
      "611\n",
      "612\n",
      "613\n",
      "614\n",
      "615\n",
      "616\n",
      "617\n",
      "618\n",
      "619\n",
      "620\n",
      "621\n",
      "622\n",
      "623\n",
      "624\n",
      "625\n",
      "626\n",
      "627\n",
      "628\n",
      "629\n",
      "630\n",
      "631\n",
      "632\n",
      "633\n",
      "634\n",
      "635\n",
      "636\n",
      "637\n",
      "638\n",
      "639\n",
      "640\n",
      "641\n",
      "642\n",
      "643\n",
      "644\n",
      "645\n",
      "646\n",
      "647\n",
      "648\n",
      "649\n",
      "650\n",
      "651\n",
      "652\n",
      "653\n",
      "654\n",
      "655\n",
      "656\n",
      "657\n",
      "658\n",
      "659\n",
      "660\n",
      "661\n",
      "662\n",
      "663\n",
      "664\n",
      "665\n",
      "666\n",
      "667\n",
      "668\n",
      "669\n",
      "670\n",
      "671\n",
      "672\n",
      "673\n",
      "674\n",
      "675\n",
      "676\n",
      "677\n",
      "678\n",
      "679\n",
      "680\n",
      "681\n",
      "682\n",
      "683\n",
      "684\n",
      "685\n",
      "686\n",
      "687\n",
      "688\n",
      "689\n",
      "690\n",
      "691\n",
      "692\n",
      "693\n",
      "694\n",
      "695\n",
      "696\n",
      "697\n",
      "698\n",
      "699\n",
      "700\n",
      "701\n",
      "702\n",
      "703\n",
      "704\n",
      "705\n",
      "706\n",
      "707\n",
      "708\n",
      "709\n",
      "710\n",
      "711\n",
      "712\n",
      "713\n",
      "714\n",
      "715\n",
      "716\n",
      "717\n",
      "718\n",
      "719\n",
      "720\n",
      "721\n",
      "722\n",
      "723\n",
      "724\n",
      "725\n",
      "726\n",
      "727\n",
      "728\n",
      "729\n",
      "730\n",
      "731\n",
      "732\n",
      "733\n",
      "734\n",
      "735\n",
      "736\n",
      "737\n",
      "738\n",
      "739\n",
      "740\n",
      "741\n",
      "742\n",
      "743\n",
      "744\n",
      "745\n",
      "746\n",
      "747\n",
      "748\n",
      "749\n",
      "750\n",
      "751\n",
      "752\n",
      "753\n",
      "754\n",
      "755\n",
      "756\n",
      "757\n",
      "758\n",
      "759\n",
      "760\n",
      "761\n",
      "762\n",
      "763\n",
      "764\n",
      "765\n",
      "766\n",
      "767\n",
      "768\n",
      "769\n",
      "770\n",
      "771\n",
      "772\n",
      "773\n",
      "774\n",
      "775\n",
      "776\n",
      "777\n",
      "778\n",
      "779\n",
      "780\n",
      "781\n",
      "782\n",
      "783\n",
      "784\n",
      "785\n",
      "786\n",
      "787\n",
      "788\n",
      "789\n",
      "790\n",
      "791\n",
      "792\n",
      "793\n",
      "794\n",
      "795\n",
      "796\n",
      "797\n",
      "798\n",
      "799\n",
      "800\n",
      "801\n",
      "802\n",
      "803\n",
      "804\n",
      "805\n",
      "806\n",
      "807\n",
      "808\n",
      "809\n",
      "810\n",
      "811\n",
      "812\n",
      "813\n",
      "814\n",
      "815\n",
      "816\n",
      "817\n",
      "818\n",
      "819\n",
      "820\n",
      "821\n",
      "822\n",
      "823\n",
      "824\n",
      "825\n",
      "826\n",
      "827\n",
      "828\n",
      "829\n",
      "830\n",
      "831\n",
      "832\n",
      "833\n",
      "834\n",
      "835\n",
      "836\n",
      "837\n",
      "838\n",
      "839\n",
      "840\n",
      "841\n",
      "842\n",
      "843\n",
      "844\n",
      "845\n",
      "846\n",
      "847\n",
      "848\n",
      "849\n",
      "850\n",
      "851\n",
      "852\n",
      "853\n",
      "854\n",
      "855\n",
      "856\n",
      "857\n",
      "858\n",
      "859\n",
      "860\n",
      "861\n",
      "862\n",
      "863\n",
      "864\n",
      "865\n",
      "866\n",
      "867\n",
      "868\n",
      "869\n",
      "870\n",
      "871\n",
      "872\n",
      "873\n",
      "874\n",
      "875\n",
      "876\n",
      "877\n",
      "878\n",
      "879\n",
      "880\n",
      "881\n",
      "882\n",
      "883\n",
      "884\n",
      "885\n",
      "886\n",
      "887\n",
      "888\n",
      "889\n",
      "890\n",
      "891\n",
      "892\n",
      "893\n",
      "894\n",
      "895\n",
      "896\n",
      "897\n",
      "898\n",
      "899\n",
      "900\n",
      "901\n",
      "902\n",
      "903\n",
      "904\n",
      "905\n",
      "906\n",
      "907\n",
      "908\n",
      "909\n",
      "910\n",
      "911\n",
      "912\n",
      "913\n",
      "914\n",
      "915\n",
      "916\n",
      "917\n",
      "918\n",
      "919\n",
      "920\n",
      "921\n",
      "922\n",
      "923\n",
      "924\n",
      "925\n",
      "926\n",
      "927\n",
      "928\n",
      "929\n",
      "930\n",
      "931\n",
      "932\n",
      "933\n",
      "934\n",
      "935\n",
      "936\n",
      "937\n",
      "938\n",
      "939\n",
      "940\n",
      "941\n",
      "942\n",
      "943\n",
      "944\n",
      "945\n",
      "946\n",
      "947\n",
      "948\n",
      "949\n",
      "950\n",
      "951\n",
      "952\n",
      "953\n",
      "954\n",
      "955\n",
      "956\n",
      "957\n",
      "958\n",
      "959\n",
      "960\n",
      "961\n",
      "962\n",
      "963\n",
      "964\n",
      "965\n",
      "966\n",
      "967\n",
      "968\n",
      "969\n",
      "970\n",
      "971\n",
      "972\n",
      "973\n",
      "974\n",
      "975\n",
      "976\n",
      "977\n",
      "978\n",
      "979\n",
      "980\n",
      "981\n",
      "982\n",
      "983\n",
      "984\n",
      "985\n",
      "986\n",
      "987\n",
      "988\n",
      "989\n",
      "990\n",
      "991\n",
      "992\n",
      "993\n",
      "994\n",
      "995\n",
      "996\n",
      "997\n",
      "998\n",
      "999\n",
      "1000\n",
      "1001\n",
      "1002\n",
      "1003\n",
      "1004\n",
      "1005\n",
      "1006\n",
      "1007\n",
      "1008\n",
      "1009\n",
      "1010\n",
      "1011\n",
      "1012\n",
      "1013\n",
      "1014\n",
      "1015\n",
      "1016\n",
      "1017\n",
      "1018\n",
      "1019\n",
      "1020\n",
      "1021\n",
      "1022\n",
      "1023\n",
      "1024\n",
      "1025\n",
      "1026\n",
      "1027\n",
      "1028\n",
      "1029\n",
      "1030\n",
      "1031\n",
      "1032\n",
      "1033\n",
      "1034\n",
      "1035\n",
      "1036\n",
      "1037\n",
      "1038\n",
      "1039\n",
      "1040\n",
      "1041\n",
      "1042\n",
      "1043\n",
      "1044\n",
      "1045\n",
      "1046\n",
      "1047\n",
      "1048\n",
      "1049\n",
      "1050\n",
      "1051\n",
      "1052\n",
      "1053\n",
      "1054\n",
      "1055\n",
      "1056\n",
      "1057\n",
      "1058\n",
      "1059\n",
      "1060\n",
      "1061\n",
      "1062\n",
      "1063\n",
      "1064\n",
      "1065\n",
      "1066\n",
      "1067\n",
      "1068\n",
      "1069\n",
      "1070\n",
      "1071\n",
      "1072\n",
      "1073\n",
      "1074\n",
      "1075\n",
      "1076\n",
      "1077\n",
      "1078\n",
      "1079\n",
      "1080\n",
      "1081\n",
      "1082\n",
      "1083\n",
      "1084\n",
      "1085\n",
      "1086\n",
      "1087\n",
      "1088\n",
      "1089\n",
      "1090\n",
      "1091\n",
      "1092\n",
      "1093\n",
      "1094\n",
      "1095\n",
      "1096\n",
      "1097\n",
      "1098\n",
      "1099\n",
      "1100\n",
      "1101\n",
      "1102\n",
      "1103\n",
      "1104\n",
      "1105\n",
      "1106\n",
      "1107\n",
      "1108\n",
      "1109\n",
      "1110\n",
      "1111\n",
      "1112\n",
      "1113\n",
      "1114\n",
      "1115\n",
      "1116\n",
      "1117\n",
      "1118\n",
      "1119\n",
      "1120\n",
      "1121\n",
      "1122\n",
      "1123\n",
      "1124\n",
      "1125\n",
      "1126\n",
      "1127\n",
      "1128\n",
      "1129\n",
      "1130\n",
      "1131\n",
      "1132\n",
      "1133\n",
      "1134\n",
      "1135\n",
      "1136\n",
      "1137\n",
      "1138\n",
      "1139\n",
      "1140\n",
      "1141\n",
      "1142\n",
      "1143\n",
      "1144\n",
      "1145\n",
      "1146\n",
      "1147\n",
      "1148\n",
      "1149\n",
      "1150\n",
      "1151\n",
      "1152\n",
      "1153\n",
      "1154\n",
      "1155\n",
      "1156\n",
      "1157\n",
      "1158\n",
      "1159\n",
      "1160\n",
      "1161\n",
      "1162\n",
      "1163\n",
      "1164\n",
      "1165\n",
      "1166\n",
      "1167\n",
      "1168\n",
      "1169\n",
      "1170\n",
      "1171\n",
      "1172\n",
      "1173\n",
      "1174\n",
      "1175\n",
      "1176\n",
      "1177\n",
      "1178\n",
      "1179\n",
      "1180\n",
      "1181\n",
      "1182\n",
      "1183\n",
      "1184\n",
      "1185\n",
      "1186\n",
      "1187\n",
      "1188\n",
      "1189\n",
      "1190\n",
      "1191\n",
      "1192\n",
      "1193\n",
      "1194\n",
      "1195\n",
      "1196\n",
      "1197\n",
      "1198\n",
      "1199\n",
      "1200\n",
      "1201\n",
      "1202\n",
      "1203\n",
      "1204\n",
      "1205\n",
      "1206\n",
      "1207\n",
      "1208\n",
      "1209\n",
      "1210\n",
      "1211\n",
      "1212\n",
      "1213\n",
      "1214\n",
      "1215\n",
      "1216\n",
      "1217\n",
      "1218\n",
      "1219\n",
      "1220\n",
      "1221\n",
      "1222\n",
      "1223\n",
      "1224\n",
      "1225\n",
      "1226\n",
      "1227\n",
      "1228\n",
      "1229\n",
      "1230\n",
      "1231\n",
      "1232\n",
      "1233\n",
      "1234\n",
      "1235\n",
      "1236\n",
      "1237\n",
      "1238\n",
      "1239\n",
      "1240\n",
      "1241\n",
      "1242\n",
      "1243\n",
      "1244\n",
      "1245\n",
      "1246\n",
      "1247\n",
      "1248\n",
      "1249\n",
      "1250\n",
      "1251\n",
      "1252\n",
      "1253\n",
      "1254\n",
      "1255\n",
      "1256\n",
      "1257\n",
      "1258\n",
      "1259\n",
      "1260\n",
      "1261\n",
      "1262\n",
      "1263\n",
      "1264\n",
      "1265\n",
      "1266\n",
      "1267\n",
      "1268\n",
      "1269\n",
      "1270\n",
      "1271\n",
      "1272\n",
      "1273\n",
      "1274\n",
      "1275\n",
      "1276\n",
      "1277\n",
      "1278\n",
      "1279\n",
      "1280\n",
      "1281\n",
      "1282\n",
      "1283\n",
      "1284\n",
      "1285\n",
      "1286\n",
      "1287\n",
      "1288\n",
      "1289\n",
      "1290\n",
      "1291\n",
      "1292\n",
      "1293\n",
      "1294\n",
      "1295\n",
      "1296\n",
      "1297\n",
      "1298\n",
      "1299\n",
      "1300\n",
      "1301\n",
      "1302\n",
      "1303\n",
      "1304\n",
      "1305\n",
      "1306\n",
      "1307\n",
      "1308\n",
      "1309\n",
      "1310\n",
      "1311\n",
      "1312\n",
      "1313\n",
      "1314\n",
      "1315\n",
      "1316\n",
      "1317\n",
      "1318\n",
      "1319\n",
      "1320\n",
      "1321\n",
      "1322\n",
      "1323\n",
      "1324\n",
      "1325\n",
      "1326\n",
      "1327\n",
      "1328\n",
      "1329\n",
      "1330\n",
      "1331\n",
      "1332\n",
      "1333\n",
      "1334\n",
      "1335\n",
      "1336\n",
      "1337\n",
      "1338\n",
      "1339\n",
      "1340\n",
      "1341\n",
      "1342\n",
      "1343\n",
      "1344\n",
      "1345\n",
      "1346\n",
      "1347\n",
      "1348\n",
      "1349\n",
      "1350\n",
      "1351\n",
      "1352\n",
      "1353\n",
      "1354\n",
      "1355\n",
      "1356\n",
      "1357\n",
      "1358\n",
      "1359\n",
      "1360\n",
      "1361\n",
      "1362\n",
      "1363\n",
      "1364\n",
      "1365\n",
      "1366\n",
      "1367\n",
      "1368\n",
      "1369\n",
      "1370\n",
      "1371\n",
      "1372\n",
      "1373\n",
      "1374\n",
      "1375\n",
      "1376\n",
      "1377\n",
      "1378\n",
      "1379\n",
      "1380\n",
      "1381\n",
      "1382\n",
      "1383\n",
      "1384\n",
      "1385\n",
      "1386\n",
      "1387\n",
      "1388\n",
      "1389\n",
      "1390\n",
      "1391\n",
      "1392\n",
      "1393\n",
      "1394\n",
      "1395\n",
      "1396\n",
      "1397\n",
      "1398\n",
      "1399\n",
      "1400\n",
      "1401\n",
      "1402\n",
      "1403\n",
      "1404\n",
      "1405\n",
      "1406\n",
      "1407\n",
      "1408\n",
      "1409\n",
      "1410\n",
      "1411\n",
      "1412\n",
      "1413\n",
      "1414\n",
      "1415\n",
      "1416\n",
      "1417\n",
      "1418\n",
      "1419\n",
      "1420\n",
      "1421\n",
      "1422\n",
      "1423\n",
      "1424\n",
      "1425\n",
      "1426\n",
      "1427\n",
      "1428\n",
      "1429\n",
      "1430\n",
      "1431\n",
      "1432\n",
      "1433\n",
      "1434\n",
      "1435\n",
      "1436\n",
      "1437\n",
      "1438\n",
      "1439\n",
      "1440\n",
      "1441\n",
      "1442\n",
      "1443\n",
      "1444\n",
      "1445\n",
      "1446\n",
      "1447\n",
      "1448\n",
      "1449\n",
      "1450\n",
      "1451\n",
      "1452\n",
      "1453\n",
      "1454\n",
      "1455\n",
      "1456\n",
      "1457\n",
      "1458\n",
      "1459\n",
      "1460\n",
      "1461\n",
      "1462\n",
      "1463\n",
      "1464\n",
      "1465\n",
      "1466\n",
      "1467\n",
      "1468\n",
      "1469\n",
      "1470\n",
      "1471\n",
      "1472\n",
      "1473\n",
      "1474\n",
      "1475\n",
      "1476\n",
      "1477\n",
      "1478\n",
      "1479\n",
      "1480\n",
      "1481\n",
      "1482\n",
      "1483\n",
      "1484\n",
      "1485\n",
      "1486\n",
      "1487\n",
      "1488\n",
      "1489\n",
      "1490\n",
      "1491\n",
      "1492\n",
      "1493\n",
      "1494\n",
      "1495\n",
      "1496\n",
      "1497\n",
      "1498\n",
      "1499\n",
      "1500\n",
      "1501\n",
      "1502\n",
      "1503\n",
      "1504\n",
      "1505\n",
      "1506\n",
      "1507\n",
      "1508\n",
      "1509\n",
      "1510\n",
      "1511\n",
      "1512\n",
      "1513\n",
      "1514\n",
      "1515\n",
      "1516\n",
      "1517\n",
      "1518\n",
      "1519\n",
      "1520\n",
      "1521\n",
      "1522\n",
      "1523\n",
      "1524\n",
      "1525\n",
      "1526\n",
      "1527\n",
      "1528\n",
      "1529\n",
      "1530\n",
      "1531\n",
      "1532\n",
      "1533\n",
      "1534\n",
      "1535\n",
      "1536\n",
      "1537\n",
      "1538\n",
      "1539\n",
      "1540\n",
      "1541\n",
      "1542\n",
      "1543\n",
      "1544\n",
      "1545\n",
      "1546\n",
      "1547\n",
      "1548\n",
      "1549\n",
      "1550\n",
      "1551\n",
      "1552\n",
      "1553\n",
      "1554\n",
      "1555\n",
      "1556\n",
      "1557\n",
      "1558\n",
      "1559\n",
      "1560\n",
      "1561\n",
      "1562\n",
      "1563\n",
      "1564\n",
      "1565\n",
      "1566\n",
      "1567\n",
      "1568\n",
      "1569\n",
      "1570\n",
      "1571\n",
      "1572\n",
      "1573\n",
      "1574\n",
      "1575\n",
      "1576\n",
      "1577\n",
      "1578\n",
      "1579\n",
      "1580\n",
      "1581\n",
      "1582\n",
      "1583\n",
      "1584\n",
      "1585\n",
      "1586\n",
      "1587\n",
      "1588\n",
      "1589\n",
      "1590\n",
      "1591\n",
      "1592\n",
      "1593\n",
      "1594\n",
      "1595\n",
      "1596\n",
      "1597\n",
      "1598\n",
      "1599\n",
      "1600\n",
      "1601\n",
      "1602\n",
      "1603\n",
      "1604\n",
      "1605\n",
      "1606\n",
      "1607\n",
      "1608\n",
      "1609\n",
      "1610\n",
      "1611\n",
      "1612\n",
      "1613\n",
      "1614\n",
      "1615\n",
      "1616\n",
      "1617\n",
      "1618\n",
      "1619\n",
      "1620\n",
      "1621\n",
      "1622\n",
      "1623\n",
      "1624\n",
      "1625\n",
      "1626\n",
      "1627\n",
      "1628\n",
      "1629\n",
      "1630\n",
      "1631\n",
      "1632\n",
      "1633\n",
      "1634\n",
      "1635\n",
      "1636\n",
      "1637\n",
      "1638\n",
      "1639\n",
      "1640\n",
      "1641\n",
      "1642\n",
      "1643\n",
      "1644\n",
      "1645\n",
      "1646\n",
      "1647\n",
      "1648\n",
      "1649\n",
      "1650\n",
      "1651\n",
      "1652\n",
      "1653\n",
      "1654\n",
      "1655\n",
      "1656\n",
      "1657\n",
      "1658\n",
      "1659\n",
      "1660\n",
      "1661\n",
      "1662\n",
      "1663\n",
      "1664\n",
      "1665\n",
      "1666\n",
      "1667\n",
      "1668\n",
      "1669\n",
      "1670\n",
      "1671\n",
      "1672\n",
      "1673\n",
      "1674\n",
      "1675\n",
      "1676\n",
      "1677\n",
      "1678\n",
      "1679\n",
      "1680\n",
      "1681\n",
      "1682\n",
      "1683\n",
      "1684\n",
      "1685\n",
      "1686\n",
      "1687\n",
      "1688\n",
      "1689\n",
      "1690\n",
      "1691\n",
      "1692\n",
      "1693\n",
      "1694\n",
      "1695\n",
      "1696\n",
      "1697\n",
      "1698\n",
      "1699\n",
      "1700\n",
      "1701\n",
      "1702\n",
      "1703\n",
      "1704\n",
      "1705\n",
      "1706\n",
      "1707\n",
      "1708\n",
      "1709\n",
      "1710\n",
      "1711\n",
      "1712\n",
      "1713\n",
      "1714\n",
      "1715\n",
      "1716\n",
      "1717\n",
      "1718\n",
      "1719\n",
      "1720\n",
      "1721\n",
      "1722\n",
      "1723\n",
      "1724\n",
      "1725\n",
      "1726\n",
      "1727\n",
      "1728\n",
      "1729\n",
      "1730\n",
      "1731\n",
      "1732\n",
      "1733\n",
      "1734\n",
      "1735\n",
      "1736\n",
      "1737\n",
      "1738\n",
      "1739\n",
      "1740\n",
      "1741\n",
      "1742\n",
      "1743\n",
      "1744\n",
      "1745\n",
      "1746\n",
      "1747\n",
      "1748\n",
      "1749\n",
      "1750\n",
      "1751\n",
      "1752\n",
      "1753\n",
      "1754\n",
      "1755\n",
      "1756\n",
      "1757\n",
      "1758\n",
      "1759\n",
      "1760\n",
      "1761\n",
      "1762\n",
      "1763\n",
      "1764\n",
      "1765\n",
      "1766\n",
      "1767\n",
      "1768\n",
      "1769\n",
      "1770\n",
      "1771\n",
      "1772\n",
      "1773\n",
      "1774\n",
      "1775\n",
      "1776\n",
      "1777\n",
      "1778\n",
      "1779\n",
      "1780\n",
      "1781\n",
      "1782\n",
      "1783\n",
      "1784\n",
      "1785\n",
      "1786\n",
      "1787\n",
      "1788\n",
      "1789\n",
      "1790\n",
      "1791\n",
      "1792\n",
      "1793\n",
      "1794\n",
      "1795\n",
      "1796\n",
      "1797\n",
      "1798\n",
      "1799\n",
      "1800\n",
      "1801\n",
      "1802\n",
      "1803\n",
      "1804\n",
      "1805\n",
      "1806\n",
      "1807\n",
      "1808\n",
      "1809\n",
      "1810\n",
      "1811\n",
      "1812\n",
      "1813\n",
      "1814\n",
      "1815\n",
      "1816\n",
      "1817\n",
      "1818\n",
      "1819\n",
      "1820\n",
      "1821\n",
      "1822\n",
      "1823\n",
      "1824\n",
      "1825\n",
      "1826\n",
      "1827\n",
      "1828\n",
      "1829\n",
      "1830\n",
      "1831\n",
      "1832\n",
      "1833\n",
      "1834\n",
      "1835\n",
      "1836\n",
      "1837\n",
      "1838\n",
      "1839\n",
      "1840\n",
      "1841\n",
      "1842\n",
      "1843\n",
      "1844\n",
      "1845\n",
      "1846\n",
      "1847\n",
      "1848\n",
      "1849\n",
      "1850\n",
      "1851\n",
      "1852\n",
      "1853\n",
      "1854\n",
      "1855\n",
      "1856\n",
      "1857\n",
      "1858\n",
      "1859\n",
      "1860\n",
      "1861\n",
      "1862\n",
      "1863\n",
      "1864\n",
      "1865\n",
      "1866\n",
      "1867\n",
      "1868\n",
      "1869\n",
      "1870\n",
      "1871\n",
      "1872\n",
      "1873\n",
      "1874\n",
      "1875\n",
      "1876\n",
      "1877\n",
      "1878\n",
      "1879\n",
      "1880\n",
      "1881\n",
      "1882\n",
      "1883\n",
      "1884\n",
      "1885\n",
      "1886\n",
      "1887\n",
      "1888\n",
      "1889\n",
      "1890\n",
      "1891\n",
      "1892\n",
      "1893\n",
      "1894\n",
      "1895\n",
      "1896\n",
      "1897\n",
      "1898\n",
      "1899\n",
      "1900\n",
      "1901\n",
      "1902\n",
      "1903\n",
      "1904\n",
      "1905\n",
      "1906\n",
      "1907\n",
      "1908\n",
      "1909\n",
      "1910\n",
      "1911\n",
      "1912\n",
      "1913\n",
      "1914\n",
      "1915\n",
      "1916\n",
      "1917\n",
      "1918\n",
      "1919\n",
      "1920\n",
      "1921\n",
      "1922\n",
      "1923\n",
      "1924\n",
      "1925\n",
      "1926\n",
      "1927\n",
      "1928\n",
      "1929\n",
      "1930\n",
      "1931\n",
      "1932\n",
      "1933\n",
      "1934\n",
      "1935\n",
      "1936\n",
      "1937\n",
      "1938\n",
      "1939\n",
      "1940\n",
      "1941\n",
      "1942\n",
      "1943\n",
      "1944\n",
      "1945\n",
      "1946\n",
      "1947\n",
      "1948\n",
      "1949\n",
      "1950\n",
      "1951\n",
      "1952\n",
      "1953\n",
      "1954\n",
      "1955\n",
      "1956\n",
      "1957\n",
      "1958\n",
      "1959\n",
      "1960\n",
      "1961\n",
      "1962\n",
      "1963\n",
      "1964\n",
      "1965\n",
      "1966\n",
      "1967\n",
      "1968\n",
      "1969\n",
      "1970\n",
      "1971\n",
      "1972\n",
      "1973\n",
      "1974\n",
      "1975\n",
      "1976\n",
      "1977\n",
      "1978\n",
      "1979\n",
      "1980\n",
      "1981\n",
      "1982\n",
      "1983\n",
      "1984\n",
      "1985\n",
      "1986\n",
      "1987\n",
      "1988\n",
      "1989\n",
      "1990\n",
      "1991\n",
      "1992\n",
      "1993\n",
      "1994\n",
      "1995\n",
      "1996\n",
      "1997\n",
      "1998\n",
      "1999\n",
      "2000\n",
      "2001\n",
      "2002\n",
      "2003\n",
      "2004\n",
      "2005\n",
      "2006\n",
      "2007\n",
      "2008\n",
      "2009\n",
      "2010\n",
      "2011\n",
      "2012\n",
      "2013\n",
      "2014\n",
      "2015\n",
      "2016\n",
      "2017\n",
      "2018\n",
      "2019\n",
      "2020\n",
      "2021\n",
      "2022\n",
      "2023\n",
      "2024\n",
      "2025\n",
      "2026\n",
      "2027\n",
      "2028\n",
      "2029\n",
      "2030\n",
      "2031\n",
      "2032\n",
      "2033\n",
      "2034\n",
      "2035\n",
      "2036\n",
      "2037\n",
      "2038\n",
      "2039\n",
      "2040\n",
      "2041\n",
      "2042\n",
      "2043\n",
      "2044\n",
      "2045\n",
      "2046\n",
      "2047\n",
      "2048\n",
      "2049\n",
      "2050\n",
      "2051\n",
      "2052\n",
      "2053\n",
      "2054\n",
      "2055\n",
      "2056\n",
      "2057\n",
      "2058\n",
      "2059\n",
      "2060\n",
      "2061\n",
      "2062\n",
      "2063\n",
      "2064\n",
      "2065\n",
      "2066\n",
      "2067\n",
      "2068\n",
      "2069\n",
      "2070\n",
      "2071\n",
      "2072\n",
      "2073\n",
      "2074\n",
      "2075\n",
      "2076\n",
      "2077\n",
      "2078\n",
      "2079\n",
      "2080\n",
      "2081\n",
      "2082\n",
      "2083\n",
      "2084\n",
      "2085\n",
      "2086\n",
      "2087\n",
      "2088\n",
      "2089\n",
      "2090\n",
      "2091\n",
      "2092\n",
      "2093\n",
      "2094\n",
      "2095\n",
      "2096\n",
      "2097\n",
      "2098\n",
      "2099\n",
      "2100\n",
      "2101\n",
      "2102\n",
      "2103\n",
      "2104\n",
      "2105\n",
      "2106\n",
      "2107\n",
      "2108\n",
      "2109\n",
      "2110\n",
      "2111\n",
      "2112\n",
      "2113\n",
      "2114\n",
      "2115\n",
      "2116\n",
      "2117\n",
      "2118\n",
      "2119\n",
      "2120\n",
      "2121\n",
      "2122\n",
      "2123\n",
      "2124\n",
      "2125\n",
      "2126\n",
      "2127\n",
      "2128\n",
      "2129\n",
      "2130\n",
      "2131\n",
      "2132\n",
      "2133\n",
      "2134\n",
      "2135\n",
      "2136\n",
      "2137\n",
      "2138\n",
      "2139\n",
      "2140\n",
      "2141\n",
      "2142\n",
      "2143\n",
      "2144\n",
      "2145\n",
      "2146\n",
      "2147\n",
      "2148\n",
      "2149\n",
      "2150\n",
      "2151\n",
      "2152\n",
      "2153\n",
      "2154\n",
      "2155\n",
      "2156\n",
      "2157\n",
      "2158\n",
      "2159\n",
      "2160\n",
      "2161\n",
      "2162\n",
      "2163\n",
      "2164\n",
      "2165\n",
      "2166\n",
      "2167\n",
      "2168\n",
      "2169\n",
      "2170\n",
      "2171\n",
      "2172\n",
      "2173\n",
      "2174\n",
      "2175\n",
      "2176\n",
      "2177\n",
      "2178\n",
      "2179\n",
      "2180\n",
      "2181\n",
      "2182\n",
      "2183\n",
      "2184\n",
      "2185\n",
      "2186\n",
      "2187\n",
      "2188\n",
      "2189\n",
      "2190\n",
      "2191\n",
      "2192\n",
      "2193\n",
      "2194\n",
      "2195\n",
      "2196\n",
      "2197\n",
      "2198\n",
      "2199\n",
      "2200\n",
      "2201\n",
      "2202\n",
      "2203\n",
      "2204\n",
      "2205\n",
      "2206\n",
      "2207\n",
      "2208\n",
      "2209\n",
      "2210\n",
      "2211\n",
      "2212\n",
      "2213\n",
      "2214\n",
      "2215\n",
      "2216\n",
      "2217\n",
      "2218\n",
      "2219\n",
      "2220\n",
      "2221\n",
      "2222\n",
      "2223\n",
      "2224\n",
      "2225\n",
      "2226\n",
      "2227\n",
      "2228\n",
      "2229\n",
      "2230\n",
      "2231\n",
      "2232\n",
      "2233\n",
      "2234\n",
      "2235\n",
      "2236\n",
      "2237\n",
      "2238\n",
      "2239\n",
      "2240\n",
      "2241\n",
      "2242\n",
      "2243\n",
      "2244\n",
      "2245\n",
      "2246\n",
      "2247\n",
      "2248\n",
      "2249\n",
      "2250\n",
      "2251\n",
      "2252\n",
      "2253\n",
      "2254\n",
      "2255\n",
      "2256\n",
      "2257\n",
      "2258\n",
      "2259\n",
      "2260\n",
      "2261\n",
      "2262\n",
      "2263\n",
      "2264\n",
      "2265\n",
      "2266\n",
      "2267\n",
      "2268\n",
      "2269\n",
      "2270\n",
      "2271\n",
      "2272\n",
      "2273\n",
      "2274\n",
      "2275\n",
      "2276\n",
      "2277\n",
      "2278\n",
      "2279\n",
      "2280\n",
      "2281\n",
      "2282\n",
      "2283\n",
      "2284\n",
      "2285\n",
      "2286\n",
      "2287\n",
      "2288\n",
      "2289\n",
      "2290\n",
      "2291\n",
      "2292\n",
      "2293\n",
      "2294\n",
      "2295\n",
      "2296\n",
      "2297\n",
      "2298\n",
      "2299\n",
      "2300\n",
      "2301\n",
      "2302\n",
      "2303\n",
      "2304\n",
      "2305\n",
      "2306\n",
      "2307\n",
      "2308\n",
      "2309\n",
      "2310\n",
      "2311\n",
      "2312\n",
      "2313\n",
      "2314\n",
      "2315\n",
      "2316\n",
      "2317\n",
      "2318\n",
      "2319\n",
      "2320\n",
      "2321\n",
      "2322\n",
      "2323\n",
      "2324\n",
      "2325\n",
      "2326\n",
      "2327\n",
      "2328\n",
      "2329\n",
      "2330\n",
      "2331\n",
      "2332\n",
      "2333\n",
      "2334\n",
      "2335\n",
      "2336\n",
      "2337\n",
      "2338\n",
      "2339\n",
      "2340\n",
      "2341\n",
      "2342\n",
      "2343\n",
      "2344\n",
      "2345\n",
      "2346\n",
      "2347\n",
      "2348\n",
      "2349\n",
      "2350\n",
      "2351\n",
      "2352\n",
      "2353\n",
      "2354\n",
      "2355\n",
      "2356\n",
      "2357\n",
      "2358\n",
      "2359\n",
      "2360\n",
      "2361\n",
      "2362\n",
      "2363\n",
      "2364\n",
      "2365\n",
      "2366\n",
      "2367\n",
      "2368\n",
      "2369\n",
      "2370\n",
      "2371\n",
      "2372\n",
      "2373\n",
      "2374\n",
      "2375\n",
      "2376\n",
      "2377\n",
      "2378\n",
      "2379\n",
      "2380\n",
      "2381\n",
      "2382\n",
      "2383\n",
      "2384\n",
      "2385\n",
      "2386\n",
      "2387\n",
      "2388\n",
      "2389\n",
      "2390\n",
      "2391\n",
      "2392\n",
      "2393\n",
      "2394\n",
      "2395\n",
      "2396\n",
      "2397\n",
      "2398\n",
      "2399\n",
      "2400\n",
      "2401\n",
      "2402\n",
      "2403\n",
      "2404\n",
      "2405\n",
      "2406\n",
      "2407\n",
      "2408\n",
      "2409\n",
      "2410\n",
      "2411\n",
      "2412\n",
      "2413\n",
      "2414\n",
      "2415\n",
      "2416\n",
      "2417\n",
      "2418\n",
      "2419\n",
      "2420\n",
      "2421\n",
      "2422\n",
      "2423\n",
      "2424\n",
      "2425\n",
      "2426\n",
      "2427\n",
      "2428\n",
      "2429\n",
      "2430\n",
      "2431\n",
      "2432\n",
      "2433\n",
      "2434\n",
      "2435\n",
      "2436\n",
      "2437\n",
      "2438\n",
      "2439\n",
      "2440\n",
      "2441\n",
      "2442\n",
      "2443\n",
      "2444\n",
      "2445\n",
      "2446\n",
      "2447\n",
      "2448\n",
      "2449\n",
      "2450\n",
      "2451\n",
      "2452\n",
      "2453\n",
      "2454\n",
      "2455\n",
      "2456\n",
      "2457\n",
      "2458\n",
      "2459\n",
      "2460\n",
      "2461\n",
      "2462\n",
      "2463\n",
      "2464\n",
      "2465\n",
      "2466\n",
      "2467\n",
      "2468\n",
      "2469\n",
      "2470\n",
      "2471\n",
      "2472\n",
      "2473\n",
      "2474\n",
      "2475\n",
      "2476\n",
      "2477\n",
      "2478\n",
      "2479\n",
      "2480\n",
      "2481\n",
      "2482\n",
      "2483\n",
      "2484\n",
      "2485\n",
      "2486\n",
      "2487\n",
      "2488\n",
      "2489\n",
      "2490\n",
      "2491\n",
      "2492\n",
      "2493\n",
      "2494\n",
      "2495\n",
      "2496\n",
      "2497\n",
      "2498\n",
      "2499\n",
      "2500\n",
      "2501\n",
      "2502\n",
      "2503\n",
      "2504\n",
      "2505\n",
      "2506\n",
      "2507\n",
      "2508\n",
      "2509\n",
      "2510\n",
      "2511\n",
      "2512\n",
      "2513\n",
      "2514\n",
      "2515\n",
      "2516\n",
      "2517\n",
      "2518\n",
      "2519\n",
      "2520\n",
      "2521\n",
      "2522\n",
      "2523\n",
      "2524\n",
      "2525\n",
      "2526\n",
      "2527\n",
      "2528\n",
      "2529\n",
      "2530\n",
      "2531\n",
      "2532\n",
      "2533\n",
      "2534\n",
      "2535\n",
      "2536\n",
      "2537\n",
      "2538\n",
      "2539\n",
      "2540\n",
      "2541\n",
      "2542\n",
      "2543\n",
      "2544\n",
      "2545\n",
      "2546\n",
      "2547\n",
      "2548\n",
      "2549\n",
      "2550\n",
      "2551\n",
      "2552\n",
      "2553\n",
      "2554\n",
      "2555\n",
      "2556\n",
      "2557\n",
      "2558\n",
      "2559\n",
      "2560\n",
      "2561\n",
      "2562\n",
      "2563\n",
      "2564\n",
      "2565\n",
      "2566\n",
      "2567\n",
      "2568\n",
      "2569\n",
      "2570\n",
      "2571\n",
      "2572\n",
      "2573\n",
      "2574\n",
      "2575\n",
      "2576\n",
      "2577\n",
      "2578\n",
      "2579\n",
      "2580\n",
      "2581\n",
      "2582\n",
      "2583\n",
      "2584\n",
      "2585\n",
      "2586\n",
      "2587\n",
      "2588\n",
      "2589\n",
      "2590\n",
      "2591\n",
      "2592\n",
      "2593\n",
      "2594\n",
      "2595\n",
      "2596\n",
      "2597\n",
      "2598\n",
      "2599\n",
      "2600\n",
      "2601\n",
      "2602\n",
      "2603\n",
      "2604\n",
      "2605\n",
      "2606\n",
      "2607\n",
      "2608\n",
      "2609\n",
      "2610\n",
      "2611\n",
      "2612\n",
      "2613\n",
      "2614\n",
      "2615\n",
      "2616\n",
      "2617\n",
      "2618\n",
      "2619\n",
      "2620\n",
      "2621\n",
      "2622\n",
      "2623\n",
      "2624\n",
      "2625\n",
      "2626\n",
      "2627\n",
      "2628\n",
      "2629\n",
      "2630\n",
      "2631\n",
      "2632\n",
      "2633\n",
      "2634\n",
      "2635\n",
      "2636\n",
      "2637\n",
      "2638\n",
      "2639\n",
      "2640\n",
      "2641\n",
      "2642\n",
      "2643\n",
      "2644\n",
      "2645\n",
      "2646\n",
      "2647\n",
      "2648\n",
      "2649\n",
      "2650\n",
      "2651\n",
      "2652\n",
      "2653\n",
      "2654\n",
      "2655\n",
      "2656\n",
      "2657\n",
      "2658\n",
      "2659\n",
      "2660\n",
      "2661\n",
      "2662\n",
      "2663\n",
      "2664\n",
      "2665\n",
      "2666\n",
      "2667\n",
      "2668\n",
      "2669\n",
      "2670\n",
      "2671\n",
      "2672\n",
      "2673\n",
      "2674\n",
      "2675\n",
      "2676\n",
      "2677\n",
      "2678\n",
      "2679\n",
      "2680\n",
      "2681\n",
      "2682\n",
      "2683\n",
      "2684\n",
      "2685\n",
      "2686\n",
      "2687\n",
      "2688\n",
      "2689\n",
      "2690\n",
      "2691\n",
      "2692\n",
      "2693\n",
      "2694\n",
      "2695\n",
      "2696\n",
      "2697\n",
      "2698\n",
      "2699\n",
      "2700\n",
      "2701\n",
      "2702\n",
      "2703\n",
      "2704\n",
      "2705\n",
      "2706\n",
      "2707\n",
      "2708\n",
      "2709\n",
      "2710\n",
      "2711\n",
      "2712\n",
      "2713\n",
      "2714\n",
      "2715\n",
      "2716\n",
      "2717\n",
      "2718\n",
      "2719\n",
      "2720\n",
      "2721\n",
      "2722\n",
      "2723\n",
      "2724\n",
      "2725\n",
      "2726\n",
      "2727\n",
      "2728\n",
      "2729\n",
      "2730\n",
      "2731\n",
      "2732\n",
      "2733\n",
      "2734\n",
      "2735\n",
      "2736\n",
      "2737\n",
      "2738\n",
      "2739\n",
      "2740\n",
      "2741\n",
      "2742\n",
      "2743\n",
      "2744\n",
      "2745\n",
      "2746\n",
      "2747\n",
      "2748\n",
      "2749\n",
      "2750\n",
      "2751\n",
      "2752\n",
      "2753\n",
      "2754\n",
      "2755\n",
      "2756\n",
      "2757\n",
      "2758\n",
      "2759\n",
      "2760\n",
      "2761\n",
      "2762\n",
      "2763\n",
      "2764\n",
      "2765\n",
      "2766\n",
      "2767\n",
      "2768\n",
      "2769\n",
      "2770\n",
      "2771\n",
      "2772\n",
      "2773\n",
      "2774\n",
      "2775\n",
      "2776\n",
      "2777\n",
      "2778\n",
      "2779\n",
      "2780\n",
      "2781\n",
      "2782\n",
      "2783\n",
      "2784\n",
      "2785\n",
      "2786\n",
      "2787\n",
      "2788\n",
      "2789\n",
      "2790\n",
      "2791\n",
      "2792\n",
      "2793\n",
      "2794\n",
      "2795\n",
      "2796\n",
      "2797\n",
      "2798\n",
      "2799\n",
      "2800\n",
      "2801\n",
      "2802\n",
      "2803\n",
      "2804\n",
      "2805\n",
      "2806\n",
      "2807\n",
      "2808\n",
      "2809\n",
      "2810\n",
      "2811\n",
      "2812\n",
      "2813\n",
      "2814\n",
      "2815\n",
      "2816\n",
      "2817\n",
      "2818\n",
      "2819\n",
      "2820\n",
      "2821\n",
      "2822\n",
      "2823\n",
      "2824\n",
      "2825\n",
      "2826\n",
      "2827\n",
      "2828\n",
      "2829\n",
      "2830\n",
      "2831\n",
      "2832\n",
      "2833\n",
      "2834\n",
      "2835\n",
      "2836\n",
      "2837\n",
      "2838\n",
      "2839\n",
      "2840\n",
      "2841\n",
      "2842\n",
      "2843\n",
      "2844\n",
      "2845\n",
      "2846\n",
      "2847\n",
      "2848\n",
      "2849\n",
      "2850\n",
      "2851\n",
      "2852\n",
      "2853\n",
      "2854\n",
      "2855\n",
      "2856\n",
      "2857\n",
      "2858\n",
      "2859\n",
      "2860\n",
      "2861\n",
      "2862\n",
      "2863\n",
      "2864\n",
      "2865\n",
      "2866\n",
      "2867\n",
      "2868\n",
      "2869\n",
      "2870\n",
      "2871\n",
      "2872\n",
      "2873\n",
      "2874\n",
      "2875\n",
      "2876\n",
      "2877\n",
      "2878\n",
      "2879\n",
      "2880\n",
      "2881\n",
      "2882\n",
      "2883\n",
      "2884\n",
      "2885\n",
      "2886\n",
      "2887\n",
      "2888\n",
      "2889\n",
      "2890\n",
      "2891\n",
      "2892\n",
      "2893\n",
      "2894\n",
      "2895\n",
      "2896\n",
      "2897\n",
      "2898\n",
      "2899\n",
      "2900\n",
      "2901\n",
      "2902\n",
      "2903\n",
      "2904\n",
      "2905\n",
      "2906\n",
      "2907\n",
      "2908\n",
      "2909\n",
      "2910\n",
      "2911\n",
      "2912\n",
      "2913\n",
      "2914\n",
      "2915\n",
      "2916\n",
      "2917\n",
      "2918\n",
      "2919\n",
      "2920\n",
      "2921\n",
      "2922\n",
      "2923\n",
      "2924\n",
      "2925\n",
      "2926\n",
      "2927\n",
      "2928\n",
      "2929\n",
      "2930\n",
      "2931\n",
      "2932\n",
      "2933\n",
      "2934\n",
      "2935\n",
      "2936\n",
      "2937\n",
      "2938\n",
      "2939\n",
      "2940\n",
      "2941\n",
      "2942\n",
      "2943\n",
      "2944\n",
      "2945\n",
      "2946\n",
      "2947\n",
      "2948\n",
      "2949\n",
      "2950\n",
      "2951\n",
      "2952\n",
      "2953\n",
      "2954\n",
      "2955\n",
      "2956\n",
      "2957\n",
      "2958\n",
      "2959\n",
      "2960\n",
      "2961\n",
      "2962\n",
      "2963\n",
      "2964\n",
      "2965\n",
      "2966\n",
      "2967\n",
      "2968\n",
      "2969\n",
      "2970\n",
      "2971\n",
      "2972\n",
      "2973\n",
      "2974\n",
      "2975\n",
      "2976\n",
      "2977\n",
      "2978\n",
      "2979\n",
      "2980\n",
      "2981\n",
      "2982\n",
      "2983\n",
      "2984\n",
      "2985\n",
      "2986\n",
      "2987\n",
      "2988\n",
      "2989\n",
      "2990\n",
      "2991\n",
      "2992\n",
      "2993\n",
      "2994\n",
      "2995\n",
      "2996\n",
      "2997\n",
      "2998\n",
      "2999\n",
      "3000\n",
      "3001\n",
      "3002\n",
      "3003\n",
      "3004\n",
      "3005\n",
      "3006\n",
      "3007\n",
      "3008\n",
      "3009\n",
      "3010\n",
      "3011\n",
      "3012\n",
      "3013\n",
      "3014\n",
      "3015\n",
      "3016\n",
      "3017\n",
      "3018\n",
      "3019\n",
      "3020\n",
      "3021\n",
      "3022\n",
      "3023\n",
      "3024\n",
      "3025\n",
      "3026\n",
      "3027\n",
      "3028\n",
      "3029\n",
      "3030\n",
      "3031\n",
      "3032\n",
      "3033\n",
      "3034\n",
      "3035\n",
      "3036\n",
      "3037\n",
      "3038\n",
      "3039\n",
      "3040\n",
      "3041\n",
      "3042\n",
      "3043\n",
      "3044\n",
      "3045\n",
      "3046\n",
      "3047\n",
      "3048\n",
      "3049\n",
      "3050\n",
      "3051\n",
      "3052\n",
      "3053\n",
      "3054\n",
      "3055\n",
      "3056\n",
      "3057\n",
      "3058\n",
      "3059\n",
      "3060\n",
      "3061\n",
      "3062\n",
      "3063\n",
      "3064\n",
      "3065\n",
      "3066\n",
      "3067\n",
      "3068\n",
      "3069\n",
      "3070\n",
      "3071\n",
      "3072\n",
      "3073\n",
      "3074\n",
      "3075\n",
      "3076\n",
      "3077\n",
      "3078\n",
      "3079\n",
      "3080\n",
      "3081\n",
      "3082\n",
      "3083\n",
      "3084\n",
      "3085\n",
      "3086\n",
      "3087\n",
      "3088\n",
      "3089\n",
      "3090\n",
      "3091\n",
      "3092\n",
      "3093\n",
      "3094\n",
      "3095\n",
      "3096\n",
      "3097\n",
      "3098\n",
      "3099\n",
      "3100\n",
      "3101\n",
      "3102\n",
      "3103\n",
      "3104\n",
      "3105\n",
      "3106\n",
      "3107\n",
      "3108\n",
      "3109\n",
      "3110\n",
      "3111\n",
      "3112\n",
      "3113\n",
      "3114\n",
      "3115\n",
      "3116\n",
      "3117\n",
      "3118\n",
      "3119\n",
      "3120\n",
      "3121\n",
      "3122\n",
      "3123\n",
      "3124\n",
      "3125\n",
      "3126\n",
      "3127\n",
      "3128\n",
      "3129\n",
      "3130\n",
      "3131\n",
      "3132\n",
      "3133\n",
      "3134\n",
      "3135\n",
      "3136\n",
      "3137\n",
      "3138\n",
      "3139\n",
      "3140\n",
      "3141\n",
      "3142\n",
      "3143\n",
      "3144\n",
      "3145\n",
      "3146\n",
      "3147\n",
      "3148\n",
      "3149\n",
      "3150\n",
      "3151\n",
      "3152\n",
      "3153\n",
      "3154\n",
      "3155\n",
      "3156\n",
      "3157\n",
      "3158\n",
      "3159\n",
      "3160\n",
      "3161\n",
      "3162\n",
      "3163\n",
      "3164\n",
      "3165\n",
      "3166\n",
      "3167\n",
      "3168\n",
      "3169\n",
      "3170\n",
      "3171\n",
      "3172\n",
      "3173\n",
      "3174\n",
      "3175\n",
      "3176\n",
      "3177\n",
      "3178\n",
      "3179\n",
      "3180\n",
      "3181\n",
      "3182\n",
      "3183\n",
      "3184\n",
      "3185\n",
      "3186\n",
      "3187\n",
      "3188\n",
      "3189\n",
      "3190\n",
      "3191\n",
      "3192\n",
      "3193\n",
      "3194\n",
      "3195\n",
      "3196\n",
      "3197\n",
      "3198\n",
      "3199\n",
      "3200\n",
      "3201\n",
      "3202\n",
      "3203\n",
      "3204\n",
      "3205\n",
      "3206\n",
      "3207\n",
      "3208\n",
      "3209\n",
      "3210\n",
      "3211\n",
      "3212\n",
      "3213\n",
      "3214\n",
      "3215\n",
      "3216\n",
      "3217\n",
      "3218\n",
      "3219\n",
      "3220\n",
      "3221\n",
      "3222\n",
      "3223\n",
      "3224\n",
      "3225\n",
      "3226\n",
      "3227\n",
      "3228\n",
      "3229\n",
      "3230\n",
      "3231\n",
      "3232\n",
      "3233\n",
      "3234\n",
      "3235\n",
      "3236\n",
      "3237\n",
      "3238\n",
      "3239\n",
      "3240\n",
      "3241\n",
      "3242\n",
      "3243\n",
      "3244\n",
      "3245\n",
      "3246\n",
      "3247\n",
      "3248\n",
      "3249\n",
      "3250\n",
      "3251\n",
      "3252\n",
      "3253\n",
      "3254\n",
      "3255\n",
      "3256\n",
      "3257\n",
      "3258\n",
      "3259\n",
      "3260\n",
      "3261\n",
      "3262\n",
      "3263\n",
      "3264\n",
      "3265\n",
      "3266\n",
      "3267\n",
      "3268\n",
      "3269\n",
      "3270\n",
      "3271\n",
      "3272\n",
      "3273\n",
      "3274\n",
      "3275\n",
      "3276\n",
      "3277\n",
      "3278\n",
      "3279\n",
      "3280\n",
      "3281\n",
      "3282\n",
      "3283\n",
      "3284\n",
      "3285\n",
      "3286\n",
      "3287\n",
      "3288\n",
      "3289\n",
      "3290\n",
      "3291\n",
      "3292\n",
      "3293\n",
      "3294\n",
      "3295\n",
      "3296\n",
      "3297\n",
      "3298\n",
      "3299\n",
      "3300\n",
      "3301\n",
      "3302\n",
      "3303\n",
      "3304\n",
      "3305\n",
      "3306\n",
      "3307\n",
      "3308\n",
      "3309\n",
      "3310\n",
      "3311\n",
      "3312\n",
      "3313\n",
      "3314\n",
      "3315\n",
      "3316\n",
      "3317\n",
      "3318\n",
      "3319\n",
      "3320\n",
      "3321\n",
      "3322\n",
      "3323\n",
      "3324\n",
      "3325\n",
      "3326\n",
      "3327\n",
      "3328\n",
      "3329\n",
      "3330\n",
      "3331\n",
      "3332\n",
      "3333\n",
      "3334\n",
      "3335\n",
      "3336\n",
      "3337\n",
      "3338\n",
      "3339\n",
      "3340\n",
      "3341\n",
      "3342\n",
      "3343\n",
      "3344\n",
      "3345\n",
      "3346\n",
      "3347\n",
      "3348\n",
      "3349\n",
      "3350\n",
      "3351\n",
      "3352\n",
      "3353\n",
      "3354\n",
      "3355\n",
      "3356\n",
      "3357\n",
      "3358\n",
      "3359\n",
      "3360\n",
      "3361\n",
      "3362\n",
      "3363\n",
      "3364\n",
      "3365\n",
      "3366\n",
      "3367\n",
      "3368\n",
      "3369\n",
      "3370\n",
      "3371\n",
      "3372\n",
      "3373\n",
      "3374\n",
      "3375\n",
      "3376\n",
      "3377\n",
      "3378\n",
      "3379\n",
      "3380\n",
      "3381\n",
      "3382\n",
      "3383\n",
      "3384\n",
      "3385\n",
      "3386\n",
      "3387\n",
      "3388\n",
      "3389\n",
      "3390\n",
      "3391\n",
      "3392\n",
      "3393\n",
      "3394\n",
      "3395\n",
      "3396\n",
      "3397\n",
      "3398\n",
      "3399\n",
      "3400\n",
      "3401\n",
      "3402\n",
      "3403\n",
      "3404\n",
      "3405\n",
      "3406\n",
      "3407\n",
      "3408\n",
      "3409\n",
      "3410\n",
      "3411\n",
      "3412\n",
      "3413\n",
      "3414\n",
      "3415\n",
      "3416\n",
      "3417\n",
      "3418\n",
      "3419\n",
      "3420\n",
      "3421\n",
      "3422\n",
      "3423\n",
      "3424\n",
      "3425\n",
      "3426\n",
      "3427\n",
      "3428\n",
      "3429\n",
      "3430\n",
      "3431\n",
      "3432\n",
      "3433\n",
      "3434\n",
      "3435\n",
      "3436\n",
      "3437\n",
      "3438\n",
      "3439\n",
      "3440\n",
      "3441\n",
      "3442\n",
      "3443\n",
      "3444\n",
      "3445\n",
      "3446\n",
      "3447\n",
      "3448\n",
      "3449\n",
      "3450\n",
      "3451\n",
      "3452\n",
      "3453\n",
      "3454\n",
      "3455\n",
      "3456\n",
      "3457\n",
      "3458\n",
      "3459\n",
      "3460\n",
      "3461\n",
      "3462\n",
      "3463\n",
      "3464\n",
      "3465\n",
      "3466\n",
      "3467\n",
      "3468\n",
      "3469\n",
      "3470\n",
      "3471\n",
      "3472\n",
      "3473\n",
      "3474\n",
      "3475\n",
      "3476\n",
      "3477\n",
      "3478\n",
      "3479\n",
      "3480\n",
      "3481\n",
      "3482\n",
      "3483\n",
      "3484\n",
      "3485\n",
      "3486\n",
      "3487\n",
      "3488\n",
      "3489\n",
      "3490\n",
      "3491\n",
      "3492\n",
      "3493\n",
      "3494\n",
      "3495\n",
      "3496\n",
      "3497\n",
      "3498\n",
      "3499\n",
      "3500\n",
      "3501\n",
      "3502\n",
      "3503\n",
      "3504\n",
      "3505\n",
      "3506\n",
      "3507\n",
      "3508\n",
      "3509\n",
      "3510\n",
      "3511\n",
      "3512\n",
      "3513\n",
      "3514\n",
      "3515\n",
      "3516\n",
      "3517\n",
      "3518\n",
      "3519\n",
      "3520\n",
      "3521\n",
      "3522\n",
      "3523\n",
      "3524\n",
      "3525\n",
      "3526\n",
      "3527\n",
      "3528\n",
      "3529\n",
      "3530\n",
      "3531\n",
      "3532\n",
      "3533\n",
      "3534\n",
      "3535\n",
      "3536\n",
      "3537\n",
      "3538\n",
      "3539\n",
      "3540\n",
      "3541\n",
      "3542\n",
      "3543\n",
      "3544\n",
      "3545\n",
      "3546\n",
      "3547\n",
      "3548\n",
      "3549\n",
      "3550\n",
      "3551\n",
      "3552\n",
      "3553\n",
      "3554\n",
      "3555\n",
      "3556\n",
      "3557\n",
      "3558\n",
      "3559\n",
      "3560\n",
      "3561\n",
      "3562\n",
      "3563\n",
      "3564\n",
      "3565\n",
      "3566\n",
      "3567\n",
      "3568\n",
      "3569\n",
      "3570\n",
      "3571\n",
      "3572\n",
      "3573\n",
      "3574\n",
      "3575\n",
      "3576\n",
      "3577\n",
      "3578\n",
      "3579\n",
      "3580\n",
      "3581\n",
      "3582\n",
      "3583\n",
      "3584\n",
      "3585\n",
      "3586\n",
      "3587\n",
      "3588\n",
      "3589\n",
      "3590\n",
      "3591\n",
      "3592\n",
      "3593\n",
      "3594\n",
      "3595\n",
      "3596\n",
      "3597\n",
      "3598\n",
      "3599\n",
      "3600\n",
      "3601\n",
      "3602\n",
      "3603\n",
      "3604\n",
      "3605\n",
      "3606\n",
      "3607\n",
      "3608\n",
      "3609\n",
      "3610\n",
      "3611\n",
      "3612\n",
      "3613\n",
      "3614\n",
      "3615\n",
      "3616\n",
      "3617\n",
      "3618\n",
      "3619\n",
      "3620\n",
      "3621\n",
      "3622\n",
      "3623\n",
      "3624\n",
      "3625\n",
      "3626\n",
      "3627\n",
      "3628\n",
      "3629\n",
      "3630\n",
      "3631\n",
      "3632\n",
      "3633\n",
      "3634\n",
      "3635\n",
      "3636\n",
      "3637\n",
      "3638\n",
      "3639\n",
      "3640\n",
      "3641\n",
      "3642\n",
      "3643\n",
      "3644\n",
      "3645\n",
      "3646\n",
      "3647\n",
      "3648\n",
      "3649\n",
      "3650\n",
      "3651\n",
      "3652\n",
      "3653\n",
      "3654\n",
      "3655\n",
      "3656\n",
      "3657\n",
      "3658\n",
      "3659\n",
      "3660\n",
      "3661\n",
      "3662\n",
      "3663\n",
      "3664\n",
      "3665\n",
      "3666\n",
      "3667\n",
      "3668\n",
      "3669\n",
      "3670\n",
      "3671\n",
      "3672\n",
      "3673\n",
      "3674\n",
      "3675\n",
      "3676\n",
      "3677\n",
      "3678\n",
      "3679\n",
      "3680\n",
      "3681\n",
      "3682\n",
      "3683\n",
      "3684\n",
      "3685\n",
      "3686\n",
      "3687\n",
      "3688\n",
      "3689\n",
      "3690\n",
      "3691\n",
      "3692\n",
      "3693\n",
      "3694\n",
      "3695\n",
      "3696\n",
      "3697\n",
      "3698\n",
      "3699\n",
      "3700\n",
      "3701\n",
      "3702\n",
      "3703\n",
      "3704\n",
      "3705\n",
      "3706\n",
      "3707\n",
      "3708\n",
      "3709\n",
      "3710\n",
      "3711\n",
      "3712\n",
      "3713\n",
      "3714\n",
      "3715\n",
      "3716\n",
      "3717\n",
      "3718\n",
      "3719\n",
      "3720\n",
      "3721\n",
      "3722\n",
      "3723\n",
      "3724\n",
      "3725\n",
      "3726\n",
      "3727\n",
      "3728\n",
      "3729\n",
      "3730\n",
      "3731\n",
      "3732\n",
      "3733\n",
      "3734\n",
      "3735\n",
      "3736\n",
      "3737\n",
      "3738\n",
      "3739\n",
      "3740\n",
      "3741\n",
      "3742\n",
      "3743\n",
      "3744\n",
      "3745\n",
      "3746\n",
      "3747\n",
      "3748\n",
      "3749\n",
      "3750\n",
      "3751\n",
      "3752\n",
      "3753\n",
      "3754\n",
      "3755\n",
      "3756\n",
      "3757\n",
      "3758\n",
      "3759\n",
      "3760\n",
      "3761\n",
      "3762\n",
      "3763\n",
      "3764\n",
      "3765\n",
      "3766\n",
      "3767\n",
      "3768\n",
      "3769\n",
      "3770\n",
      "3771\n",
      "3772\n",
      "3773\n",
      "3774\n",
      "3775\n",
      "3776\n",
      "3777\n",
      "3778\n",
      "3779\n",
      "3780\n",
      "3781\n",
      "3782\n",
      "3783\n",
      "3784\n",
      "3785\n",
      "3786\n",
      "3787\n",
      "3788\n",
      "3789\n",
      "3790\n",
      "3791\n",
      "3792\n",
      "3793\n",
      "3794\n",
      "3795\n",
      "3796\n",
      "3797\n",
      "3798\n",
      "3799\n",
      "3800\n",
      "3801\n",
      "3802\n",
      "3803\n",
      "3804\n",
      "3805\n",
      "3806\n",
      "3807\n",
      "3808\n",
      "3809\n",
      "3810\n",
      "3811\n",
      "3812\n",
      "3813\n",
      "3814\n",
      "3815\n",
      "3816\n",
      "3817\n",
      "3818\n",
      "3819\n",
      "3820\n",
      "3821\n",
      "3822\n",
      "3823\n",
      "3824\n",
      "3825\n",
      "3826\n",
      "3827\n",
      "3828\n",
      "3829\n",
      "3830\n",
      "3831\n",
      "3832\n",
      "3833\n",
      "3834\n",
      "3835\n",
      "3836\n",
      "3837\n",
      "3838\n",
      "3839\n",
      "3840\n",
      "3841\n",
      "3842\n",
      "3843\n",
      "3844\n",
      "3845\n",
      "3846\n",
      "3847\n",
      "3848\n",
      "3849\n",
      "3850\n",
      "3851\n",
      "3852\n",
      "3853\n",
      "3854\n",
      "3855\n",
      "3856\n",
      "3857\n",
      "3858\n",
      "3859\n",
      "3860\n",
      "3861\n",
      "3862\n",
      "3863\n",
      "3864\n",
      "3865\n",
      "3866\n",
      "3867\n",
      "3868\n",
      "3869\n",
      "3870\n",
      "3871\n",
      "3872\n",
      "3873\n",
      "3874\n",
      "3875\n",
      "3876\n",
      "3877\n",
      "3878\n",
      "3879\n",
      "3880\n",
      "3881\n",
      "3882\n",
      "3883\n",
      "3884\n",
      "3885\n",
      "3886\n",
      "3887\n",
      "3888\n",
      "3889\n",
      "3890\n",
      "3891\n",
      "3892\n",
      "3893\n",
      "3894\n",
      "3895\n",
      "3896\n",
      "3897\n",
      "3898\n",
      "3899\n",
      "3900\n",
      "3901\n",
      "3902\n",
      "3903\n",
      "3904\n",
      "3905\n",
      "3906\n",
      "3907\n",
      "3908\n",
      "3909\n",
      "3910\n",
      "3911\n",
      "3912\n",
      "3913\n",
      "3914\n",
      "3915\n",
      "3916\n",
      "3917\n",
      "3918\n",
      "3919\n",
      "3920\n",
      "3921\n",
      "3922\n",
      "3923\n",
      "3924\n",
      "3925\n",
      "3926\n",
      "3927\n",
      "3928\n",
      "3929\n",
      "3930\n",
      "3931\n",
      "3932\n",
      "3933\n",
      "3934\n",
      "3935\n",
      "3936\n",
      "3937\n",
      "3938\n",
      "3939\n",
      "3940\n",
      "3941\n",
      "3942\n",
      "3943\n",
      "3944\n",
      "3945\n",
      "3946\n",
      "3947\n",
      "3948\n",
      "3949\n",
      "3950\n",
      "3951\n",
      "3952\n",
      "3953\n",
      "3954\n",
      "3955\n",
      "3956\n",
      "3957\n",
      "3958\n",
      "3959\n",
      "3960\n",
      "3961\n",
      "3962\n",
      "3963\n",
      "3964\n",
      "3965\n",
      "3966\n",
      "3967\n",
      "3968\n",
      "3969\n",
      "3970\n",
      "3971\n",
      "3972\n",
      "3973\n",
      "3974\n",
      "3975\n",
      "3976\n",
      "3977\n",
      "3978\n",
      "3979\n",
      "3980\n",
      "3981\n",
      "3982\n",
      "3983\n",
      "3984\n",
      "3985\n",
      "3986\n",
      "3987\n",
      "3988\n",
      "3989\n",
      "3990\n",
      "3991\n",
      "3992\n",
      "3993\n",
      "3994\n",
      "3995\n",
      "3996\n",
      "3997\n",
      "3998\n",
      "3999\n",
      "4000\n",
      "4001\n",
      "4002\n",
      "4003\n",
      "4004\n",
      "4005\n",
      "4006\n",
      "4007\n",
      "4008\n",
      "4009\n",
      "4010\n",
      "4011\n",
      "4012\n",
      "4013\n",
      "4014\n",
      "4015\n",
      "4016\n",
      "4017\n",
      "4018\n",
      "4019\n",
      "4020\n",
      "4021\n",
      "4022\n",
      "4023\n",
      "4024\n",
      "4025\n",
      "4026\n",
      "4027\n",
      "4028\n",
      "4029\n",
      "4030\n",
      "4031\n",
      "4032\n",
      "4033\n",
      "4034\n",
      "4035\n",
      "4036\n",
      "4037\n",
      "4038\n",
      "4039\n",
      "4040\n",
      "4041\n",
      "4042\n",
      "4043\n",
      "4044\n",
      "4045\n",
      "4046\n",
      "4047\n",
      "4048\n",
      "4049\n",
      "4050\n",
      "4051\n",
      "4052\n",
      "4053\n",
      "4054\n",
      "4055\n",
      "4056\n",
      "4057\n",
      "4058\n",
      "4059\n",
      "4060\n",
      "4061\n",
      "4062\n",
      "4063\n",
      "4064\n",
      "4065\n",
      "4066\n",
      "4067\n",
      "4068\n",
      "4069\n",
      "4070\n",
      "4071\n",
      "4072\n",
      "4073\n",
      "4074\n",
      "4075\n",
      "4076\n",
      "4077\n",
      "4078\n",
      "4079\n",
      "4080\n",
      "4081\n",
      "4082\n",
      "4083\n",
      "4084\n",
      "4085\n",
      "4086\n",
      "4087\n",
      "4088\n",
      "4089\n",
      "4090\n",
      "4091\n",
      "4092\n",
      "4093\n",
      "4094\n",
      "4095\n",
      "4096\n",
      "4097\n",
      "4098\n",
      "4099\n",
      "4100\n",
      "4101\n",
      "4102\n",
      "4103\n",
      "4104\n",
      "4105\n",
      "4106\n",
      "4107\n",
      "4108\n",
      "4109\n",
      "4110\n",
      "4111\n",
      "4112\n",
      "4113\n",
      "4114\n",
      "4115\n",
      "4116\n",
      "4117\n",
      "4118\n",
      "4119\n",
      "4120\n",
      "4121\n",
      "4122\n",
      "4123\n",
      "4124\n",
      "4125\n",
      "4126\n",
      "4127\n",
      "4128\n",
      "4129\n",
      "4130\n",
      "4131\n",
      "4132\n",
      "4133\n",
      "4134\n",
      "4135\n",
      "4136\n",
      "4137\n",
      "4138\n",
      "4139\n",
      "4140\n",
      "4141\n",
      "4142\n",
      "4143\n",
      "4144\n",
      "4145\n",
      "4146\n",
      "4147\n",
      "4148\n",
      "4149\n",
      "4150\n",
      "4151\n",
      "4152\n",
      "4153\n",
      "4154\n",
      "4155\n",
      "4156\n",
      "4157\n",
      "4158\n",
      "4159\n",
      "4160\n",
      "4161\n",
      "4162\n",
      "4163\n",
      "4164\n",
      "4165\n",
      "4166\n",
      "4167\n",
      "4168\n",
      "4169\n",
      "4170\n",
      "4171\n",
      "4172\n",
      "4173\n",
      "4174\n",
      "4175\n",
      "4176\n",
      "4177\n",
      "4178\n",
      "4179\n",
      "4180\n",
      "4181\n",
      "4182\n",
      "4183\n",
      "4184\n",
      "4185\n",
      "4186\n",
      "4187\n",
      "4188\n",
      "4189\n",
      "4190\n",
      "4191\n",
      "4192\n",
      "4193\n",
      "4194\n",
      "4195\n",
      "4196\n",
      "4197\n",
      "4198\n",
      "4199\n",
      "4200\n",
      "4201\n",
      "4202\n",
      "4203\n",
      "4204\n",
      "4205\n",
      "4206\n",
      "4207\n",
      "4208\n",
      "4209\n",
      "4210\n",
      "4211\n",
      "4212\n",
      "4213\n",
      "4214\n",
      "4215\n",
      "4216\n",
      "4217\n",
      "4218\n",
      "4219\n",
      "4220\n",
      "4221\n",
      "4222\n",
      "4223\n",
      "4224\n",
      "4225\n",
      "4226\n",
      "4227\n",
      "4228\n",
      "4229\n",
      "4230\n",
      "4231\n",
      "4232\n",
      "4233\n",
      "4234\n",
      "4235\n",
      "4236\n",
      "4237\n",
      "4238\n",
      "4239\n",
      "4240\n",
      "4241\n",
      "4242\n",
      "4243\n",
      "4244\n",
      "4245\n",
      "4246\n",
      "4247\n",
      "4248\n",
      "4249\n",
      "4250\n",
      "4251\n",
      "4252\n",
      "4253\n",
      "4254\n",
      "4255\n",
      "4256\n",
      "4257\n",
      "4258\n",
      "4259\n",
      "4260\n",
      "4261\n",
      "4262\n",
      "4263\n",
      "4264\n",
      "4265\n",
      "4266\n",
      "4267\n",
      "4268\n",
      "4269\n",
      "4270\n",
      "4271\n",
      "4272\n",
      "4273\n",
      "4274\n",
      "4275\n",
      "4276\n",
      "4277\n",
      "4278\n",
      "4279\n",
      "4280\n",
      "4281\n",
      "4282\n",
      "4283\n",
      "4284\n",
      "4285\n",
      "4286\n",
      "4287\n",
      "4288\n",
      "4289\n",
      "4290\n",
      "4291\n",
      "4292\n",
      "4293\n",
      "4294\n",
      "4295\n",
      "4296\n",
      "4297\n",
      "4298\n",
      "4299\n",
      "4300\n",
      "4301\n",
      "4302\n",
      "4303\n",
      "4304\n",
      "4305\n",
      "4306\n",
      "4307\n",
      "4308\n",
      "4309\n",
      "4310\n",
      "4311\n",
      "4312\n",
      "4313\n",
      "4314\n",
      "4315\n",
      "4316\n",
      "4317\n",
      "4318\n",
      "4319\n",
      "4320\n",
      "4321\n",
      "4322\n",
      "4323\n",
      "4324\n",
      "4325\n",
      "4326\n",
      "4327\n",
      "4328\n",
      "4329\n",
      "4330\n",
      "4331\n",
      "4332\n",
      "4333\n",
      "4334\n",
      "4335\n",
      "4336\n",
      "4337\n",
      "4338\n",
      "4339\n",
      "4340\n",
      "4341\n",
      "4342\n",
      "4343\n",
      "4344\n",
      "4345\n",
      "4346\n",
      "4347\n",
      "4348\n",
      "4349\n",
      "4350\n",
      "4351\n",
      "4352\n",
      "4353\n",
      "4354\n",
      "4355\n",
      "4356\n",
      "4357\n",
      "4358\n",
      "4359\n",
      "4360\n",
      "4361\n",
      "4362\n",
      "4363\n",
      "4364\n",
      "4365\n",
      "4366\n",
      "4367\n",
      "4368\n",
      "4369\n",
      "4370\n",
      "4371\n",
      "4372\n",
      "4373\n",
      "4374\n",
      "4375\n",
      "4376\n",
      "4377\n",
      "4378\n",
      "4379\n",
      "4380\n",
      "4381\n",
      "4382\n",
      "4383\n",
      "4384\n",
      "4385\n",
      "4386\n",
      "4387\n",
      "4388\n",
      "4389\n",
      "4390\n",
      "4391\n",
      "4392\n",
      "4393\n",
      "4394\n",
      "4395\n",
      "4396\n",
      "4397\n",
      "4398\n",
      "4399\n",
      "4400\n",
      "4401\n",
      "4402\n",
      "4403\n",
      "4404\n",
      "4405\n",
      "4406\n",
      "4407\n",
      "4408\n",
      "4409\n",
      "4410\n",
      "4411\n",
      "4412\n",
      "4413\n",
      "4414\n",
      "4415\n",
      "4416\n",
      "4417\n",
      "4418\n",
      "4419\n",
      "4420\n",
      "4421\n",
      "4422\n",
      "4423\n",
      "4424\n",
      "4425\n",
      "4426\n",
      "4427\n",
      "4428\n",
      "4429\n",
      "4430\n",
      "4431\n",
      "4432\n",
      "4433\n",
      "4434\n",
      "4435\n",
      "4436\n",
      "4437\n",
      "4438\n",
      "4439\n",
      "4440\n",
      "4441\n",
      "4442\n",
      "4443\n",
      "4444\n",
      "4445\n",
      "4446\n",
      "4447\n",
      "4448\n",
      "4449\n",
      "4450\n",
      "4451\n",
      "4452\n",
      "4453\n",
      "4454\n",
      "4455\n",
      "4456\n",
      "4457\n",
      "4458\n",
      "4459\n",
      "4460\n",
      "4461\n",
      "4462\n",
      "4463\n",
      "4464\n",
      "4465\n",
      "4466\n",
      "4467\n",
      "4468\n",
      "4469\n",
      "4470\n",
      "4471\n",
      "4472\n",
      "4473\n",
      "4474\n",
      "4475\n",
      "4476\n",
      "4477\n",
      "4478\n",
      "4479\n",
      "4480\n",
      "4481\n",
      "4482\n",
      "4483\n",
      "4484\n",
      "4485\n",
      "4486\n",
      "4487\n",
      "4488\n",
      "4489\n",
      "4490\n",
      "4491\n",
      "4492\n",
      "4493\n",
      "4494\n",
      "4495\n",
      "4496\n",
      "4497\n",
      "4498\n",
      "4499\n",
      "4500\n",
      "4501\n",
      "4502\n",
      "4503\n",
      "4504\n",
      "4505\n",
      "4506\n",
      "4507\n",
      "4508\n",
      "4509\n",
      "4510\n",
      "4511\n",
      "4512\n",
      "4513\n",
      "4514\n",
      "4515\n",
      "4516\n",
      "4517\n",
      "4518\n",
      "4519\n",
      "4520\n",
      "4521\n",
      "4522\n",
      "4523\n",
      "4524\n",
      "4525\n",
      "4526\n",
      "4527\n",
      "4528\n",
      "4529\n",
      "4530\n",
      "4531\n",
      "4532\n",
      "4533\n",
      "4534\n",
      "4535\n",
      "4536\n",
      "4537\n",
      "4538\n",
      "4539\n",
      "4540\n",
      "4541\n",
      "4542\n",
      "4543\n",
      "4544\n",
      "4545\n",
      "4546\n",
      "4547\n",
      "4548\n",
      "4549\n",
      "4550\n",
      "4551\n",
      "4552\n",
      "4553\n",
      "4554\n",
      "4555\n",
      "4556\n",
      "4557\n",
      "4558\n",
      "4559\n",
      "4560\n",
      "4561\n",
      "4562\n",
      "4563\n",
      "4564\n",
      "4565\n",
      "4566\n",
      "4567\n",
      "4568\n",
      "4569\n",
      "4570\n",
      "4571\n",
      "4572\n",
      "4573\n",
      "4574\n",
      "4575\n",
      "4576\n",
      "4577\n",
      "4578\n",
      "4579\n",
      "4580\n",
      "4581\n",
      "4582\n",
      "4583\n",
      "4584\n",
      "4585\n",
      "4586\n",
      "4587\n",
      "4588\n",
      "4589\n",
      "4590\n",
      "4591\n",
      "4592\n",
      "4593\n",
      "4594\n",
      "4595\n",
      "4596\n",
      "4597\n",
      "4598\n",
      "4599\n",
      "4600\n",
      "4601\n",
      "4602\n",
      "4603\n",
      "4604\n",
      "4605\n",
      "4606\n",
      "4607\n",
      "4608\n",
      "4609\n",
      "4610\n",
      "4611\n",
      "4612\n",
      "4613\n",
      "4614\n",
      "4615\n",
      "4616\n",
      "4617\n",
      "4618\n",
      "4619\n",
      "4620\n",
      "4621\n",
      "4622\n",
      "4623\n",
      "4624\n",
      "4625\n",
      "4626\n",
      "4627\n",
      "4628\n",
      "4629\n",
      "4630\n",
      "4631\n",
      "4632\n",
      "4633\n",
      "4634\n",
      "4635\n",
      "4636\n",
      "4637\n",
      "4638\n",
      "4639\n",
      "4640\n",
      "4641\n",
      "4642\n",
      "4643\n",
      "4644\n",
      "4645\n",
      "4646\n",
      "4647\n",
      "4648\n",
      "4649\n",
      "4650\n",
      "4651\n",
      "4652\n",
      "4653\n",
      "4654\n",
      "4655\n",
      "4656\n",
      "4657\n",
      "4658\n",
      "4659\n",
      "4660\n",
      "4661\n",
      "4662\n",
      "4663\n",
      "4664\n",
      "4665\n",
      "4666\n",
      "4667\n",
      "4668\n",
      "4669\n",
      "4670\n",
      "4671\n",
      "4672\n",
      "4673\n",
      "4674\n",
      "4675\n",
      "4676\n",
      "4677\n",
      "4678\n",
      "4679\n",
      "4680\n",
      "4681\n",
      "4682\n",
      "4683\n",
      "4684\n",
      "4685\n",
      "4686\n",
      "4687\n",
      "4688\n",
      "4689\n",
      "4690\n",
      "4691\n",
      "4692\n",
      "4693\n",
      "4694\n",
      "4695\n",
      "4696\n",
      "4697\n",
      "4698\n",
      "4699\n",
      "4700\n",
      "4701\n",
      "4702\n",
      "4703\n",
      "4704\n",
      "4705\n",
      "4706\n",
      "4707\n",
      "4708\n",
      "4709\n",
      "4710\n",
      "4711\n",
      "4712\n",
      "4713\n",
      "4714\n",
      "4715\n",
      "4716\n",
      "4717\n",
      "4718\n",
      "4719\n",
      "4720\n",
      "4721\n",
      "4722\n",
      "4723\n",
      "4724\n",
      "4725\n",
      "4726\n",
      "4727\n",
      "4728\n",
      "4729\n",
      "4730\n",
      "4731\n",
      "4732\n",
      "4733\n",
      "4734\n",
      "4735\n",
      "4736\n",
      "4737\n",
      "4738\n",
      "4739\n",
      "4740\n",
      "4741\n",
      "4742\n",
      "4743\n",
      "4744\n",
      "4745\n",
      "4746\n",
      "4747\n",
      "4748\n",
      "4749\n",
      "4750\n",
      "4751\n",
      "4752\n",
      "4753\n",
      "4754\n",
      "4755\n",
      "4756\n",
      "4757\n",
      "4758\n",
      "4759\n",
      "4760\n",
      "4761\n",
      "4762\n",
      "4763\n",
      "4764\n",
      "4765\n",
      "4766\n",
      "4767\n",
      "4768\n",
      "4769\n",
      "4770\n",
      "4771\n",
      "4772\n",
      "4773\n",
      "4774\n",
      "4775\n",
      "4776\n",
      "4777\n",
      "4778\n",
      "4779\n",
      "4780\n",
      "4781\n",
      "4782\n",
      "4783\n",
      "4784\n",
      "4785\n",
      "4786\n",
      "4787\n",
      "4788\n",
      "4789\n",
      "4790\n",
      "4791\n",
      "4792\n",
      "4793\n",
      "4794\n",
      "4795\n",
      "4796\n",
      "4797\n",
      "4798\n",
      "4799\n",
      "4800\n",
      "4801\n",
      "4802\n",
      "4803\n",
      "4804\n",
      "4805\n",
      "4806\n",
      "4807\n",
      "4808\n",
      "4809\n",
      "4810\n",
      "4811\n",
      "4812\n",
      "4813\n",
      "4814\n",
      "4815\n",
      "4816\n",
      "4817\n",
      "4818\n",
      "4819\n",
      "4820\n",
      "4821\n",
      "4822\n",
      "4823\n",
      "4824\n",
      "4825\n",
      "4826\n",
      "4827\n",
      "4828\n",
      "4829\n",
      "4830\n",
      "4831\n",
      "4832\n",
      "4833\n",
      "4834\n",
      "4835\n",
      "4836\n",
      "4837\n",
      "4838\n",
      "4839\n",
      "4840\n",
      "4841\n",
      "4842\n",
      "4843\n",
      "4844\n",
      "4845\n",
      "4846\n",
      "4847\n",
      "4848\n",
      "4849\n",
      "4850\n",
      "4851\n",
      "4852\n",
      "4853\n",
      "4854\n",
      "4855\n",
      "4856\n",
      "4857\n",
      "4858\n",
      "4859\n",
      "4860\n",
      "4861\n",
      "4862\n",
      "4863\n",
      "4864\n",
      "4865\n",
      "4866\n",
      "4867\n",
      "4868\n",
      "4869\n",
      "4870\n",
      "4871\n",
      "4872\n",
      "4873\n",
      "4874\n",
      "4875\n",
      "4876\n",
      "4877\n",
      "4878\n",
      "4879\n",
      "4880\n",
      "4881\n",
      "4882\n",
      "4883\n",
      "4884\n",
      "4885\n",
      "4886\n",
      "4887\n",
      "4888\n",
      "4889\n",
      "4890\n",
      "4891\n",
      "4892\n",
      "4893\n",
      "4894\n",
      "4895\n",
      "4896\n",
      "4897\n",
      "4898\n",
      "4899\n",
      "4900\n",
      "4901\n",
      "4902\n",
      "4903\n",
      "4904\n",
      "4905\n",
      "4906\n",
      "4907\n",
      "4908\n",
      "4909\n",
      "4910\n",
      "4911\n",
      "4912\n",
      "4913\n",
      "4914\n",
      "4915\n",
      "4916\n",
      "4917\n",
      "4918\n",
      "4919\n",
      "4920\n",
      "4921\n",
      "4922\n",
      "4923\n",
      "4924\n",
      "4925\n",
      "4926\n",
      "4927\n",
      "4928\n",
      "4929\n",
      "4930\n",
      "4931\n",
      "4932\n",
      "4933\n",
      "4934\n",
      "4935\n",
      "4936\n",
      "4937\n",
      "4938\n",
      "4939\n",
      "4940\n",
      "4941\n",
      "4942\n",
      "4943\n",
      "4944\n",
      "4945\n",
      "4946\n",
      "4947\n",
      "4948\n",
      "4949\n",
      "4950\n",
      "4951\n",
      "4952\n",
      "4953\n",
      "4954\n",
      "4955\n",
      "4956\n",
      "4957\n",
      "4958\n",
      "4959\n",
      "4960\n",
      "4961\n",
      "4962\n",
      "4963\n",
      "4964\n",
      "4965\n",
      "4966\n",
      "4967\n",
      "4968\n",
      "4969\n",
      "4970\n",
      "4971\n",
      "4972\n",
      "4973\n",
      "4974\n",
      "4975\n",
      "4976\n",
      "4977\n",
      "4978\n",
      "4979\n",
      "4980\n",
      "4981\n",
      "4982\n",
      "4983\n",
      "4984\n",
      "4985\n",
      "4986\n",
      "4987\n",
      "4988\n",
      "4989\n",
      "4990\n",
      "4991\n",
      "4992\n",
      "4993\n",
      "4994\n",
      "4995\n",
      "4996\n",
      "4997\n",
      "4998\n",
      "4999\n",
      "5000\n",
      "5001\n",
      "5002\n",
      "5003\n",
      "5004\n",
      "5005\n",
      "5006\n",
      "5007\n",
      "5008\n",
      "5009\n",
      "5010\n",
      "5011\n",
      "5012\n",
      "5013\n",
      "5014\n",
      "5015\n",
      "5016\n",
      "5017\n",
      "5018\n",
      "5019\n",
      "5020\n",
      "5021\n",
      "5022\n",
      "5023\n",
      "5024\n",
      "5025\n",
      "5026\n",
      "5027\n",
      "5028\n",
      "5029\n",
      "5030\n",
      "5031\n",
      "5032\n",
      "5033\n",
      "5034\n",
      "5035\n",
      "5036\n",
      "5037\n",
      "5038\n",
      "5039\n",
      "5040\n",
      "5041\n",
      "5042\n",
      "5043\n",
      "5044\n",
      "5045\n",
      "5046\n",
      "5047\n",
      "5048\n",
      "5049\n",
      "5050\n",
      "5051\n",
      "5052\n",
      "5053\n",
      "5054\n",
      "5055\n",
      "5056\n",
      "5057\n",
      "5058\n",
      "5059\n",
      "5060\n",
      "5061\n",
      "5062\n",
      "5063\n",
      "5064\n",
      "5065\n",
      "5066\n",
      "5067\n",
      "5068\n",
      "5069\n",
      "5070\n",
      "5071\n",
      "5072\n",
      "5073\n",
      "5074\n",
      "5075\n",
      "5076\n",
      "5077\n",
      "5078\n",
      "5079\n",
      "5080\n",
      "5081\n",
      "5082\n",
      "5083\n",
      "5084\n",
      "5085\n",
      "5086\n",
      "5087\n",
      "5088\n",
      "5089\n",
      "5090\n",
      "5091\n",
      "5092\n",
      "5093\n",
      "5094\n",
      "5095\n",
      "5096\n",
      "5097\n",
      "5098\n",
      "5099\n",
      "5100\n",
      "5101\n",
      "5102\n",
      "5103\n",
      "5104\n",
      "5105\n",
      "5106\n",
      "5107\n",
      "5108\n",
      "5109\n",
      "5110\n",
      "5111\n",
      "5112\n",
      "5113\n",
      "5114\n",
      "5115\n",
      "5116\n",
      "5117\n",
      "5118\n",
      "5119\n",
      "5120\n",
      "5121\n",
      "5122\n",
      "5123\n",
      "5124\n",
      "5125\n",
      "5126\n",
      "5127\n",
      "5128\n",
      "5129\n",
      "5130\n",
      "5131\n",
      "5132\n",
      "5133\n",
      "5134\n",
      "5135\n",
      "5136\n",
      "5137\n",
      "5138\n",
      "5139\n",
      "5140\n",
      "5141\n",
      "5142\n",
      "5143\n",
      "5144\n",
      "5145\n",
      "5146\n",
      "5147\n",
      "5148\n",
      "5149\n",
      "5150\n",
      "5151\n",
      "5152\n",
      "5153\n",
      "5154\n",
      "5155\n",
      "5156\n",
      "5157\n",
      "5158\n",
      "5159\n",
      "5160\n",
      "5161\n",
      "5162\n",
      "5163\n",
      "5164\n",
      "5165\n",
      "5166\n",
      "5167\n",
      "5168\n",
      "5169\n",
      "5170\n",
      "5171\n",
      "5172\n",
      "5173\n",
      "5174\n",
      "5175\n",
      "5176\n",
      "5177\n",
      "5178\n",
      "5179\n",
      "5180\n",
      "5181\n",
      "5182\n",
      "5183\n",
      "5184\n",
      "5185\n",
      "5186\n",
      "5187\n",
      "5188\n",
      "5189\n",
      "5190\n",
      "5191\n",
      "5192\n",
      "5193\n",
      "5194\n",
      "5195\n",
      "5196\n",
      "5197\n",
      "5198\n",
      "5199\n",
      "5200\n",
      "5201\n",
      "5202\n",
      "5203\n",
      "5204\n",
      "5205\n",
      "5206\n",
      "5207\n",
      "5208\n",
      "5209\n",
      "5210\n",
      "5211\n",
      "5212\n",
      "5213\n",
      "5214\n",
      "5215\n",
      "5216\n",
      "5217\n",
      "5218\n",
      "5219\n",
      "5220\n",
      "5221\n",
      "5222\n",
      "5223\n",
      "5224\n",
      "5225\n",
      "5226\n",
      "5227\n",
      "5228\n",
      "5229\n",
      "5230\n",
      "5231\n",
      "5232\n",
      "5233\n",
      "5234\n",
      "5235\n",
      "5236\n",
      "5237\n",
      "5238\n",
      "5239\n",
      "5240\n",
      "5241\n",
      "5242\n",
      "5243\n",
      "5244\n",
      "5245\n",
      "5246\n",
      "5247\n",
      "5248\n",
      "5249\n",
      "5250\n",
      "5251\n",
      "5252\n",
      "5253\n",
      "5254\n",
      "5255\n",
      "5256\n",
      "5257\n",
      "5258\n",
      "5259\n",
      "5260\n",
      "5261\n",
      "5262\n",
      "5263\n",
      "5264\n",
      "5265\n",
      "5266\n",
      "5267\n",
      "5268\n",
      "5269\n",
      "5270\n",
      "5271\n",
      "5272\n",
      "5273\n",
      "5274\n",
      "5275\n",
      "5276\n",
      "5277\n",
      "5278\n",
      "5279\n",
      "5280\n",
      "5281\n",
      "5282\n",
      "5283\n",
      "5284\n",
      "5285\n",
      "5286\n",
      "5287\n",
      "5288\n",
      "5289\n",
      "5290\n",
      "5291\n",
      "5292\n",
      "5293\n",
      "5294\n",
      "5295\n",
      "5296\n",
      "5297\n",
      "5298\n",
      "5299\n",
      "5300\n",
      "5301\n",
      "5302\n",
      "5303\n",
      "5304\n",
      "5305\n",
      "5306\n",
      "5307\n",
      "5308\n",
      "5309\n",
      "5310\n",
      "5311\n",
      "5312\n",
      "5313\n",
      "5314\n",
      "5315\n",
      "5316\n",
      "5317\n",
      "5318\n",
      "5319\n",
      "5320\n",
      "5321\n",
      "5322\n",
      "5323\n",
      "5324\n",
      "5325\n",
      "5326\n",
      "5327\n",
      "5328\n",
      "5329\n",
      "5330\n",
      "5331\n",
      "5332\n",
      "5333\n",
      "5334\n",
      "5335\n",
      "5336\n",
      "5337\n",
      "5338\n",
      "5339\n",
      "5340\n",
      "5341\n",
      "5342\n",
      "5343\n",
      "5344\n",
      "5345\n",
      "5346\n",
      "5347\n",
      "5348\n",
      "5349\n",
      "5350\n",
      "5351\n",
      "5352\n",
      "5353\n",
      "5354\n",
      "5355\n",
      "5356\n",
      "5357\n",
      "5358\n",
      "5359\n",
      "5360\n",
      "5361\n",
      "5362\n",
      "5363\n",
      "5364\n",
      "5365\n",
      "5366\n",
      "5367\n",
      "5368\n",
      "5369\n",
      "5370\n",
      "5371\n",
      "5372\n",
      "5373\n",
      "5374\n",
      "5375\n",
      "5376\n",
      "5377\n",
      "5378\n",
      "5379\n",
      "5380\n",
      "5381\n",
      "5382\n",
      "5383\n",
      "5384\n",
      "5385\n",
      "5386\n",
      "5387\n",
      "5388\n",
      "5389\n",
      "5390\n",
      "5391\n",
      "5392\n",
      "5393\n",
      "5394\n",
      "5395\n",
      "5396\n",
      "5397\n",
      "5398\n",
      "5399\n",
      "5400\n",
      "5401\n",
      "5402\n",
      "5403\n",
      "5404\n",
      "5405\n",
      "5406\n",
      "5407\n",
      "5408\n",
      "5409\n",
      "5410\n",
      "5411\n",
      "5412\n",
      "5413\n",
      "5414\n",
      "5415\n",
      "5416\n",
      "5417\n",
      "5418\n",
      "5419\n",
      "5420\n",
      "5421\n",
      "5422\n",
      "5423\n",
      "5424\n",
      "5425\n",
      "5426\n",
      "5427\n",
      "5428\n",
      "5429\n",
      "5430\n",
      "5431\n",
      "5432\n",
      "5433\n",
      "5434\n",
      "5435\n",
      "5436\n",
      "5437\n",
      "5438\n",
      "5439\n",
      "5440\n",
      "5441\n",
      "5442\n",
      "5443\n",
      "5444\n",
      "5445\n",
      "5446\n",
      "5447\n",
      "5448\n",
      "5449\n",
      "5450\n",
      "5451\n",
      "5452\n",
      "5453\n",
      "5454\n",
      "5455\n",
      "5456\n",
      "5457\n",
      "5458\n",
      "5459\n",
      "5460\n",
      "5461\n",
      "5462\n",
      "5463\n",
      "5464\n",
      "5465\n",
      "5466\n",
      "5467\n",
      "5468\n",
      "5469\n",
      "5470\n",
      "5471\n",
      "5472\n",
      "5473\n",
      "5474\n",
      "5475\n",
      "5476\n",
      "5477\n",
      "5478\n",
      "5479\n",
      "5480\n",
      "5481\n",
      "5482\n",
      "5483\n",
      "5484\n",
      "5485\n",
      "5486\n",
      "5487\n",
      "5488\n",
      "5489\n",
      "5490\n",
      "5491\n",
      "5492\n",
      "5493\n",
      "5494\n",
      "5495\n",
      "5496\n",
      "5497\n",
      "5498\n",
      "5499\n",
      "5500\n",
      "5501\n",
      "5502\n",
      "5503\n",
      "5504\n",
      "5505\n",
      "5506\n",
      "5507\n",
      "5508\n",
      "5509\n",
      "5510\n",
      "5511\n",
      "5512\n",
      "5513\n",
      "5514\n",
      "5515\n",
      "5516\n",
      "5517\n",
      "5518\n",
      "5519\n",
      "5520\n",
      "5521\n",
      "5522\n",
      "5523\n",
      "5524\n",
      "5525\n",
      "5526\n",
      "5527\n",
      "5528\n",
      "5529\n",
      "5530\n",
      "5531\n",
      "5532\n",
      "5533\n",
      "5534\n",
      "5535\n",
      "5536\n",
      "5537\n",
      "5538\n",
      "5539\n",
      "5540\n",
      "5541\n",
      "5542\n",
      "5543\n",
      "5544\n",
      "5545\n",
      "5546\n",
      "5547\n",
      "5548\n",
      "5549\n",
      "5550\n",
      "5551\n",
      "5552\n",
      "5553\n",
      "5554\n",
      "5555\n",
      "5556\n",
      "5557\n",
      "5558\n",
      "5559\n",
      "5560\n",
      "5561\n",
      "5562\n",
      "5563\n",
      "5564\n",
      "5565\n",
      "5566\n",
      "5567\n",
      "5568\n",
      "5569\n",
      "5570\n",
      "5571\n",
      "5572\n",
      "5573\n",
      "5574\n",
      "5575\n",
      "5576\n",
      "5577\n",
      "5578\n",
      "5579\n",
      "5580\n",
      "5581\n",
      "5582\n",
      "5583\n",
      "5584\n",
      "5585\n",
      "5586\n",
      "5587\n",
      "5588\n",
      "5589\n",
      "5590\n",
      "5591\n",
      "5592\n",
      "5593\n",
      "5594\n",
      "5595\n",
      "5596\n",
      "5597\n",
      "5598\n",
      "5599\n",
      "5600\n",
      "5601\n",
      "5602\n",
      "5603\n",
      "5604\n",
      "5605\n",
      "5606\n",
      "5607\n",
      "5608\n",
      "5609\n",
      "5610\n",
      "5611\n",
      "5612\n",
      "5613\n",
      "5614\n",
      "5615\n",
      "5616\n",
      "5617\n",
      "5618\n",
      "5619\n",
      "5620\n",
      "5621\n",
      "5622\n",
      "5623\n",
      "5624\n",
      "5625\n",
      "5626\n",
      "5627\n",
      "5628\n",
      "5629\n",
      "5630\n",
      "5631\n",
      "5632\n",
      "5633\n",
      "5634\n",
      "5635\n",
      "5636\n",
      "5637\n",
      "5638\n",
      "5639\n",
      "5640\n",
      "5641\n",
      "5642\n",
      "5643\n",
      "5644\n",
      "5645\n",
      "5646\n",
      "5647\n",
      "5648\n",
      "5649\n",
      "5650\n",
      "5651\n",
      "5652\n",
      "5653\n",
      "5654\n",
      "5655\n",
      "5656\n",
      "5657\n",
      "5658\n",
      "5659\n",
      "5660\n",
      "5661\n",
      "5662\n",
      "5663\n",
      "5664\n",
      "5665\n",
      "5666\n",
      "5667\n",
      "5668\n",
      "5669\n",
      "5670\n",
      "5671\n",
      "5672\n",
      "5673\n",
      "5674\n",
      "5675\n",
      "5676\n",
      "5677\n",
      "5678\n",
      "5679\n",
      "5680\n",
      "5681\n",
      "5682\n",
      "5683\n",
      "5684\n",
      "5685\n",
      "5686\n",
      "5687\n",
      "5688\n",
      "5689\n",
      "5690\n",
      "5691\n",
      "5692\n",
      "5693\n",
      "5694\n",
      "5695\n",
      "5696\n",
      "5697\n",
      "5698\n",
      "5699\n",
      "5700\n",
      "5701\n",
      "5702\n",
      "5703\n",
      "5704\n",
      "5705\n",
      "5706\n",
      "5707\n",
      "5708\n",
      "5709\n",
      "5710\n",
      "5711\n",
      "5712\n",
      "5713\n",
      "5714\n",
      "5715\n",
      "5716\n",
      "5717\n",
      "5718\n",
      "5719\n",
      "5720\n",
      "5721\n",
      "5722\n",
      "5723\n",
      "5724\n",
      "5725\n",
      "5726\n",
      "5727\n",
      "5728\n",
      "5729\n",
      "5730\n",
      "5731\n",
      "5732\n",
      "5733\n",
      "5734\n",
      "5735\n",
      "5736\n",
      "5737\n",
      "5738\n",
      "5739\n",
      "5740\n",
      "5741\n",
      "5742\n",
      "5743\n",
      "5744\n",
      "5745\n",
      "5746\n",
      "5747\n",
      "5748\n",
      "5749\n",
      "5750\n",
      "5751\n",
      "5752\n",
      "5753\n",
      "5754\n",
      "5755\n",
      "5756\n",
      "5757\n",
      "5758\n",
      "5759\n",
      "5760\n",
      "5761\n",
      "5762\n",
      "5763\n",
      "5764\n",
      "5765\n",
      "5766\n",
      "5767\n",
      "5768\n",
      "5769\n",
      "5770\n",
      "5771\n",
      "5772\n",
      "5773\n",
      "5774\n",
      "5775\n",
      "5776\n",
      "5777\n",
      "5778\n",
      "5779\n",
      "5780\n",
      "5781\n",
      "5782\n",
      "5783\n",
      "5784\n",
      "5785\n",
      "5786\n",
      "5787\n",
      "5788\n",
      "5789\n",
      "5790\n",
      "5791\n",
      "5792\n",
      "5793\n",
      "5794\n",
      "5795\n",
      "5796\n",
      "5797\n",
      "5798\n",
      "5799\n",
      "5800\n",
      "5801\n",
      "5802\n",
      "5803\n",
      "5804\n",
      "5805\n",
      "5806\n",
      "5807\n",
      "5808\n",
      "5809\n",
      "5810\n",
      "5811\n",
      "5812\n",
      "5813\n",
      "5814\n",
      "5815\n",
      "5816\n",
      "5817\n",
      "5818\n",
      "5819\n",
      "5820\n",
      "5821\n",
      "5822\n",
      "5823\n",
      "5824\n",
      "5825\n",
      "5826\n",
      "5827\n",
      "5828\n",
      "5829\n",
      "5830\n",
      "5831\n",
      "5832\n",
      "5833\n",
      "5834\n",
      "5835\n",
      "5836\n",
      "5837\n",
      "5838\n",
      "5839\n",
      "5840\n",
      "5841\n",
      "5842\n",
      "5843\n",
      "5844\n",
      "5845\n",
      "5846\n",
      "5847\n",
      "5848\n",
      "5849\n",
      "5850\n",
      "5851\n",
      "5852\n",
      "5853\n",
      "5854\n",
      "5855\n",
      "5856\n",
      "5857\n",
      "5858\n",
      "5859\n",
      "5860\n",
      "5861\n",
      "5862\n",
      "5863\n",
      "5864\n",
      "5865\n",
      "5866\n",
      "5867\n",
      "5868\n",
      "5869\n",
      "5870\n",
      "5871\n",
      "5872\n",
      "5873\n",
      "5874\n",
      "5875\n",
      "5876\n",
      "5877\n",
      "5878\n",
      "5879\n",
      "5880\n",
      "5881\n",
      "5882\n",
      "5883\n",
      "5884\n",
      "5885\n",
      "5886\n",
      "5887\n",
      "5888\n",
      "5889\n",
      "5890\n",
      "5891\n",
      "5892\n",
      "5893\n",
      "5894\n",
      "5895\n",
      "5896\n",
      "5897\n",
      "5898\n",
      "5899\n",
      "5900\n",
      "5901\n",
      "5902\n",
      "5903\n",
      "5904\n",
      "5905\n",
      "5906\n",
      "5907\n",
      "5908\n",
      "5909\n",
      "5910\n",
      "5911\n",
      "5912\n",
      "5913\n",
      "5914\n",
      "5915\n",
      "5916\n",
      "5917\n",
      "5918\n",
      "5919\n",
      "5920\n",
      "5921\n",
      "5922\n",
      "5923\n",
      "5924\n",
      "5925\n",
      "5926\n",
      "5927\n",
      "5928\n",
      "5929\n",
      "5930\n",
      "5931\n",
      "5932\n",
      "5933\n",
      "5934\n",
      "5935\n",
      "5936\n",
      "5937\n",
      "5938\n",
      "5939\n",
      "5940\n",
      "5941\n",
      "5942\n",
      "5943\n",
      "5944\n",
      "5945\n",
      "5946\n",
      "5947\n",
      "5948\n",
      "5949\n",
      "5950\n",
      "5951\n",
      "5952\n",
      "5953\n",
      "5954\n",
      "5955\n",
      "5956\n",
      "5957\n",
      "5958\n",
      "5959\n",
      "5960\n",
      "5961\n",
      "5962\n",
      "5963\n",
      "5964\n",
      "5965\n",
      "5966\n",
      "5967\n",
      "5968\n",
      "5969\n",
      "5970\n",
      "5971\n",
      "5972\n",
      "5973\n",
      "5974\n",
      "5975\n",
      "5976\n",
      "5977\n",
      "5978\n",
      "5979\n",
      "5980\n",
      "5981\n",
      "5982\n",
      "5983\n",
      "5984\n",
      "5985\n",
      "5986\n",
      "5987\n",
      "5988\n",
      "5989\n",
      "5990\n",
      "5991\n",
      "5992\n",
      "5993\n",
      "5994\n",
      "5995\n",
      "5996\n",
      "5997\n",
      "5998\n",
      "5999\n",
      "6000\n",
      "6001\n",
      "6002\n",
      "6003\n",
      "6004\n",
      "6005\n",
      "6006\n",
      "6007\n",
      "6008\n",
      "6009\n",
      "6010\n",
      "6011\n",
      "6012\n",
      "6013\n",
      "6014\n",
      "6015\n",
      "6016\n",
      "6017\n",
      "6018\n",
      "6019\n",
      "6020\n",
      "6021\n",
      "6022\n",
      "6023\n",
      "6024\n",
      "6025\n",
      "6026\n",
      "6027\n",
      "6028\n",
      "6029\n",
      "6030\n",
      "6031\n",
      "6032\n",
      "6033\n",
      "6034\n",
      "6035\n",
      "1234567891011121314151617181920212223242526272829303132333435363738394041424344454647484950515253545556575859606162636465666768697071727374757677787980818283848586878889909192939495969798991001011021031041051061071081091101111121131141151161171181191201211221231241251261271281291301311321331341351361371381391401411421431441451461471481491501511521\n",
      "2\n",
      "3\n",
      "4\n",
      "5\n",
      "6\n",
      "7\n",
      "8\n",
      "9\n",
      "10\n",
      "11\n",
      "12\n",
      "13\n",
      "14\n",
      "15\n",
      "16\n",
      "17\n",
      "18\n",
      "19\n",
      "20\n",
      "21\n",
      "22\n",
      "23\n",
      "24\n",
      "25\n",
      "26\n",
      "27\n",
      "28\n",
      "29\n",
      "30\n",
      "31\n",
      "32\n",
      "33\n",
      "34\n",
      "35\n",
      "36\n",
      "37\n",
      "38\n",
      "39\n",
      "40\n",
      "41\n",
      "42\n",
      "43\n",
      "44\n",
      "45\n",
      "46\n",
      "47\n",
      "48\n",
      "49\n",
      "50\n",
      "51\n",
      "52\n",
      "53\n",
      "54\n",
      "55\n",
      "56\n",
      "57\n",
      "58\n",
      "59\n",
      "60\n",
      "61\n",
      "62\n",
      "63\n",
      "64\n",
      "65\n",
      "66\n",
      "67\n",
      "68\n",
      "69\n",
      "70\n",
      "71\n",
      "72\n",
      "73\n",
      "74\n",
      "75\n",
      "76\n",
      "77\n",
      "78\n",
      "79\n",
      "80\n",
      "81\n",
      "82\n",
      "83\n",
      "84\n",
      "85\n",
      "86\n",
      "87\n",
      "88\n",
      "89\n",
      "90\n",
      "91\n",
      "92\n",
      "93\n",
      "94\n",
      "95\n",
      "96\n",
      "97\n",
      "98\n",
      "99\n",
      "100\n",
      "101\n",
      "102\n",
      "103\n",
      "104\n",
      "105\n",
      "106\n",
      "107\n",
      "108\n",
      "109\n",
      "110\n",
      "111\n",
      "112\n",
      "113\n",
      "114\n",
      "115\n",
      "116\n",
      "117\n",
      "118\n",
      "119\n",
      "120\n",
      "121\n",
      "122\n",
      "123\n",
      "124\n",
      "125\n",
      "126\n",
      "127\n",
      "128\n",
      "129\n",
      "130\n",
      "131\n",
      "132\n",
      "133\n",
      "134\n",
      "135\n",
      "136\n",
      "137\n",
      "138\n",
      "139\n",
      "140\n",
      "141\n",
      "142\n",
      "143\n",
      "144\n",
      "145\n",
      "146\n",
      "147\n",
      "148\n",
      "149\n",
      "150\n",
      "151\n",
      "1234567891011121314151617181920212223242526272829303132333435363738394041424344454647484950515253545556575859606162636465666768697071727374757677787980818283848586878889909192939495969798991001011021031041051061071081091101111121131141151161171181191201211221231241251261271281291301311321331341351361371381391401411421431441451461471481491501511521531541551561571581591601611621631641651661671681691701711721731741751761771781791801811821831841851861871881891901911921931941951961971981992002012022032042052062072082092102112122132142152162172182192202212222232242252262272282292302312322332342352362372382392402412422432442452462472482492502512522532542552562572582592602612622632642652662672682692702712722732742752762772782792802812822832842852862872882892902912922932942952962972982993003013023033043053063073083093103113123133143153163173183193201\n",
      "2\n",
      "3\n",
      "4\n",
      "5\n",
      "6\n",
      "7\n",
      "8\n",
      "9\n",
      "10\n",
      "11\n",
      "12\n",
      "13\n",
      "14\n",
      "15\n",
      "16\n",
      "17\n",
      "18\n",
      "19\n",
      "20\n",
      "21\n",
      "22\n",
      "23\n",
      "24\n",
      "25\n",
      "26\n",
      "27\n",
      "28\n",
      "29\n",
      "30\n",
      "31\n",
      "32\n",
      "33\n",
      "34\n",
      "35\n",
      "36\n",
      "37\n",
      "38\n",
      "39\n",
      "40\n",
      "41\n",
      "42\n",
      "43\n",
      "44\n",
      "45\n",
      "46\n",
      "47\n",
      "48\n",
      "49\n",
      "50\n",
      "51\n",
      "52\n",
      "53\n",
      "54\n",
      "55\n",
      "56\n",
      "57\n",
      "58\n",
      "59\n",
      "60\n",
      "61\n",
      "62\n",
      "63\n",
      "64\n",
      "65\n",
      "66\n",
      "67\n",
      "68\n",
      "69\n",
      "70\n",
      "71\n",
      "72\n",
      "73\n",
      "74\n",
      "75\n",
      "76\n",
      "77\n",
      "78\n",
      "79\n",
      "80\n",
      "81\n",
      "82\n",
      "83\n",
      "84\n",
      "85\n",
      "86\n",
      "87\n",
      "88\n",
      "89\n",
      "90\n",
      "91\n",
      "92\n",
      "93\n",
      "94\n",
      "95\n",
      "96\n",
      "97\n",
      "98\n",
      "99\n",
      "100\n",
      "101\n",
      "102\n",
      "103\n",
      "104\n",
      "105\n",
      "106\n",
      "107\n",
      "108\n",
      "109\n",
      "110\n",
      "111\n",
      "112\n",
      "113\n",
      "114\n",
      "115\n",
      "116\n",
      "117\n",
      "118\n",
      "119\n",
      "120\n",
      "121\n",
      "122\n",
      "123\n",
      "124\n",
      "125\n",
      "126\n",
      "127\n",
      "128\n",
      "129\n",
      "130\n",
      "131\n",
      "132\n",
      "133\n",
      "134\n",
      "135\n",
      "136\n",
      "137\n",
      "138\n",
      "139\n",
      "140\n",
      "141\n",
      "142\n",
      "143\n",
      "144\n",
      "145\n",
      "146\n",
      "147\n",
      "148\n",
      "149\n",
      "150\n",
      "151\n",
      "152\n",
      "153\n",
      "154\n",
      "155\n",
      "156\n",
      "157\n",
      "158\n",
      "159\n",
      "160\n",
      "161\n",
      "162\n",
      "163\n",
      "164\n",
      "165\n",
      "166\n",
      "167\n",
      "168\n",
      "169\n",
      "170\n",
      "171\n",
      "172\n",
      "173\n",
      "174\n",
      "175\n",
      "176\n",
      "177\n",
      "178\n",
      "179\n",
      "180\n",
      "181\n",
      "182\n",
      "183\n",
      "184\n",
      "185\n",
      "186\n",
      "187\n",
      "188\n",
      "189\n",
      "190\n",
      "191\n",
      "192\n",
      "193\n",
      "194\n",
      "195\n",
      "196\n",
      "197\n",
      "198\n",
      "199\n",
      "200\n",
      "201\n",
      "202\n",
      "203\n",
      "204\n",
      "205\n",
      "206\n",
      "207\n",
      "208\n",
      "209\n",
      "210\n",
      "211\n",
      "212\n",
      "213\n",
      "214\n",
      "215\n",
      "216\n",
      "217\n",
      "218\n",
      "219\n",
      "220\n",
      "221\n",
      "222\n",
      "223\n",
      "224\n",
      "225\n",
      "226\n",
      "227\n",
      "228\n",
      "229\n",
      "230\n",
      "231\n",
      "232\n",
      "233\n",
      "234\n",
      "235\n",
      "236\n",
      "237\n",
      "238\n",
      "239\n",
      "240\n",
      "241\n",
      "242\n",
      "243\n",
      "244\n",
      "245\n",
      "246\n",
      "247\n",
      "248\n",
      "249\n",
      "250\n",
      "251\n",
      "252\n",
      "253\n",
      "254\n",
      "255\n",
      "256\n",
      "257\n",
      "258\n",
      "259\n",
      "260\n",
      "261\n",
      "262\n",
      "263\n",
      "264\n",
      "265\n",
      "266\n",
      "267\n",
      "268\n",
      "269\n",
      "270\n",
      "271\n",
      "272\n",
      "273\n",
      "274\n",
      "275\n",
      "276\n",
      "277\n",
      "278\n",
      "279\n",
      "280\n",
      "281\n",
      "282\n",
      "283\n",
      "284\n",
      "285\n",
      "286\n",
      "287\n",
      "288\n",
      "289\n",
      "290\n",
      "291\n",
      "292\n",
      "293\n",
      "294\n",
      "295\n",
      "296\n",
      "297\n",
      "298\n",
      "299\n",
      "300\n",
      "301\n",
      "302\n",
      "303\n",
      "304\n",
      "305\n",
      "306\n",
      "307\n",
      "308\n",
      "309\n",
      "310\n",
      "311\n",
      "312\n",
      "313\n",
      "314\n",
      "315\n",
      "316\n",
      "317\n",
      "318\n",
      "319\n",
      "1234567891011121314151617181920212223242526272829303132333435363738394041424344454647484950515253545556575859606162636465666768697071727374757677787980818283848586878889909192939495969798991001011021031041051061071081091101111121131141151161171181191201211221231241251261271281291301311321331341351361371381391401411421431441451461471481491501511521531541551561571581591601611621631641651661671681691701711721731741751761771781791801811821831841851861871881891901911921931941951961971981992002012022032042052062072082092102112122132142152162172182192202212222232242252262272282292302312322332342352362372382392402412422432442452462472482492502512522532542552562572582592602612622632642652662672682692702712722732742752762772782792802812822832842852862872882892902912922932942952962972982993003013023033043053063073083093103113123133143153163173183193203213223233243253263273283293303313323333343353363373383393403413423433443453463473483493503513523533543553563573583593603613623633643653663673683693703713723733743753763773783793803813823833843853863873883893903913923933943953963973983994004014024034044054064074084094104114124134144154164174184194204214224234244254264274284294304314324334344354364374384394404414424434444454464474484494504514524534544554564574584594604614624634644654664674684694704714724734744754764774784794804814824834844854864874884894904914924934944954964974984995005015025035045055065075085095105115125135145155165175185195205215225235245255265275285295305315325335345355365375385395405415425435445455465475485495505515525535545555565575585595605615625635645655665675685695705715725735745755765775785795805815825835845855865875885895905915925935945955965975985996006016026036046056066076086096106116126136146156166176186196206216226236246256266276286296306316326336346356366376386396406416426436446456466476486496506516526536546556566576586596606616626636646656666676686696706716726736746756766776786796806816826836846856866876886896906916926936946956966976986997007017027037047057067077087097107117127137147157167177187197207217227237247257267277287297307317327337347357367377387397407417427437447457467477487497507517527537547557567577587597607617627637647657667677687697707717727737747757767777787797807817827837847857867877887897907917927937947957967977987998008018028038048058068078088098108118128138148158168178188198208218228238248258268278288298308318328338348358368378388398408418428438448458468478488498508518528538548558568578588598608618628638648658668678688698708718728738748758768778788798808818828838848858868878888898908918928938948958968978988999009019029039049059069079089099109119129139149159169179189199209219229239249259269279289299309319329339349359369379389399409419429439449459469479489499509519529539549559569579589599609619629639649659669679689699709719729739749759769779789799809819829839849859869879889899909919929939949959969979989991000100110021003100410051006100710081009101010111012101310141015101610171018101910201021102210231024102510261027102810291030103110321033103410351036103710381039104010411042104310441045104610471048104910501051105210531054105510561057105810591060106110621063106410651066106710681069107010711072107310741075107610771078107910801081108210831084108510861087108810891090109110921093109410951096109710981099110011011102110311041105110611071108110911101111111211131114111511161117111811191120112111221123112411251126112711281129113011311132113311341135113611371138113911401141114211431144114511461147114811491150115111521153115411551156115711581159116011611162116311641165116611671168116911701171117211731174117511761177117811791180118111821183118411851186118711881189119011911192119311941195119611971198119912001201120212031204120512061207120812091210121112121213121412151216121712181219122012211222122312241225122612271228122912301231123212331234123512361237123812391240124112421243124412451246124712481249125012511252125312541255125612571258125912601261126212631264126512661267126812691270127112721273127412751276127712781279128012811282128312841285128612871288128912901291129212931294129512961297129812991300130113021303130413051306130713081309131013111312131313141315131613171318131913201321132213231324132513261327132813291330133113321333133413351336133713381339134013411342134313441345134613471348134913501351135213531354135513561357135813591360136113621363136413651366136713681369137013711372137313741375137613771378137913801381138213831384138513861387138813891390139113921393139413951396139713981399140014011402140314041405140614071408140914101411141214131414141514161417141814191420142114221423142414251426142714281429143014311432143314341435143614371438143914401441144214431444144514461447144814491450145114521453145414551456145714581459146014611462146314641465146614671468146914701471147214731474147514761477147814791480148114821483148414851486148714881489149014911492149314941495149614971498149915001501150215031504150515061507150815091510151115121513151415151516151715181519152015211522152315241525152615271528152915301531153215331534153515361537153815391540154115421543154415451546154715481549155015511552155315541555155615571558155915601561156215631564156515661567156815691570157115721573157415751576157715781579158015811582158315841585158615871588158915901591159215931594159515961597159815991600160116021603160416051606160716081609161016111612161316141615161616171618161916201621162216231624162516261627162816291630163116321633163416351636163716381639164016411642164316441645164616471648164916501651165216531654165516561657165816591660166116621663166416651666166716681669167016711672167316741675167616771678167916801681168216831684168516861687168816891690169116921693169416951696169716981699170017011702170317041705170617071708170917101711171217131714171517161717171817191720172117221723172417251726172717281729173017311732173317341735173617371738173917401741174217431744174517461747174817491750175117521753175417551756175717581759176017611762176317641765176617671768176917701771177217731774177517761777177817791780178117821783178417851786178717881789179017911792179317941795179617971798179918001801180218031804180518061807180818091810181118121813181418151816181718181\n",
      "2\n",
      "3\n",
      "4\n",
      "5\n",
      "6\n",
      "7\n",
      "8\n",
      "9\n",
      "10\n",
      "11\n",
      "12\n",
      "13\n",
      "14\n",
      "15\n",
      "16\n",
      "17\n",
      "18\n",
      "19\n",
      "20\n",
      "21\n",
      "22\n",
      "23\n",
      "24\n",
      "25\n",
      "26\n",
      "27\n",
      "28\n",
      "29\n",
      "30\n",
      "31\n",
      "32\n",
      "33\n",
      "34\n",
      "35\n",
      "36\n",
      "37\n",
      "38\n",
      "39\n",
      "40\n",
      "41\n",
      "42\n",
      "43\n",
      "44\n",
      "45\n",
      "46\n",
      "47\n",
      "48\n",
      "49\n",
      "50\n",
      "51\n",
      "52\n",
      "53\n",
      "54\n",
      "55\n",
      "56\n",
      "57\n",
      "58\n",
      "59\n",
      "60\n",
      "61\n",
      "62\n",
      "63\n",
      "64\n",
      "65\n",
      "66\n",
      "67\n",
      "68\n",
      "69\n",
      "70\n",
      "71\n",
      "72\n",
      "73\n",
      "74\n",
      "75\n",
      "76\n",
      "77\n",
      "78\n",
      "79\n",
      "80\n",
      "81\n",
      "82\n",
      "83\n",
      "84\n",
      "85\n",
      "86\n",
      "87\n",
      "88\n",
      "89\n",
      "90\n",
      "91\n",
      "92\n",
      "93\n",
      "94\n",
      "95\n",
      "96\n",
      "97\n",
      "98\n",
      "99\n",
      "100\n",
      "101\n",
      "102\n",
      "103\n",
      "104\n",
      "105\n",
      "106\n",
      "107\n",
      "108\n",
      "109\n",
      "110\n",
      "111\n",
      "112\n",
      "113\n",
      "114\n",
      "115\n",
      "116\n",
      "117\n",
      "118\n",
      "119\n",
      "120\n",
      "121\n",
      "122\n",
      "123\n",
      "124\n",
      "125\n",
      "126\n",
      "127\n",
      "128\n",
      "129\n",
      "130\n",
      "131\n",
      "132\n",
      "133\n",
      "134\n",
      "135\n",
      "136\n",
      "137\n",
      "138\n",
      "139\n",
      "140\n",
      "141\n",
      "142\n",
      "143\n",
      "144\n",
      "145\n",
      "146\n",
      "147\n",
      "148\n",
      "149\n",
      "150\n",
      "151\n",
      "152\n",
      "153\n",
      "154\n",
      "155\n",
      "156\n",
      "157\n",
      "158\n",
      "159\n",
      "160\n",
      "161\n",
      "162\n",
      "163\n",
      "164\n",
      "165\n",
      "166\n",
      "167\n",
      "168\n",
      "169\n",
      "170\n",
      "171\n",
      "172\n",
      "173\n",
      "174\n",
      "175\n",
      "176\n",
      "177\n",
      "178\n",
      "179\n",
      "180\n",
      "181\n",
      "182\n",
      "183\n",
      "184\n",
      "185\n",
      "186\n",
      "187\n",
      "188\n",
      "189\n",
      "190\n",
      "191\n",
      "192\n",
      "193\n",
      "194\n",
      "195\n",
      "196\n",
      "197\n",
      "198\n",
      "199\n",
      "200\n",
      "201\n",
      "202\n",
      "203\n",
      "204\n",
      "205\n",
      "206\n",
      "207\n",
      "208\n",
      "209\n",
      "210\n",
      "211\n",
      "212\n",
      "213\n",
      "214\n",
      "215\n",
      "216\n",
      "217\n",
      "218\n",
      "219\n",
      "220\n",
      "221\n",
      "222\n",
      "223\n",
      "224\n",
      "225\n",
      "226\n",
      "227\n",
      "228\n",
      "229\n",
      "230\n",
      "231\n",
      "232\n",
      "233\n",
      "234\n",
      "235\n",
      "236\n",
      "237\n",
      "238\n",
      "239\n",
      "240\n",
      "241\n",
      "242\n",
      "243\n",
      "244\n",
      "245\n",
      "246\n",
      "247\n",
      "248\n",
      "249\n",
      "250\n",
      "251\n",
      "252\n",
      "253\n",
      "254\n",
      "255\n",
      "256\n",
      "257\n",
      "258\n",
      "259\n",
      "260\n",
      "261\n",
      "262\n",
      "263\n",
      "264\n",
      "265\n",
      "266\n",
      "267\n",
      "268\n",
      "269\n",
      "270\n",
      "271\n",
      "272\n",
      "273\n",
      "274\n",
      "275\n",
      "276\n",
      "277\n",
      "278\n",
      "279\n",
      "280\n",
      "281\n",
      "282\n",
      "283\n",
      "284\n",
      "285\n",
      "286\n",
      "287\n",
      "288\n",
      "289\n",
      "290\n",
      "291\n",
      "292\n",
      "293\n",
      "294\n",
      "295\n",
      "296\n",
      "297\n",
      "298\n",
      "299\n",
      "300\n",
      "301\n",
      "302\n",
      "303\n",
      "304\n",
      "305\n",
      "306\n",
      "307\n",
      "308\n",
      "309\n",
      "310\n",
      "311\n",
      "312\n",
      "313\n",
      "314\n",
      "315\n",
      "316\n",
      "317\n",
      "318\n",
      "319\n",
      "320\n",
      "321\n",
      "322\n",
      "323\n",
      "324\n",
      "325\n",
      "326\n",
      "327\n",
      "328\n",
      "329\n",
      "330\n",
      "331\n",
      "332\n",
      "333\n",
      "334\n",
      "335\n",
      "336\n",
      "337\n",
      "338\n",
      "339\n",
      "340\n",
      "341\n",
      "342\n",
      "343\n",
      "344\n",
      "345\n",
      "346\n",
      "347\n",
      "348\n",
      "349\n",
      "350\n",
      "351\n",
      "352\n",
      "353\n",
      "354\n",
      "355\n",
      "356\n",
      "357\n",
      "358\n",
      "359\n",
      "360\n",
      "361\n",
      "362\n",
      "363\n",
      "364\n",
      "365\n",
      "366\n",
      "367\n",
      "368\n",
      "369\n",
      "370\n",
      "371\n",
      "372\n",
      "373\n",
      "374\n",
      "375\n",
      "376\n",
      "377\n",
      "378\n",
      "379\n",
      "380\n",
      "381\n",
      "382\n",
      "383\n",
      "384\n",
      "385\n",
      "386\n",
      "387\n",
      "388\n",
      "389\n",
      "390\n",
      "391\n",
      "392\n",
      "393\n",
      "394\n",
      "395\n",
      "396\n",
      "397\n",
      "398\n",
      "399\n",
      "400\n",
      "401\n",
      "402\n",
      "403\n",
      "404\n",
      "405\n",
      "406\n",
      "407\n",
      "408\n",
      "409\n",
      "410\n",
      "411\n",
      "412\n",
      "413\n",
      "414\n",
      "415\n",
      "416\n",
      "417\n",
      "418\n",
      "419\n",
      "420\n",
      "421\n",
      "422\n",
      "423\n",
      "424\n",
      "425\n",
      "426\n",
      "427\n",
      "428\n",
      "429\n",
      "430\n",
      "431\n",
      "432\n",
      "433\n",
      "434\n",
      "435\n",
      "436\n",
      "437\n",
      "438\n",
      "439\n",
      "440\n",
      "441\n",
      "442\n",
      "443\n",
      "444\n",
      "445\n",
      "446\n",
      "447\n",
      "448\n",
      "449\n",
      "450\n",
      "451\n",
      "452\n",
      "453\n",
      "454\n",
      "455\n",
      "456\n",
      "457\n",
      "458\n",
      "459\n",
      "460\n",
      "461\n",
      "462\n",
      "463\n",
      "464\n",
      "465\n",
      "466\n",
      "467\n",
      "468\n",
      "469\n",
      "470\n",
      "471\n",
      "472\n",
      "473\n",
      "474\n",
      "475\n",
      "476\n",
      "477\n",
      "478\n",
      "479\n",
      "480\n",
      "481\n",
      "482\n",
      "483\n",
      "484\n",
      "485\n",
      "486\n",
      "487\n",
      "488\n",
      "489\n",
      "490\n",
      "491\n",
      "492\n",
      "493\n",
      "494\n",
      "495\n",
      "496\n",
      "497\n",
      "498\n",
      "499\n",
      "500\n",
      "501\n",
      "502\n",
      "503\n",
      "504\n",
      "505\n",
      "506\n",
      "507\n",
      "508\n",
      "509\n",
      "510\n",
      "511\n",
      "512\n",
      "513\n",
      "514\n",
      "515\n",
      "516\n",
      "517\n",
      "518\n",
      "519\n",
      "520\n",
      "521\n",
      "522\n",
      "523\n",
      "524\n",
      "525\n",
      "526\n",
      "527\n",
      "528\n",
      "529\n",
      "530\n",
      "531\n",
      "532\n",
      "533\n",
      "534\n",
      "535\n",
      "536\n",
      "537\n",
      "538\n",
      "539\n",
      "540\n",
      "541\n",
      "542\n",
      "543\n",
      "544\n",
      "545\n",
      "546\n",
      "547\n",
      "548\n",
      "549\n",
      "550\n",
      "551\n",
      "552\n",
      "553\n",
      "554\n",
      "555\n",
      "556\n",
      "557\n",
      "558\n",
      "559\n",
      "560\n",
      "561\n",
      "562\n",
      "563\n",
      "564\n",
      "565\n",
      "566\n",
      "567\n",
      "568\n",
      "569\n",
      "570\n",
      "571\n",
      "572\n",
      "573\n",
      "574\n",
      "575\n",
      "576\n",
      "577\n",
      "578\n",
      "579\n",
      "580\n",
      "581\n",
      "582\n",
      "583\n",
      "584\n",
      "585\n",
      "586\n",
      "587\n",
      "588\n",
      "589\n",
      "590\n",
      "591\n",
      "592\n",
      "593\n",
      "594\n",
      "595\n",
      "596\n",
      "597\n",
      "598\n",
      "599\n",
      "600\n",
      "601\n",
      "602\n",
      "603\n",
      "604\n",
      "605\n",
      "606\n",
      "607\n",
      "608\n",
      "609\n",
      "610\n",
      "611\n",
      "612\n",
      "613\n",
      "614\n",
      "615\n",
      "616\n",
      "617\n",
      "618\n",
      "619\n",
      "620\n",
      "621\n",
      "622\n",
      "623\n",
      "624\n",
      "625\n",
      "626\n",
      "627\n",
      "628\n",
      "629\n",
      "630\n",
      "631\n",
      "632\n",
      "633\n",
      "634\n",
      "635\n",
      "636\n",
      "637\n",
      "638\n",
      "639\n",
      "640\n",
      "641\n",
      "642\n",
      "643\n",
      "644\n",
      "645\n",
      "646\n",
      "647\n",
      "648\n",
      "649\n",
      "650\n",
      "651\n",
      "652\n",
      "653\n",
      "654\n",
      "655\n",
      "656\n",
      "657\n",
      "658\n",
      "659\n",
      "660\n",
      "661\n",
      "662\n",
      "663\n",
      "664\n",
      "665\n",
      "666\n",
      "667\n",
      "668\n",
      "669\n",
      "670\n",
      "671\n",
      "672\n",
      "673\n",
      "674\n",
      "675\n",
      "676\n",
      "677\n",
      "678\n",
      "679\n",
      "680\n",
      "681\n",
      "682\n",
      "683\n",
      "684\n",
      "685\n",
      "686\n",
      "687\n",
      "688\n",
      "689\n",
      "690\n",
      "691\n",
      "692\n",
      "693\n",
      "694\n",
      "695\n",
      "696\n",
      "697\n",
      "698\n",
      "699\n",
      "700\n",
      "701\n",
      "702\n",
      "703\n",
      "704\n",
      "705\n",
      "706\n",
      "707\n",
      "708\n",
      "709\n",
      "710\n",
      "711\n",
      "712\n",
      "713\n",
      "714\n",
      "715\n",
      "716\n",
      "717\n",
      "718\n",
      "719\n",
      "720\n",
      "721\n",
      "722\n",
      "723\n",
      "724\n",
      "725\n",
      "726\n",
      "727\n",
      "728\n",
      "729\n",
      "730\n",
      "731\n",
      "732\n",
      "733\n",
      "734\n",
      "735\n",
      "736\n",
      "737\n",
      "738\n",
      "739\n",
      "740\n",
      "741\n",
      "742\n",
      "743\n",
      "744\n",
      "745\n",
      "746\n",
      "747\n",
      "748\n",
      "749\n",
      "750\n",
      "751\n",
      "752\n",
      "753\n",
      "754\n",
      "755\n",
      "756\n",
      "757\n",
      "758\n",
      "759\n",
      "760\n",
      "761\n",
      "762\n",
      "763\n",
      "764\n",
      "765\n",
      "766\n",
      "767\n",
      "768\n",
      "769\n",
      "770\n",
      "771\n",
      "772\n",
      "773\n",
      "774\n",
      "775\n",
      "776\n",
      "777\n",
      "778\n",
      "779\n",
      "780\n",
      "781\n",
      "782\n",
      "783\n",
      "784\n",
      "785\n",
      "786\n",
      "787\n",
      "788\n",
      "789\n",
      "790\n",
      "791\n",
      "792\n",
      "793\n",
      "794\n",
      "795\n",
      "796\n",
      "797\n",
      "798\n",
      "799\n",
      "800\n",
      "801\n",
      "802\n",
      "803\n",
      "804\n",
      "805\n",
      "806\n",
      "807\n",
      "808\n",
      "809\n",
      "810\n",
      "811\n",
      "812\n",
      "813\n",
      "814\n",
      "815\n",
      "816\n",
      "817\n",
      "818\n",
      "819\n",
      "820\n",
      "821\n",
      "822\n",
      "823\n",
      "824\n",
      "825\n",
      "826\n",
      "827\n",
      "828\n",
      "829\n",
      "830\n",
      "831\n",
      "832\n",
      "833\n",
      "834\n",
      "835\n",
      "836\n",
      "837\n",
      "838\n",
      "839\n",
      "840\n",
      "841\n",
      "842\n",
      "843\n",
      "844\n",
      "845\n",
      "846\n",
      "847\n",
      "848\n",
      "849\n",
      "850\n",
      "851\n",
      "852\n",
      "853\n",
      "854\n",
      "855\n",
      "856\n",
      "857\n",
      "858\n",
      "859\n",
      "860\n",
      "861\n",
      "862\n",
      "863\n",
      "864\n",
      "865\n",
      "866\n",
      "867\n",
      "868\n",
      "869\n",
      "870\n",
      "871\n",
      "872\n",
      "873\n",
      "874\n",
      "875\n",
      "876\n",
      "877\n",
      "878\n",
      "879\n",
      "880\n",
      "881\n",
      "882\n",
      "883\n",
      "884\n",
      "885\n",
      "886\n",
      "887\n",
      "888\n",
      "889\n",
      "890\n",
      "891\n",
      "892\n",
      "893\n",
      "894\n",
      "895\n",
      "896\n",
      "897\n",
      "898\n",
      "899\n",
      "900\n",
      "901\n",
      "902\n",
      "903\n",
      "904\n",
      "905\n",
      "906\n",
      "907\n",
      "908\n",
      "909\n",
      "910\n",
      "911\n",
      "912\n",
      "913\n",
      "914\n",
      "915\n",
      "916\n",
      "917\n",
      "918\n",
      "919\n",
      "920\n",
      "921\n",
      "922\n",
      "923\n",
      "924\n",
      "925\n",
      "926\n",
      "927\n",
      "928\n",
      "929\n",
      "930\n",
      "931\n",
      "932\n",
      "933\n",
      "934\n",
      "935\n",
      "936\n",
      "937\n",
      "938\n",
      "939\n",
      "940\n",
      "941\n",
      "942\n",
      "943\n",
      "944\n",
      "945\n",
      "946\n",
      "947\n",
      "948\n",
      "949\n",
      "950\n",
      "951\n",
      "952\n",
      "953\n",
      "954\n",
      "955\n",
      "956\n",
      "957\n",
      "958\n",
      "959\n",
      "960\n",
      "961\n",
      "962\n",
      "963\n",
      "964\n",
      "965\n",
      "966\n",
      "967\n",
      "968\n",
      "969\n",
      "970\n",
      "971\n",
      "972\n",
      "973\n",
      "974\n",
      "975\n",
      "976\n",
      "977\n",
      "978\n",
      "979\n",
      "980\n",
      "981\n",
      "982\n",
      "983\n",
      "984\n",
      "985\n",
      "986\n",
      "987\n",
      "988\n",
      "989\n",
      "990\n",
      "991\n",
      "992\n",
      "993\n",
      "994\n",
      "995\n",
      "996\n",
      "997\n",
      "998\n",
      "999\n",
      "1000\n",
      "1001\n",
      "1002\n",
      "1003\n",
      "1004\n",
      "1005\n",
      "1006\n",
      "1007\n",
      "1008\n",
      "1009\n",
      "1010\n",
      "1011\n",
      "1012\n",
      "1013\n",
      "1014\n",
      "1015\n",
      "1016\n",
      "1017\n",
      "1018\n",
      "1019\n",
      "1020\n",
      "1021\n",
      "1022\n",
      "1023\n",
      "1024\n",
      "1025\n",
      "1026\n",
      "1027\n",
      "1028\n",
      "1029\n",
      "1030\n",
      "1031\n",
      "1032\n",
      "1033\n",
      "1034\n",
      "1035\n",
      "1036\n",
      "1037\n",
      "1038\n",
      "1039\n",
      "1040\n",
      "1041\n",
      "1042\n",
      "1043\n",
      "1044\n",
      "1045\n",
      "1046\n",
      "1047\n",
      "1048\n",
      "1049\n",
      "1050\n",
      "1051\n",
      "1052\n",
      "1053\n",
      "1054\n",
      "1055\n",
      "1056\n",
      "1057\n",
      "1058\n",
      "1059\n",
      "1060\n",
      "1061\n",
      "1062\n",
      "1063\n",
      "1064\n",
      "1065\n",
      "1066\n",
      "1067\n",
      "1068\n",
      "1069\n",
      "1070\n",
      "1071\n",
      "1072\n",
      "1073\n",
      "1074\n",
      "1075\n",
      "1076\n",
      "1077\n",
      "1078\n",
      "1079\n",
      "1080\n",
      "1081\n",
      "1082\n",
      "1083\n",
      "1084\n",
      "1085\n",
      "1086\n",
      "1087\n",
      "1088\n",
      "1089\n",
      "1090\n",
      "1091\n",
      "1092\n",
      "1093\n",
      "1094\n",
      "1095\n",
      "1096\n",
      "1097\n",
      "1098\n",
      "1099\n",
      "1100\n",
      "1101\n",
      "1102\n",
      "1103\n",
      "1104\n",
      "1105\n",
      "1106\n",
      "1107\n",
      "1108\n",
      "1109\n",
      "1110\n",
      "1111\n",
      "1112\n",
      "1113\n",
      "1114\n",
      "1115\n",
      "1116\n",
      "1117\n",
      "1118\n",
      "1119\n",
      "1120\n",
      "1121\n",
      "1122\n",
      "1123\n",
      "1124\n",
      "1125\n",
      "1126\n",
      "1127\n",
      "1128\n",
      "1129\n",
      "1130\n",
      "1131\n",
      "1132\n",
      "1133\n",
      "1134\n",
      "1135\n",
      "1136\n",
      "1137\n",
      "1138\n",
      "1139\n",
      "1140\n",
      "1141\n",
      "1142\n",
      "1143\n",
      "1144\n",
      "1145\n",
      "1146\n",
      "1147\n",
      "1148\n",
      "1149\n",
      "1150\n",
      "1151\n",
      "1152\n",
      "1153\n",
      "1154\n",
      "1155\n",
      "1156\n",
      "1157\n",
      "1158\n",
      "1159\n",
      "1160\n",
      "1161\n",
      "1162\n",
      "1163\n",
      "1164\n",
      "1165\n",
      "1166\n",
      "1167\n",
      "1168\n",
      "1169\n",
      "1170\n",
      "1171\n",
      "1172\n",
      "1173\n",
      "1174\n",
      "1175\n",
      "1176\n",
      "1177\n",
      "1178\n",
      "1179\n",
      "1180\n",
      "1181\n",
      "1182\n",
      "1183\n",
      "1184\n",
      "1185\n",
      "1186\n",
      "1187\n",
      "1188\n",
      "1189\n",
      "1190\n",
      "1191\n",
      "1192\n",
      "1193\n",
      "1194\n",
      "1195\n",
      "1196\n",
      "1197\n",
      "1198\n",
      "1199\n",
      "1200\n",
      "1201\n",
      "1202\n",
      "1203\n",
      "1204\n",
      "1205\n",
      "1206\n",
      "1207\n",
      "1208\n",
      "1209\n",
      "1210\n",
      "1211\n",
      "1212\n",
      "1213\n",
      "1214\n",
      "1215\n",
      "1216\n",
      "1217\n",
      "1218\n",
      "1219\n",
      "1220\n",
      "1221\n",
      "1222\n",
      "1223\n",
      "1224\n",
      "1225\n",
      "1226\n",
      "1227\n",
      "1228\n",
      "1229\n",
      "1230\n",
      "1231\n",
      "1232\n",
      "1233\n",
      "1234\n",
      "1235\n",
      "1236\n",
      "1237\n",
      "1238\n",
      "1239\n",
      "1240\n",
      "1241\n",
      "1242\n",
      "1243\n",
      "1244\n",
      "1245\n",
      "1246\n",
      "1247\n",
      "1248\n",
      "1249\n",
      "1250\n",
      "1251\n",
      "1252\n",
      "1253\n",
      "1254\n",
      "1255\n",
      "1256\n",
      "1257\n",
      "1258\n",
      "1259\n",
      "1260\n",
      "1261\n",
      "1262\n",
      "1263\n",
      "1264\n",
      "1265\n",
      "1266\n",
      "1267\n",
      "1268\n",
      "1269\n",
      "1270\n",
      "1271\n",
      "1272\n",
      "1273\n",
      "1274\n",
      "1275\n",
      "1276\n",
      "1277\n",
      "1278\n",
      "1279\n",
      "1280\n",
      "1281\n",
      "1282\n",
      "1283\n",
      "1284\n",
      "1285\n",
      "1286\n",
      "1287\n",
      "1288\n",
      "1289\n",
      "1290\n",
      "1291\n",
      "1292\n",
      "1293\n",
      "1294\n",
      "1295\n",
      "1296\n",
      "1297\n",
      "1298\n",
      "1299\n",
      "1300\n",
      "1301\n",
      "1302\n",
      "1303\n",
      "1304\n",
      "1305\n",
      "1306\n",
      "1307\n",
      "1308\n",
      "1309\n",
      "1310\n",
      "1311\n",
      "1312\n",
      "1313\n",
      "1314\n",
      "1315\n",
      "1316\n",
      "1317\n",
      "1318\n",
      "1319\n",
      "1320\n",
      "1321\n",
      "1322\n",
      "1323\n",
      "1324\n",
      "1325\n",
      "1326\n",
      "1327\n",
      "1328\n",
      "1329\n",
      "1330\n",
      "1331\n",
      "1332\n",
      "1333\n",
      "1334\n",
      "1335\n",
      "1336\n",
      "1337\n",
      "1338\n",
      "1339\n",
      "1340\n",
      "1341\n",
      "1342\n",
      "1343\n",
      "1344\n",
      "1345\n",
      "1346\n",
      "1347\n",
      "1348\n",
      "1349\n",
      "1350\n",
      "1351\n",
      "1352\n",
      "1353\n",
      "1354\n",
      "1355\n",
      "1356\n",
      "1357\n",
      "1358\n",
      "1359\n",
      "1360\n",
      "1361\n",
      "1362\n",
      "1363\n",
      "1364\n",
      "1365\n",
      "1366\n",
      "1367\n",
      "1368\n",
      "1369\n",
      "1370\n",
      "1371\n",
      "1372\n",
      "1373\n",
      "1374\n",
      "1375\n",
      "1376\n",
      "1377\n",
      "1378\n",
      "1379\n",
      "1380\n",
      "1381\n",
      "1382\n",
      "1383\n",
      "1384\n",
      "1385\n",
      "1386\n",
      "1387\n",
      "1388\n",
      "1389\n",
      "1390\n",
      "1391\n",
      "1392\n",
      "1393\n",
      "1394\n",
      "1395\n",
      "1396\n",
      "1397\n",
      "1398\n",
      "1399\n",
      "1400\n",
      "1401\n",
      "1402\n",
      "1403\n",
      "1404\n",
      "1405\n",
      "1406\n",
      "1407\n",
      "1408\n",
      "1409\n",
      "1410\n",
      "1411\n",
      "1412\n",
      "1413\n",
      "1414\n",
      "1415\n",
      "1416\n",
      "1417\n",
      "1418\n",
      "1419\n",
      "1420\n",
      "1421\n",
      "1422\n",
      "1423\n",
      "1424\n",
      "1425\n",
      "1426\n",
      "1427\n",
      "1428\n",
      "1429\n",
      "1430\n",
      "1431\n",
      "1432\n",
      "1433\n",
      "1434\n",
      "1435\n",
      "1436\n",
      "1437\n",
      "1438\n",
      "1439\n",
      "1440\n",
      "1441\n",
      "1442\n",
      "1443\n",
      "1444\n",
      "1445\n",
      "1446\n",
      "1447\n",
      "1448\n",
      "1449\n",
      "1450\n",
      "1451\n",
      "1452\n",
      "1453\n",
      "1454\n",
      "1455\n",
      "1456\n",
      "1457\n",
      "1458\n",
      "1459\n",
      "1460\n",
      "1461\n",
      "1462\n",
      "1463\n",
      "1464\n",
      "1465\n",
      "1466\n",
      "1467\n",
      "1468\n",
      "1469\n",
      "1470\n",
      "1471\n",
      "1472\n",
      "1473\n",
      "1474\n",
      "1475\n",
      "1476\n",
      "1477\n",
      "1478\n",
      "1479\n",
      "1480\n",
      "1481\n",
      "1482\n",
      "1483\n",
      "1484\n",
      "1485\n",
      "1486\n",
      "1487\n",
      "1488\n",
      "1489\n",
      "1490\n",
      "1491\n",
      "1492\n",
      "1493\n",
      "1494\n",
      "1495\n",
      "1496\n",
      "1497\n",
      "1498\n",
      "1499\n",
      "1500\n",
      "1501\n",
      "1502\n",
      "1503\n",
      "1504\n",
      "1505\n",
      "1506\n",
      "1507\n",
      "1508\n",
      "1509\n",
      "1510\n",
      "1511\n",
      "1512\n",
      "1513\n",
      "1514\n",
      "1515\n",
      "1516\n",
      "1517\n",
      "1518\n",
      "1519\n",
      "1520\n",
      "1521\n",
      "1522\n",
      "1523\n",
      "1524\n",
      "1525\n",
      "1526\n",
      "1527\n",
      "1528\n",
      "1529\n",
      "1530\n",
      "1531\n",
      "1532\n",
      "1533\n",
      "1534\n",
      "1535\n",
      "1536\n",
      "1537\n",
      "1538\n",
      "1539\n",
      "1540\n",
      "1541\n",
      "1542\n",
      "1543\n",
      "1544\n",
      "1545\n",
      "1546\n",
      "1547\n",
      "1548\n",
      "1549\n",
      "1550\n",
      "1551\n",
      "1552\n",
      "1553\n",
      "1554\n",
      "1555\n",
      "1556\n",
      "1557\n",
      "1558\n",
      "1559\n",
      "1560\n",
      "1561\n",
      "1562\n",
      "1563\n",
      "1564\n",
      "1565\n",
      "1566\n",
      "1567\n",
      "1568\n",
      "1569\n",
      "1570\n",
      "1571\n",
      "1572\n",
      "1573\n",
      "1574\n",
      "1575\n",
      "1576\n",
      "1577\n",
      "1578\n",
      "1579\n",
      "1580\n",
      "1581\n",
      "1582\n",
      "1583\n",
      "1584\n",
      "1585\n",
      "1586\n",
      "1587\n",
      "1588\n",
      "1589\n",
      "1590\n",
      "1591\n",
      "1592\n",
      "1593\n",
      "1594\n",
      "1595\n",
      "1596\n",
      "1597\n",
      "1598\n",
      "1599\n",
      "1600\n",
      "1601\n",
      "1602\n",
      "1603\n",
      "1604\n",
      "1605\n",
      "1606\n",
      "1607\n",
      "1608\n",
      "1609\n",
      "1610\n",
      "1611\n",
      "1612\n",
      "1613\n",
      "1614\n",
      "1615\n",
      "1616\n",
      "1617\n",
      "1618\n",
      "1619\n",
      "1620\n",
      "1621\n",
      "1622\n",
      "1623\n",
      "1624\n",
      "1625\n",
      "1626\n",
      "1627\n",
      "1628\n",
      "1629\n",
      "1630\n",
      "1631\n",
      "1632\n",
      "1633\n",
      "1634\n",
      "1635\n",
      "1636\n",
      "1637\n",
      "1638\n",
      "1639\n",
      "1640\n",
      "1641\n",
      "1642\n",
      "1643\n",
      "1644\n",
      "1645\n",
      "1646\n",
      "1647\n",
      "1648\n",
      "1649\n",
      "1650\n",
      "1651\n",
      "1652\n",
      "1653\n",
      "1654\n",
      "1655\n",
      "1656\n",
      "1657\n",
      "1658\n",
      "1659\n",
      "1660\n",
      "1661\n",
      "1662\n",
      "1663\n",
      "1664\n",
      "1665\n",
      "1666\n",
      "1667\n",
      "1668\n",
      "1669\n",
      "1670\n",
      "1671\n",
      "1672\n",
      "1673\n",
      "1674\n",
      "1675\n",
      "1676\n",
      "1677\n",
      "1678\n",
      "1679\n",
      "1680\n",
      "1681\n",
      "1682\n",
      "1683\n",
      "1684\n",
      "1685\n",
      "1686\n",
      "1687\n",
      "1688\n",
      "1689\n",
      "1690\n",
      "1691\n",
      "1692\n",
      "1693\n",
      "1694\n",
      "1695\n",
      "1696\n",
      "1697\n",
      "1698\n",
      "1699\n",
      "1700\n",
      "1701\n",
      "1702\n",
      "1703\n",
      "1704\n",
      "1705\n",
      "1706\n",
      "1707\n",
      "1708\n",
      "1709\n",
      "1710\n",
      "1711\n",
      "1712\n",
      "1713\n",
      "1714\n",
      "1715\n",
      "1716\n",
      "1717\n",
      "1718\n",
      "1719\n",
      "1720\n",
      "1721\n",
      "1722\n",
      "1723\n",
      "1724\n",
      "1725\n",
      "1726\n",
      "1727\n",
      "1728\n",
      "1729\n",
      "1730\n",
      "1731\n",
      "1732\n",
      "1733\n",
      "1734\n",
      "1735\n",
      "1736\n",
      "1737\n",
      "1738\n",
      "1739\n",
      "1740\n",
      "1741\n",
      "1742\n",
      "1743\n",
      "1744\n",
      "1745\n",
      "1746\n",
      "1747\n",
      "1748\n",
      "1749\n",
      "1750\n",
      "1751\n",
      "1752\n",
      "1753\n",
      "1754\n",
      "1755\n",
      "1756\n",
      "1757\n",
      "1758\n",
      "1759\n",
      "1760\n",
      "1761\n",
      "1762\n",
      "1763\n",
      "1764\n",
      "1765\n",
      "1766\n",
      "1767\n",
      "1768\n",
      "1769\n",
      "1770\n",
      "1771\n",
      "1772\n",
      "1773\n",
      "1774\n",
      "1775\n",
      "1776\n",
      "1777\n",
      "1778\n",
      "1779\n",
      "1780\n",
      "1781\n",
      "1782\n",
      "1783\n",
      "1784\n",
      "1785\n",
      "1786\n",
      "1787\n",
      "1788\n",
      "1789\n",
      "1790\n",
      "1791\n",
      "1792\n",
      "1793\n",
      "1794\n",
      "1795\n",
      "1796\n",
      "1797\n",
      "1798\n",
      "1799\n",
      "1800\n",
      "1801\n",
      "1802\n",
      "1803\n",
      "1804\n",
      "1805\n",
      "1806\n",
      "1807\n",
      "1808\n",
      "1809\n",
      "1810\n",
      "1811\n",
      "1812\n",
      "1813\n",
      "1814\n",
      "1815\n",
      "1816\n",
      "1817\n",
      "[1] \"filtering out lowly expressed genes...\"\n",
      "[1] \"getting thematic genesets...\"\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Read 129 records\n",
       "Read 8627 items\n",
       "Read 215 records\n",
       "Read 3223 items\n",
       "Read 267 records\n",
       "Read 4147 items\n",
       "Read 6036 records\n",
       "Read 215704 items\n",
       "Read 152 records\n",
       "Read 2356 items\n",
       "Read 320 records\n",
       "Read 33449 items\n",
       "Read 1818 records\n",
       "Read 116666 items\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R -i pathway_gene_threshold -o pw_data\n",
    "\n",
    "conflicts_prefer(base::intersect)\n",
    "\n",
    "########## get pathway databases for use in analysis ##########\n",
    "###############################################################\n",
    "print('|| getting pathway databases... ||')\n",
    "\n",
    "##### required libraries ####\n",
    "source('../scripts/functions/pathway_setup.r')\n",
    "\n",
    "pw_data = list()\n",
    "\n",
    "# load the data\n",
    "# expressed = readRDS('../data/single_cell_data/expressed_genes_per_celltype.rds')\n",
    "\n",
    "# load all the pathway terms\n",
    "gb = read.geneset('../data/pathway_databases/gabitto_2023_curated_gene_sets/gs.gmt')\n",
    "gl = read.geneset('../data/pathway_databases/galea_2022_curated_gene_sets/gs.gmt')\n",
    "sy = read.geneset('../data/pathway_databases/SynGO_bulk_download_release_20231201/gs.gmt')\n",
    "bp = read.geneset('../data/pathway_databases/GO_Biological_Process_2021.txt')\n",
    "cy = read.geneset('../data/pathway_databases/HumanCyc_2016.txt')\n",
    "ke = read.geneset('../data/pathway_databases/KEGG_2021_Human.txt')\n",
    "re = read.geneset('../data/pathway_databases/Reactome_2022.txt')\n",
    "\n",
    "all_paths = c(bp, cy, ke, re)\n",
    "cura_paths = c(gb, gl, sy)\n",
    "\n",
    "# filter out the lowly expressed genes (10%, for each celltype)\n",
    "print('filtering out lowly expressed genes...')\n",
    "\n",
    "low_removed_gabitto = filter_lowly_exp_genes(expressed_genes_per_celltype, gb, pathway_gene_threshold)\n",
    "low_removed_galea = filter_lowly_exp_genes(expressed_genes_per_celltype, gl, pathway_gene_threshold)\n",
    "low_removed_syngo = filter_lowly_exp_genes(expressed_genes_per_celltype, sy, pathway_gene_threshold)\n",
    "low_removed_bp = filter_lowly_exp_genes(expressed_genes_per_celltype, bp, pathway_gene_threshold)\n",
    "low_removed_humancyc = filter_lowly_exp_genes(expressed_genes_per_celltype, cy, pathway_gene_threshold)\n",
    "low_removed_kegg = filter_lowly_exp_genes(expressed_genes_per_celltype, ke, pathway_gene_threshold)\n",
    "low_removed_reactome = filter_lowly_exp_genes(expressed_genes_per_celltype, re, pathway_gene_threshold)\n",
    "\n",
    "low_removed_all_terms = filter_lowly_exp_genes(expressed_genes_per_celltype, all_paths, pathway_gene_threshold)\n",
    "low_removed_cura_terms = filter_lowly_exp_genes(expressed_genes_per_celltype, cura_paths, pathway_gene_threshold)\n",
    "\n",
    "\n",
    "# add the filtered pathways to the dataset to be saved\n",
    "pw_data[['pathways']][['low_removed_gabitto']] = low_removed_gabitto\n",
    "pw_data[['pathways']][['low_removed_galea']] = low_removed_galea\n",
    "pw_data[['pathways']][['low_removed_syngo']] = low_removed_syngo\n",
    "pw_data[['pathways']][['low_removed_bp']] = low_removed_bp\n",
    "pw_data[['pathways']][['low_removed_reactome']] = low_removed_reactome\n",
    "pw_data[['pathways']][['low_removed_kegg']] = low_removed_kegg\n",
    "pw_data[['pathways']][['low_removed_humancyc']] = low_removed_humancyc\n",
    "pw_data[['pathways']][['low_removed_cura_terms']] = low_removed_cura_terms\n",
    "\n",
    "pw_data[['pathways']][['all_gabitto']] = gb\n",
    "pw_data[['pathways']][['all_galea']] = gl\n",
    "pw_data[['pathways']][['all_syngo']] = sy\n",
    "pw_data[['pathways']][['all_reactome']] = re\n",
    "pw_data[['pathways']][['all_kegg']] = ke\n",
    "pw_data[['pathways']][['all_humancyc']] = cy\n",
    "pw_data[['pathways']][['all_bp']] = bp\n",
    "\n",
    "pw_data[['pathways']][['all']] = all_paths\n",
    "pw_data[['pathways']][['cura']] = cura_paths\n",
    "\n",
    "# get apoe-associated genesets\n",
    "print('getting thematic genesets...')\n",
    "\n",
    "gene = 'APOE'\n",
    "index = unlist(lapply(names(all_paths), function(x) c(gene)%in%all_paths[[x]]))\n",
    "apoe_gsets = all_paths[index]\n",
    "gs = lapply(names(apoe_gsets), function(x) apoe_gsets[[x]][apoe_gsets[[x]]!=''])\n",
    "names(gs) = names(apoe_gsets)\n",
    "apoe_gsets = gs\n",
    "apoe_gsets_low_removed = filter_lowly_exp_genes(expressed_genes_per_celltype, apoe_gsets, pathway_gene_threshold)\n",
    "pw_data[['pathways']][['apoe_gsets_low_removed']] = apoe_gsets_low_removed\n",
    "pw_data[['pathways']][['apoe_gsets_all']] = apoe_gsets\n",
    "\n",
    "# get lipid-associated genesets\n",
    "lipid_keys = c('sterol','athero','cholest','LDL','HDL','lipoprotein','triglyceride','TAG','DAG','lipid','steroid','fatty acid', 'ceramide')\n",
    "lipid_paths = get_gset_names_by_category(lipid_keys, names(all_paths))\n",
    "low_removed_lipid_paths = filter_lowly_exp_genes(expressed_genes_per_celltype, all_paths[lipid_paths], pathway_gene_threshold)\n",
    "pw_data[['pathways']][['all_lipid_associated']] = lipid_paths\n",
    "pw_data[['pathways']][['low_removed_lipid_associated']] = low_removed_lipid_paths\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "652c6839",
   "metadata": {},
   "source": [
    "### **Estimate Pathway Activity Scores**\n",
    "\n",
    "After estimating individual-cell-type-level normalized gene expression profile averages, we next calculate pathway activity scores, as described in [**Joel W. Blanchard et. al.**](https://doi.org/10.1038/s41586-022-05439-w) and previously implemented in the R package (Gene Set Variational Analysis) GSVA (v.1.42.0)41. \n",
    "\n",
    "Briefly, GSVA estimates a normalized relative expression level per gene across samples. This expression level is then rank-ordered for each sample and aggregated into gene sets by calculating sample-wise enrichment scores using a Kolmogorov–Smirnov-like rank statistic.\n",
    "\n",
    "As described in [**Joel W. Blanchard et. al.**](https://doi.org/10.1038/s41586-022-05439-w), we used the parameters to evaluate the \n",
    "GSVA function:\n",
    "\n",
    "- `mx.diff=TRUE` \n",
    "- `kcdf=c(\"Gaussian\")`\n",
    "- `min.sz=5` \n",
    "- `max.sz=150`\n",
    "\n",
    "Here, we instead set `max.sz=500`, as described in the [**GSVA Tutorial**](http://www.bioconductor.org/packages/release/bioc/vignettes/GSVA/inst/doc/GSVA.html#5_Quantification_of_pathway_activity_in_bulk_microarray_and_RNA-seq_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "d82607c6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Estimating Activity Scores of low_removed_cura_terms Pathways in COL8A1 Cells\"\n",
      "[1] \"........\"\n",
      "Estimating GSVA scores for 281 gene sets.\n",
      "Estimating ECDFs with Poisson kernels\n",
      "  |======================================================================| 100%\n",
      "\n",
      "[1] \"Estimating Activity Scores of low_removed_bp Pathways in COL8A1 Cells\"\n",
      "[1] \"........\"\n",
      "Estimating GSVA scores for 3813 gene sets.\n",
      "Estimating ECDFs with Poisson kernels\n",
      "  |======================================================================| 100%\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "In addition: Warning messages:\n",
       "1: In .local(expr, gset.idx.list, ...) :\n",
       "  Using 'dgCMatrix' objects as input is still in an experimental stage.\n",
       "2: In .local(expr, gset.idx.list, ...) :\n",
       "  Using 'dgCMatrix' objects as input is still in an experimental stage.\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R -o gsva_scores \n",
    "\n",
    "gsva_scores = list()\n",
    "gene_sets <- c('low_removed_cura_terms', 'low_removed_bp')\n",
    "\n",
    "# iterate over celltypes and perform gsva\n",
    "for (gs in gene_sets){\n",
    "\n",
    "    gsva_scores[[gs]] <- list()\n",
    "    \n",
    "    for (cell_type in names(summed_counts_per_celltype)){\n",
    "\n",
    "        # get gene sets\n",
    "        geneSets = pw_data[['pathways']][[gs]][[cell_type]]\n",
    "\n",
    "        # read in expression data\n",
    "        mat = assays(summed_counts_per_celltype[[cell_type]])[[\"counts\"]]\n",
    "\n",
    "        # run gsva\n",
    "        print(paste0(\"Estimating Activity Scores of \", gs, \" Pathways in \", cell_type, \" Cells\"))\n",
    "        print(\"........\")\n",
    "\n",
    "        tryCatch({\n",
    "            exc_gsva <- gsva(mat, geneSets, mx.diff=TRUE, kcdf=c(\"Poisson\"), min.sz=3, max.sz = Inf)\n",
    "            \n",
    "            exc_gsva <- as.data.frame(exc_gsva, stringsAsFactors = FALSE)\n",
    "            \n",
    "            # Store the output in a list\n",
    "            gsva_scores[[gs]][[cell_type]] <- exc_gsva\n",
    "\n",
    "        }, error = function(e) {\n",
    "            if (grepl(\"The gene set list is empty\", e$message)) {\n",
    "                cat(\"Skipping loop due to empty gene set list.\\n\")\n",
    "        } else {\n",
    "            stop(e)\n",
    "        }\n",
    "        })\n",
    "    }\n",
    "}"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "13bc27b5",
   "metadata": {},
   "source": [
    "### **Differential pathway activity analysis along diseaase pseudo-progression**\n",
    "\n",
    "To investigate differential pathway activity along disease pseudoprogression across different cell types, we use a multivariate linear model approach, as described in [**Joel W. Blanchard et. al.**](https://doi.org/10.1038/s41586-022-05439-w), using pathway activity scores. \n",
    "\n",
    "The design matrix will appear as consequently: \n",
    "\n",
    "`pathway activity ~ β0 × CPS + β1 × Coavriate1 + β2 × Covariate2 + ...` while including `subjedt_id` as `random_effect`\n",
    "\n",
    "The linear models were fitted using the `lmfit()` function and t-statistics were computed using the `eBayes()` function from the Limma R package (v.3.50.3). \n",
    "\n",
    "As outlined in [Gabitto et al. 2023](https://doi.org/10.1101/2023.05.08.539485), we analyzed pseudocell objects based on disease phase along the Continuous Pseudo-progression Score (CPS), dividing them into `early (<0.55) and late (>0.45) phases`. \n",
    "\n",
    "These two phases represent: \n",
    "\n",
    "1) A period where donors had normal cognition and relatively low levels of plaque and tangle pathology (but changes in other quantitative neuropathology variables) and\n",
    "2) a period where donors had markedly increased levels of plaque and tangle pathology and cognitive decline.\n",
    "\n",
    "To prioritize candidate AD-dysregulated processes, pathways with snRNA-seq evidence of association with AD at a `nominal P value of 0.01` are considered to be potential candidates. This procedure will result in prioritized candidate pathways for all cell types. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "416c958a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[conflicted] Will prefer base::unname over any other package.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[conflicted] Will prefer dplyr::arrange over any other package.\n",
      "[1] \"getting linear model fits...\"\n",
      "Coefficients not estimable: QC_MT.pct anno_age \n",
      "Note: design matrix not of full rank (2 coef not estimable).\n",
      "Error in .ebayes(fit = fit, proportion = proportion, stdev.coef.lim = stdev.coef.lim,  : \n",
      "  No finite residual standard deviations\n",
      "In addition: Warning messages:\n",
      "1: In max(abs(QtBlock[-(1:QR$rank), ])) :\n",
      "  no non-missing arguments to max; returning -Inf\n",
      "2: In duplicateCorrelation(expression, mod1, block = as.character(meta[,  :\n",
      "  Block factor already encoded in the design matrix: setting intrablock correlation to zero.\n",
      "3: Partial NA coefficients for 281 probe(s)\n",
      "Error in .ebayes(fit = fit, proportion = proportion, stdev.coef.lim = stdev.coef.lim,  : \n",
      "  No finite residual standard deviations\n"
     ]
    },
    {
     "ename": "RInterpreterError",
     "evalue": "Failed to parse and evaluate line '\\nconflicts_prefer(base::unname)\\nconflicts_prefer(dplyr::arrange)\\n\\nsource(\\'../scripts/functions/pathway_analyses.R\\')\\n \\nsummary <- list()\\nfor (cell_type in names(summed_counts_per_celltype)) {\\n    summary[[cell_type]] <- colData(summed_counts_per_celltype[[cell_type]])\\n    summary[[cell_type]][, random_effect] <- as.factor(summary[[cell_type]][, random_effect])\\n\\n    if (numeric_covariates!=\\'None\\'){\\n      for (cov in numeric_covariates){\\n        #cov <- gsub(\" \", \".\", cov)\\n        summary[[cell_type]][, cov] <- as.numeric(summary[[cell_type]][, cov])\\n      } \\n    }\\n  \\n  if (factor == \\'ContinuousPseudoprogressionScore\\'){\\n    summary[[cell_type]]$CPS <- ifelse(summary[[cell_type]][, \\'ContinuousPseudoprogressionScore\\'] < 0.5, \"early\", \"late\")    \\n  }\\n}\\n\\n\\nall_data = list()\\n# get linear model fits\\nprint(\\'getting linear model fits...\\')\\n\\nif (factor == \\'ContinuousPseudoprogressionScore\\'){\\n\\n  # Replace \"ContinuousPseudoprogressionScore\" with \"CPS\"\\n  covariates <- gsub(\"ContinuousPseudoprogressionScore\", \"CPS\", covariates)\\n\\n  for (gs in gene_sets){\\n    fits = get_fits_cps(gsva_scores[[gs]], summary, covariates, random_effect)\\n    all_data[[gs]] <- list()\\n    all_data[[gs]][[\\'fits_all\\']] = fits\\n\\n    for (cell_type in names(all_data[[gs]][[\\'fits_all\\']])){\\n      all_data[[gs]][[\\'fits_all\\']][cell_type][[\\'late_vs_early\\']] = get_scores(all_data[[gs]][[\\'fits_all\\']][cell_type][[\\'late_vs_early\\']])\\n    }\\n  }\\n\\n} else if (factor == \\'pathology.group\\'){\\n\\n  for (gs in gene_sets){\\n\\n    fits = get_fits(gsva_scores[[gs]], summary, covariates, random_effect)\\n\\n    all_data[[gs]] <- list()\\n    all_data[[gs]][[\\'fits_all\\']] = fits\\n\\n    # get matrix of scores for heatmap\\n    print(\\'get matrix of scores\\')\\n    scores = get_scores(fits)\\n\\n    # Create a new empty list to store the reorganized scores\\n    new_scores <- list()\\n\\n    tests <- test_names\\n    for (test in tests) {\\n      new_scores[[test]] <- list(\\'all\\' = list())\\n      for (cell_type in names(summed_counts_per_celltype)) {\\n            new_scores[[test]][[\\'all\\']][[cell_type]] <- scores[[\\'all\\']][[cell_type]][[test]] \\n      }\\n\\n      all_data[[gs]][[\\'scores_all\\']] = new_scores\\n\\n      print(\\'filter by score 1.3\\')\\n      \\n      all_data[[gs]][[\\'scores_filtered\\']] = list()\\n\\n      for(i in names(new_scores)){\\n        names = unique(unname(unlist(lapply(names(new_scores[[i]]$all), function(x) rownames(new_scores[[i]]$all[[x]])))))\\n\\n        mat = get_matrix(new_scores[[i]]$all, names)\\n        \\n        mat = mat[unname(rowSums(abs(mat)>1.3)>0), ]\\n\\n        all_data[[gs]][[\\'scores_filtered\\']][[i]] = mat\\n      }\\n    }\\n  }\\n}\\n'.\nR error message: 'Error in .ebayes(fit = fit, proportion = proportion, stdev.coef.lim = stdev.coef.lim,  : \\n  No finite residual standard deviations'\nR stdout:\nNote: design matrix not of full rank (2 coef not estimable).\nError in .ebayes(fit = fit, proportion = proportion, stdev.coef.lim = stdev.coef.lim,  : \n  No finite residual standard deviations\nIn addition: Warning messages:\n1: In max(abs(QtBlock[-(1:QR$rank), ])) :\n  no non-missing arguments to max; returning -Inf\n2: In duplicateCorrelation(expression, mod1, block = as.character(meta[,  :\n  Block factor already encoded in the design matrix: setting intrablock correlation to zero.\n3: Partial NA coefficients for 281 probe(s)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mRRuntimeError\u001b[0m                             Traceback (most recent call last)",
      "File \u001b[0;32m~/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/rpy2/ipython/rmagic.py:401\u001b[0m, in \u001b[0;36mRMagics.eval\u001b[0;34m(self, code)\u001b[0m\n\u001b[1;32m    400\u001b[0m     r_expr \u001b[38;5;241m=\u001b[39m ri\u001b[38;5;241m.\u001b[39mparse(code)\n\u001b[0;32m--> 401\u001b[0m     value, visible \u001b[38;5;241m=\u001b[39m \u001b[43mri\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mevalr_expr_with_visible\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    402\u001b[0m \u001b[43m        \u001b[49m\u001b[43mr_expr\u001b[49m\n\u001b[1;32m    403\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    404\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m (ri\u001b[38;5;241m.\u001b[39membedded\u001b[38;5;241m.\u001b[39mRRuntimeError, \u001b[38;5;167;01mValueError\u001b[39;00m) \u001b[38;5;28;01mas\u001b[39;00m exception:\n\u001b[1;32m    405\u001b[0m     \u001b[38;5;66;03m# Otherwise next return seems to have copy of error.\u001b[39;00m\n",
      "File \u001b[0;32m~/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/rpy2/rinterface.py:183\u001b[0m, in \u001b[0;36mevalr_expr_with_visible\u001b[0;34m(expr, envir)\u001b[0m\n\u001b[1;32m    182\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m error_occured[\u001b[38;5;241m0\u001b[39m]:\n\u001b[0;32m--> 183\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m embedded\u001b[38;5;241m.\u001b[39mRRuntimeError(_rinterface\u001b[38;5;241m.\u001b[39m_geterrmessage())\n\u001b[1;32m    184\u001b[0m res \u001b[38;5;241m=\u001b[39m conversion\u001b[38;5;241m.\u001b[39m_cdata_to_rinterface(r_res)\n",
      "\u001b[0;31mRRuntimeError\u001b[0m: Error in .ebayes(fit = fit, proportion = proportion, stdev.coef.lim = stdev.coef.lim,  : \n  No finite residual standard deviations\n",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mRInterpreterError\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[24], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mget_ipython\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun_cell_magic\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mR\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m-o all_data -i numeric_covariates -i covariates -i factor -i cell_supertype\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43mconflicts_prefer(base::unname)\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43mconflicts_prefer(dplyr::arrange)\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43msource(\u001b[39;49m\u001b[38;5;130;43;01m\\'\u001b[39;49;00m\u001b[38;5;124;43m../scripts/functions/pathway_analyses.R\u001b[39;49m\u001b[38;5;130;43;01m\\'\u001b[39;49;00m\u001b[38;5;124;43m)\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m \u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43msummary <- list()\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43mfor (cell_type in names(summed_counts_per_celltype)) \u001b[39;49m\u001b[38;5;124;43m{\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m    summary[[cell_type]] <- colData(summed_counts_per_celltype[[cell_type]])\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m    summary[[cell_type]][, random_effect] <- as.factor(summary[[cell_type]][, random_effect])\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m    if (numeric_covariates!=\u001b[39;49m\u001b[38;5;130;43;01m\\'\u001b[39;49;00m\u001b[38;5;124;43mNone\u001b[39;49m\u001b[38;5;130;43;01m\\'\u001b[39;49;00m\u001b[38;5;124;43m)\u001b[39;49m\u001b[38;5;124;43m{\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m      for (cov in numeric_covariates)\u001b[39;49m\u001b[38;5;124;43m{\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m        #cov <- gsub(\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m \u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m, \u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m.\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m, cov)\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m        summary[[cell_type]][, cov] <- as.numeric(summary[[cell_type]][, cov])\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m      } \u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m    }\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m  \u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m  if (factor == \u001b[39;49m\u001b[38;5;130;43;01m\\'\u001b[39;49;00m\u001b[38;5;124;43mContinuousPseudoprogressionScore\u001b[39;49m\u001b[38;5;130;43;01m\\'\u001b[39;49;00m\u001b[38;5;124;43m)\u001b[39;49m\u001b[38;5;124;43m{\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m    summary[[cell_type]]$CPS <- ifelse(summary[[cell_type]][, \u001b[39;49m\u001b[38;5;130;43;01m\\'\u001b[39;49;00m\u001b[38;5;124;43mContinuousPseudoprogressionScore\u001b[39;49m\u001b[38;5;130;43;01m\\'\u001b[39;49;00m\u001b[38;5;124;43m] < 0.5, \u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mearly\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m, \u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mlate\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m)    \u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m  }\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m}\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43mall_data = list()\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m# get linear model fits\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43mprint(\u001b[39;49m\u001b[38;5;130;43;01m\\'\u001b[39;49;00m\u001b[38;5;124;43mgetting linear model fits...\u001b[39;49m\u001b[38;5;130;43;01m\\'\u001b[39;49;00m\u001b[38;5;124;43m)\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43mif (factor == \u001b[39;49m\u001b[38;5;130;43;01m\\'\u001b[39;49;00m\u001b[38;5;124;43mContinuousPseudoprogressionScore\u001b[39;49m\u001b[38;5;130;43;01m\\'\u001b[39;49;00m\u001b[38;5;124;43m)\u001b[39;49m\u001b[38;5;124;43m{\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m  # Replace \u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mContinuousPseudoprogressionScore\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m with \u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mCPS\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m  covariates <- gsub(\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mContinuousPseudoprogressionScore\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m, \u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mCPS\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m, covariates)\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m  for (gs in gene_sets)\u001b[39;49m\u001b[38;5;124;43m{\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m    fits = get_fits_cps(gsva_scores[[gs]], summary, covariates, random_effect)\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m    all_data[[gs]] <- list()\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m    all_data[[gs]][[\u001b[39;49m\u001b[38;5;130;43;01m\\'\u001b[39;49;00m\u001b[38;5;124;43mfits_all\u001b[39;49m\u001b[38;5;130;43;01m\\'\u001b[39;49;00m\u001b[38;5;124;43m]] = fits\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m    for (cell_type in names(all_data[[gs]][[\u001b[39;49m\u001b[38;5;130;43;01m\\'\u001b[39;49;00m\u001b[38;5;124;43mfits_all\u001b[39;49m\u001b[38;5;130;43;01m\\'\u001b[39;49;00m\u001b[38;5;124;43m]]))\u001b[39;49m\u001b[38;5;124;43m{\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m      all_data[[gs]][[\u001b[39;49m\u001b[38;5;130;43;01m\\'\u001b[39;49;00m\u001b[38;5;124;43mfits_all\u001b[39;49m\u001b[38;5;130;43;01m\\'\u001b[39;49;00m\u001b[38;5;124;43m]][cell_type][[\u001b[39;49m\u001b[38;5;130;43;01m\\'\u001b[39;49;00m\u001b[38;5;124;43mlate_vs_early\u001b[39;49m\u001b[38;5;130;43;01m\\'\u001b[39;49;00m\u001b[38;5;124;43m]] = get_scores(all_data[[gs]][[\u001b[39;49m\u001b[38;5;130;43;01m\\'\u001b[39;49;00m\u001b[38;5;124;43mfits_all\u001b[39;49m\u001b[38;5;130;43;01m\\'\u001b[39;49;00m\u001b[38;5;124;43m]][cell_type][[\u001b[39;49m\u001b[38;5;130;43;01m\\'\u001b[39;49;00m\u001b[38;5;124;43mlate_vs_early\u001b[39;49m\u001b[38;5;130;43;01m\\'\u001b[39;49;00m\u001b[38;5;124;43m]])\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m    }\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m  }\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m} else if (factor == \u001b[39;49m\u001b[38;5;130;43;01m\\'\u001b[39;49;00m\u001b[38;5;124;43mpathology.group\u001b[39;49m\u001b[38;5;130;43;01m\\'\u001b[39;49;00m\u001b[38;5;124;43m)\u001b[39;49m\u001b[38;5;124;43m{\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m  for (gs in gene_sets)\u001b[39;49m\u001b[38;5;124;43m{\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m    fits = get_fits(gsva_scores[[gs]], summary, covariates, random_effect)\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m    all_data[[gs]] <- list()\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m    all_data[[gs]][[\u001b[39;49m\u001b[38;5;130;43;01m\\'\u001b[39;49;00m\u001b[38;5;124;43mfits_all\u001b[39;49m\u001b[38;5;130;43;01m\\'\u001b[39;49;00m\u001b[38;5;124;43m]] = fits\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m    # get matrix of scores for heatmap\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m    print(\u001b[39;49m\u001b[38;5;130;43;01m\\'\u001b[39;49;00m\u001b[38;5;124;43mget matrix of scores\u001b[39;49m\u001b[38;5;130;43;01m\\'\u001b[39;49;00m\u001b[38;5;124;43m)\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m    scores = get_scores(fits)\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m    # Create a new empty list to store the reorganized scores\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m    new_scores <- list()\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m    tests <- test_names\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m    for (test in tests) \u001b[39;49m\u001b[38;5;124;43m{\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m      new_scores[[test]] <- list(\u001b[39;49m\u001b[38;5;130;43;01m\\'\u001b[39;49;00m\u001b[38;5;124;43mall\u001b[39;49m\u001b[38;5;130;43;01m\\'\u001b[39;49;00m\u001b[38;5;124;43m = list())\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m      for (cell_type in names(summed_counts_per_celltype)) \u001b[39;49m\u001b[38;5;124;43m{\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m            new_scores[[test]][[\u001b[39;49m\u001b[38;5;130;43;01m\\'\u001b[39;49;00m\u001b[38;5;124;43mall\u001b[39;49m\u001b[38;5;130;43;01m\\'\u001b[39;49;00m\u001b[38;5;124;43m]][[cell_type]] <- scores[[\u001b[39;49m\u001b[38;5;130;43;01m\\'\u001b[39;49;00m\u001b[38;5;124;43mall\u001b[39;49m\u001b[38;5;130;43;01m\\'\u001b[39;49;00m\u001b[38;5;124;43m]][[cell_type]][[test]] \u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m      }\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m      all_data[[gs]][[\u001b[39;49m\u001b[38;5;130;43;01m\\'\u001b[39;49;00m\u001b[38;5;124;43mscores_all\u001b[39;49m\u001b[38;5;130;43;01m\\'\u001b[39;49;00m\u001b[38;5;124;43m]] = new_scores\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m      print(\u001b[39;49m\u001b[38;5;130;43;01m\\'\u001b[39;49;00m\u001b[38;5;124;43mfilter by score 1.3\u001b[39;49m\u001b[38;5;130;43;01m\\'\u001b[39;49;00m\u001b[38;5;124;43m)\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m      \u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m      all_data[[gs]][[\u001b[39;49m\u001b[38;5;130;43;01m\\'\u001b[39;49;00m\u001b[38;5;124;43mscores_filtered\u001b[39;49m\u001b[38;5;130;43;01m\\'\u001b[39;49;00m\u001b[38;5;124;43m]] = list()\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m      for(i in names(new_scores))\u001b[39;49m\u001b[38;5;124;43m{\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m        names = unique(unname(unlist(lapply(names(new_scores[[i]]$all), function(x) rownames(new_scores[[i]]$all[[x]])))))\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m        mat = get_matrix(new_scores[[i]]$all, names)\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m        \u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m        mat = mat[unname(rowSums(abs(mat)>1.3)>0), ]\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m        all_data[[gs]][[\u001b[39;49m\u001b[38;5;130;43;01m\\'\u001b[39;49;00m\u001b[38;5;124;43mscores_filtered\u001b[39;49m\u001b[38;5;130;43;01m\\'\u001b[39;49;00m\u001b[38;5;124;43m]][[i]] = mat\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m      }\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m    }\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m  }\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m}\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/IPython/core/interactiveshell.py:2478\u001b[0m, in \u001b[0;36mInteractiveShell.run_cell_magic\u001b[0;34m(self, magic_name, line, cell)\u001b[0m\n\u001b[1;32m   2476\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mbuiltin_trap:\n\u001b[1;32m   2477\u001b[0m     args \u001b[38;5;241m=\u001b[39m (magic_arg_s, cell)\n\u001b[0;32m-> 2478\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[43mfn\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   2480\u001b[0m \u001b[38;5;66;03m# The code below prevents the output from being displayed\u001b[39;00m\n\u001b[1;32m   2481\u001b[0m \u001b[38;5;66;03m# when using magics with decodator @output_can_be_silenced\u001b[39;00m\n\u001b[1;32m   2482\u001b[0m \u001b[38;5;66;03m# when the last Python token in the expression is a ';'.\u001b[39;00m\n\u001b[1;32m   2483\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mgetattr\u001b[39m(fn, magic\u001b[38;5;241m.\u001b[39mMAGIC_OUTPUT_CAN_BE_SILENCED, \u001b[38;5;28;01mFalse\u001b[39;00m):\n",
      "File \u001b[0;32m~/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/rpy2/ipython/rmagic.py:978\u001b[0m, in \u001b[0;36mRMagics.R\u001b[0;34m(self, line, cell, local_ns)\u001b[0m\n\u001b[1;32m    976\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m e\u001b[38;5;241m.\u001b[39mstdout\u001b[38;5;241m.\u001b[39mendswith(e\u001b[38;5;241m.\u001b[39merr):\n\u001b[1;32m    977\u001b[0m         \u001b[38;5;28mprint\u001b[39m(e\u001b[38;5;241m.\u001b[39merr)\n\u001b[0;32m--> 978\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m e\n\u001b[1;32m    979\u001b[0m \u001b[38;5;28;01mfinally\u001b[39;00m:\n\u001b[1;32m    980\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mdevice \u001b[38;5;129;01min\u001b[39;00m DEVICES_STATIC:\n",
      "File \u001b[0;32m~/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/rpy2/ipython/rmagic.py:943\u001b[0m, in \u001b[0;36mRMagics.R\u001b[0;34m(self, line, cell, local_ns)\u001b[0m\n\u001b[1;32m    941\u001b[0m         return_output \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mFalse\u001b[39;00m\n\u001b[1;32m    942\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 943\u001b[0m     text_result, result, visible \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43meval\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcode\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    944\u001b[0m     text_output \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m text_result\n\u001b[1;32m    945\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m visible:\n",
      "File \u001b[0;32m~/miniconda3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/rpy2/ipython/rmagic.py:407\u001b[0m, in \u001b[0;36mRMagics.eval\u001b[0;34m(self, code)\u001b[0m\n\u001b[1;32m    404\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m (ri\u001b[38;5;241m.\u001b[39membedded\u001b[38;5;241m.\u001b[39mRRuntimeError, \u001b[38;5;167;01mValueError\u001b[39;00m) \u001b[38;5;28;01mas\u001b[39;00m exception:\n\u001b[1;32m    405\u001b[0m     \u001b[38;5;66;03m# Otherwise next return seems to have copy of error.\u001b[39;00m\n\u001b[1;32m    406\u001b[0m     warning_or_other_msg \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mflush()\n\u001b[0;32m--> 407\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m RInterpreterError(code, \u001b[38;5;28mstr\u001b[39m(exception),\n\u001b[1;32m    408\u001b[0m                             warning_or_other_msg)\n\u001b[1;32m    409\u001b[0m \u001b[38;5;28;01mfinally\u001b[39;00m:\n\u001b[1;32m    410\u001b[0m     ro\u001b[38;5;241m.\u001b[39m_print_deferred_warnings()\n",
      "\u001b[0;31mRInterpreterError\u001b[0m: Failed to parse and evaluate line '\\nconflicts_prefer(base::unname)\\nconflicts_prefer(dplyr::arrange)\\n\\nsource(\\'../scripts/functions/pathway_analyses.R\\')\\n \\nsummary <- list()\\nfor (cell_type in names(summed_counts_per_celltype)) {\\n    summary[[cell_type]] <- colData(summed_counts_per_celltype[[cell_type]])\\n    summary[[cell_type]][, random_effect] <- as.factor(summary[[cell_type]][, random_effect])\\n\\n    if (numeric_covariates!=\\'None\\'){\\n      for (cov in numeric_covariates){\\n        #cov <- gsub(\" \", \".\", cov)\\n        summary[[cell_type]][, cov] <- as.numeric(summary[[cell_type]][, cov])\\n      } \\n    }\\n  \\n  if (factor == \\'ContinuousPseudoprogressionScore\\'){\\n    summary[[cell_type]]$CPS <- ifelse(summary[[cell_type]][, \\'ContinuousPseudoprogressionScore\\'] < 0.5, \"early\", \"late\")    \\n  }\\n}\\n\\n\\nall_data = list()\\n# get linear model fits\\nprint(\\'getting linear model fits...\\')\\n\\nif (factor == \\'ContinuousPseudoprogressionScore\\'){\\n\\n  # Replace \"ContinuousPseudoprogressionScore\" with \"CPS\"\\n  covariates <- gsub(\"ContinuousPseudoprogressionScore\", \"CPS\", covariates)\\n\\n  for (gs in gene_sets){\\n    fits = get_fits_cps(gsva_scores[[gs]], summary, covariates, random_effect)\\n    all_data[[gs]] <- list()\\n    all_data[[gs]][[\\'fits_all\\']] = fits\\n\\n    for (cell_type in names(all_data[[gs]][[\\'fits_all\\']])){\\n      all_data[[gs]][[\\'fits_all\\']][cell_type][[\\'late_vs_early\\']] = get_scores(all_data[[gs]][[\\'fits_all\\']][cell_type][[\\'late_vs_early\\']])\\n    }\\n  }\\n\\n} else if (factor == \\'pathology.group\\'){\\n\\n  for (gs in gene_sets){\\n\\n    fits = get_fits(gsva_scores[[gs]], summary, covariates, random_effect)\\n\\n    all_data[[gs]] <- list()\\n    all_data[[gs]][[\\'fits_all\\']] = fits\\n\\n    # get matrix of scores for heatmap\\n    print(\\'get matrix of scores\\')\\n    scores = get_scores(fits)\\n\\n    # Create a new empty list to store the reorganized scores\\n    new_scores <- list()\\n\\n    tests <- test_names\\n    for (test in tests) {\\n      new_scores[[test]] <- list(\\'all\\' = list())\\n      for (cell_type in names(summed_counts_per_celltype)) {\\n            new_scores[[test]][[\\'all\\']][[cell_type]] <- scores[[\\'all\\']][[cell_type]][[test]] \\n      }\\n\\n      all_data[[gs]][[\\'scores_all\\']] = new_scores\\n\\n      print(\\'filter by score 1.3\\')\\n      \\n      all_data[[gs]][[\\'scores_filtered\\']] = list()\\n\\n      for(i in names(new_scores)){\\n        names = unique(unname(unlist(lapply(names(new_scores[[i]]$all), function(x) rownames(new_scores[[i]]$all[[x]])))))\\n\\n        mat = get_matrix(new_scores[[i]]$all, names)\\n        \\n        mat = mat[unname(rowSums(abs(mat)>1.3)>0), ]\\n\\n        all_data[[gs]][[\\'scores_filtered\\']][[i]] = mat\\n      }\\n    }\\n  }\\n}\\n'.\nR error message: 'Error in .ebayes(fit = fit, proportion = proportion, stdev.coef.lim = stdev.coef.lim,  : \\n  No finite residual standard deviations'\nR stdout:\nNote: design matrix not of full rank (2 coef not estimable).\nError in .ebayes(fit = fit, proportion = proportion, stdev.coef.lim = stdev.coef.lim,  : \n  No finite residual standard deviations\nIn addition: Warning messages:\n1: In max(abs(QtBlock[-(1:QR$rank), ])) :\n  no non-missing arguments to max; returning -Inf\n2: In duplicateCorrelation(expression, mod1, block = as.character(meta[,  :\n  Block factor already encoded in the design matrix: setting intrablock correlation to zero.\n3: Partial NA coefficients for 281 probe(s)"
     ]
    }
   ],
   "source": [
    "%%R -o all_data -i numeric_covariates -i covariates -i factor -i cell_supertype\n",
    "\n",
    "conflicts_prefer(base::unname)\n",
    "conflicts_prefer(dplyr::arrange)\n",
    "\n",
    "source('../scripts/functions/pathway_analyses.R')\n",
    " \n",
    "summary <- list()\n",
    "for (cell_type in names(summed_counts_per_celltype)) {\n",
    "    summary[[cell_type]] <- colData(summed_counts_per_celltype[[cell_type]])\n",
    "    summary[[cell_type]][, random_effect] <- as.factor(summary[[cell_type]][, random_effect])\n",
    "\n",
    "    if (numeric_covariates!='None'){\n",
    "      for (cov in numeric_covariates){\n",
    "        #cov <- gsub(\" \", \".\", cov)\n",
    "        summary[[cell_type]][, cov] <- as.numeric(summary[[cell_type]][, cov])\n",
    "      } \n",
    "    }\n",
    "  \n",
    "  if (factor == 'ContinuousPseudoprogressionScore'){\n",
    "    summary[[cell_type]]$CPS <- ifelse(summary[[cell_type]][, 'ContinuousPseudoprogressionScore'] < 0.5, \"early\", \"late\")    \n",
    "  }\n",
    "}\n",
    "\n",
    "\n",
    "all_data = list()\n",
    "# get linear model fits\n",
    "print('getting linear model fits...')\n",
    "\n",
    "if (factor == 'ContinuousPseudoprogressionScore'){\n",
    "\n",
    "  # Replace \"ContinuousPseudoprogressionScore\" with \"CPS\"\n",
    "  covariates <- gsub(\"ContinuousPseudoprogressionScore\", \"CPS\", covariates)\n",
    "\n",
    "  for (gs in gene_sets){\n",
    "    fits = get_fits_cps(gsva_scores[[gs]], summary, covariates, random_effect)\n",
    "    all_data[[gs]] <- list()\n",
    "    all_data[[gs]][['fits_all']] = fits\n",
    "\n",
    "    for (cell_type in names(all_data[[gs]][['fits_all']])){\n",
    "      all_data[[gs]][['fits_all']][cell_type][['late_vs_early']] = get_scores(all_data[[gs]][['fits_all']][cell_type][['late_vs_early']])\n",
    "    }\n",
    "  }\n",
    "\n",
    "} else if (factor == 'pathology.group'){\n",
    "\n",
    "  for (gs in gene_sets){\n",
    "\n",
    "    fits = get_fits(gsva_scores[[gs]], summary, covariates, random_effect)\n",
    "\n",
    "    all_data[[gs]] <- list()\n",
    "    all_data[[gs]][['fits_all']] = fits\n",
    "\n",
    "    # get matrix of scores for heatmap\n",
    "    print('get matrix of scores')\n",
    "    scores = get_scores(fits)\n",
    "\n",
    "    # Create a new empty list to store the reorganized scores\n",
    "    new_scores <- list()\n",
    "\n",
    "    tests <- test_names\n",
    "    for (test in tests) {\n",
    "      new_scores[[test]] <- list('all' = list())\n",
    "      for (cell_type in names(summed_counts_per_celltype)) {\n",
    "            new_scores[[test]][['all']][[cell_type]] <- scores[['all']][[cell_type]][[test]] \n",
    "      }\n",
    "\n",
    "      all_data[[gs]][['scores_all']] = new_scores\n",
    "\n",
    "      print('filter by score 1.3')\n",
    "      \n",
    "      all_data[[gs]][['scores_filtered']] = list()\n",
    "\n",
    "      for(i in names(new_scores)){\n",
    "        names = unique(unname(unlist(lapply(names(new_scores[[i]]$all), function(x) rownames(new_scores[[i]]$all[[x]])))))\n",
    "\n",
    "        mat = get_matrix(new_scores[[i]]$all, names)\n",
    "        \n",
    "        mat = mat[unname(rowSums(abs(mat)>1.3)>0), ]\n",
    "\n",
    "        all_data[[gs]][['scores_filtered']][[i]] = mat\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "79a4b56a",
   "metadata": {},
   "source": [
    "### **Save Aggregated Data, Filtered Gene Sets, and Expressed Genes**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc438456",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"saving filtered gene sets...\"\n",
      "[1] \"done.\"\n",
      "[1] \"saving aggregated count data...\"\n",
      "[1] \"done.\"\n",
      "[1] \"saving expressed genes...\"\n",
      "[1] \"done.\"\n",
      "[1] \"saving GSVA scores...\"\n",
      "[1] \"done.\"\n",
      "[1] \"saving Fit scores...\"\n",
      "[1] \"done.\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i cell_supertype\n",
    "\n",
    "# save all the data\n",
    "save_path <- paste0('../results/', save_prefix, '/pathway_analysis/', cell_group, '/')\n",
    "\n",
    "if (!dir.exists(save_path)) {\n",
    "  dir.create(save_path, recursive = TRUE)\n",
    "}\n",
    "\n",
    "cell_supertype <- gsub(\" |/\", \"_\", cell_supertype)\n",
    "\n",
    "print('saving filtered gene sets...')\n",
    "saveRDS(pw_data, paste0(save_path, cell_type_column, '_', cell_supertype, '_filtered_genes_sets.rds'))\n",
    "print('done.')\n",
    "\n",
    "print('saving aggregated count data...')\n",
    "saveRDS(summed_counts_per_celltype, paste0(save_path, cell_type_column, '_', cell_supertype, '_aggregated_counts.rds'))\n",
    "print('done.')\n",
    "\n",
    "print('saving expressed genes...')\n",
    "saveRDS(expressed_genes_per_celltype, paste0(save_path, cell_type_column, '_', cell_supertype, '_expressed_genes.rds'))\n",
    "print('done.')\n",
    "\n",
    "print('saving GSVA scores...')\n",
    "saveRDS(gsva_scores, paste0(save_path, cell_type_column, '_', cell_supertype, '_gsva_scores.rds'))\n",
    "print('done.')\n",
    "\n",
    "print('saving Fit scores...')\n",
    "saveRDS(all_data, paste0(save_path, cell_type_column, '_', cell_supertype, '_pathway_scores.rds'))\n",
    "print('done.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "731fc0b2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>logFC</th>\n",
       "      <th>CI.L</th>\n",
       "      <th>CI.R</th>\n",
       "      <th>AveExpr</th>\n",
       "      <th>t</th>\n",
       "      <th>P.Value</th>\n",
       "      <th>adj.P.Val</th>\n",
       "      <th>B</th>\n",
       "      <th>celltype</th>\n",
       "      <th>names</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Transcription factors-zf-MIZ</th>\n",
       "      <td>0.616689</td>\n",
       "      <td>0.308159</td>\n",
       "      <td>0.925220</td>\n",
       "      <td>-0.046028</td>\n",
       "      <td>4.062944</td>\n",
       "      <td>0.000272</td>\n",
       "      <td>0.075464</td>\n",
       "      <td>0.372537</td>\n",
       "      <td>HRK</td>\n",
       "      <td>Transcription factors-zf-MIZ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>synaptic vesicle priming (GO:0016082)</th>\n",
       "      <td>0.463072</td>\n",
       "      <td>0.187528</td>\n",
       "      <td>0.738616</td>\n",
       "      <td>-0.010139</td>\n",
       "      <td>3.416096</td>\n",
       "      <td>0.001670</td>\n",
       "      <td>0.231297</td>\n",
       "      <td>-1.170968</td>\n",
       "      <td>HRK</td>\n",
       "      <td>synaptic vesicle priming (GO:0016082)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>postsynaptic endocytic zone (GO:0098843)</th>\n",
       "      <td>0.241252</td>\n",
       "      <td>0.090035</td>\n",
       "      <td>0.392468</td>\n",
       "      <td>-0.015540</td>\n",
       "      <td>3.242981</td>\n",
       "      <td>0.002663</td>\n",
       "      <td>0.231956</td>\n",
       "      <td>-1.565982</td>\n",
       "      <td>HRK</td>\n",
       "      <td>postsynaptic endocytic zone (GO:0098843)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Signaling: GTPase</th>\n",
       "      <td>0.342737</td>\n",
       "      <td>0.114281</td>\n",
       "      <td>0.571194</td>\n",
       "      <td>-0.034236</td>\n",
       "      <td>3.049516</td>\n",
       "      <td>0.004434</td>\n",
       "      <td>0.231956</td>\n",
       "      <td>-1.995762</td>\n",
       "      <td>HRK</td>\n",
       "      <td>Signaling: GTPase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Plasticity: EGF receptor</th>\n",
       "      <td>0.573344</td>\n",
       "      <td>0.189756</td>\n",
       "      <td>0.956932</td>\n",
       "      <td>-0.087557</td>\n",
       "      <td>3.038245</td>\n",
       "      <td>0.004566</td>\n",
       "      <td>0.231956</td>\n",
       "      <td>-2.020386</td>\n",
       "      <td>HRK</td>\n",
       "      <td>Plasticity: EGF receptor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Plasticity: Pluripotency</th>\n",
       "      <td>0.525491</td>\n",
       "      <td>0.169594</td>\n",
       "      <td>0.881388</td>\n",
       "      <td>-0.038371</td>\n",
       "      <td>3.001323</td>\n",
       "      <td>0.005024</td>\n",
       "      <td>0.231956</td>\n",
       "      <td>-2.100716</td>\n",
       "      <td>HRK</td>\n",
       "      <td>Plasticity: Pluripotency</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>regulation of presynaptic dense core vesicle exocytosis (GO:0099161)</th>\n",
       "      <td>0.441428</td>\n",
       "      <td>0.124473</td>\n",
       "      <td>0.758382</td>\n",
       "      <td>-0.038684</td>\n",
       "      <td>2.830969</td>\n",
       "      <td>0.007759</td>\n",
       "      <td>0.307054</td>\n",
       "      <td>-2.464371</td>\n",
       "      <td>HRK</td>\n",
       "      <td>regulation of presynaptic dense core vesicle e...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PAP: PAP Plasticity</th>\n",
       "      <td>-0.534948</td>\n",
       "      <td>-0.929165</td>\n",
       "      <td>-0.140731</td>\n",
       "      <td>-0.056728</td>\n",
       "      <td>-2.758343</td>\n",
       "      <td>0.009306</td>\n",
       "      <td>0.322225</td>\n",
       "      <td>-2.615713</td>\n",
       "      <td>HRK</td>\n",
       "      <td>PAP: PAP Plasticity</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                       logFC      CI.L  \\\n",
       "Transcription factors-zf-MIZ                        0.616689  0.308159   \n",
       "synaptic vesicle priming (GO:0016082)               0.463072  0.187528   \n",
       "postsynaptic endocytic zone (GO:0098843)            0.241252  0.090035   \n",
       "Signaling: GTPase                                   0.342737  0.114281   \n",
       "Plasticity: EGF receptor                            0.573344  0.189756   \n",
       "Plasticity: Pluripotency                            0.525491  0.169594   \n",
       "regulation of presynaptic dense core vesicle ex...  0.441428  0.124473   \n",
       "PAP: PAP Plasticity                                -0.534948 -0.929165   \n",
       "\n",
       "                                                        CI.R   AveExpr  \\\n",
       "Transcription factors-zf-MIZ                        0.925220 -0.046028   \n",
       "synaptic vesicle priming (GO:0016082)               0.738616 -0.010139   \n",
       "postsynaptic endocytic zone (GO:0098843)            0.392468 -0.015540   \n",
       "Signaling: GTPase                                   0.571194 -0.034236   \n",
       "Plasticity: EGF receptor                            0.956932 -0.087557   \n",
       "Plasticity: Pluripotency                            0.881388 -0.038371   \n",
       "regulation of presynaptic dense core vesicle ex...  0.758382 -0.038684   \n",
       "PAP: PAP Plasticity                                -0.140731 -0.056728   \n",
       "\n",
       "                                                           t   P.Value  \\\n",
       "Transcription factors-zf-MIZ                        4.062944  0.000272   \n",
       "synaptic vesicle priming (GO:0016082)               3.416096  0.001670   \n",
       "postsynaptic endocytic zone (GO:0098843)            3.242981  0.002663   \n",
       "Signaling: GTPase                                   3.049516  0.004434   \n",
       "Plasticity: EGF receptor                            3.038245  0.004566   \n",
       "Plasticity: Pluripotency                            3.001323  0.005024   \n",
       "regulation of presynaptic dense core vesicle ex...  2.830969  0.007759   \n",
       "PAP: PAP Plasticity                                -2.758343  0.009306   \n",
       "\n",
       "                                                    adj.P.Val         B  \\\n",
       "Transcription factors-zf-MIZ                         0.075464  0.372537   \n",
       "synaptic vesicle priming (GO:0016082)                0.231297 -1.170968   \n",
       "postsynaptic endocytic zone (GO:0098843)             0.231956 -1.565982   \n",
       "Signaling: GTPase                                    0.231956 -1.995762   \n",
       "Plasticity: EGF receptor                             0.231956 -2.020386   \n",
       "Plasticity: Pluripotency                             0.231956 -2.100716   \n",
       "regulation of presynaptic dense core vesicle ex...   0.307054 -2.464371   \n",
       "PAP: PAP Plasticity                                  0.322225 -2.615713   \n",
       "\n",
       "                                                   celltype  \\\n",
       "Transcription factors-zf-MIZ                            HRK   \n",
       "synaptic vesicle priming (GO:0016082)                   HRK   \n",
       "postsynaptic endocytic zone (GO:0098843)                HRK   \n",
       "Signaling: GTPase                                       HRK   \n",
       "Plasticity: EGF receptor                                HRK   \n",
       "Plasticity: Pluripotency                                HRK   \n",
       "regulation of presynaptic dense core vesicle ex...      HRK   \n",
       "PAP: PAP Plasticity                                     HRK   \n",
       "\n",
       "                                                                                                names  \n",
       "Transcription factors-zf-MIZ                                             Transcription factors-zf-MIZ  \n",
       "synaptic vesicle priming (GO:0016082)                           synaptic vesicle priming (GO:0016082)  \n",
       "postsynaptic endocytic zone (GO:0098843)                     postsynaptic endocytic zone (GO:0098843)  \n",
       "Signaling: GTPase                                                                   Signaling: GTPase  \n",
       "Plasticity: EGF receptor                                                     Plasticity: EGF receptor  \n",
       "Plasticity: Pluripotency                                                     Plasticity: Pluripotency  \n",
       "regulation of presynaptic dense core vesicle ex...  regulation of presynaptic dense core vesicle e...  \n",
       "PAP: PAP Plasticity                                                               PAP: PAP Plasticity  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ct = cell_supertype\n",
    "pw = 'low_removed_cura_terms'\n",
    "dt = 'ad_vs_no'\n",
    "\n",
    "tmp2 = all_data[pw]['fits_all'][ct][dt][(abs(all_data[pw]['fits_all'][ct][dt]['P.Value']<0.01))]\n",
    "tmp2.head(50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11e4f393",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>logFC</th>\n",
       "      <th>CI.L</th>\n",
       "      <th>CI.R</th>\n",
       "      <th>AveExpr</th>\n",
       "      <th>t</th>\n",
       "      <th>P.Value</th>\n",
       "      <th>adj.P.Val</th>\n",
       "      <th>B</th>\n",
       "      <th>celltype</th>\n",
       "      <th>names</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>negative regulation of stress-activated protein kinase signaling cascade (GO:0070303)</th>\n",
       "      <td>-0.542201</td>\n",
       "      <td>-0.805352</td>\n",
       "      <td>-0.279050</td>\n",
       "      <td>-0.002864</td>\n",
       "      <td>-4.188976</td>\n",
       "      <td>0.000191</td>\n",
       "      <td>0.363446</td>\n",
       "      <td>0.048592</td>\n",
       "      <td>HRK</td>\n",
       "      <td>negative regulation of stress-activated protei...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cellular response to virus (GO:0098586)</th>\n",
       "      <td>-0.341615</td>\n",
       "      <td>-0.519055</td>\n",
       "      <td>-0.164176</td>\n",
       "      <td>0.000652</td>\n",
       "      <td>-3.914170</td>\n",
       "      <td>0.000419</td>\n",
       "      <td>0.363446</td>\n",
       "      <td>-0.495876</td>\n",
       "      <td>HRK</td>\n",
       "      <td>cellular response to virus (GO:0098586)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>positive regulation of cytoplasmic translation (GO:2000767)</th>\n",
       "      <td>-0.501110</td>\n",
       "      <td>-0.762715</td>\n",
       "      <td>-0.239506</td>\n",
       "      <td>-0.012460</td>\n",
       "      <td>-3.894412</td>\n",
       "      <td>0.000444</td>\n",
       "      <td>0.363446</td>\n",
       "      <td>-0.534781</td>\n",
       "      <td>HRK</td>\n",
       "      <td>positive regulation of cytoplasmic translation...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>caveola assembly (GO:0070836)</th>\n",
       "      <td>-0.469296</td>\n",
       "      <td>-0.716437</td>\n",
       "      <td>-0.222156</td>\n",
       "      <td>-0.063081</td>\n",
       "      <td>-3.860611</td>\n",
       "      <td>0.000488</td>\n",
       "      <td>0.363446</td>\n",
       "      <td>-0.601247</td>\n",
       "      <td>HRK</td>\n",
       "      <td>caveola assembly (GO:0070836)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>negative regulation of establishment of protein localization (GO:1904950)</th>\n",
       "      <td>-0.456388</td>\n",
       "      <td>-0.696744</td>\n",
       "      <td>-0.216032</td>\n",
       "      <td>-0.023514</td>\n",
       "      <td>-3.860404</td>\n",
       "      <td>0.000488</td>\n",
       "      <td>0.363446</td>\n",
       "      <td>-0.601652</td>\n",
       "      <td>HRK</td>\n",
       "      <td>negative regulation of establishment of protei...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>positive regulation of transcription regulatory region DNA binding (GO:2000679)</th>\n",
       "      <td>-0.346670</td>\n",
       "      <td>-0.532055</td>\n",
       "      <td>-0.161285</td>\n",
       "      <td>0.004294</td>\n",
       "      <td>-3.801845</td>\n",
       "      <td>0.000576</td>\n",
       "      <td>0.363446</td>\n",
       "      <td>-0.716505</td>\n",
       "      <td>HRK</td>\n",
       "      <td>positive regulation of transcription regulator...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>regulation of secretion by cell (GO:1903530)</th>\n",
       "      <td>-0.243568</td>\n",
       "      <td>-0.376268</td>\n",
       "      <td>-0.110867</td>\n",
       "      <td>-0.007719</td>\n",
       "      <td>-3.731637</td>\n",
       "      <td>0.000702</td>\n",
       "      <td>0.379214</td>\n",
       "      <td>-0.853673</td>\n",
       "      <td>HRK</td>\n",
       "      <td>regulation of secretion by cell (GO:1903530)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>positive regulation of deacetylase activity (GO:0090045)</th>\n",
       "      <td>0.554247</td>\n",
       "      <td>0.244806</td>\n",
       "      <td>0.863689</td>\n",
       "      <td>-0.000473</td>\n",
       "      <td>3.641482</td>\n",
       "      <td>0.000902</td>\n",
       "      <td>0.379214</td>\n",
       "      <td>-1.028864</td>\n",
       "      <td>HRK</td>\n",
       "      <td>positive regulation of deacetylase activity (G...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>protein insertion into ER membrane (GO:0045048)</th>\n",
       "      <td>-0.356374</td>\n",
       "      <td>-0.556861</td>\n",
       "      <td>-0.155887</td>\n",
       "      <td>-0.018634</td>\n",
       "      <td>-3.613872</td>\n",
       "      <td>0.000974</td>\n",
       "      <td>0.379214</td>\n",
       "      <td>-1.082284</td>\n",
       "      <td>HRK</td>\n",
       "      <td>protein insertion into ER membrane (GO:0045048)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>endoplasmic reticulum organization (GO:0007029)</th>\n",
       "      <td>-0.248904</td>\n",
       "      <td>-0.389325</td>\n",
       "      <td>-0.108484</td>\n",
       "      <td>-0.022258</td>\n",
       "      <td>-3.603756</td>\n",
       "      <td>0.001002</td>\n",
       "      <td>0.379214</td>\n",
       "      <td>-1.101828</td>\n",
       "      <td>HRK</td>\n",
       "      <td>endoplasmic reticulum organization (GO:0007029)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>negative regulation of cellular component movement (GO:0051271)</th>\n",
       "      <td>-0.350024</td>\n",
       "      <td>-0.552181</td>\n",
       "      <td>-0.147868</td>\n",
       "      <td>-0.002283</td>\n",
       "      <td>-3.520165</td>\n",
       "      <td>0.001262</td>\n",
       "      <td>0.434133</td>\n",
       "      <td>-1.262699</td>\n",
       "      <td>HRK</td>\n",
       "      <td>negative regulation of cellular component move...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>positive regulation of protein localization to centrosome (GO:1904781)</th>\n",
       "      <td>0.571485</td>\n",
       "      <td>0.232578</td>\n",
       "      <td>0.910392</td>\n",
       "      <td>-0.043352</td>\n",
       "      <td>3.428288</td>\n",
       "      <td>0.001622</td>\n",
       "      <td>0.503393</td>\n",
       "      <td>-1.438139</td>\n",
       "      <td>HRK</td>\n",
       "      <td>positive regulation of protein localization to...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MyD88-independent toll-like receptor signaling pathway (GO:0002756)</th>\n",
       "      <td>-0.291130</td>\n",
       "      <td>-0.466862</td>\n",
       "      <td>-0.115397</td>\n",
       "      <td>-0.028807</td>\n",
       "      <td>-3.368121</td>\n",
       "      <td>0.001910</td>\n",
       "      <td>0.503393</td>\n",
       "      <td>-1.552174</td>\n",
       "      <td>HRK</td>\n",
       "      <td>MyD88-independent toll-like receptor signaling...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>positive regulation of leukocyte chemotaxis (GO:0002690)</th>\n",
       "      <td>-0.261426</td>\n",
       "      <td>-0.420990</td>\n",
       "      <td>-0.101861</td>\n",
       "      <td>-0.008022</td>\n",
       "      <td>-3.330933</td>\n",
       "      <td>0.002111</td>\n",
       "      <td>0.503393</td>\n",
       "      <td>-1.622296</td>\n",
       "      <td>HRK</td>\n",
       "      <td>positive regulation of leukocyte chemotaxis (G...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>protein insertion into membrane (GO:0051205)</th>\n",
       "      <td>-0.281639</td>\n",
       "      <td>-0.455926</td>\n",
       "      <td>-0.107351</td>\n",
       "      <td>-0.029968</td>\n",
       "      <td>-3.285337</td>\n",
       "      <td>0.002386</td>\n",
       "      <td>0.503393</td>\n",
       "      <td>-1.707875</td>\n",
       "      <td>HRK</td>\n",
       "      <td>protein insertion into membrane (GO:0051205)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>regulation of cell projection organization (GO:0031344)</th>\n",
       "      <td>0.481681</td>\n",
       "      <td>0.181380</td>\n",
       "      <td>0.781982</td>\n",
       "      <td>-0.014370</td>\n",
       "      <td>3.261039</td>\n",
       "      <td>0.002546</td>\n",
       "      <td>0.503393</td>\n",
       "      <td>-1.753297</td>\n",
       "      <td>HRK</td>\n",
       "      <td>regulation of cell projection organization (GO...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>regulation of protein localization to centrosome (GO:1904779)</th>\n",
       "      <td>0.474222</td>\n",
       "      <td>0.175779</td>\n",
       "      <td>0.772665</td>\n",
       "      <td>-0.040933</td>\n",
       "      <td>3.230526</td>\n",
       "      <td>0.002762</td>\n",
       "      <td>0.503393</td>\n",
       "      <td>-1.810150</td>\n",
       "      <td>HRK</td>\n",
       "      <td>regulation of protein localization to centroso...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>negative regulation of trophoblast cell migration (GO:1901164)</th>\n",
       "      <td>-0.546252</td>\n",
       "      <td>-0.892517</td>\n",
       "      <td>-0.199987</td>\n",
       "      <td>-0.002468</td>\n",
       "      <td>-3.207283</td>\n",
       "      <td>0.002938</td>\n",
       "      <td>0.503393</td>\n",
       "      <td>-1.853311</td>\n",
       "      <td>HRK</td>\n",
       "      <td>negative regulation of trophoblast cell migrat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Notch signaling involved in heart development (GO:0061314)</th>\n",
       "      <td>-0.306714</td>\n",
       "      <td>-0.501979</td>\n",
       "      <td>-0.111449</td>\n",
       "      <td>-0.014230</td>\n",
       "      <td>-3.193460</td>\n",
       "      <td>0.003048</td>\n",
       "      <td>0.503393</td>\n",
       "      <td>-1.878922</td>\n",
       "      <td>HRK</td>\n",
       "      <td>Notch signaling involved in heart development ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>positive regulation of transmembrane transport (GO:0034764)</th>\n",
       "      <td>0.421186</td>\n",
       "      <td>0.152345</td>\n",
       "      <td>0.690027</td>\n",
       "      <td>-0.058193</td>\n",
       "      <td>3.185162</td>\n",
       "      <td>0.003116</td>\n",
       "      <td>0.503393</td>\n",
       "      <td>-1.894273</td>\n",
       "      <td>HRK</td>\n",
       "      <td>positive regulation of transmembrane transport...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>negative regulation of ubiquitin-protein transferase activity (GO:0051444)</th>\n",
       "      <td>-0.348098</td>\n",
       "      <td>-0.571019</td>\n",
       "      <td>-0.125177</td>\n",
       "      <td>-0.033611</td>\n",
       "      <td>-3.174706</td>\n",
       "      <td>0.003203</td>\n",
       "      <td>0.503393</td>\n",
       "      <td>-1.913595</td>\n",
       "      <td>HRK</td>\n",
       "      <td>negative regulation of ubiquitin-protein trans...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>apoptotic mitochondrial changes (GO:0008637)</th>\n",
       "      <td>-0.325130</td>\n",
       "      <td>-0.537266</td>\n",
       "      <td>-0.112995</td>\n",
       "      <td>-0.019424</td>\n",
       "      <td>-3.115997</td>\n",
       "      <td>0.003738</td>\n",
       "      <td>0.503393</td>\n",
       "      <td>-2.021580</td>\n",
       "      <td>HRK</td>\n",
       "      <td>apoptotic mitochondrial changes (GO:0008637)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>release of cytochrome c from mitochondria (GO:0001836)</th>\n",
       "      <td>-0.392873</td>\n",
       "      <td>-0.650477</td>\n",
       "      <td>-0.135269</td>\n",
       "      <td>-0.008697</td>\n",
       "      <td>-3.100651</td>\n",
       "      <td>0.003892</td>\n",
       "      <td>0.503393</td>\n",
       "      <td>-2.049662</td>\n",
       "      <td>HRK</td>\n",
       "      <td>release of cytochrome c from mitochondria (GO:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>positive regulation of NIK/NF-kappaB signaling (GO:1901224)</th>\n",
       "      <td>-0.217835</td>\n",
       "      <td>-0.360900</td>\n",
       "      <td>-0.074769</td>\n",
       "      <td>-0.011845</td>\n",
       "      <td>-3.095606</td>\n",
       "      <td>0.003943</td>\n",
       "      <td>0.503393</td>\n",
       "      <td>-2.058882</td>\n",
       "      <td>HRK</td>\n",
       "      <td>positive regulation of NIK/NF-kappaB signaling...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>protein palmitoylation (GO:0018345)</th>\n",
       "      <td>0.228423</td>\n",
       "      <td>0.077940</td>\n",
       "      <td>0.378905</td>\n",
       "      <td>0.016558</td>\n",
       "      <td>3.086071</td>\n",
       "      <td>0.004043</td>\n",
       "      <td>0.503393</td>\n",
       "      <td>-2.076288</td>\n",
       "      <td>HRK</td>\n",
       "      <td>protein palmitoylation (GO:0018345)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>positive regulation of type I interferon production (GO:0032481)</th>\n",
       "      <td>-0.228814</td>\n",
       "      <td>-0.379922</td>\n",
       "      <td>-0.077707</td>\n",
       "      <td>-0.020382</td>\n",
       "      <td>-3.078573</td>\n",
       "      <td>0.004123</td>\n",
       "      <td>0.503393</td>\n",
       "      <td>-2.089959</td>\n",
       "      <td>HRK</td>\n",
       "      <td>positive regulation of type I interferon produ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ear development (GO:0043583)</th>\n",
       "      <td>0.576667</td>\n",
       "      <td>0.195298</td>\n",
       "      <td>0.958037</td>\n",
       "      <td>-0.038880</td>\n",
       "      <td>3.074207</td>\n",
       "      <td>0.004170</td>\n",
       "      <td>0.503393</td>\n",
       "      <td>-2.097912</td>\n",
       "      <td>HRK</td>\n",
       "      <td>ear development (GO:0043583)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>skeletal muscle organ development (GO:0060538)</th>\n",
       "      <td>-0.303389</td>\n",
       "      <td>-0.504367</td>\n",
       "      <td>-0.102410</td>\n",
       "      <td>-0.037422</td>\n",
       "      <td>-3.069040</td>\n",
       "      <td>0.004226</td>\n",
       "      <td>0.503393</td>\n",
       "      <td>-2.107319</td>\n",
       "      <td>HRK</td>\n",
       "      <td>skeletal muscle organ development (GO:0060538)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mature ribosome assembly (GO:0042256)</th>\n",
       "      <td>-0.475651</td>\n",
       "      <td>-0.793871</td>\n",
       "      <td>-0.157432</td>\n",
       "      <td>0.001060</td>\n",
       "      <td>-3.038888</td>\n",
       "      <td>0.004571</td>\n",
       "      <td>0.503393</td>\n",
       "      <td>-2.162067</td>\n",
       "      <td>HRK</td>\n",
       "      <td>mature ribosome assembly (GO:0042256)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>histone H4-K16 acetylation (GO:0043984)</th>\n",
       "      <td>0.253542</td>\n",
       "      <td>0.083639</td>\n",
       "      <td>0.423444</td>\n",
       "      <td>0.006033</td>\n",
       "      <td>3.033911</td>\n",
       "      <td>0.004631</td>\n",
       "      <td>0.503393</td>\n",
       "      <td>-2.171079</td>\n",
       "      <td>HRK</td>\n",
       "      <td>histone H4-K16 acetylation (GO:0043984)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>peptidyl-L-cysteine S-palmitoylation (GO:0018230)</th>\n",
       "      <td>0.259847</td>\n",
       "      <td>0.085199</td>\n",
       "      <td>0.434495</td>\n",
       "      <td>0.020371</td>\n",
       "      <td>3.024869</td>\n",
       "      <td>0.004740</td>\n",
       "      <td>0.503393</td>\n",
       "      <td>-2.187437</td>\n",
       "      <td>HRK</td>\n",
       "      <td>peptidyl-L-cysteine S-palmitoylation (GO:0018230)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>histone H4-K5 acetylation (GO:0043981)</th>\n",
       "      <td>0.362629</td>\n",
       "      <td>0.117979</td>\n",
       "      <td>0.607280</td>\n",
       "      <td>0.015651</td>\n",
       "      <td>3.013489</td>\n",
       "      <td>0.004882</td>\n",
       "      <td>0.503393</td>\n",
       "      <td>-2.207992</td>\n",
       "      <td>HRK</td>\n",
       "      <td>histone H4-K5 acetylation (GO:0043981)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>positive regulation of DNA binding (GO:0043388)</th>\n",
       "      <td>-0.344155</td>\n",
       "      <td>-0.577108</td>\n",
       "      <td>-0.111202</td>\n",
       "      <td>-0.013005</td>\n",
       "      <td>-3.003581</td>\n",
       "      <td>0.005009</td>\n",
       "      <td>0.503393</td>\n",
       "      <td>-2.225859</td>\n",
       "      <td>HRK</td>\n",
       "      <td>positive regulation of DNA binding (GO:0043388)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>long-term memory (GO:0007616)</th>\n",
       "      <td>-0.391963</td>\n",
       "      <td>-0.659603</td>\n",
       "      <td>-0.124324</td>\n",
       "      <td>-0.000705</td>\n",
       "      <td>-2.977475</td>\n",
       "      <td>0.005357</td>\n",
       "      <td>0.503393</td>\n",
       "      <td>-2.272805</td>\n",
       "      <td>HRK</td>\n",
       "      <td>long-term memory (GO:0007616)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>positive regulation of binding (GO:0051099)</th>\n",
       "      <td>-0.165004</td>\n",
       "      <td>-0.277765</td>\n",
       "      <td>-0.052243</td>\n",
       "      <td>-0.010414</td>\n",
       "      <td>-2.975014</td>\n",
       "      <td>0.005391</td>\n",
       "      <td>0.503393</td>\n",
       "      <td>-2.277220</td>\n",
       "      <td>HRK</td>\n",
       "      <td>positive regulation of binding (GO:0051099)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>response to interferon-alpha (GO:0035455)</th>\n",
       "      <td>-0.352297</td>\n",
       "      <td>-0.595227</td>\n",
       "      <td>-0.109367</td>\n",
       "      <td>-0.049735</td>\n",
       "      <td>-2.948361</td>\n",
       "      <td>0.005773</td>\n",
       "      <td>0.503393</td>\n",
       "      <td>-2.324928</td>\n",
       "      <td>HRK</td>\n",
       "      <td>response to interferon-alpha (GO:0035455)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DNA topological change (GO:0006265)</th>\n",
       "      <td>-0.404327</td>\n",
       "      <td>-0.684834</td>\n",
       "      <td>-0.123820</td>\n",
       "      <td>-0.027384</td>\n",
       "      <td>-2.930504</td>\n",
       "      <td>0.006043</td>\n",
       "      <td>0.503393</td>\n",
       "      <td>-2.356777</td>\n",
       "      <td>HRK</td>\n",
       "      <td>DNA topological change (GO:0006265)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>regulation of epidermal growth factor receptor signaling pathway (GO:0042058)</th>\n",
       "      <td>0.244380</td>\n",
       "      <td>0.073810</td>\n",
       "      <td>0.414951</td>\n",
       "      <td>-0.002798</td>\n",
       "      <td>2.912833</td>\n",
       "      <td>0.006321</td>\n",
       "      <td>0.503393</td>\n",
       "      <td>-2.388200</td>\n",
       "      <td>HRK</td>\n",
       "      <td>regulation of epidermal growth factor receptor...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mitochondrial electron transport, cytochrome c to oxygen (GO:0006123)</th>\n",
       "      <td>-0.501468</td>\n",
       "      <td>-0.852392</td>\n",
       "      <td>-0.150543</td>\n",
       "      <td>-0.034879</td>\n",
       "      <td>-2.905244</td>\n",
       "      <td>0.006445</td>\n",
       "      <td>0.503393</td>\n",
       "      <td>-2.401666</td>\n",
       "      <td>HRK</td>\n",
       "      <td>mitochondrial electron transport, cytochrome c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>regulation of ventricular cardiac muscle cell action potential (GO:0098911)</th>\n",
       "      <td>-0.369135</td>\n",
       "      <td>-0.627931</td>\n",
       "      <td>-0.110339</td>\n",
       "      <td>0.018386</td>\n",
       "      <td>-2.899887</td>\n",
       "      <td>0.006533</td>\n",
       "      <td>0.503393</td>\n",
       "      <td>-2.411161</td>\n",
       "      <td>HRK</td>\n",
       "      <td>regulation of ventricular cardiac muscle cell ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>negative regulation of ion transmembrane transporter activity (GO:0032413)</th>\n",
       "      <td>0.310817</td>\n",
       "      <td>0.092339</td>\n",
       "      <td>0.529295</td>\n",
       "      <td>-0.009918</td>\n",
       "      <td>2.892342</td>\n",
       "      <td>0.006660</td>\n",
       "      <td>0.503393</td>\n",
       "      <td>-2.424520</td>\n",
       "      <td>HRK</td>\n",
       "      <td>negative regulation of ion transmembrane trans...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>negative regulation of potassium ion transmembrane transporter activity (GO:1901017)</th>\n",
       "      <td>0.402931</td>\n",
       "      <td>0.119311</td>\n",
       "      <td>0.686551</td>\n",
       "      <td>0.019980</td>\n",
       "      <td>2.888331</td>\n",
       "      <td>0.006728</td>\n",
       "      <td>0.503393</td>\n",
       "      <td>-2.431614</td>\n",
       "      <td>HRK</td>\n",
       "      <td>negative regulation of potassium ion transmemb...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>fructose metabolic process (GO:0006000)</th>\n",
       "      <td>-0.599763</td>\n",
       "      <td>-1.022296</td>\n",
       "      <td>-0.177229</td>\n",
       "      <td>0.006361</td>\n",
       "      <td>-2.885832</td>\n",
       "      <td>0.006771</td>\n",
       "      <td>0.503393</td>\n",
       "      <td>-2.436033</td>\n",
       "      <td>HRK</td>\n",
       "      <td>fructose metabolic process (GO:0006000)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>positive regulation of protein sumoylation (GO:0033235)</th>\n",
       "      <td>0.389092</td>\n",
       "      <td>0.114320</td>\n",
       "      <td>0.663863</td>\n",
       "      <td>0.039847</td>\n",
       "      <td>2.878941</td>\n",
       "      <td>0.006890</td>\n",
       "      <td>0.503393</td>\n",
       "      <td>-2.448205</td>\n",
       "      <td>HRK</td>\n",
       "      <td>positive regulation of protein sumoylation (GO...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ER-associated misfolded protein catabolic process (GO:0071712)</th>\n",
       "      <td>-0.299129</td>\n",
       "      <td>-0.510373</td>\n",
       "      <td>-0.087885</td>\n",
       "      <td>0.020263</td>\n",
       "      <td>-2.878910</td>\n",
       "      <td>0.006890</td>\n",
       "      <td>0.503393</td>\n",
       "      <td>-2.448258</td>\n",
       "      <td>HRK</td>\n",
       "      <td>ER-associated misfolded protein catabolic proc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>central nervous system projection neuron axonogenesis (GO:0021952)</th>\n",
       "      <td>0.409161</td>\n",
       "      <td>0.119379</td>\n",
       "      <td>0.698943</td>\n",
       "      <td>-0.033785</td>\n",
       "      <td>2.870623</td>\n",
       "      <td>0.007037</td>\n",
       "      <td>0.503393</td>\n",
       "      <td>-2.462877</td>\n",
       "      <td>HRK</td>\n",
       "      <td>central nervous system projection neuron axono...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>negative regulation of cell size (GO:0045792)</th>\n",
       "      <td>-0.352537</td>\n",
       "      <td>-0.602868</td>\n",
       "      <td>-0.102206</td>\n",
       "      <td>-0.003693</td>\n",
       "      <td>-2.863143</td>\n",
       "      <td>0.007171</td>\n",
       "      <td>0.503393</td>\n",
       "      <td>-2.476053</td>\n",
       "      <td>HRK</td>\n",
       "      <td>negative regulation of cell size (GO:0045792)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>heart process (GO:0003015)</th>\n",
       "      <td>-0.417689</td>\n",
       "      <td>-0.715604</td>\n",
       "      <td>-0.119775</td>\n",
       "      <td>0.019632</td>\n",
       "      <td>-2.850460</td>\n",
       "      <td>0.007404</td>\n",
       "      <td>0.503393</td>\n",
       "      <td>-2.498356</td>\n",
       "      <td>HRK</td>\n",
       "      <td>heart process (GO:0003015)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>glial cell migration (GO:0008347)</th>\n",
       "      <td>-0.527087</td>\n",
       "      <td>-0.903644</td>\n",
       "      <td>-0.150530</td>\n",
       "      <td>0.028544</td>\n",
       "      <td>-2.845800</td>\n",
       "      <td>0.007492</td>\n",
       "      <td>0.503393</td>\n",
       "      <td>-2.506536</td>\n",
       "      <td>HRK</td>\n",
       "      <td>glial cell migration (GO:0008347)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA modification (GO:0009451)</th>\n",
       "      <td>-0.217435</td>\n",
       "      <td>-0.373442</td>\n",
       "      <td>-0.061428</td>\n",
       "      <td>0.015951</td>\n",
       "      <td>-2.833596</td>\n",
       "      <td>0.007726</td>\n",
       "      <td>0.503393</td>\n",
       "      <td>-2.527930</td>\n",
       "      <td>HRK</td>\n",
       "      <td>RNA modification (GO:0009451)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                       logFC      CI.L  \\\n",
       "negative regulation of stress-activated protein... -0.542201 -0.805352   \n",
       "cellular response to virus (GO:0098586)            -0.341615 -0.519055   \n",
       "positive regulation of cytoplasmic translation ... -0.501110 -0.762715   \n",
       "caveola assembly (GO:0070836)                      -0.469296 -0.716437   \n",
       "negative regulation of establishment of protein... -0.456388 -0.696744   \n",
       "positive regulation of transcription regulatory... -0.346670 -0.532055   \n",
       "regulation of secretion by cell (GO:1903530)       -0.243568 -0.376268   \n",
       "positive regulation of deacetylase activity (GO...  0.554247  0.244806   \n",
       "protein insertion into ER membrane (GO:0045048)    -0.356374 -0.556861   \n",
       "endoplasmic reticulum organization (GO:0007029)    -0.248904 -0.389325   \n",
       "negative regulation of cellular component movem... -0.350024 -0.552181   \n",
       "positive regulation of protein localization to ...  0.571485  0.232578   \n",
       "MyD88-independent toll-like receptor signaling ... -0.291130 -0.466862   \n",
       "positive regulation of leukocyte chemotaxis (GO... -0.261426 -0.420990   \n",
       "protein insertion into membrane (GO:0051205)       -0.281639 -0.455926   \n",
       "regulation of cell projection organization (GO:...  0.481681  0.181380   \n",
       "regulation of protein localization to centrosom...  0.474222  0.175779   \n",
       "negative regulation of trophoblast cell migrati... -0.546252 -0.892517   \n",
       "Notch signaling involved in heart development (... -0.306714 -0.501979   \n",
       "positive regulation of transmembrane transport ...  0.421186  0.152345   \n",
       "negative regulation of ubiquitin-protein transf... -0.348098 -0.571019   \n",
       "apoptotic mitochondrial changes (GO:0008637)       -0.325130 -0.537266   \n",
       "release of cytochrome c from mitochondria (GO:0... -0.392873 -0.650477   \n",
       "positive regulation of NIK/NF-kappaB signaling ... -0.217835 -0.360900   \n",
       "protein palmitoylation (GO:0018345)                 0.228423  0.077940   \n",
       "positive regulation of type I interferon produc... -0.228814 -0.379922   \n",
       "ear development (GO:0043583)                        0.576667  0.195298   \n",
       "skeletal muscle organ development (GO:0060538)     -0.303389 -0.504367   \n",
       "mature ribosome assembly (GO:0042256)              -0.475651 -0.793871   \n",
       "histone H4-K16 acetylation (GO:0043984)             0.253542  0.083639   \n",
       "peptidyl-L-cysteine S-palmitoylation (GO:0018230)   0.259847  0.085199   \n",
       "histone H4-K5 acetylation (GO:0043981)              0.362629  0.117979   \n",
       "positive regulation of DNA binding (GO:0043388)    -0.344155 -0.577108   \n",
       "long-term memory (GO:0007616)                      -0.391963 -0.659603   \n",
       "positive regulation of binding (GO:0051099)        -0.165004 -0.277765   \n",
       "response to interferon-alpha (GO:0035455)          -0.352297 -0.595227   \n",
       "DNA topological change (GO:0006265)                -0.404327 -0.684834   \n",
       "regulation of epidermal growth factor receptor ...  0.244380  0.073810   \n",
       "mitochondrial electron transport, cytochrome c ... -0.501468 -0.852392   \n",
       "regulation of ventricular cardiac muscle cell a... -0.369135 -0.627931   \n",
       "negative regulation of ion transmembrane transp...  0.310817  0.092339   \n",
       "negative regulation of potassium ion transmembr...  0.402931  0.119311   \n",
       "fructose metabolic process (GO:0006000)            -0.599763 -1.022296   \n",
       "positive regulation of protein sumoylation (GO:...  0.389092  0.114320   \n",
       "ER-associated misfolded protein catabolic proce... -0.299129 -0.510373   \n",
       "central nervous system projection neuron axonog...  0.409161  0.119379   \n",
       "negative regulation of cell size (GO:0045792)      -0.352537 -0.602868   \n",
       "heart process (GO:0003015)                         -0.417689 -0.715604   \n",
       "glial cell migration (GO:0008347)                  -0.527087 -0.903644   \n",
       "RNA modification (GO:0009451)                      -0.217435 -0.373442   \n",
       "\n",
       "                                                        CI.R   AveExpr  \\\n",
       "negative regulation of stress-activated protein... -0.279050 -0.002864   \n",
       "cellular response to virus (GO:0098586)            -0.164176  0.000652   \n",
       "positive regulation of cytoplasmic translation ... -0.239506 -0.012460   \n",
       "caveola assembly (GO:0070836)                      -0.222156 -0.063081   \n",
       "negative regulation of establishment of protein... -0.216032 -0.023514   \n",
       "positive regulation of transcription regulatory... -0.161285  0.004294   \n",
       "regulation of secretion by cell (GO:1903530)       -0.110867 -0.007719   \n",
       "positive regulation of deacetylase activity (GO...  0.863689 -0.000473   \n",
       "protein insertion into ER membrane (GO:0045048)    -0.155887 -0.018634   \n",
       "endoplasmic reticulum organization (GO:0007029)    -0.108484 -0.022258   \n",
       "negative regulation of cellular component movem... -0.147868 -0.002283   \n",
       "positive regulation of protein localization to ...  0.910392 -0.043352   \n",
       "MyD88-independent toll-like receptor signaling ... -0.115397 -0.028807   \n",
       "positive regulation of leukocyte chemotaxis (GO... -0.101861 -0.008022   \n",
       "protein insertion into membrane (GO:0051205)       -0.107351 -0.029968   \n",
       "regulation of cell projection organization (GO:...  0.781982 -0.014370   \n",
       "regulation of protein localization to centrosom...  0.772665 -0.040933   \n",
       "negative regulation of trophoblast cell migrati... -0.199987 -0.002468   \n",
       "Notch signaling involved in heart development (... -0.111449 -0.014230   \n",
       "positive regulation of transmembrane transport ...  0.690027 -0.058193   \n",
       "negative regulation of ubiquitin-protein transf... -0.125177 -0.033611   \n",
       "apoptotic mitochondrial changes (GO:0008637)       -0.112995 -0.019424   \n",
       "release of cytochrome c from mitochondria (GO:0... -0.135269 -0.008697   \n",
       "positive regulation of NIK/NF-kappaB signaling ... -0.074769 -0.011845   \n",
       "protein palmitoylation (GO:0018345)                 0.378905  0.016558   \n",
       "positive regulation of type I interferon produc... -0.077707 -0.020382   \n",
       "ear development (GO:0043583)                        0.958037 -0.038880   \n",
       "skeletal muscle organ development (GO:0060538)     -0.102410 -0.037422   \n",
       "mature ribosome assembly (GO:0042256)              -0.157432  0.001060   \n",
       "histone H4-K16 acetylation (GO:0043984)             0.423444  0.006033   \n",
       "peptidyl-L-cysteine S-palmitoylation (GO:0018230)   0.434495  0.020371   \n",
       "histone H4-K5 acetylation (GO:0043981)              0.607280  0.015651   \n",
       "positive regulation of DNA binding (GO:0043388)    -0.111202 -0.013005   \n",
       "long-term memory (GO:0007616)                      -0.124324 -0.000705   \n",
       "positive regulation of binding (GO:0051099)        -0.052243 -0.010414   \n",
       "response to interferon-alpha (GO:0035455)          -0.109367 -0.049735   \n",
       "DNA topological change (GO:0006265)                -0.123820 -0.027384   \n",
       "regulation of epidermal growth factor receptor ...  0.414951 -0.002798   \n",
       "mitochondrial electron transport, cytochrome c ... -0.150543 -0.034879   \n",
       "regulation of ventricular cardiac muscle cell a... -0.110339  0.018386   \n",
       "negative regulation of ion transmembrane transp...  0.529295 -0.009918   \n",
       "negative regulation of potassium ion transmembr...  0.686551  0.019980   \n",
       "fructose metabolic process (GO:0006000)            -0.177229  0.006361   \n",
       "positive regulation of protein sumoylation (GO:...  0.663863  0.039847   \n",
       "ER-associated misfolded protein catabolic proce... -0.087885  0.020263   \n",
       "central nervous system projection neuron axonog...  0.698943 -0.033785   \n",
       "negative regulation of cell size (GO:0045792)      -0.102206 -0.003693   \n",
       "heart process (GO:0003015)                         -0.119775  0.019632   \n",
       "glial cell migration (GO:0008347)                  -0.150530  0.028544   \n",
       "RNA modification (GO:0009451)                      -0.061428  0.015951   \n",
       "\n",
       "                                                           t   P.Value  \\\n",
       "negative regulation of stress-activated protein... -4.188976  0.000191   \n",
       "cellular response to virus (GO:0098586)            -3.914170  0.000419   \n",
       "positive regulation of cytoplasmic translation ... -3.894412  0.000444   \n",
       "caveola assembly (GO:0070836)                      -3.860611  0.000488   \n",
       "negative regulation of establishment of protein... -3.860404  0.000488   \n",
       "positive regulation of transcription regulatory... -3.801845  0.000576   \n",
       "regulation of secretion by cell (GO:1903530)       -3.731637  0.000702   \n",
       "positive regulation of deacetylase activity (GO...  3.641482  0.000902   \n",
       "protein insertion into ER membrane (GO:0045048)    -3.613872  0.000974   \n",
       "endoplasmic reticulum organization (GO:0007029)    -3.603756  0.001002   \n",
       "negative regulation of cellular component movem... -3.520165  0.001262   \n",
       "positive regulation of protein localization to ...  3.428288  0.001622   \n",
       "MyD88-independent toll-like receptor signaling ... -3.368121  0.001910   \n",
       "positive regulation of leukocyte chemotaxis (GO... -3.330933  0.002111   \n",
       "protein insertion into membrane (GO:0051205)       -3.285337  0.002386   \n",
       "regulation of cell projection organization (GO:...  3.261039  0.002546   \n",
       "regulation of protein localization to centrosom...  3.230526  0.002762   \n",
       "negative regulation of trophoblast cell migrati... -3.207283  0.002938   \n",
       "Notch signaling involved in heart development (... -3.193460  0.003048   \n",
       "positive regulation of transmembrane transport ...  3.185162  0.003116   \n",
       "negative regulation of ubiquitin-protein transf... -3.174706  0.003203   \n",
       "apoptotic mitochondrial changes (GO:0008637)       -3.115997  0.003738   \n",
       "release of cytochrome c from mitochondria (GO:0... -3.100651  0.003892   \n",
       "positive regulation of NIK/NF-kappaB signaling ... -3.095606  0.003943   \n",
       "protein palmitoylation (GO:0018345)                 3.086071  0.004043   \n",
       "positive regulation of type I interferon produc... -3.078573  0.004123   \n",
       "ear development (GO:0043583)                        3.074207  0.004170   \n",
       "skeletal muscle organ development (GO:0060538)     -3.069040  0.004226   \n",
       "mature ribosome assembly (GO:0042256)              -3.038888  0.004571   \n",
       "histone H4-K16 acetylation (GO:0043984)             3.033911  0.004631   \n",
       "peptidyl-L-cysteine S-palmitoylation (GO:0018230)   3.024869  0.004740   \n",
       "histone H4-K5 acetylation (GO:0043981)              3.013489  0.004882   \n",
       "positive regulation of DNA binding (GO:0043388)    -3.003581  0.005009   \n",
       "long-term memory (GO:0007616)                      -2.977475  0.005357   \n",
       "positive regulation of binding (GO:0051099)        -2.975014  0.005391   \n",
       "response to interferon-alpha (GO:0035455)          -2.948361  0.005773   \n",
       "DNA topological change (GO:0006265)                -2.930504  0.006043   \n",
       "regulation of epidermal growth factor receptor ...  2.912833  0.006321   \n",
       "mitochondrial electron transport, cytochrome c ... -2.905244  0.006445   \n",
       "regulation of ventricular cardiac muscle cell a... -2.899887  0.006533   \n",
       "negative regulation of ion transmembrane transp...  2.892342  0.006660   \n",
       "negative regulation of potassium ion transmembr...  2.888331  0.006728   \n",
       "fructose metabolic process (GO:0006000)            -2.885832  0.006771   \n",
       "positive regulation of protein sumoylation (GO:...  2.878941  0.006890   \n",
       "ER-associated misfolded protein catabolic proce... -2.878910  0.006890   \n",
       "central nervous system projection neuron axonog...  2.870623  0.007037   \n",
       "negative regulation of cell size (GO:0045792)      -2.863143  0.007171   \n",
       "heart process (GO:0003015)                         -2.850460  0.007404   \n",
       "glial cell migration (GO:0008347)                  -2.845800  0.007492   \n",
       "RNA modification (GO:0009451)                      -2.833596  0.007726   \n",
       "\n",
       "                                                    adj.P.Val         B  \\\n",
       "negative regulation of stress-activated protein...   0.363446  0.048592   \n",
       "cellular response to virus (GO:0098586)              0.363446 -0.495876   \n",
       "positive regulation of cytoplasmic translation ...   0.363446 -0.534781   \n",
       "caveola assembly (GO:0070836)                        0.363446 -0.601247   \n",
       "negative regulation of establishment of protein...   0.363446 -0.601652   \n",
       "positive regulation of transcription regulatory...   0.363446 -0.716505   \n",
       "regulation of secretion by cell (GO:1903530)         0.379214 -0.853673   \n",
       "positive regulation of deacetylase activity (GO...   0.379214 -1.028864   \n",
       "protein insertion into ER membrane (GO:0045048)      0.379214 -1.082284   \n",
       "endoplasmic reticulum organization (GO:0007029)      0.379214 -1.101828   \n",
       "negative regulation of cellular component movem...   0.434133 -1.262699   \n",
       "positive regulation of protein localization to ...   0.503393 -1.438139   \n",
       "MyD88-independent toll-like receptor signaling ...   0.503393 -1.552174   \n",
       "positive regulation of leukocyte chemotaxis (GO...   0.503393 -1.622296   \n",
       "protein insertion into membrane (GO:0051205)         0.503393 -1.707875   \n",
       "regulation of cell projection organization (GO:...   0.503393 -1.753297   \n",
       "regulation of protein localization to centrosom...   0.503393 -1.810150   \n",
       "negative regulation of trophoblast cell migrati...   0.503393 -1.853311   \n",
       "Notch signaling involved in heart development (...   0.503393 -1.878922   \n",
       "positive regulation of transmembrane transport ...   0.503393 -1.894273   \n",
       "negative regulation of ubiquitin-protein transf...   0.503393 -1.913595   \n",
       "apoptotic mitochondrial changes (GO:0008637)         0.503393 -2.021580   \n",
       "release of cytochrome c from mitochondria (GO:0...   0.503393 -2.049662   \n",
       "positive regulation of NIK/NF-kappaB signaling ...   0.503393 -2.058882   \n",
       "protein palmitoylation (GO:0018345)                  0.503393 -2.076288   \n",
       "positive regulation of type I interferon produc...   0.503393 -2.089959   \n",
       "ear development (GO:0043583)                         0.503393 -2.097912   \n",
       "skeletal muscle organ development (GO:0060538)       0.503393 -2.107319   \n",
       "mature ribosome assembly (GO:0042256)                0.503393 -2.162067   \n",
       "histone H4-K16 acetylation (GO:0043984)              0.503393 -2.171079   \n",
       "peptidyl-L-cysteine S-palmitoylation (GO:0018230)    0.503393 -2.187437   \n",
       "histone H4-K5 acetylation (GO:0043981)               0.503393 -2.207992   \n",
       "positive regulation of DNA binding (GO:0043388)      0.503393 -2.225859   \n",
       "long-term memory (GO:0007616)                        0.503393 -2.272805   \n",
       "positive regulation of binding (GO:0051099)          0.503393 -2.277220   \n",
       "response to interferon-alpha (GO:0035455)            0.503393 -2.324928   \n",
       "DNA topological change (GO:0006265)                  0.503393 -2.356777   \n",
       "regulation of epidermal growth factor receptor ...   0.503393 -2.388200   \n",
       "mitochondrial electron transport, cytochrome c ...   0.503393 -2.401666   \n",
       "regulation of ventricular cardiac muscle cell a...   0.503393 -2.411161   \n",
       "negative regulation of ion transmembrane transp...   0.503393 -2.424520   \n",
       "negative regulation of potassium ion transmembr...   0.503393 -2.431614   \n",
       "fructose metabolic process (GO:0006000)              0.503393 -2.436033   \n",
       "positive regulation of protein sumoylation (GO:...   0.503393 -2.448205   \n",
       "ER-associated misfolded protein catabolic proce...   0.503393 -2.448258   \n",
       "central nervous system projection neuron axonog...   0.503393 -2.462877   \n",
       "negative regulation of cell size (GO:0045792)        0.503393 -2.476053   \n",
       "heart process (GO:0003015)                           0.503393 -2.498356   \n",
       "glial cell migration (GO:0008347)                    0.503393 -2.506536   \n",
       "RNA modification (GO:0009451)                        0.503393 -2.527930   \n",
       "\n",
       "                                                   celltype  \\\n",
       "negative regulation of stress-activated protein...      HRK   \n",
       "cellular response to virus (GO:0098586)                 HRK   \n",
       "positive regulation of cytoplasmic translation ...      HRK   \n",
       "caveola assembly (GO:0070836)                           HRK   \n",
       "negative regulation of establishment of protein...      HRK   \n",
       "positive regulation of transcription regulatory...      HRK   \n",
       "regulation of secretion by cell (GO:1903530)            HRK   \n",
       "positive regulation of deacetylase activity (GO...      HRK   \n",
       "protein insertion into ER membrane (GO:0045048)         HRK   \n",
       "endoplasmic reticulum organization (GO:0007029)         HRK   \n",
       "negative regulation of cellular component movem...      HRK   \n",
       "positive regulation of protein localization to ...      HRK   \n",
       "MyD88-independent toll-like receptor signaling ...      HRK   \n",
       "positive regulation of leukocyte chemotaxis (GO...      HRK   \n",
       "protein insertion into membrane (GO:0051205)            HRK   \n",
       "regulation of cell projection organization (GO:...      HRK   \n",
       "regulation of protein localization to centrosom...      HRK   \n",
       "negative regulation of trophoblast cell migrati...      HRK   \n",
       "Notch signaling involved in heart development (...      HRK   \n",
       "positive regulation of transmembrane transport ...      HRK   \n",
       "negative regulation of ubiquitin-protein transf...      HRK   \n",
       "apoptotic mitochondrial changes (GO:0008637)            HRK   \n",
       "release of cytochrome c from mitochondria (GO:0...      HRK   \n",
       "positive regulation of NIK/NF-kappaB signaling ...      HRK   \n",
       "protein palmitoylation (GO:0018345)                     HRK   \n",
       "positive regulation of type I interferon produc...      HRK   \n",
       "ear development (GO:0043583)                            HRK   \n",
       "skeletal muscle organ development (GO:0060538)          HRK   \n",
       "mature ribosome assembly (GO:0042256)                   HRK   \n",
       "histone H4-K16 acetylation (GO:0043984)                 HRK   \n",
       "peptidyl-L-cysteine S-palmitoylation (GO:0018230)       HRK   \n",
       "histone H4-K5 acetylation (GO:0043981)                  HRK   \n",
       "positive regulation of DNA binding (GO:0043388)         HRK   \n",
       "long-term memory (GO:0007616)                           HRK   \n",
       "positive regulation of binding (GO:0051099)             HRK   \n",
       "response to interferon-alpha (GO:0035455)               HRK   \n",
       "DNA topological change (GO:0006265)                     HRK   \n",
       "regulation of epidermal growth factor receptor ...      HRK   \n",
       "mitochondrial electron transport, cytochrome c ...      HRK   \n",
       "regulation of ventricular cardiac muscle cell a...      HRK   \n",
       "negative regulation of ion transmembrane transp...      HRK   \n",
       "negative regulation of potassium ion transmembr...      HRK   \n",
       "fructose metabolic process (GO:0006000)                 HRK   \n",
       "positive regulation of protein sumoylation (GO:...      HRK   \n",
       "ER-associated misfolded protein catabolic proce...      HRK   \n",
       "central nervous system projection neuron axonog...      HRK   \n",
       "negative regulation of cell size (GO:0045792)           HRK   \n",
       "heart process (GO:0003015)                              HRK   \n",
       "glial cell migration (GO:0008347)                       HRK   \n",
       "RNA modification (GO:0009451)                           HRK   \n",
       "\n",
       "                                                                                                names  \n",
       "negative regulation of stress-activated protein...  negative regulation of stress-activated protei...  \n",
       "cellular response to virus (GO:0098586)                       cellular response to virus (GO:0098586)  \n",
       "positive regulation of cytoplasmic translation ...  positive regulation of cytoplasmic translation...  \n",
       "caveola assembly (GO:0070836)                                           caveola assembly (GO:0070836)  \n",
       "negative regulation of establishment of protein...  negative regulation of establishment of protei...  \n",
       "positive regulation of transcription regulatory...  positive regulation of transcription regulator...  \n",
       "regulation of secretion by cell (GO:1903530)             regulation of secretion by cell (GO:1903530)  \n",
       "positive regulation of deacetylase activity (GO...  positive regulation of deacetylase activity (G...  \n",
       "protein insertion into ER membrane (GO:0045048)       protein insertion into ER membrane (GO:0045048)  \n",
       "endoplasmic reticulum organization (GO:0007029)       endoplasmic reticulum organization (GO:0007029)  \n",
       "negative regulation of cellular component movem...  negative regulation of cellular component move...  \n",
       "positive regulation of protein localization to ...  positive regulation of protein localization to...  \n",
       "MyD88-independent toll-like receptor signaling ...  MyD88-independent toll-like receptor signaling...  \n",
       "positive regulation of leukocyte chemotaxis (GO...  positive regulation of leukocyte chemotaxis (G...  \n",
       "protein insertion into membrane (GO:0051205)             protein insertion into membrane (GO:0051205)  \n",
       "regulation of cell projection organization (GO:...  regulation of cell projection organization (GO...  \n",
       "regulation of protein localization to centrosom...  regulation of protein localization to centroso...  \n",
       "negative regulation of trophoblast cell migrati...  negative regulation of trophoblast cell migrat...  \n",
       "Notch signaling involved in heart development (...  Notch signaling involved in heart development ...  \n",
       "positive regulation of transmembrane transport ...  positive regulation of transmembrane transport...  \n",
       "negative regulation of ubiquitin-protein transf...  negative regulation of ubiquitin-protein trans...  \n",
       "apoptotic mitochondrial changes (GO:0008637)             apoptotic mitochondrial changes (GO:0008637)  \n",
       "release of cytochrome c from mitochondria (GO:0...  release of cytochrome c from mitochondria (GO:...  \n",
       "positive regulation of NIK/NF-kappaB signaling ...  positive regulation of NIK/NF-kappaB signaling...  \n",
       "protein palmitoylation (GO:0018345)                               protein palmitoylation (GO:0018345)  \n",
       "positive regulation of type I interferon produc...  positive regulation of type I interferon produ...  \n",
       "ear development (GO:0043583)                                             ear development (GO:0043583)  \n",
       "skeletal muscle organ development (GO:0060538)         skeletal muscle organ development (GO:0060538)  \n",
       "mature ribosome assembly (GO:0042256)                           mature ribosome assembly (GO:0042256)  \n",
       "histone H4-K16 acetylation (GO:0043984)                       histone H4-K16 acetylation (GO:0043984)  \n",
       "peptidyl-L-cysteine S-palmitoylation (GO:0018230)   peptidyl-L-cysteine S-palmitoylation (GO:0018230)  \n",
       "histone H4-K5 acetylation (GO:0043981)                         histone H4-K5 acetylation (GO:0043981)  \n",
       "positive regulation of DNA binding (GO:0043388)       positive regulation of DNA binding (GO:0043388)  \n",
       "long-term memory (GO:0007616)                                           long-term memory (GO:0007616)  \n",
       "positive regulation of binding (GO:0051099)               positive regulation of binding (GO:0051099)  \n",
       "response to interferon-alpha (GO:0035455)                   response to interferon-alpha (GO:0035455)  \n",
       "DNA topological change (GO:0006265)                               DNA topological change (GO:0006265)  \n",
       "regulation of epidermal growth factor receptor ...  regulation of epidermal growth factor receptor...  \n",
       "mitochondrial electron transport, cytochrome c ...  mitochondrial electron transport, cytochrome c...  \n",
       "regulation of ventricular cardiac muscle cell a...  regulation of ventricular cardiac muscle cell ...  \n",
       "negative regulation of ion transmembrane transp...  negative regulation of ion transmembrane trans...  \n",
       "negative regulation of potassium ion transmembr...  negative regulation of potassium ion transmemb...  \n",
       "fructose metabolic process (GO:0006000)                       fructose metabolic process (GO:0006000)  \n",
       "positive regulation of protein sumoylation (GO:...  positive regulation of protein sumoylation (GO...  \n",
       "ER-associated misfolded protein catabolic proce...  ER-associated misfolded protein catabolic proc...  \n",
       "central nervous system projection neuron axonog...  central nervous system projection neuron axono...  \n",
       "negative regulation of cell size (GO:0045792)           negative regulation of cell size (GO:0045792)  \n",
       "heart process (GO:0003015)                                                 heart process (GO:0003015)  \n",
       "glial cell migration (GO:0008347)                                   glial cell migration (GO:0008347)  \n",
       "RNA modification (GO:0009451)                                           RNA modification (GO:0009451)  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ct = cell_supertype\n",
    "pw = 'low_removed_bp'\n",
    "dt = 'ad_vs_no'\n",
    "\n",
    "tmp2 = all_data[pw]['fits_all'][ct][dt][(abs(all_data[pw]['fits_all'][ct][dt]['P.Value']<0.01))]\n",
    "tmp2.head(50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b1fda07",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%R -o merged_df\n",
    "\n",
    "# library(dplyr)\n",
    "# conflicts_prefer(dplyr::summarise)\n",
    "# conflicts_prefer(dplyr::mutate)\n",
    "\n",
    "\n",
    "# merged_df <- merge(t(gsva_scores[['low_removed_cura_terms']][['Oligodendrocyte']]), \n",
    "#                     summary[['Oligodendrocyte']][, c(factor, subject_id)], \n",
    "#                     by = NULL,\n",
    "#                     by.x = 'row.names', \n",
    "#                     by.y = 'row.names')\n",
    "\n",
    "# rownames(merged_df) <- merged_df$Row.names\n",
    "\n",
    "# merged_df <- merged_df[, -which(names(merged_df) == 'Row.names')]\n",
    "\n",
    "# merged_df <- merged_df %>%\n",
    "#             group_by(.data[[subject_id]]) %>%\n",
    "#             summarise(across(where(is.numeric), mean, na.rm = TRUE))\n",
    "\n",
    "\n",
    "# merged_df <- merged_df[order(merged_df[[factor]]), ]\n",
    "# #colnames(merged_df) <- gsub(\" \", \".\", colnames(merged_df))\n",
    "# merged_df <- merged_df %>% column_to_rownames(\"DonorID\")\n",
    "\n",
    "# cps_df <- all_data[['low_removed_cura_terms']][['fits_all']][['Oligodendrocyte']][['CPS']]\n",
    "# valid_paths <- cps_df[cps_df$adj.P.Val < 0.01, ] %>% column_to_rownames(\"Row.names\")\n",
    "\n",
    "# cps_df <- all_data[['low_removed_cura_terms']][['fits_all']][['Oligodendrocyte']][['late_vs_early']]\n",
    "# valid_paths_lve <- cps_df[cps_df$P.Value < 0.01, ]\n",
    "\n",
    "\n",
    "# merged_df <- merged_df[, c(intersect(rownames(valid_paths), rownames(valid_paths_lve)), factor)]\n",
    "\n",
    "\n",
    "# x <- merged_df[[factor]]\n",
    "# y <- merged_df[[18]]\n",
    "\n",
    "# # Fit the loess curve\n",
    "# loess_fit <- loess(y ~ x, span = 0.75, degree = 2)\n",
    "\n",
    "# # Plot the data and loess curve\n",
    "# plot(x, y)\n",
    "# lines(loess_fit$x, loess_fit$fitted, col = \"red\", lwd = 2)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b56d2708",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%R -o long_df\n",
    "\n",
    "# long_df <- merged_df %>%\n",
    "#   pivot_longer(cols = -ContinuousPseudoprogressionScore,\n",
    "#                names_to = \"pathways\",\n",
    "#                values_to = \"activity.score\")\n",
    "\n",
    "# p1 <- ggplot(long_df, aes(x = .data[[factor]], y = activity.score, color=pathways)) +\n",
    "#   #geom_point() +\n",
    "#   geom_smooth(alpha = 0.01,) +\n",
    "#   #scale_color_manual(values=mod_colors)+\n",
    "#   xlab('CPS') +\n",
    "#   ylab('Activity Score') +\n",
    "#   theme(\n",
    "#     axis.ticks.x = element_blank(),\n",
    "#     axis.text.x = element_blank()\n",
    "#   ) +\n",
    "#   NoLegend()\n",
    "  \n",
    "# p1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "993ba3ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%R\n",
    "\n",
    "# library(ggplot2)\n",
    "# library(dplyr)\n",
    "\n",
    "# data <- pbmc2\n",
    "# disease_group_col_name <- 'condition'\n",
    "# cell_type_col_name <- 'collapsed'\n",
    "\n",
    "# # Calculate the relative abundance of each disease group within each cell type\n",
    "# cell_type_abundance <- data@meta.data %>%\n",
    "#   group_by(!!sym(cell_type_col_name), !!sym(disease_group_col_name)) %>%\n",
    "#   summarize(n = n()) %>%\n",
    "#   group_by(!!sym(cell_type_col_name)) %>%\n",
    "#   mutate(prop = n / sum(n))\n",
    "\n",
    "# # Create the stacked barchart\n",
    "# ggplot(cell_type_abundance, aes(x = !!sym(cell_type_col_name), y = prop, fill = !!sym(disease_group_col_name))) +\n",
    "#   geom_bar(stat = \"identity\", position = \"stack\") +\n",
    "#   labs(x = \"Cell Type\", y = \"Relative Abundance\", fill = \"Disease Group\") +\n",
    "#   theme_classic() +\n",
    "#   theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "99ad781e",
   "metadata": {},
   "source": [
    "### **Jack Kniffing**\n",
    "\n",
    "Jack-knifing is a resampling technique in which we repeat the analysis on a subset of samples to assess the robustness of the observed patterns. If results are driven by outlier samples, we expect to see a bimodal or multi-modal pattern when we jack-knife the analysis. `This technique is most useful when we dealing with large datasets with +10 controls and +10 experiment conditions.`\n",
    "\n",
    "We can perform jack-knife analysis at two levels:\n",
    "\n",
    "- `Leave-one-out-validation (LOOV)`: In this type of analysis, we iteratively exclude one subject/donor from the analysis and examine the p-value distribution of the genes.\n",
    "\n",
    "- `50% resampling:` In this analysis, iterating 100 times, we randomly select 50% of donors from the cohort and repeat the analysis to see if the logFC pattern of genes remain the same or change. The number of iterations that the logFC remain same as the full cohort provide a measure of robustness on the patterns\n",
    "\n",
    "Below we perform jack-knifing to assess the robustness of the DE patterns."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab90cc74",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%R\n",
    "\n",
    "# myJKloov_arrangeFn = function(inputList) {\n",
    "\n",
    "#     # Extract the column name containing p-values\n",
    "#     pval_colname = colnames(inputList[[1]])[grepl(\"P\\\\.Value\", colnames(inputList[[1]]))]\n",
    "    \n",
    "#     # Extract the column name containing log-fold changes\n",
    "#     logFC_colname = colnames(inputList[[1]])[grepl(\"logFC\", colnames(inputList[[1]]))]\n",
    "\n",
    "#     # Calculate z-scores for each data frame in the input list\n",
    "#     for (i in 1:length(inputList)) {\n",
    "#         inputList[[i]]$zscore = qnorm(inputList[[i]][, pval_colname]/2, lower.tail = FALSE) * sign(inputList[[i]][, logFC_colname])\n",
    "#     }\n",
    "    \n",
    "#     # Initialize the result data frame with the pathway and zscore columns from the first data frame\n",
    "#     res = inputList[[1]][, c(\"name\", \"zscore\")]\n",
    "\n",
    "#     # Merge the zscore columns from the remaining data frames based on the \"pathway\" column\n",
    "#     for (i in 2:length(inputList)) {\n",
    "#         res = suppressWarnings(merge(res, inputList[[i]][, c(\"name\", \"zscore\")], by = \"pathway\"))\n",
    "#     }\n",
    "\n",
    "#     # Calculate the maximum p-value for each pathway across all data frames\n",
    "#     res_max_pval = apply(res[,-1], 1, function(x) {\n",
    "#         y = table(sign(x))\n",
    "#         y = y[order(as.numeric(y), decreasing = TRUE)]\n",
    "#         y = names(y)[1]\n",
    "#         x = pnorm(abs(x), lower.tail = FALSE) * 2\n",
    "#         x[sign(x) != y] = 1\n",
    "#         return(max(x))\n",
    "#     })\n",
    "\n",
    "#     # Create a data frame with pathway and corresponding maximum p-value\n",
    "#     res_max_pval = data.frame(pathway = res[, 1], max_pval = as.numeric(res_max_pval), stringsAsFactors = FALSE)\n",
    "\n",
    "#     # Return the result data frame\n",
    "#     return(res_max_pval)\n",
    "# }\n",
    "\n",
    "\n",
    "\n",
    "# myJK50pct_arrangeFn = function(inputList) {\n",
    "\n",
    "#     # Extract the column name containing log-fold changes\n",
    "#     logFC_colname = colnames(inputList[[1]])[grepl(\"logFC\", colnames(inputList[[1]]))]\n",
    "\n",
    "#     # Initialize the result data frame with the \"pathway\" and log-fold change columns from the first data frame\n",
    "#     res = inputList[[1]][, c(\"pathway\", logFC_colname)]\n",
    "\n",
    "#     # Merge the log-fold change columns from the remaining data frames based on the \"gene\" column\n",
    "#     for (i in 2:length(inputList)) {\n",
    "#         res = suppressWarnings(merge(res, inputList[[i]][, c(\"pathway\", logFC_colname)], by = \"pathway\"))\n",
    "#     }\n",
    "\n",
    "#     # Calculate concordance scores for each gene\n",
    "#     res_concordance_score = lapply(1:nrow(res), function(i) {\n",
    "#         x = as.numeric(res[i, -1])\n",
    "#         y = table(sign(x))/length(x)\n",
    "#         y = y[order(as.numeric(y), decreasing = TRUE)]\n",
    "\n",
    "#         return(data.frame(gene = res[i, 1], pattern = names(y)[1], score = y[1], jk_count = length(x), stringsAsFactors = FALSE))\n",
    "#     })\n",
    "\n",
    "#     # Combine the individual concordance score data frames into a single data frame\n",
    "#     res_concordance_score = do.call(\"rbind\", res_concordance_score)\n",
    "\n",
    "#     # Return the result data frame with concordance scores\n",
    "#     return(res_concordance_score)\n",
    "# }\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6c88c80",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%R -o res_list_ad\n",
    "\n",
    "\n",
    "# ########### Setting the parameters ################\n",
    "# jack_knifing = TRUE\n",
    "\n",
    "# if (jack_knifing){\n",
    "\n",
    "#     res_w_jk <- list()\n",
    "    \n",
    "#     for(cell_type in names(summed_counts_per_celltype)){\n",
    "        \n",
    "#         summary <- list()\n",
    "#         tmp_data <- list()\n",
    "    \n",
    "#         res_list_early=list()\n",
    "#         res_list_late=list()\n",
    "#         res_list_ad=list()\n",
    "#         res_list_lve=list()\n",
    "\n",
    "#         for(isbj in unique(colData(summed_counts_per_celltype[[cell_type]])[, subject_id])){\n",
    "#             meta <- colData(summed_counts_per_celltype[[cell_type]])\n",
    "#             meta <- meta[meta[, subject_id]!=isbj, ]\n",
    "#             print('done')\n",
    "\n",
    "#             meta[, random_effect] <- as.factor(meta[, random_effect])\n",
    "#             summary[[cell_type]] <- meta\n",
    "\n",
    "#             tmp_data[[cell_type]] <- gsva_scores[[cell_type]][colnames(gsva_scores[[cell_type]]) %in% as.vector(rownames(summary[[cell_type]]))]\n",
    "                     \n",
    "\n",
    "#             # get linear model fits\n",
    "#             print('getting linear model fits...')\n",
    "#             fits = get_fits(tmp_data, summary, covariates, random_effect)\n",
    "\n",
    "            \n",
    "#             # get matrix of scores for heatmap\n",
    "#             print('get matrix of scores')\n",
    "#             temp_res = get_scores(fits)\n",
    "                                            \n",
    "#             if(isbj %in% unique(meta[, subject_id][meta$pathology.group %in% c(\"no\", \"early\")])){\n",
    "#                 res_list_early = c(res_list_early, list(tmp_res[['all']][[cell_type]][['early_vs_no']][, c(\"names\", \"logFC\", \"P.Value\")]))\n",
    "#             }\n",
    "#             if(isbj %in% unique(meta[, subject_id][meta$pathology.group %in% c(\"no\", \"late\")])){\n",
    "#                 res_list_late = c(res_list_late, list(tmp_res[['all']][[cell_type]][['late_vs_no']][, c(\"names\", \"logFC\", \"P.Value\")]))\n",
    "#             }\n",
    "#             if(isbj %in% unique(meta[, subject_id][meta$pathology.group %in% c(\"no\", \"late\", \"early\")])){\n",
    "#                 res_list_ad = c(res_list_ad, list(tmp_res[['all']][[cell_type]][['ad_vs_no']][, c(\"names\", \"logFC\", \"P.Value\")]))\n",
    "#             }\n",
    "#             if(isbj %in% unique(meta[, subject_id][meta$pathology.group %in% c(\"late\", \"early\")])){\n",
    "#                 res_list_lve = c(res_list_lve, list(tmp_res[['all']][[cell_type]][['late_vs_early']][, c(\"names\", \"logFC\", \"P.Value\")]))\n",
    "#             } \n",
    "#         }\n",
    "\n",
    "#         # res_jk_early=myJKloov_arrangeFn(inputList=res_list_early)\n",
    "#         # colnames(res_jk_early)[2]=paste0(\"jk_early.\",colnames(res_jk_early)[2])\n",
    "        \n",
    "#         # res_jk_late=myJKloov_arrangeFn(inputList=res_list_late)\n",
    "#         # colnames(res_jk_late)[2]=paste0(\"jk_late.\",colnames(res_jk_late)[2])\n",
    "        \n",
    "#         # res_jk_ad=myJKloov_arrangeFn(inputList=res_list_ad)\n",
    "#         # colnames(res_jk_ad)[2]=paste0(\"jk_ad.\",colnames(res_jk_ad)[2])\n",
    "\n",
    "#         # res_jk_lve=myJKloov_arrangeFn(inputList=res_list_lve)\n",
    "#         # colnames(res_jk_lve)[2]=paste0(\"jk_lve.\",colnames(res_jk_lve)[2])\n",
    "\n",
    "#         # res_jk=merge(res_jk_early, res_jk_late, by=\"pathway\")\n",
    "#         # res_jk=merge(res_jk, res_jk_ad, by='pathway')\n",
    "#         # res_jk=merge(res_jk, res_lve, by='pathway')\n",
    "                    \n",
    "#         # res_jk$jk_early_partial_fdr=p.adjust(res_jk$k_early.max_pval[res_jk$early.adj.P.Val<0.1],method=\"BH\")\n",
    "#         # res_jk$jk_late_partial_fdr=p.adjust(res_jk$k_late.max_pval[res_jk$late.adj.P.Val<0.1],method=\"BH\")\n",
    "#         # res_jk$jk_ad_partial_fdr=p.adjust(res_jk$k_ad.max_pval[res_jk$ad.adj.P.Val<0.1],method=\"BH\")\n",
    "#         # res_jk$jk_lve_partial_fdr=p.adjust(res_jk$k_lve.max_pval[res_jk$lve.adj.P.Val<0.1],method=\"BH\")\n",
    "\n",
    "#         # res_w_jk=c(res_w_jk, list(merge(DE_results, res_jk, by=\"pathway\")))\n",
    "                                            \n",
    "#         # rm(res_jk, res_jk_early, res_jk_late, res_jk_ad, res_jk_lve, res_list_early, res_list_late, res_list_ad, res_list_lve)\n",
    "        \n",
    "#     }\n",
    "# } else {\n",
    "#     print('Number of unique subjects <= 10 or `jack_knifing` param set to `FALSE`....')\n",
    "#     print('Warning.....')\n",
    "#     print('Skipping Jack Knifing')\n",
    "#     res_w_jk <- NULL\n",
    "# }\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26ed113c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# valid_pathways = {}\n",
    "# all_pathways = {}\n",
    "\n",
    "# for test_name in test_names:\n",
    "#     valid_pathways[test_name] = {}\n",
    "#     all_pathways[test_name] = pd.DataFrame()\n",
    "#     # Loop through each unique cell type in the cell_type column of adata_annot and filter statistically significant pathways\n",
    "#     for cell_type in celltypes:\n",
    "#         try:\n",
    "#             # Create a dataframe for the current cell type from a matrix in all_data, rename and set index.\n",
    "#             valid_pathways[test_name][cell_type] = pd.DataFrame(all_data['scores_filtered'][test_name][cell_type])\n",
    "#             all_data['scores_all'][test_name]['all'][cell_type].index.name = 'pathway'\n",
    "#             valid_pathways[test_name][cell_type].index.name = 'pathway'\n",
    "#             valid_pathways[test_name][cell_type].rename(columns={cell_type: 'score_adj'}, inplace=True)\n",
    "\n",
    "#             # Merge current cell type dataframe with its corresponding scores_all dataframe on \"pathway\" and remove missing rows\n",
    "#             valid_pathways[test_name][cell_type] = valid_pathways[test_name][cell_type].merge(all_data['scores_all'][test_name]['all'][cell_type],\n",
    "#                                                      how='outer', on='pathway')\n",
    "#             valid_pathways[test_name][cell_type].dropna(axis=0, inplace=True)\n",
    "\n",
    "#             # Filter rows with P.Value less than 0.05 from the current cell type dataframe and sort by P.Value    \n",
    "#             # valid_pathways[test_name][cell_type] = valid_pathways[test_name][cell_type][valid_pathways[test_name][cell_type]['P.Value']<0.05]\n",
    "#             valid_pathways[test_name][cell_type].sort_values(by='P.Value', inplace=True)\n",
    "            \n",
    "#             # Append the current cell type's dataframe to the all_pathways dataframe and sort by pathway\n",
    "#             all_pathways[test_name] = pd.concat([all_pathways[test_name], valid_pathways[test_name][cell_type]], axis=0)\n",
    "#             all_pathways[test_name].sort_values(by='pathway', inplace=True)\n",
    "#         except KeyError:\n",
    "#             continue"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "48d98f9c",
   "metadata": {},
   "source": [
    "### Renaming convention"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2ba11c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Read in a CSV file containing the renaming convention for pathway names\n",
    "# # Remove duplicate names in the renaming convention dataframe\n",
    "# renaming_convention = pd.read_csv('../data/pathway_databases/renaming_convention.csv', dtype=str)\n",
    "# renaming_convention.drop_duplicates(subset='names', inplace=True)\n",
    "\n",
    "# for test_name in test_names:\n",
    "#     # Reset the index of the all_pathways dataframe to allow merging with renamed convention\n",
    "#     all_pathways[test_name].reset_index(inplace=True)\n",
    "#     all_pathways[test_name] = pd.merge(all_pathways[test_name], renaming_convention[['names', 'shortened', 'highlight']], how='left', on='names')\n",
    "#     all_pathways[test_name].set_index(keys='pathway', inplace=True)\n",
    "#     all_pathways[test_name] =  all_pathways[test_name].astype(object).replace(np.nan, 'None')\n",
    "\n",
    "\n",
    "#     dat_dir = f'../results/{test_name}/{save_prefix}/Data/'\n",
    "\n",
    "#     if not os.path.exists(dat_dir):\n",
    "#         os.makedirs(dat_dir)\n",
    "\n",
    "#     # Write the all_pathways dataframe to a CSV file\n",
    "#     all_pathways[test_name].to_csv(dat_dir+f'differentially_expressed_pathways.csv')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "f7dd5249",
   "metadata": {},
   "source": [
    "### Preliminary Heat Map Visualization"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "9894567c",
   "metadata": {},
   "source": [
    "Plot Select Cell-type-specific Pathways"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d330986",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for test_name in test_names:\n",
    "#     print(f'heatmap for select cell-type-specific pathways in {test_name} test')\n",
    "#     list_of_paths_to_annotate = list(all_pathways[test_name].shortened[all_pathways[test_name].highlight=='yes'])\n",
    "#     df = all_pathways[test_name][all_pathways[test_name]['P.Value']<0.05]\n",
    "\n",
    "#     fig_dir = f'../results/{test_name}/{save_prefix}/HeatMap/'\n",
    "#     if not os.path.exists(fig_dir):\n",
    "#         os.makedirs(fig_dir)\n",
    "        \n",
    "#     save_path = fig_dir+'_cell_type_specific_diff_exp_paths.pdf'\n",
    "\n",
    "#     pathway_analyses.plot_and_select_top_deps(df, list_of_paths_to_annotate, save_prefix=save_prefix,\n",
    "#                          save_path=save_path, filter=True, cell_type_specific=True, \n",
    "#                          test_name=test_name, cell_types=celltypes)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "74a487d8",
   "metadata": {},
   "source": [
    "Plot Select Shared Pathways"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "936ac4de",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for test_name in test_names:\n",
    "#     print(f'heatmap for select shared pathways in {test_name} test')\n",
    "#     list_of_paths_to_annotate = list(all_pathways[test_name].shortened[all_pathways[test_name].highlight=='yes'])\n",
    "#     df = all_pathways[test_name][all_pathways[test_name]['P.Value']<0.05]\n",
    "\n",
    "\n",
    "#     fig_dir = f'../results/{test_name}/{save_prefix}/HeatMap/'\n",
    "#     if not os.path.exists(fig_dir):\n",
    "#         os.makedirs(fig_dir)\n",
    "        \n",
    "#     save_path = fig_dir+'filtered_broad_diff_exp_paths.pdf'\n",
    "\n",
    "#     pathway_analyses.plot_and_select_top_deps(df, list_of_paths_to_annotate, save_prefix=save_prefix,\n",
    "#                          save_path=save_path, filter=True, cell_type_specific=False, \n",
    "#                          test_name=test_name, cell_types=celltypes)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19a94ae9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%R -o meta_obj -o expressed_genes_per_celltype -o aggregated_gene_normed_counts_per_celltype\n",
    "\n",
    "# if (pseudobulking_strategy == 'metacell'){\n",
    "\n",
    "#   seurat_obj <- as.Seurat(adata_annot, counts = \"counts\")\n",
    "\n",
    "#   print(seurat_obj)\n",
    "\n",
    "#   # create a hdWGCNA experiment for each celltype\n",
    "#   seurat_dat <- seurat_obj\n",
    "#   aggregated_gene_normed_counts_per_celltype <- list()\n",
    "#   expressed_genes_per_celltype <- list()\n",
    "\n",
    "#   for (cur_cell_type in celltypes){\n",
    "\n",
    "#     seurat_dat <- \n",
    "#     seurat_obj <- as.Seurat(adata_annot[olData(adata_annot)[, cell_type_column]==cur_cell_type], counts = 'counts')\n",
    "\n",
    "#     print(paste0('Creating hdWGNA Experiment for ', toupper(cur_cell_type)))\n",
    "\n",
    "#     if (gene_selection == 'custom') {\n",
    "#       seurat_obj <- SetupForWGCNA(\n",
    "#         seurat_obj,\n",
    "#         gene_select = \"custom\",                                            # the gene selection approach\n",
    "#         gene_list = as.vector(geneSet),    # list of genes to be included\n",
    "#         group.by = cell_type_column,                                            # grouping parameter\n",
    "#         wgcna_name = cur_cell_type                  # the name of the hdWGCNA experiment\n",
    "#       )\n",
    "#     } else {\n",
    "#       seurat_obj <- SetupForWGCNA(\n",
    "#         seurat_obj,\n",
    "#         gene_select = \"fraction\",                               # the gene selection approach\n",
    "#         fraction = gene_celltype_threshold,                     # fraction of cells for gene inclusion\n",
    "#         group.by = cell_type_column',                                 # grouping parameter\n",
    "#         wgcna_name = cur_cell_type       # the name of the hdWGCNA experiment\n",
    "#       )\n",
    "#     }\n",
    "\n",
    "#     print(paste0('Constructing MetaCells in hdWGCNA Experiment for ', toupper(cur_cell_type)))\n",
    "    \n",
    "#     seurat_obj <- MetacellsByGroups(\n",
    "#             seurat_obj = seurat_obj,\n",
    "#             group.by = c(cell_type_column, subject_id), # specify the columns in seurat_obj@meta.data to group by\n",
    "#             reduction = 'harmony',  # select the dimensionality reduction to perform KNN on\n",
    "#             k = 25, # nearest-neighbors parameter\n",
    "#             max_shared = 10, # maximum number of shared cells between two metacells\n",
    "#             ident.group = cell_type_column, # set the Idents of the metacell seurat object\n",
    "#             wgcna_name = cur_cell_type,        # the name of the hdWGCNA experiment\n",
    "#             )\n",
    "\n",
    "#     # normalize metacell expression matrix:\n",
    "#     seurat_obj  <- NormalizeMetacells(seurat_obj, wgcna_name = toupper(substr(cur_cell_type, 1, 3)))\n",
    "\n",
    "#     aggregated_gene_normed_counts_per_celltype[[cur_cell_type]] <- as.SingleCellExperiment(GetMetacellObject(seurat_obj,\n",
    "#                                                       wgcna_name = toupper(substr(cur_cell_type, 1, 3))))   \n",
    "\n",
    "#     expressed_genes_per_celltype[[cur_cell_type]] <- GetWGCNAGenes(seurat_obj)                                       \n",
    "#   }\n",
    "\n",
    "#   rm(seurat_obj)\n",
    "#   rm(seurat_dat)\n",
    "#   rm(adata_annot)\n",
    "# } else {\n",
    "#   aggregated_gene_normed_counts_per_celltype <- NULL\n",
    "#   expressed_genes_per_celltype <- NULL\n",
    "# }\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f76c7da",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%R -i metadata\n",
    "\n",
    "# conflicts_prefer(dplyr::rename)\n",
    "# conflicts_prefer(dplyr::select)\n",
    "\n",
    "\n",
    "# dat_dir <- paste0(\"../results/\", save_prefix, \"/general/\")\n",
    "\n",
    "# if (!dir.exists(dat_dir)) {\n",
    "#     dir.create(dat_dir, recursive = TRUE)\n",
    "# }\n",
    "\n",
    "# summary <- read.csv(metadata, check.names = FALSE)\n",
    "\n",
    "# result_df <- lapply(names(avs_by_ind_out), function(cell_type) {\n",
    "#   cell_df <- avs_by_ind_out[[cell_type]]\n",
    "#   merged_df <- merge(t(cell_df), summary, by.x = 0, by.y = subject_id, all.x = TRUE)\n",
    "#   percentages <- prop.table(table(merged_df$pathology.group))\n",
    "\n",
    "#   data.frame(\n",
    "#     category = names(percentages),\n",
    "#     cell_type = rep(cell_type, length(percentages)),\n",
    "#     percentage = as.numeric(percentages)\n",
    "#   )\n",
    "# }) %>% bind_rows()\n",
    "\n",
    "# result_df <- result_df %>% \n",
    "#             pivot_wider(names_from = cell_type, values_from = percentage) %>%\n",
    "#             pivot_longer(cols = !category, names_to = \"cell_type\", values_to = \"percentage\")\n",
    "\n",
    "# no_reorder <- result_df %>% \n",
    "#   ggplot(aes(x = cell_type, y = percentage)) +\n",
    "#   geom_bar(stat = \"identity\", aes(fill = category)) +\n",
    "#   scale_fill_manual(values = brewer.pal(8, \"Set2\")) +\n",
    "#   labs(x = \"Cell Types\", y = \"Fraction\", fill = \"Pathology Group\") +  # Rename X and Y-axis labels\n",
    "#   scale_x_discrete(labels = result_df$cell_type) +  # Include corresponding cell types on X-ticks\n",
    "#   theme_classic() +\n",
    "#   theme(axis.text.x = element_blank()) +\n",
    "#   theme(axis.text.x = element_text(angle = 45, hjust = 1))  # Adjust X-axis text angle for better visibility\n",
    "\n",
    "# no_reorder\n",
    "\n",
    "# ggsave(paste0(dat_dir, \"cell_type_composition_stacked_bars.svg\"), height = 4, width = 6, bg = \"white\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scRNA_seq_meta_analysis",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.1.undefined"
  },
  "vscode": {
   "interpreter": {
    "hash": "2f414a000012e1980f5fd1119ed6b9bf8cc73b51e37b52eb52a16ff857effd1c"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
