{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import rpy2\n",
    "import logging\n",
    "import warnings\n",
    "import anndata2ri\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import anndata as ad\n",
    "import numpy as np\n",
    "import rpy2.robjects as robjects\n",
    "from rpy2.robjects import pandas2ri\n",
    "from matplotlib.pyplot import rcParams"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-----\n",
      "anndata     0.8.0\n",
      "scanpy      1.9.3\n",
      "-----\n",
      "PIL                         9.5.0\n",
      "anndata2ri                  1.1\n",
      "appnope                     0.1.3\n",
      "asttokens                   NA\n",
      "backcall                    0.2.0\n",
      "cffi                        1.15.1\n",
      "comm                        0.1.3\n",
      "cycler                      0.10.0\n",
      "cython_runtime              NA\n",
      "dateutil                    2.8.2\n",
      "debugpy                     1.6.7\n",
      "decorator                   5.1.1\n",
      "executing                   1.2.0\n",
      "google                      NA\n",
      "h5py                        3.9.0\n",
      "igraph                      0.10.4\n",
      "ipykernel                   6.23.2\n",
      "ipywidgets                  8.0.6\n",
      "jedi                        0.18.2\n",
      "jinja2                      3.1.2\n",
      "joblib                      1.2.0\n",
      "kiwisolver                  1.4.4\n",
      "leidenalg                   0.9.1\n",
      "llvmlite                    0.39.1\n",
      "louvain                     0.8.0\n",
      "markupsafe                  2.1.3\n",
      "matplotlib                  3.7.1\n",
      "mpl_toolkits                NA\n",
      "natsort                     8.4.0\n",
      "numba                       0.56.4\n",
      "numexpr                     2.8.4\n",
      "numpy                       1.23.5\n",
      "packaging                   23.1\n",
      "pandas                      2.0.2\n",
      "parso                       0.8.3\n",
      "pexpect                     4.8.0\n",
      "pickleshare                 0.7.5\n",
      "pkg_resources               NA\n",
      "platformdirs                3.7.0\n",
      "prompt_toolkit              3.0.38\n",
      "psutil                      5.9.5\n",
      "ptyprocess                  0.7.0\n",
      "pure_eval                   0.2.2\n",
      "pyarrow                     12.0.1\n",
      "pydev_ipython               NA\n",
      "pydevconsole                NA\n",
      "pydevd                      2.9.5\n",
      "pydevd_file_utils           NA\n",
      "pydevd_plugins              NA\n",
      "pydevd_tracing              NA\n",
      "pygments                    2.15.1\n",
      "pyparsing                   3.1.0\n",
      "pytz                        2023.3\n",
      "pytz_deprecation_shim       NA\n",
      "rpy2                        3.5.12\n",
      "scipy                       1.10.1\n",
      "session_info                1.0.0\n",
      "setuptools                  68.0.0\n",
      "six                         1.16.0\n",
      "sklearn                     1.2.2\n",
      "stack_data                  0.6.2\n",
      "texttable                   1.6.7\n",
      "threadpoolctl               3.1.0\n",
      "tornado                     6.3.2\n",
      "traitlets                   5.9.0\n",
      "typing_extensions           NA\n",
      "tzlocal                     NA\n",
      "wcwidth                     0.2.6\n",
      "yaml                        6.0\n",
      "zmq                         25.1.0\n",
      "zoneinfo                    NA\n",
      "-----\n",
      "IPython             8.14.0\n",
      "jupyter_client      8.2.0\n",
      "jupyter_core        5.3.1\n",
      "-----\n",
      "Python 3.10.6 (main, Oct 24 2022, 11:04:34) [Clang 12.0.0 ]\n",
      "macOS-10.16-x86_64-i386-64bit\n",
      "-----\n",
      "Session information updated at 2024-02-23 15:12\n"
     ]
    }
   ],
   "source": [
    "# # Ignore R warning messages\n",
    "#Note: this can be commented out to get more verbose R output\n",
    "rpy2.rinterface_lib.callbacks.logger.setLevel(logging.ERROR)\n",
    "\n",
    "warnings.filterwarnings(\"ignore\", category=PendingDeprecationWarning)\n",
    "warnings.filterwarnings(\"ignore\", category=UserWarning)\n",
    "warnings.filterwarnings(\"ignore\", category=FutureWarning)\n",
    "\n",
    "# Automatically convert rpy2 outputs to pandas dataframes\n",
    "pandas2ri.activate()\n",
    "anndata2ri.activate()\n",
    "%load_ext rpy2.ipython\n",
    "\n",
    "# rcParams['figure.dpi'] = get_sys_dpi(1512, 982, 14.125)\n",
    "# rcParams['figure.figsize']=(4,4) #rescale figures\n",
    "\n",
    "sc.settings.verbosity = 3\n",
    "#sc.set_figure_params(dpi=200, dpi_save=300)\n",
    "sc.logging.print_versions()\n",
    "\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# **Load input data in raw form and save in `.h5ad` format**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **List of all datasets present in Gazestani meta-analysis**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "    WARNING: The R package \"reticulate\" only fixed recently\n",
      "    an issue that caused a segfault when used with rpy2:\n",
      "    https://github.com/rstudio/reticulate/pull/1188\n",
      "    Make sure that you use a version of that package that includes\n",
      "    the fix.\n",
      "     [1] \"human_Lake_29227469_VisualCortex\" \"human_Li_nuclei_30545854\"        \n",
      " [3] \"human_Hodge_31435019\"             \"human_Bakken_34616062\"           \n",
      " [5] \"mouse_Chen_28355573\"              \"mouse_Keren-Shaul_28602351\"      \n",
      " [7] \"mouse_SalaFrigerio_exp1_31018141\" \"mouse_SalaFrigerio_exp2_31018141\"\n",
      " [9] \"mouse_Masuda_GSE120745\"           \"mouse_Masuda_GSE120629\"          \n",
      "[11] \"mouse_Masuda_GSE120744\"           \"mouse_Sierksma_31951107\"         \n",
      "[13] \"mouse_Yao_34616066_BICCN_sn\"      \"mouse_Zhou_31932797\"             \n",
      "[15] \"mouse_dulken_31270459\"            \"mouse_Zywitza_30485812\"          \n",
      "[17] \"mouse_Hammond_30471926\"           \"mouse_Hammond_30471926_LPC\"      \n",
      "[19] \"mouse_Bhattacherjee_31519873\"     \"human_Zhou_31932797\"             \n",
      "[21] \"human_Leng_33432193\"              \"human_Gerrits_33609158\"          \n",
      "[23] \"human_Smajic_34919646\"            \"human_NPH\"                       \n",
      "[25] \"human_Kamath_35513515\"            \"human_Grubman_31768052\"          \n",
      "[27] \"human_Velmeshev_31097668\"         \"human_Mathys_31042697\"           \n",
      "[29] \"human_Lau_32989152\"               \"human_Jakel_30747918\"            \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "qs 0.25.5\n",
       "\n",
       "Attaching package: ‘dplyr’\n",
       "\n",
       "The following objects are masked from ‘package:stats’:\n",
       "\n",
       "    filter, lag\n",
       "\n",
       "The following objects are masked from ‘package:base’:\n",
       "\n",
       "    intersect, setdiff, setequal, union\n",
       "\n",
       "Loading required package: SummarizedExperiment\n",
       "Loading required package: MatrixGenerics\n",
       "Loading required package: matrixStats\n",
       "\n",
       "Attaching package: ‘matrixStats’\n",
       "\n",
       "The following object is masked from ‘package:dplyr’:\n",
       "\n",
       "    count\n",
       "\n",
       "\n",
       "Attaching package: ‘MatrixGenerics’\n",
       "\n",
       "The following objects are masked from ‘package:matrixStats’:\n",
       "\n",
       "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
       "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
       "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
       "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
       "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
       "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
       "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
       "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
       "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
       "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
       "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
       "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
       "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
       "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
       "    rowWeightedSds, rowWeightedVars\n",
       "\n",
       "Loading required package: GenomicRanges\n",
       "Loading required package: stats4\n",
       "Loading required package: BiocGenerics\n",
       "\n",
       "Attaching package: ‘BiocGenerics’\n",
       "\n",
       "The following objects are masked from ‘package:dplyr’:\n",
       "\n",
       "    combine, intersect, setdiff, union\n",
       "\n",
       "The following objects are masked from ‘package:stats’:\n",
       "\n",
       "    IQR, mad, sd, var, xtabs\n",
       "\n",
       "The following objects are masked from ‘package:base’:\n",
       "\n",
       "    Filter, Find, Map, Position, Reduce, anyDuplicated, aperm, append,\n",
       "    as.data.frame, basename, cbind, colnames, dirname, do.call,\n",
       "    duplicated, eval, evalq, get, grep, grepl, intersect, is.unsorted,\n",
       "    lapply, mapply, match, mget, order, paste, pmax, pmax.int, pmin,\n",
       "    pmin.int, rank, rbind, rownames, sapply, setdiff, sort, table,\n",
       "    tapply, union, unique, unsplit, which.max, which.min\n",
       "\n",
       "Loading required package: S4Vectors\n",
       "\n",
       "Attaching package: ‘S4Vectors’\n",
       "\n",
       "The following objects are masked from ‘package:dplyr’:\n",
       "\n",
       "    first, rename\n",
       "\n",
       "The following object is masked from ‘package:utils’:\n",
       "\n",
       "    findMatches\n",
       "\n",
       "The following objects are masked from ‘package:base’:\n",
       "\n",
       "    I, expand.grid, unname\n",
       "\n",
       "Loading required package: IRanges\n",
       "\n",
       "Attaching package: ‘IRanges’\n",
       "\n",
       "The following objects are masked from ‘package:dplyr’:\n",
       "\n",
       "    collapse, desc, slice\n",
       "\n",
       "Loading required package: GenomeInfoDb\n",
       "Loading required package: Biobase\n",
       "Welcome to Bioconductor\n",
       "\n",
       "    Vignettes contain introductory material; view with\n",
       "    'browseVignettes()'. To cite Bioconductor, see\n",
       "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
       "\n",
       "\n",
       "Attaching package: ‘Biobase’\n",
       "\n",
       "The following object is masked from ‘package:MatrixGenerics’:\n",
       "\n",
       "    rowMedians\n",
       "\n",
       "The following objects are masked from ‘package:matrixStats’:\n",
       "\n",
       "    anyMissing, rowMedians\n",
       "\n",
       "The legacy packages maptools, rgdal, and rgeos, underpinning the sp package,\n",
       "which was just loaded, were retired in October 2023.\n",
       "Please refer to R-spatial evolution reports for details, especially\n",
       "https://r-spatial.org/r/2023/05/15/evolution4.html.\n",
       "It may be desirable to make the sf package available;\n",
       "package maintainers should consider adding sf to Suggests:.\n",
       "Attaching SeuratObject\n",
       "\n",
       "Attaching package: ‘Seurat’\n",
       "\n",
       "The following object is masked from ‘package:SummarizedExperiment’:\n",
       "\n",
       "    Assays\n",
       "\n",
       "In addition: Warning messages:\n",
       "1: package ‘MatrixGenerics’ was built under R version 4.3.1 \n",
       "2: package ‘S4Vectors’ was built under R version 4.3.1 \n",
       "3: package ‘IRanges’ was built under R version 4.3.1 \n",
       "4: package ‘GenomeInfoDb’ was built under R version 4.3.1 \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "library(qs)\n",
    "library(dplyr)\n",
    "library(SingleCellExperiment)\n",
    "library(readxl)\n",
    "library(scMerge)\n",
    "library(Seurat)\n",
    "\n",
    "mic_annot <- qread(\"../data/raw/gazestani_pfc/annotations/MG_Final_anno.qs\")\n",
    "\n",
    "unique(mic_annot$ds_batch)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_columns_to_string(sce, obs_cols=None, var_cols=None):\n",
    "    \"\"\"\n",
    "    Convert specified columns in the .obs and .var DataFrames of an anndata object to string type.\n",
    "\n",
    "    Parameters:\n",
    "    sce (anndata.AnnData): The single-cell AnnData object to modify.\n",
    "    obs_cols (list of str): Columns in sce.obs to convert to strings.\n",
    "    var_cols (list of str): Columns in sce.var to convert to strings.\n",
    "    \"\"\"\n",
    "    if obs_cols is not None:\n",
    "        for col in obs_cols:\n",
    "            sce.obs[col] = sce.obs[col].astype(str)\n",
    "    \n",
    "    if var_cols is not None:\n",
    "        for col in var_cols:\n",
    "            sce.var[col] = sce.var[col].astype(str)\n",
    "\n",
    "def save_anndata(sce, file_path):\n",
    "    \"\"\"\n",
    "    Save an AnnData object to a file.\n",
    "\n",
    "    Parameters:\n",
    "    sce (anndata.AnnData): The single-cell AnnData object to save.\n",
    "    file_path (str): The path to save the file to.\n",
    "    \"\"\"\n",
    "    sce.write_h5ad(file_path, compression='gzip')\n",
    "\n",
    "\n",
    "\n",
    "def create_minimal_adata(adata_obj, output_file_path, gene_id_column=None):\n",
    "    \"\"\"\n",
    "    Load an .h5ad file, create a minimal AnnData object from it, and save the minimal object.\n",
    "\n",
    "    Parameters:\n",
    "    input_file_path (str): Path to the original .h5ad file.\n",
    "    output_file_path (str): Path where the minimal .h5ad file will be saved.\n",
    "    gene_id_column (str, optional): Column name in adata.var where gene identifiers are stored. \n",
    "                                    If None, it is assumed gene identifiers are stored in adata.var.index.\n",
    "    \"\"\"\n",
    "    try:\n",
    "        # Load the original .h5ad file\n",
    "        my_too_large_adata = adata_obj\n",
    "        \n",
    "        # Create a minimal AnnData object\n",
    "        minimal_adata = ad.AnnData(my_too_large_adata.X)\n",
    "        \n",
    "        # Set gene identifiers\n",
    "        if gene_id_column:\n",
    "            # Option (b): If gene identifiers are stored in a specific column\n",
    "            minimal_adata.var_names = my_too_large_adata.var[gene_id_column]\n",
    "        else:\n",
    "            # Option (a): If gene identifiers are stored in adata.var.index\n",
    "            minimal_adata.var_names = my_too_large_adata.var_names\n",
    "        \n",
    "        # Set observation names\n",
    "        minimal_adata.obs_names = my_too_large_adata.obs_names\n",
    "        \n",
    "        # Save the minimal AnnData object with gzip compression\n",
    "        minimal_adata.write(output_file_path, compression='gzip')\n",
    "        \n",
    "        print(f\"Minimal AnnData object has been saved to {output_file_path}\")\n",
    "    except Exception as e:\n",
    "        print(f\"An error occurred: {e}\")\n",
    "\n",
    "\n",
    "# Function to filter and update AnnData object based on cell barcode annotations\n",
    "def filter_and_update_anndata(ad, annot):\n",
    "\n",
    "    # Filter valid cell barcodes\n",
    "    #valid_barcodes = ad.obs_names.intersection()\n",
    "    ad_filtered = ad[annot['cell_barcode'].to_list()].copy()\n",
    "    \n",
    "    # Merge annotations\n",
    "    ad_filtered.obs = ad_filtered.obs.merge(annot, left_on=ad_filtered.obs_names, right_on='cell_barcode', how='right')\n",
    "    \n",
    "    # Update obs_names with cell barcodes\n",
    "    ad_filtered.obs_names = ad_filtered.obs['cell_barcode']\n",
    "    \n",
    "    return ad_filtered"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **[Gazestani et. al. 2019](https://doi.org/10.1016/j.cell.2023.08.005) (Prefrontal Cortex)**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -o sce_list -o annot_list\n",
    "\n",
    "\n",
    "\n",
    "# Function to read data\n",
    "read_data <- function(path) {\n",
    "  qread(path)\n",
    "}\n",
    "\n",
    "# Function to filter annotations based on ds_batch\n",
    "filter_annotations <- function(annot, batch = 'human_NPH') {\n",
    "  annot %>% filter(ds_batch == batch)\n",
    "}\n",
    "\n",
    "# Function to transform sample column\n",
    "transform_sample <- function(sample_column, start_ind=3) {\n",
    "  sapply(sample_column, function(x) {\n",
    "    paste(unlist(strsplit(x, \"_\"))[start_ind:length(unlist(strsplit(x, \"_\")))], collapse = \"_\")\n",
    "  })\n",
    "}\n",
    "\n",
    "# Function to subset SCE objects based on valid columns\n",
    "subset_sce <- function(sce, annot) {\n",
    "  valid_cols <- colnames(sce)[!is.na(match(colnames(sce), annot$cell_barcode))]\n",
    "  sce[, valid_cols]\n",
    "}\n",
    "\n",
    "# Function to merge annotations with SCE object\n",
    "merge_annotations <- function(sce, annot) {\n",
    "  colData(sce)$cell_barcode <- colnames(sce)\n",
    "  colData(sce) <- DataFrame(as(colData(sce), \"data.frame\") %>% \n",
    "                            right_join(annot, by = \"cell_barcode\"))\n",
    "  colnames(sce) <- colData(sce)$cell_barcode\n",
    "  return(sce)\n",
    "}\n",
    "\n",
    "# Paths\n",
    "base_path <- \"../data/raw/gazestani_pfc/organized_data/Human/brain/snRNA/Gazestani_0001\"\n",
    "annotation_path <- \"../data/raw/gazestani_pfc/annotations\"\n",
    "\n",
    "# Cell types and annotations\n",
    "cell_types <- c(\"Astro\", \"Endo\", \"ExN\", \"InN\", \"MG\", \"Oligo\", \"OPC\")\n",
    "annotations <- paste0(cell_types, \"_Final_anno.qs\")\n",
    "\n",
    "# Read and filter data\n",
    "sce_list <- lapply(paste0(base_path, \"/\", cell_types, \"_data_arranged_updatedId_final_batches.qs\"), read_data)\n",
    "annot_list <- lapply(paste0(annotation_path, \"/\", annotations), read_data)\n",
    "annot_list <- lapply(annot_list, filter_annotations)\n",
    "\n",
    "# Assign names to lists\n",
    "names(sce_list) <- cell_types\n",
    "names(annot_list) <- cell_types\n",
    "\n",
    "# Transform sample column, subset SCE objects, and merge annotations\n",
    "for (cell_type in cell_types) {\n",
    "  annot_list[[cell_type]]$cell_barcode <- transform_sample(annot_list[[cell_type]]$sample)\n",
    "  sce_list[[cell_type]] <- subset_sce(sce_list[[cell_type]], annot_list[[cell_type]])\n",
    "  sce_list[[cell_type]] <- merge_annotations(sce_list[[cell_type]], annot_list[[cell_type]])\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "dat_dir = f'../data/raw/gazestani_pfc/anndata/minimal/'\n",
    "\n",
    "if not os.path.exists(dat_dir):\n",
    "    os.makedirs(dat_dir)\n",
    "\n",
    "\n",
    "# Define the columns to convert for obs and var\n",
    "obs_columns_to_convert = ['anno_braak_score']\n",
    "var_columns_to_convert = ['entrezid']\n",
    "\n",
    "# Define the SCE objects and their corresponding file paths\n",
    "sce_objects = sce_list\n",
    "\n",
    "file_paths = {\n",
    "    'ExN': \"../data/raw/gazestani_pfc/anndata/excitatory_raw_anndata.h5ad\",\n",
    "    'InN': \"../data/raw/gazestani_pfc/anndata/inhibitory_raw_anndata.h5ad\",\n",
    "    'Astro': \"../data/raw/gazestani_pfc/anndata/astrocyte_raw_anndata.h5ad\",\n",
    "    'MG': \"../data/raw/gazestani_pfc/anndata/microglia_raw_anndata.h5ad\",\n",
    "    'Oligo': \"../data/raw/gazestani_pfc/anndata/oligodendrocyte_raw_anndata.h5ad\",\n",
    "    'OPC': \"../data/raw/gazestani_pfc/anndata/opc_raw_anndata.h5ad\",\n",
    "    'Endo': \"../data/raw/gazestani_pfc/anndata/endothelial_raw_anndata.h5ad\"\n",
    "}\n",
    "\n",
    "# Loop through SCE objects to convert column types and save them\n",
    "for sce_name, sce_obj in sce_objects.items():\n",
    "    convert_columns_to_string(sce_obj, obs_cols=obs_columns_to_convert, var_cols=var_columns_to_convert)\n",
    "    save_anndata(sce_obj, file_paths[sce_name])\n",
    "    # create_minimal_adata(sce_obj, '/'.join(file_paths[sce_name].split('/')[:-1] + ['minimal/minimal_'+file_paths[sce_name].split('/')[-1]]))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata = pd.read_excel('../data/raw/gazestani_pfc/full_meta.xlsx')\n",
    "metadata['pathology.group'] = metadata.Status.map({'Abeta': \"early\", 'AbetaTau': 'late', 'Ctrl': 'no'})\n",
    "metadata['individualID'] = metadata['Sbj code'].astype(str)\n",
    "\n",
    "metadata.to_csv('../data/raw/gazestani_pfc/gazestani_pfc_metadata.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **[Gerrits et. al. 2021](https://link.springer.com/article/10.1007/s00401-021-02263-w) (Occipital Cortex)**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -o sce_list -o annot_list\n",
    "\n",
    "# Function to transform sample column\n",
    "transform_sample <- function(sample_column, start_ind=3) {\n",
    "  sapply(sample_column, function(x) {\n",
    "    paste(unlist(strsplit(x, \"_\"))[start_ind:length(unlist(strsplit(x, \"_\")))], collapse = \"_\")\n",
    "  })\n",
    "}\n",
    "\n",
    "# Paths\n",
    "base_path <- \"../data/raw/gazestani_pfc/organized_data/Human/brain/snRNA/Gerrits_33609158\"\n",
    "annotation_path <- \"../data/raw/gazestani_pfc/annotations\"\n",
    "\n",
    "# Cell types and annotations\n",
    "cell_types <- c(\"MG\")\n",
    "annotations <- paste0(cell_types, \"_Final_anno.qs\")\n",
    "\n",
    "# Read and filter data\n",
    "sce_list <- lapply(paste0(base_path, \"/\", cell_types, \"_data_arranged_updatedId_final_batches.qs\"), read_data)\n",
    "annot_list <- lapply(paste0(annotation_path, \"/\", annotations), read_data)\n",
    "annot_list <- lapply(annot_list, function(x) filter_annotations(x, batch = 'human_Gerrits_33609158'))\n",
    "\n",
    "\n",
    "# Assign names to lists\n",
    "names(sce_list) <- cell_types\n",
    "names(annot_list) <- cell_types\n",
    "\n",
    "# Transform sample column, subset SCE objects, and merge annotations\n",
    "for (cell_type in cell_types) {\n",
    "  annot_list[[cell_type]]$cell_barcode <- transform_sample(annot_list[[cell_type]]$sample, 4)\n",
    "  sce_list[[cell_type]] <- subset_sce(sce_list[[cell_type]], annot_list[[cell_type]])\n",
    "  sce_list[[cell_type]] <- merge_annotations(sce_list[[cell_type]], annot_list[[cell_type]])\n",
    "}\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Minimal AnnData object has been saved to ../data/raw/gerrits_otc/anndata/minimal_microglia_raw_anndata.h5ad\n"
     ]
    }
   ],
   "source": [
    "dat_dir = f'../data/raw/gerrits_otc/anndata/minimal/'\n",
    "\n",
    "if not os.path.exists(dat_dir):\n",
    "    os.makedirs(dat_dir)\n",
    "\n",
    "# Define the columns to convert for obs and var\n",
    "obs_columns_to_convert = ['anno_braak_score', 'anno_orig_cellState']\n",
    "var_columns_to_convert = ['entrezid']\n",
    "\n",
    "# Define the SCE objects and their corresponding file paths\n",
    "sce_objects = sce_list\n",
    "\n",
    "file_paths = {\n",
    "    'MG': \"../data/raw/gerrits_otc/anndata/microglia_raw_anndata.h5ad\",\n",
    "}\n",
    "\n",
    "# Loop through SCE objects to convert column types and save them\n",
    "for sce_name, sce_obj in sce_objects.items():\n",
    "    convert_columns_to_string(sce_obj, obs_cols=obs_columns_to_convert, var_cols=var_columns_to_convert)\n",
    "    save_anndata(sce_obj, file_paths[sce_name])\n",
    "    create_minimal_adata(sce_obj, '/'.join(file_paths[sce_name].split('/')[:-1] + ['minimal/minimal_'+file_paths[sce_name].split('/')[-1]]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/70/42xwxjh94x5f164djlw1hwpc0000gv/T/ipykernel_27680/2133377725.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  metadata['pathology_group'] = metadata['sample.group.ch1'].map({'CTR+': \"early\", 'AD': 'late', 'CTR': 'no'})\n",
      "/var/folders/70/42xwxjh94x5f164djlw1hwpc0000gv/T/ipykernel_27680/2133377725.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  metadata['individualID'] = metadata.title.astype(str)\n"
     ]
    }
   ],
   "source": [
    "metadata = sce_list['MG'].obs.drop_duplicates(subset='title', keep='first')\n",
    "metadata['pathology.group'] = metadata['sample.group.ch1'].map({'CTR+': \"early\", 'AD': 'late', 'CTR': 'no'})\n",
    "metadata['individualID'] = metadata.title.astype(str)\n",
    "\n",
    "metadata.to_csv('../data/raw/gerrits_otc/gerrits_otc_metadata.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **[Mathys et. al. 2019](https://www.nature.com/articles/s41586-019-1195-2) (Prefrontal Cortex)**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -o annot_list\n",
    "\n",
    "# Paths\n",
    "annotation_path <- \"../data/raw/gazestani_pfc/annotations\"\n",
    "\n",
    "# Cell types and annotations\n",
    "cell_types <- c(\"Astro\", \"Endo\", \"ExN\", \"InN\", \"MG\", \"Oligo\", \"OPC\")\n",
    "annotations <- paste0(cell_types, \"_Final_anno.qs\")\n",
    "\n",
    "# Read and filter data\n",
    "annot_list <- lapply(paste0(annotation_path, \"/\", annotations), read_data)\n",
    "annot_list <- lapply(annot_list, function(x) filter_annotations(x, batch = 'human_Mathys_31042697'))\n",
    "\n",
    "# Assign names to lists\n",
    "names(annot_list) <- cell_types\n",
    "\n",
    "# Transform sample column, subset SCE objects, and merge annotations\n",
    "for (cell_type in cell_types) {\n",
    "  annot_list[[cell_type]]$cell_barcode <- transform_sample(annot_list[[cell_type]]$sample, 4)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read_mtx('../data/raw/mathys_pfc/notfiltered_count_matrix.mtx').T\n",
    "\n",
    "obs_names = pd.read_csv('../data/raw/mathys_pfc/notfiltered_column_metadata.txt', sep='\\t')\n",
    "obs_names.TAG = obs_names.TAG.str.replace('.', '-')\n",
    "obs_names.set_index('TAG', inplace=True)\n",
    "obs_names.index.rename('index', inplace=True)\n",
    "\n",
    "gene_names = pd.read_csv('../data/raw/mathys_pfc/notfiltered_gene_row_names.txt', sep='\\t', header=None)\n",
    "\n",
    "adata.obs = obs_names\n",
    "adata.var_names = gene_names[1]\n",
    "\n",
    "\n",
    "sce_list = {}\n",
    "for cell_type, annot in annot_list.items():\n",
    "    if len(annot_list[cell_type])>0:\n",
    "        sce_list[cell_type] = filter_and_update_anndata(adata, annot_list[cell_type])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Minimal AnnData object has been saved to ../data/raw/mathys_pfc/anndata/minimal_astrocyte_raw_anndata.h5ad\n",
      "Minimal AnnData object has been saved to ../data/raw/mathys_pfc/anndata/minimal_excitatory_raw_anndata.h5ad\n",
      "Minimal AnnData object has been saved to ../data/raw/mathys_pfc/anndata/minimal_inhibitory_raw_anndata.h5ad\n",
      "Minimal AnnData object has been saved to ../data/raw/mathys_pfc/anndata/minimal_microglia_raw_anndata.h5ad\n",
      "Minimal AnnData object has been saved to ../data/raw/mathys_pfc/anndata/minimal_oligodendrocyte_raw_anndata.h5ad\n",
      "Minimal AnnData object has been saved to ../data/raw/mathys_pfc/anndata/minimal_opc_raw_anndata.h5ad\n"
     ]
    }
   ],
   "source": [
    "dat_dir = f'../data/raw/mathys_pfc/anndata/minimal/'\n",
    "\n",
    "if not os.path.exists(dat_dir):\n",
    "    os.makedirs(dat_dir)\n",
    "\n",
    "# Define the columns to convert for obs and var\n",
    "obs_columns_to_convert = ['anno_braak_score']\n",
    "var_columns_to_convert = None\n",
    "\n",
    "# Define the SCE objects and their corresponding file paths\n",
    "sce_objects = sce_list\n",
    "\n",
    "file_paths = {\n",
    "    'ExN': \"../data/raw/mathys_pfc/anndata/excitatory_raw_anndata.h5ad\",\n",
    "    'InN': \"../data/raw/mathys_pfc/anndata/inhibitory_raw_anndata.h5ad\",\n",
    "    'Astro': \"../data/raw/mathys_pfc/anndata/astrocyte_raw_anndata.h5ad\",\n",
    "    'MG': \"../data/raw/mathys_pfc/anndata/microglia_raw_anndata.h5ad\",\n",
    "    'Oligo': \"../data/raw/mathys_pfc/anndata/oligodendrocyte_raw_anndata.h5ad\",\n",
    "    'OPC': \"../data/raw/mathys_pfc/anndata/opc_raw_anndata.h5ad\",\n",
    "    'Endo': \"../data/raw/mathys_pfc/anndata/endothelial_raw_anndata.h5ad\"\n",
    "}\n",
    "\n",
    "# Loop through SCE objects to convert column types and save them\n",
    "for sce_name, sce_obj in sce_objects.items():\n",
    "    sce_obj.var_names_make_unique()\n",
    "    convert_columns_to_string(sce_obj, obs_cols=obs_columns_to_convert, var_cols=var_columns_to_convert)\n",
    "    save_anndata(sce_obj, file_paths[sce_name])   \n",
    "    create_minimal_adata(sce_obj, '/'.join(file_paths[sce_name].split('/')[:-1] + ['minimal/minimal_'+file_paths[sce_name].split('/')[-1]]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/70/42xwxjh94x5f164djlw1hwpc0000gv/T/ipykernel_27680/4054192223.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  metadata['pathology_group'] = metadata['anno_braak_score'].map({'3': \"early\", '4': 'early', '5': 'late', '6': 'late', '0': 'no', '1': 'no', '2': 'no'})\n",
      "/var/folders/70/42xwxjh94x5f164djlw1hwpc0000gv/T/ipykernel_27680/4054192223.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  metadata['individualID'] = metadata.projid.astype(str)\n"
     ]
    }
   ],
   "source": [
    "metadata = sce_list['Astro'].obs.drop_duplicates(subset='projid', keep='first')\n",
    "metadata['pathology.group'] = metadata['anno_braak_score'].map({'3': \"early\", '4': 'early', '5': 'late', '6': 'late', '0': 'no', '1': 'no', '2': 'no'})\n",
    "metadata['individualID'] = metadata.projid.astype(str)\n",
    "\n",
    "metadata.to_csv('../data/raw/mathys_pfc/mathys_pfc_metadata.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **[Leng et. al. 2021](https://www.nature.com/articles/s41593-020-00764-7) (Superior Frontal Gyrus & Entorhinal Cortex)**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -o annot_list\n",
    "\n",
    "# Paths\n",
    "annotation_path <- \"../data/raw/gazestani_pfc/annotations\"\n",
    "\n",
    "# Cell types and annotations\n",
    "cell_types <- c(\"Astro\", \"Endo\", \"ExN\", \"InN\", \"MG\", \"Oligo\", \"OPC\")\n",
    "annotations <- paste0(cell_types, \"_Final_anno.qs\")\n",
    "\n",
    "# Read and filter data\n",
    "annot_list <- lapply(paste0(annotation_path, \"/\", annotations), read_data)\n",
    "annot_list <- lapply(annot_list, function(x) filter_annotations(x, batch = 'human_Leng_33432193'))\n",
    "\n",
    "# Assign names to lists\n",
    "names(annot_list) <- cell_types\n",
    "\n",
    "# Transform sample column, subset SCE objects, and merge annotations\n",
    "for (cell_type in cell_types) {\n",
    "  annot_list[[cell_type]]$cell_barcode <- transform_sample(annot_list[[cell_type]]$sample, 4)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# processed data obtained \n",
    "adata_leng = {}\n",
    "\n",
    "readRDS = robjects.r['readRDS']\n",
    "df_etc = readRDS('../data/raw/leng_etc/sce.EC.scAlign.assigned.rds')\n",
    "adata_leng['etc'] = df_etc\n",
    "\n",
    "# processed data obtained \n",
    "readRDS = robjects.r['readRDS']\n",
    "df_sfg = readRDS('../data/raw/leng_sfg/sce.SFG.scAlign.assigned.rds')\n",
    "adata_leng['sfg'] = df_sfg\n",
    "\n",
    "cell_types = ['Exc', 'Inh', 'Astro', 'Endo', 'Micro', 'OPC', 'Oligo']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Minimal AnnData object has been saved to ../data/raw/leng_etc/anndata/minimal_excitatory_raw_anndata.h5ad\n",
      "Minimal AnnData object has been saved to ../data/raw/leng_etc/anndata/minimal_inhibitory_raw_anndata.h5ad\n",
      "Minimal AnnData object has been saved to ../data/raw/leng_etc/anndata/minimal_astrocyte_raw_anndata.h5ad\n",
      "Minimal AnnData object has been saved to ../data/raw/leng_etc/anndata/minimal_endothelial_raw_anndata.h5ad\n",
      "Minimal AnnData object has been saved to ../data/raw/leng_etc/anndata/minimal_microglia_raw_anndata.h5ad\n",
      "Minimal AnnData object has been saved to ../data/raw/leng_etc/anndata/minimal_opc_raw_anndata.h5ad\n",
      "Minimal AnnData object has been saved to ../data/raw/leng_etc/anndata/minimal_oligodendrocyte_raw_anndata.h5ad\n",
      "Minimal AnnData object has been saved to ../data/raw/leng_sfg/anndata/minimal_excitatory_raw_anndata.h5ad\n",
      "Minimal AnnData object has been saved to ../data/raw/leng_sfg/anndata/minimal_inhibitory_raw_anndata.h5ad\n",
      "Minimal AnnData object has been saved to ../data/raw/leng_sfg/anndata/minimal_astrocyte_raw_anndata.h5ad\n",
      "Minimal AnnData object has been saved to ../data/raw/leng_sfg/anndata/minimal_endothelial_raw_anndata.h5ad\n",
      "Minimal AnnData object has been saved to ../data/raw/leng_sfg/anndata/minimal_microglia_raw_anndata.h5ad\n",
      "Minimal AnnData object has been saved to ../data/raw/leng_sfg/anndata/minimal_opc_raw_anndata.h5ad\n",
      "Minimal AnnData object has been saved to ../data/raw/leng_sfg/anndata/minimal_oligodendrocyte_raw_anndata.h5ad\n"
     ]
    }
   ],
   "source": [
    "dat_dir = {'etc': f'../data/raw/leng_etc/anndata/minimal/', 'sfg': f'../data/raw/leng_sfg/anndata/minimal/'}\n",
    "\n",
    "\n",
    "# Define the columns to convert for obs and var\n",
    "obs_columns_to_convert = None\n",
    "var_columns_to_convert = None\n",
    "\n",
    "file_paths = {\n",
    "    'etc': {\n",
    "        'Exc': \"../data/raw/leng_etc/anndata/excitatory_raw_anndata.h5ad\",\n",
    "        'Inh': \"../data/raw/leng_etc/anndata/inhibitory_raw_anndata.h5ad\",\n",
    "        'Astro': \"../data/raw/leng_etc/anndata/astrocyte_raw_anndata.h5ad\",\n",
    "        'Micro': \"../data/raw/leng_etc/anndata/microglia_raw_anndata.h5ad\",\n",
    "        'Oligo': \"../data/raw/leng_etc/anndata/oligodendrocyte_raw_anndata.h5ad\",\n",
    "        'OPC': \"../data/raw/leng_etc/anndata/opc_raw_anndata.h5ad\",\n",
    "        'Endo': \"../data/raw/leng_etc/anndata/endothelial_raw_anndata.h5ad\"\n",
    "        },\n",
    "\n",
    "    'sfg':\n",
    "        {   \n",
    "        'Exc': \"../data/raw/leng_sfg/anndata/excitatory_raw_anndata.h5ad\",\n",
    "        'Inh': \"../data/raw/leng_sfg/anndata/inhibitory_raw_anndata.h5ad\",\n",
    "        'Astro': \"../data/raw/leng_sfg/anndata/astrocyte_raw_anndata.h5ad\",\n",
    "        'Micro': \"../data/raw/leng_sfg/anndata/microglia_raw_anndata.h5ad\",\n",
    "        'Oligo': \"../data/raw/leng_sfg/anndata/oligodendrocyte_raw_anndata.h5ad\",\n",
    "        'OPC': \"../data/raw/leng_sfg/anndata/opc_raw_anndata.h5ad\",\n",
    "        'Endo': \"../data/raw/leng_sfg/anndata/endothelial_raw_anndata.h5ad\"\n",
    "        }\n",
    "    }\n",
    "\n",
    "for brain_region in ['etc', 'sfg']:\n",
    "\n",
    "    if not os.path.exists(dat_dir[brain_region]):\n",
    "        os.makedirs(dat_dir[brain_region])\n",
    "\n",
    "    for sce_name in cell_types:\n",
    "        sce_obj = adata_leng[brain_region][adata_leng[brain_region].obs.clusterCellType==sce_name]\n",
    "        convert_columns_to_string(sce_obj, obs_cols=obs_columns_to_convert, var_cols=var_columns_to_convert)\n",
    "        save_anndata(sce_obj, file_paths[brain_region][sce_name])   \n",
    "        create_minimal_adata(sce_obj, '/'.join(file_paths[brain_region][sce_name].split('/')[:-1] + ['minimal/minimal_'+file_paths[brain_region][sce_name].split('/')[-1]]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **[Gabitto et. al. 2021](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC10246227/) (Middle Temporal Gyrus)**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read_h5ad('../data/raw/SEA-AD/MTG/RNASeq/SEAAD_MTG_RNAseq_final-nuclei.2024-02-13.h5ad')\n",
    "\n",
    "adata.obs['cell_type'] = adata.obs['Subclass'].copy()\n",
    "adata.obs['cell_type'] = adata.obs['cell_type'].astype(str)\n",
    "adata.obs.loc[adata.obs['Class'].str.startswith(\"Neuronal:\"), 'cell_type'] = adata.obs['Class'][adata.obs['Class'].str.startswith(\"Neuronal:\")].map({\"Neuronal: Glutamatergic\": \"Excitatory\",\n",
    "                                                                \"Neuronal: GABAergic\": \"Inhibitory\"})\n",
    "\n",
    "adata.obs.loc[adata.obs.cell_type=='Microglia-PVM', 'cell_type'] = 'Microglia'\n",
    "adata.obs['cell_type'] = adata.obs['cell_type'].astype('category')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "dat_dir = f'../data/raw/seaad_mtg/anndata/'\n",
    "\n",
    "if not os.path.exists(dat_dir):\n",
    "    os.makedirs(dat_dir)\n",
    "\n",
    "# Define the columns to convert for obs and var\n",
    "obs_columns_to_convert = None\n",
    "var_columns_to_convert = None\n",
    "\n",
    "# Define the SCE objects and their corresponding file paths\n",
    "\n",
    "file_paths = {\n",
    "    'Excitatory': \"../data/raw/seaad_mtg/anndata/excitatory_raw_anndata.h5ad\",\n",
    "    'Inhibitory': \"../data/raw/seaad_mtg/anndata/inhibitory_raw_anndata.h5ad\",\n",
    "    'Astrocyte': \"../data/raw/seaad_mtg/anndata/astrocyte_raw_anndata.h5ad\",\n",
    "    'Microglia': \"../data/raw/seaad_mtg/anndata/microglia_raw_anndata.h5ad\",\n",
    "    'Oligodendrocyte': \"../data/raw/seaad_mtg/anndata/oligodendrocyte_raw_anndata.h5ad\",\n",
    "    'OPC': \"../data/raw/seaad_mtg/anndata/opc_raw_anndata.h5ad\",\n",
    "    'Endothelial': \"../data/raw/seaad_mtg/anndata/endothelial_raw_anndata.h5ad\",\n",
    "    'VLMC': \"../data/raw/seaad_mtg/anndata/vlmc_raw_anndata.h5ad\"\n",
    "}\n",
    "\n",
    "\n",
    "for sce_name in adata.obs.cell_type.unique():\n",
    "    sce_obj = adata[adata.obs.cell_type==sce_name]\n",
    "    convert_columns_to_string(sce_obj, obs_cols=obs_columns_to_convert, var_cols=var_columns_to_convert)\n",
    "    save_anndata(sce_obj, file_paths[sce_name])   \n",
    "    # create_minimal_adata(sce_obj, '/'.join(file_paths[sce_name].split('/')[:-1] + ['minimal/minimal_'+file_paths[sce_name].split('/')[-1]]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **[Gabitto et. al. 2021](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC10246227/) (Dorso Lateral Prefrontal Cortex)**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read_h5ad('../data/raw/SEA-AD/DLPFC/RNASeq/SEAAD_DLPFC_RNAseq_final-nuclei.2024-02-13.h5ad')\n",
    "\n",
    "adata.obs['cell_type'] = adata.obs['Subclass'].copy()\n",
    "adata.obs['cell_type'] = adata.obs['cell_type'].astype(str)\n",
    "adata.obs.loc[adata.obs['Class'].str.startswith(\"Neuronal:\"), 'cell_type'] = adata.obs['Class'][adata.obs['Class'].str.startswith(\"Neuronal:\")].map({\"Neuronal: Glutamatergic\": \"Excitatory\",\n",
    "                                                                \"Neuronal: GABAergic\": \"Inhibitory\"})\n",
    "\n",
    "adata.obs.loc[adata.obs.cell_type=='Microglia-PVM', 'cell_type'] = 'Microglia'\n",
    "adata.obs['cell_type'] = adata.obs['cell_type'].astype('category')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "dat_dir = f'../data/raw/seaad_pfc/anndata/'\n",
    "\n",
    "if not os.path.exists(dat_dir):\n",
    "    os.makedirs(dat_dir)\n",
    "\n",
    "# Define the columns to convert for obs and var\n",
    "obs_columns_to_convert = None\n",
    "var_columns_to_convert = None\n",
    "\n",
    "# Define the SCE objects and their corresponding file paths\n",
    "\n",
    "file_paths = {\n",
    "    'Excitatory': \"../data/raw/seaad_pfc/anndata/excitatory_raw_anndata.h5ad\",\n",
    "    'Inhibitory': \"../data/raw/seaad_pfc/anndata/inhibitory_raw_anndata.h5ad\",\n",
    "    'Astrocyte': \"../data/raw/seaad_pfc/anndata/astrocyte_raw_anndata.h5ad\",\n",
    "    'Microglia': \"../data/raw/seaad_pfc/anndata/microglia_raw_anndata.h5ad\",\n",
    "    'Oligodendrocyte': \"../data/raw/seaad_pfc/anndata/oligodendrocyte_raw_anndata.h5ad\",\n",
    "    'OPC': \"../data/raw/seaad_pfc/anndata/opc_raw_anndata.h5ad\",\n",
    "    'Endothelial': \"../data/raw/seaad_pfc/anndata/endothelial_raw_anndata.h5ad\",\n",
    "    'VLMC': \"../data/raw/seaad_pfc/anndata/vlmc_raw_anndata.h5ad\"\n",
    "}\n",
    "\n",
    "\n",
    "for sce_name in adata.obs.cell_type.unique():\n",
    "    sce_obj = adata[adata.obs.cell_type==sce_name]\n",
    "    convert_columns_to_string(sce_obj, obs_cols=obs_columns_to_convert, var_cols=var_columns_to_convert)\n",
    "    save_anndata(sce_obj, file_paths[sce_name])   \n",
    "    # create_minimal_adata(sce_obj, '/'.join(file_paths[sce_name].split('/')[:-1] + ['minimal/minimal_'+file_paths[sce_name].split('/')[-1]]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scRNA_seq_meta_analysis",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "2f414a000012e1980f5fd1119ed6b9bf8cc73b51e37b52eb52a16ff857effd1c"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
